// tdesign-mobile-vue/cjs/skeleton/skeleton.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var config = require('../config.js');
var skeleton_props = require('./props.js');
var isArray = require('../_chunks/dep-3cf5587c.js');
var isNumber = require('../_chunks/dep-055d28fd.js');
var isFunction = require('../_chunks/dep-0bd122a0.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var ThemeMap = {
  avatar: [{
    type: "circle",
    size: "48px"
  }],
  image: [{
    type: "rect",
    size: "72px"
  }],
  text: [[{
    width: "24%",
    height: "16px",
    marginRight: "16px"
  }, {
    width: "76%",
    height: "16px"
  }], 1],
  paragraph: [1, 1, 1, {
    width: "55%"
  }]
};
var getColItemStyle = function getColItemStyle(obj) {
  var styleName = ["width", "height", "marginRight", "marginLeft", "margin", "size", "background", "backgroundColor", "borderRadius"];
  var style = {};
  styleName.forEach(function (name) {
    if (name in obj) {
      var px = isNumber.isNumber(obj[name]) ? "".concat(obj[name], "px") : obj[name];
      if (name === "size") {
        var _ref = [px, px];
        style.width = _ref[0];
        style.height = _ref[1];
      } else {
        style[name] = px;
      }
    }
  });
  return style;
};
var _Skeleton = Vue.defineComponent({
  name: "".concat(prefix, "-skeleton"),
  props: skeleton_props["default"],
  setup: function setup(props, _ref2) {
    var slots = _ref2.slots;
    var isShow = Vue.ref(false);
    var renderContent = hooks_tnode.useContent();
    var skeletonClass = hooks_useClass.usePrefixClass("skeleton");
    var getColItemClass = function getColItemClass(obj) {
      return ["".concat(skeletonClass.value, "__col"), "".concat(skeletonClass.value, "--type-").concat(obj.type || "text"), _defineProperty__default["default"]({}, "".concat(skeletonClass.value, "--animation-").concat(props.animation), props.animation)];
    };
    var renderCols = function renderCols(_cols) {
      var cols = [];
      if (isArray.isArray(_cols)) {
        cols = _cols;
      } else if (isNumber.isNumber(_cols)) {
        cols = new Array(_cols).fill({
          type: "text"
        });
      } else {
        cols = [_cols];
      }
      return cols.map(function (obj) {
        return Vue.createVNode("div", {
          "class": getColItemClass(obj),
          "style": getColItemStyle(obj)
        }, [isFunction.isFunction(obj.content) ? obj.content(Vue.h) : obj.content]);
      });
    };
    var renderRowCol = function renderRowCol(_rowCol) {
      var rowCol = _rowCol || props.rowCol;
      return rowCol.map(function (item) {
        return Vue.createVNode("div", {
          "class": "".concat(skeletonClass.value, "__row")
        }, [renderCols(item)]);
      });
    };
    Vue.watch(function () {
      return props.loading;
    }, function (val) {
      if (!val || props.delay === 0) {
        isShow.value = val;
        return;
      }
      setTimeout(function () {
        isShow.value = val;
      }, props.delay);
    }, {
      immediate: true
    });
    return function () {
      var content = renderContent("default", "content");
      if (slots.default) {
        return Vue.createVNode("div", null, [content]);
      }
      if (!isShow.value) {
        return;
      }
      var children = [];
      if (props.rowCol) {
        children.push(renderRowCol(props.rowCol));
      } else if (props.theme) {
        children.push(renderRowCol(ThemeMap[props.theme]));
      } else if (!props.theme && !props.rowCol) {
        children.push(renderRowCol([[{
          width: "24%",
          height: "16px",
          marginRight: "16px"
        }, {
          width: "76%",
          height: "16px"
        }], 1]));
      }
      return Vue.createVNode("div", {
        "class": skeletonClass.value
      }, [children]);
    };
  }
});

exports["default"] = _Skeleton;
