// tdesign-mobile-vue/es/_common/js/message/consts.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var DISTANCE = "32px";
var THEME_LIST = ["info", "success", "warning", "error", "question", "loading"];
var PLACEMENT_OFFSET = {
  center: {
    left: "50%",
    top: "50%",
    transform: "translateX(-50%) translateY(-50%)"
  },
  left: {
    left: DISTANCE,
    top: "50%",
    transform: "translateY(-50%)"
  },
  bottom: {
    bottom: DISTANCE,
    left: "50%",
    transform: "translateX(-50%)"
  },
  right: {
    right: DISTANCE,
    top: "50%",
    transform: "translateY(-50%)"
  },
  top: {
    top: DISTANCE,
    left: "50%",
    transform: "translateX(-50%)"
  },
  "top-left": {
    left: DISTANCE,
    top: DISTANCE
  },
  "top-right": {
    right: DISTANCE,
    top: DISTANCE
  },
  "bottom-left": {
    left: DISTANCE,
    bottom: DISTANCE
  },
  "bottom-right": {
    right: DISTANCE,
    bottom: DISTANCE
  }
};
var PLACEMENT_LIST = Object.keys(PLACEMENT_OFFSET);

export { PLACEMENT_LIST, PLACEMENT_OFFSET, THEME_LIST };
