// tdesign-mobile-vue/cjs/color-picker/color-picker.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var colorPicker_props = require('./props.js');
var config = require('../config.js');
var popup_index = require('../popup/index.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var colorPicker_helper = require('./helper.js');
var _common_js_colorPicker_color = require('../_common/js/color-picker/color.js');
var _common_js_colorPicker_constants = require('../_common/js/color-picker/constants.js');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/defineProperty');
require('@babel/runtime/helpers/toConsumableArray');
require('../_common/js/color-picker/gradient.js');
var colorPicker_constants = require('./constants.js');
require('../popup/popup.js');
require('tdesign-icons-vue-next');
require('../popup/props.js');
require('../overlay/index.js');
require('../overlay/overlay.js');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../overlay/props.js');
require('../shared/component.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../hooks/useLockScroll.js');
require('../_util/useTouch.js');
require('../_util/getScrollParent.js');
require('../hooks/useTeleport.js');
require('../shared/render-tnode.js');
require('../shared/util.js');
require('../_chunks/dep-055d28fd.js');
require('tinycolor2');
require('../_common/js/color-picker/cmyk.js');
require('../_chunks/dep-af0b484e.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

var prefix = config["default"].prefix;
var _ColorPicker = Vue.defineComponent({
  name: "".concat(prefix, "-color-picker"),
  components: {
    TPopup: popup_index.Popup
  },
  props: colorPicker_props["default"],
  setup: function setup(props2) {
    var colorPickerClass = hooks_useClass.usePrefixClass("color-picker");
    var panelRect = Vue.ref();
    var hueSliderRect = Vue.ref();
    var alphaSliderRect = Vue.ref();
    var _toRefs = Vue.toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      _useVModel2$ = _useVModel2[0],
      innerValue = _useVModel2$ === void 0 ? Vue.ref(_common_js_colorPicker_constants.DEFAULT_COLOR) : _useVModel2$,
      setPickerValue = _useVModel2[1];
    var saturationThumbStyle = Vue.ref({
      left: "0",
      top: "0"
    });
    var hueSliderStyle = Vue.ref({
      left: "0",
      color: ""
    });
    var alphaSliderStyle = Vue.ref({
      left: "0",
      color: ""
    });
    var sliderInfo = Vue.ref({
      value: 0
    });
    var color = new _common_js_colorPicker_color.Color(props2.defaultValue || props2.value || _common_js_colorPicker_constants.DEFAULT_COLOR);
    var formatList = Vue.ref(colorPicker_helper.getFormatList(props2.format, color));
    var innerSwatchList = Vue.computed(function () {
      return colorPicker_helper.genSwatchList(props2.swatchColors);
    });
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var getSliderThumbStyle = function getSliderThumbStyle(_ref) {
      var value2 = _ref.value,
        maxValue = _ref.maxValue;
      var dom = hueSliderRect.value;
      var _dom$getBoundingClien = dom.getBoundingClientRect(),
        width = _dom$getBoundingClien.width;
      if (!width) {
        return;
      }
      var left = Math.round(value2 / maxValue * 100);
      return {
        left: "".concat(left, "%"),
        color: color.rgb
      };
    };
    var getSaturationThumbStyle = function getSaturationThumbStyle(_ref2) {
      var saturation = _ref2.saturation,
        value2 = _ref2.value;
      var _panelRect$value$getB = panelRect.value.getBoundingClientRect(),
        width = _panelRect$value$getB.width,
        height = _panelRect$value$getB.height;
      var top = Math.round((1 - value2) * height);
      var left = Math.round(saturation * width);
      return {
        color: color.rgb,
        left: "".concat(left, "px"),
        top: "".concat(top, "px")
      };
    };
    var setCoreStyle = function setCoreStyle() {
      sliderInfo.value = {
        value: color.hue
      };
      formatList.value = colorPicker_helper.getFormatList(props2.format, color);
      if (isMultiple.value) {
        var hue = getSliderThumbStyle({
          value: color.hue,
          maxValue: colorPicker_constants.HUE_MAX
        });
        if (hue) {
          hueSliderStyle.value = hue;
        }
        var alpha = getSliderThumbStyle({
          value: color.alpha * 100,
          maxValue: colorPicker_constants.ALPHA_MAX
        });
        if (alpha) {
          alphaSliderStyle.value = alpha;
        }
        saturationThumbStyle.value = getSaturationThumbStyle({
          saturation: color.saturation,
          value: color.value
        });
      }
    };
    var formatValue = function formatValue() {
      return color.getFormatsColorMap()[props2.format] || color.css;
    };
    var emitColorChange = function emitColorChange(trigger) {
      setPickerValue === null || setPickerValue === void 0 || setPickerValue(formatValue(), {
        trigger: trigger,
        color: _common_js_colorPicker_color.getColorObject(color)
      });
    };
    var clickSwatch = function clickSwatch(swatch) {
      color.update(swatch);
      emitColorChange("preset");
      setCoreStyle();
    };
    var isMultiple = Vue.computed(function () {
      return props2.type === "multiple";
    });
    var getSaturationAndValueByCoordinate = function getSaturationAndValueByCoordinate(coordinate) {
      var _panelRect$value$getB2 = panelRect.value.getBoundingClientRect(),
        width = _panelRect$value$getB2.width,
        height = _panelRect$value$getB2.height;
      var x = coordinate.x,
        y = coordinate.y;
      var saturation = x / width;
      var value2 = 1 - y / height;
      saturation = Math.min(1, Math.max(0, saturation));
      value2 = Math.min(1, Math.max(0, value2));
      return {
        saturation: saturation,
        value: value2
      };
    };
    var onChangeSaturation = function onChangeSaturation(_ref3) {
      var _props2$onPaletteBarC;
      var saturation = _ref3.saturation,
        value2 = _ref3.value;
      var sat = color.saturation,
        val = color.value;
      if (value2 !== val && saturation !== sat) {
        color.saturation = saturation;
        color.value = value2;
      } else if (saturation !== sat) {
        color.saturation = saturation;
      } else if (value2 !== val) {
        color.value = value2;
      } else {
        return;
      }
      (_props2$onPaletteBarC = props2.onPaletteBarChange) === null || _props2$onPaletteBarC === void 0 || _props2$onPaletteBarC.call(props2, {
        color: _common_js_colorPicker_color.getColorObject(color)
      });
      setCoreStyle();
    };
    var handleSaturationDrag = function handleSaturationDrag(e) {
      var coordinate = colorPicker_helper.getCoordinate(e, panelRect.value.getBoundingClientRect(), props2.fixed);
      var _getSaturationAndValu = getSaturationAndValueByCoordinate(coordinate),
        saturation = _getSaturationAndValu.saturation,
        value2 = _getSaturationAndValu.value;
      onChangeSaturation({
        saturation: saturation,
        value: value2
      });
    };
    var onChangeSlider = function onChangeSlider(_ref4) {
      var value2 = _ref4.value,
        isAlpha = _ref4.isAlpha;
      if (isAlpha) {
        color.alpha = value2 / 100;
      } else {
        color.hue = value2;
      }
      emitColorChange(isAlpha ? "palette-alpha-bar" : "palette-hue-bar");
      setCoreStyle();
    };
    var handleSliderDrag = function handleSliderDrag(e) {
      var isAlpha = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var dom = isAlpha ? alphaSliderRect.value : hueSliderRect.value;
      var rect = dom.getBoundingClientRect();
      var width = rect.width;
      var coordinate = colorPicker_helper.getCoordinate(e, rect);
      var x = coordinate.x;
      var maxValue = isAlpha ? colorPicker_constants.ALPHA_MAX : colorPicker_constants.HUE_MAX;
      var value2 = Math.min(maxValue, Math.max(0, Math.round(x / width * maxValue * 100) / 100));
      onChangeSlider({
        value: value2,
        isAlpha: isAlpha
      });
    };
    var handleDiffDrag = function handleDiffDrag(e, dragType) {
      var _callbackMap$dragType;
      e.preventDefault();
      e.stopPropagation();
      var callbackMap = {
        saturation: function saturation() {
          return handleSaturationDrag(e);
        },
        "hue-slider": function hueSlider() {
          return handleSliderDrag(e);
        },
        "alpha-slider": function alphaSlider() {
          return handleSliderDrag(e, true);
        }
      };
      (_callbackMap$dragType = callbackMap[dragType]) === null || _callbackMap$dragType === void 0 || _callbackMap$dragType.call(callbackMap);
    };
    var onTouchStart = function onTouchStart(e, dragType) {
      handleDiffDrag(e, dragType);
    };
    var onTouchMove = function onTouchMove(e, dragType) {
      handleDiffDrag(e, dragType);
    };
    var onTouchEnd = function onTouchEnd(e, dragType) {
      handleDiffDrag(e, dragType);
    };
    var updateColor = function updateColor() {
      var result = innerValue.value || _common_js_colorPicker_constants.DEFAULT_COLOR;
      color.update(result);
    };
    var init = function init() {
      updateColor();
      setCoreStyle();
    };
    Vue.onMounted(function () {
      init();
    });
    return function () {
      var _alphaSliderStyle$val;
      var type = props2.type;
        props2.format;
        var enableAlpha = props2.enableAlpha;
      renderTNodeJSX("header");
      renderTNodeJSX("footer");
      var inner = Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__panel")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__body ").concat(colorPickerClass.value, "__body--").concat(type)
      }, [isMultiple.value && Vue.createVNode("div", {
        "ref": panelRect,
        "class": "".concat(colorPickerClass.value, "__saturation"),
        "style": {
          background: "hsl(".concat(sliderInfo.value.value, ", 100%, 50%)")
        },
        "onTouchstart": function onTouchstart(e) {
          return onTouchStart(e, "saturation");
        },
        "onTouchmove": function onTouchmove(e) {
          return onTouchMove(e, "saturation");
        },
        "onTouchend": function onTouchend(e) {
          return onTouchEnd(e, "saturation");
        }
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__thumb"),
        "style": saturationThumbStyle.value
      }, null)]), isMultiple.value && Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__sliders-wrapper")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__sliders")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__slider-wrapper ").concat(colorPickerClass.value, "__slider-wrapper--hue-type")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__slider"),
        "ref": hueSliderRect,
        "onTouchstart": function onTouchstart(e) {
          return onTouchStart(e, "hue-slider");
        },
        "onTouchmove": function onTouchmove(e) {
          return onTouchMove(e, "hue-slider");
        },
        "onTouchend": function onTouchend(e) {
          return onTouchEnd(e, "hue-slider");
        }
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__rail")
      }, null), Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__thumb"),
        "style": hueSliderStyle.value
      }, null)])]), enableAlpha && Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__slider-wrapper ").concat(colorPickerClass.value, "__slider-wrapper--alpha-type")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__slider-padding ")
      }, null), Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__slider"),
        "ref": alphaSliderRect,
        "onTouchstart": function onTouchstart(e) {
          return onTouchStart(e, "alpha-slider");
        },
        "onTouchmove": function onTouchmove(e) {
          return onTouchMove(e, "alpha-slider");
        },
        "onTouchend": function onTouchend(e) {
          return onTouchEnd(e, "alpha-slider");
        }
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__rail"),
        "style": {
          background: "linear-gradient(to right, rgba(0, 0, 0, 0), ".concat((_alphaSliderStyle$val = alphaSliderStyle.value) === null || _alphaSliderStyle$val === void 0 ? void 0 : _alphaSliderStyle$val.color, ")")
        }
      }, null), Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__thumb"),
        "style": alphaSliderStyle.value
      }, null)])])])]), isMultiple.value && Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__format")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__format-item ").concat(colorPickerClass.value, "__format-item--first")
      }, [formatList.value[0]]), Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__format-item ").concat(colorPickerClass.value, "__format-item--second")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__format-inputs")
      }, [formatList.value[1].map(function (formatItem, formatIndex) {
        return Vue.createVNode("div", {
          "class": "".concat(colorPickerClass.value, "__format-input ").concat(colorPickerClass.value, "__format-input--").concat(formatIndex === formatList.value[1].length - 1 && formatList.value[1].length === 2 ? "fixed" : "base"),
          "key": formatIndex
        }, [formatItem]);
      })])])]), innerSwatchList.value.length && Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__swatches-wrap")
      }, [Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__swatches")
      }, [isMultiple.value && Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__swatches-title")
      }, ["\u7CFB\u7EDF\u9884\u8BBE\u8272\u5F69"]), Vue.createVNode("div", {
        "class": "".concat(colorPickerClass.value, "__swatches-items")
      }, [innerSwatchList.value.map(function (swatch) {
        return Vue.createVNode("div", {
          "class": "".concat(colorPickerClass.value, "__swatches-item"),
          "key": swatch,
          "onClick": function onClick() {
            return clickSwatch(swatch);
          }
        }, [Vue.createVNode("div", {
          "class": "".concat(colorPickerClass.value, "__swatches-inner"),
          "style": {
            backgroundColor: swatch
          }
        }, null)]);
      })])])])])]);
      return inner;
    };
  }
});

exports["default"] = _ColorPicker;
