// tdesign-mobile-vue/cjs/color-picker/helper.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _common_js_colorPicker_constants = require('../_common/js/color-picker/constants.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);

var getCoordinate = function getCoordinate(e, rect, isFixed) {
  var _e$changedTouches;
  var _ref = (e === null || e === void 0 || (_e$changedTouches = e.changedTouches) === null || _e$changedTouches === void 0 ? void 0 : _e$changedTouches[0]) || {},
    pageX = _ref.pageX,
    pageY = _ref.pageY,
    clientY = _ref.clientY;
  return {
    x: Math.min(Math.max(0, pageX - rect.left), rect.width),
    y: Math.min(Math.max(0, (isFixed ? clientY : pageY) - rect.top), rect.height)
  };
};
var getFormatList = function getFormatList(format, color) {
  var FORMAT_MAP = {
    HSV: Object.values(color.getHsva()),
    HSVA: Object.values(color.getHsva()),
    HSL: Object.values(color.getHsla()),
    HSLA: Object.values(color.getHsla()),
    HSB: Object.values(color.getHsla()),
    RGB: Object.values(color.getRgba()),
    RGBA: Object.values(color.getRgba()),
    CMYK: [].concat(_toConsumableArray__default["default"](Object.values(color.getCmyk())), [0]),
    HEX8: [color.hex8, 0],
    CSS: [color.css, 0],
    HEX: [color.hex, 0]
  };
  var cur = FORMAT_MAP[format];
  if (cur) {
    return [format, [].concat(_toConsumableArray__default["default"](cur.slice(0, cur.length - 1)), ["".concat(Math.round(color.alpha * 100), "%")])];
  }
  return ["RGB", FORMAT_MAP.RGB];
};
var genSwatchList = function genSwatchList(prop) {
  if (prop === void 0) {
    return _common_js_colorPicker_constants.DEFAULT_SYSTEM_SWATCH_COLORS.slice(0, 10);
  }
  if (!prop || !prop.length) {
    return [];
  }
  return prop;
};

exports.genSwatchList = genSwatchList;
exports.getCoordinate = getCoordinate;
exports.getFormatList = getFormatList;
