// tdesign-mobile-vue/es/side-bar/side-bar-item.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, getCurrentInstance, inject, computed, onUnmounted, createVNode, mergeProps } from 'vue';
import props from './side-bar-item-props.mjs';
import { Badge } from '../badge/index.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import config from '../config.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../badge/badge.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'lodash-es';
import '../badge/props.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../badge/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';

var prefix = config.prefix;
var _SideBarItem = defineComponent({
  name: "".concat(prefix, "-side-bar-item"),
  props: props,
  setup: function setup(props2) {
    var sideBarItemClass = usePrefixClass("side-bar-item");
    var renderTNodeJSX = useTNodeJSX();
    var internalInstance = getCurrentInstance();
    var proxy = internalInstance.proxy;
    var sideBarProvide = inject("sideBarProvide", void 0);
    sideBarProvide.relation(proxy);
    var isActive = computed(function () {
      return proxy.value === sideBarProvide.currentValue.value;
    });
    var rootClassName = computed(function () {
      return [sideBarItemClass.value, _defineProperty({}, "".concat(sideBarItemClass.value, "--active"), isActive.value), _defineProperty({}, "".concat(sideBarItemClass.value, "--disabled"), props2.disabled)];
    });
    var onClick = function onClick(e) {
      if (props2.disabled) return;
      sideBarProvide.onClickItem(proxy.value, props2.label);
    };
    onUnmounted(function () {
      sideBarProvide.removeRelation(proxy);
    });
    return function () {
      var renderIconNode = function renderIconNode() {
        var iconNode = renderTNodeJSX("icon");
        if (!iconNode) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(sideBarItemClass.value, "__icon")
        }, [iconNode]);
      };
      return createVNode("div", {
        "class": rootClassName.value,
        "onClick": onClick
      }, [isActive.value && createVNode("div", null, [createVNode("div", {
        "class": "".concat(sideBarItemClass.value, "__line")
      }, null), createVNode("div", {
        "class": "".concat(sideBarItemClass.value, "__prefix")
      }, null), createVNode("div", {
        "class": "".concat(sideBarItemClass.value, "__suffix")
      }, null)]), renderIconNode(), props2.badgeProps ? createVNode(Badge, mergeProps(props2.badgeProps, {
        "content": props2.label
      }), null) : createVNode("div", null, [props2.label])]);
    };
  }
});

export { _SideBarItem as default };
