// tdesign-mobile-vue/esm/avatar/avatar.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, inject, computed, createVNode, mergeProps, Fragment } from 'vue';
import { Badge } from '../badge/index.js';
import { Image } from '../image/index.js';
import config from '../config.js';
import AvatarProps from './props.js';
import { useTNodeJSX, useContent } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { isValidSize } from '../_common/js/avatar/utils.js';
import '../badge/badge.js';
import '@babel/runtime/helpers/slicedToArray';
import '@babel/runtime/helpers/defineProperty';
import 'lodash-es';
import '../badge/props.js';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/dom.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/hover.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../badge/style';
import '../badge/type.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../image/image.js';
import 'tdesign-icons-vue-next';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/setStyle.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../loading/icon/spinner.js';
import '../loading/props.js';
import '../loading/style';
import '../loading/type.js';
import '../loading/plugin.js';
import '../image/props.js';
import '../image/style';
import '../image/type.js';

var prefix = config.prefix;
var _Avatar = defineComponent({
  name: "".concat(prefix, "-avatar"),
  props: AvatarProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var avatarClass = usePrefixClass("avatar");
    var avatarGroupProps = inject("avatarGroup", {});
    var hasAvatarGroupProps = Object.keys(avatarGroupProps).length > 0;
    var shapeValue = computed(function () {
      return props.shape || (avatarGroupProps === null || avatarGroupProps === void 0 ? void 0 : avatarGroupProps.shape) || "circle";
    });
    var sizeValue = computed(function () {
      return props.size || (avatarGroupProps === null || avatarGroupProps === void 0 ? void 0 : avatarGroupProps.size);
    });
    var isCustomSize = computed(function () {
      return !isValidSize(sizeValue.value);
    });
    var avatarClasses = computed(function () {
      return ["".concat(avatarClass.value), "".concat(avatarClass.value, "--").concat(isCustomSize.value ? "medium" : sizeValue.value), "".concat(avatarClass.value, "--").concat(shapeValue.value), hasAvatarGroupProps ? "".concat(avatarClass.value, "--border ").concat(avatarClass.value, "--border-").concat(isCustomSize.value ? "medium" : sizeValue.value) : ""];
    });
    var customSize = computed(function () {
      return isCustomSize.value ? {
        height: sizeValue.value,
        width: sizeValue.value,
        "font-size": "".concat(Number.parseInt(sizeValue.value, 10) / 8 * 3 + 2, "px")
      } : {};
    });
    var handleImgLoadError = function handleImgLoadError(e) {
      var _props$onError;
      (_props$onError = props.onError) === null || _props$onError === void 0 || _props$onError.call(props, e);
    };
    return function () {
      var icon = renderTNodeJSX("icon");
      var TNodeContent = renderTNodeContent("default", "content");
      var readerAvatar = function readerAvatar() {
        if (props.image && !props.hideOnLoadFailed) {
          return createVNode(Image, mergeProps({
            "src": props.image,
            "alt": props.alt
          }, props.imageProps, {
            "onError": handleImgLoadError
          }), null);
        }
        if (icon) {
          return createVNode("div", {
            "class": "".concat(avatarClass.value, "__icon")
          }, [icon]);
        }
        return createVNode(Fragment, null, [TNodeContent]);
      };
      return createVNode("div", {
        "class": "".concat(avatarClass.value, "__wrapper")
      }, [createVNode("div", {
        "class": "".concat(avatarClass.value, "__badge")
      }, [createVNode(Badge, props.badgeProps, {
        default: function _default() {
          return [createVNode("div", {
            "class": avatarClasses.value,
            "style": customSize.value
          }, [readerAvatar()])];
        }
      })])]);
    };
  }
});

export { _Avatar as default };
