// tdesign-mobile-vue/lib/button/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var ButtonProps = {
  block: Boolean,
  content: {
    type: [String, Function]
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  ghost: Boolean,
  icon: {
    type: Function
  },
  loading: Boolean,
  loadingProps: {
    type: Object
  },
  shape: {
    type: String,
    default: "rectangle",
    validator: function validator(val) {
      if (!val) return true;
      return ["rectangle", "square", "round", "circle"].includes(val);
    }
  },
  size: {
    type: String,
    default: "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["extra-small", "small", "medium", "large"].includes(val);
    }
  },
  suffix: {
    type: Function
  },
  theme: {
    type: String,
    default: "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "primary", "danger", "light"].includes(val);
    }
  },
  type: {
    type: String,
    default: "button",
    validator: function validator(val) {
      if (!val) return true;
      return ["submit", "reset", "button"].includes(val);
    }
  },
  variant: {
    type: String,
    default: "base",
    validator: function validator(val) {
      if (!val) return true;
      return ["base", "outline", "dashed", "text"].includes(val);
    }
  },
  onClick: Function
};

export { ButtonProps as default };
