// tdesign-mobile-vue/cjs/swipe-cell/useSwipe.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var core = require('@vueuse/core');
var shared_dom = require('../shared/dom.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
var shared_util = require('../shared/util.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../hooks/useClass.js');
require('../_chunks/dep-605cd804.js');
require('@babel/runtime/helpers/defineProperty');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-055d28fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

var noop = function noop() {};
function useSwipe(target) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var _options$threshold = options.threshold,
    threshold = _options$threshold === void 0 ? 50 : _options$threshold,
    onSwipe = options.onSwipe,
    onSwipeEnd = options.onSwipeEnd,
    onSwipeStart = options.onSwipeStart,
    _options$listenerOpti = options.listenerOptions,
    listenerOptions = _options$listenerOpti === void 0 ? false : _options$listenerOpti;
  var coordsStart = Vue.reactive({
    x: 0,
    y: 0
  });
  var coordsEnd = Vue.reactive({
    x: 0,
    y: 0
  });
  var diffX = Vue.computed(function () {
    return coordsStart.x - coordsEnd.x;
  });
  var diffY = Vue.computed(function () {
    return coordsStart.y - coordsEnd.y;
  });
  var max = Math.max,
    abs = Math.abs;
  var isThresholdExceeded = Vue.computed(function () {
    return max(abs(diffX.value), abs(diffY.value)) >= threshold;
  });
  var isSwiping = Vue.ref(false);
  var direction = Vue.computed(function () {
    if (!isThresholdExceeded.value) return "none";
    if (abs(diffX.value) > abs(diffY.value)) {
      return diffX.value > 0 ? "left" : "right";
    }
    return diffY.value > 0 ? "up" : "down";
  });
  var getTouchEventCoords = function getTouchEventCoords(e) {
    return [e.touches[0].clientX, e.touches[0].clientY];
  };
  var updateCoordsStart = function updateCoordsStart(x, y) {
    coordsStart.x = x;
    coordsStart.y = y;
  };
  var updateCoordsEnd = function updateCoordsEnd(x, y) {
    coordsEnd.x = x;
    coordsEnd.y = y;
  };
  var isPassiveEventSupported = checkPassiveEventSupport();
  var onTouchEnd = function onTouchEnd(e) {
    if (isSwiping.value) onSwipeEnd === null || onSwipeEnd === void 0 || onSwipeEnd(e, direction.value);
    isSwiping.value = false;
  };
  var stops = [core.useEventListener(target, "touchstart", function (e) {
    if (e.touches.length !== 1) return;
    if (listenerOptions === true || isObject.isObject(listenerOptions) && listenerOptions.capture && !listenerOptions.passive) shared_dom.preventDefault(e, false);
    var _getTouchEventCoords = getTouchEventCoords(e),
      _getTouchEventCoords2 = _slicedToArray__default["default"](_getTouchEventCoords, 2),
      x = _getTouchEventCoords2[0],
      y = _getTouchEventCoords2[1];
    updateCoordsStart(x, y);
    updateCoordsEnd(x, y);
    onSwipeStart === null || onSwipeStart === void 0 || onSwipeStart(e);
  }, listenerOptions), core.useEventListener(target, "touchmove", function (e) {
    if (e.touches.length !== 1) return;
    var _getTouchEventCoords3 = getTouchEventCoords(e),
      _getTouchEventCoords4 = _slicedToArray__default["default"](_getTouchEventCoords3, 2),
      x = _getTouchEventCoords4[0],
      y = _getTouchEventCoords4[1];
    updateCoordsEnd(x, y);
    if (!isSwiping.value && isThresholdExceeded.value) isSwiping.value = true;
    if (isSwiping.value) onSwipe === null || onSwipe === void 0 || onSwipe(e);
  }, listenerOptions), core.useEventListener(target, "touchend", onTouchEnd, listenerOptions), core.useEventListener(target, "touchcancel", onTouchEnd, listenerOptions)];
  var stop = function stop() {
    return stops.forEach(function (s) {
      return s();
    });
  };
  return {
    isPassiveEventSupported: isPassiveEventSupported,
    isSwiping: isSwiping,
    direction: direction,
    coordsStart: coordsStart,
    coordsEnd: coordsEnd,
    lengthX: diffX,
    lengthY: diffY,
    stop: stop
  };
}
function checkPassiveEventSupport() {
  var document = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : shared_util.isBrowser ? window.document : void 0;
  if (!document) return false;
  var supportsPassive = false;
  var optionsBlock = {
    get passive() {
      supportsPassive = true;
      return false;
    }
  };
  document.addEventListener("x", noop, optionsBlock);
  document.removeEventListener("x", noop);
  return supportsPassive;
}

exports.useSwipe = useSwipe;
