// tdesign-mobile-vue/cjs/notice-bar/notice-bar.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var swiper_index = require('../swiper/index.js');
var noticeBar_props = require('./props.js');
var config = require('../config.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
var isArray = require('../_chunks/dep-3cf5587c.js');
require('../swiper/swiper.js');
require('@babel/runtime/helpers/typeof');
require('../swipe-cell/useSwipe.js');
require('@vueuse/core');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-f84e6e17.js');
require('../shared/util.js');
require('../_chunks/dep-055d28fd.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../swiper/props.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../swiper/swiper-item.js');
require('../shared/component.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var prefix = config["default"].prefix;
var iconDefault = {
  info: Vue.createVNode(tdesignIconsVueNext.InfoCircleFilledIcon, null, null),
  success: Vue.createVNode(tdesignIconsVueNext.CheckCircleFilledIcon, null, null),
  warning: Vue.createVNode(tdesignIconsVueNext.ErrorCircleFilledIcon, null, null),
  error: Vue.createVNode(tdesignIconsVueNext.ErrorCircleFilledIcon, null, null)
};
var _NoticeBar = Vue.defineComponent({
  name: "".concat(prefix, "-notice-bar"),
  props: noticeBar_props["default"],
  emits: ["change"],
  setup: function setup(props2) {
    var noticeBarClass = hooks_useClass.usePrefixClass("notice-bar");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var state = Vue.reactive({
      duration: 0,
      offset: 0,
      listWidth: 0,
      itemWidth: 0,
      timer: null,
      nextTimer: null,
      scroll: {
        marquee: false,
        speed: 50,
        loop: -1,
        delay: 0
      }
    });
    var rootClasses = Vue.computed(function () {
      return ["".concat(noticeBarClass.value), "".concat(noticeBarClass.value, "--").concat(props2.theme)];
    });
    var handleClick = function handleClick(trigger) {
      var _props2$onClick;
      (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, trigger);
    };
    var animateStyle = Vue.computed(function () {
      return {
        transform: state.offset ? "translateX(".concat(state.offset, "px)") : "",
        transitionDuration: "".concat(state.duration, "s"),
        transitionTimingFunction: "linear"
      };
    });
    var listDOM = Vue.ref();
    var itemDOM = Vue.ref();
    var _toRefs = Vue.toRefs(props2),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](visible, modelValue, props2.defaultVisible),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      isShow = _useVModel2[0];
      _useVModel2[1];
    var handleScrolling = function handleScrolling() {
      var _ref;
      if (!(props2 !== null && props2 !== void 0 && props2.marquee) || ((_ref = isObject.isObject(props2 === null || props2 === void 0 ? void 0 : props2.marquee) && (props2 === null || props2 === void 0 ? void 0 : props2.marquee)) === null || _ref === void 0 ? void 0 : _ref.loop) === 0) {
        return;
      }
      if (typeof props2.marquee === "boolean") {
        state.scroll = _objectSpread(_objectSpread({}, state.scroll), {}, {
          marquee: props2.marquee
        });
      }
      if (isObject.isObject(props2.marquee)) {
        var _marquee$speed, _marquee$delay;
        var marquee = props2.marquee;
        state.scroll = {
          marquee: true,
          loop: typeof (marquee === null || marquee === void 0 ? void 0 : marquee.loop) === "undefined" ? state.scroll.loop : marquee.loop,
          speed: (_marquee$speed = marquee.speed) !== null && _marquee$speed !== void 0 ? _marquee$speed : state.scroll.speed,
          delay: (_marquee$delay = marquee.delay) !== null && _marquee$delay !== void 0 ? _marquee$delay : state.scroll.delay
        };
      }
      setTimeout(function () {
        var _listDOM$value, _itemDOM$value;
        var listDOMWidth = (_listDOM$value = listDOM.value) === null || _listDOM$value === void 0 ? void 0 : _listDOM$value.getBoundingClientRect().width;
        var itemDOMWidth = (_itemDOM$value = itemDOM.value) === null || _itemDOM$value === void 0 ? void 0 : _itemDOM$value.getBoundingClientRect().width;
        if (itemDOMWidth >= listDOMWidth) {
          state.offset = -itemDOMWidth;
          state.duration = itemDOMWidth / state.scroll.speed;
          state.listWidth = listDOMWidth;
          state.itemWidth = itemDOMWidth;
        }
      }, state.scroll.delay);
    };
    var handleTransitionend = function handleTransitionend() {
      state.scroll.loop = --state.scroll.loop;
      if (state.scroll.loop === 0) {
        state.scroll = _objectSpread(_objectSpread({}, state.scroll), {}, {
          marquee: false
        });
        return;
      }
      state.offset = state.listWidth;
      state.duration = 0;
      setTimeout(function () {
        state.offset = -state.itemWidth;
        state.duration = (state.itemWidth + state.listWidth) / state.scroll.speed;
      }, 0);
    };
    Vue.onMounted(function () {
      Vue.nextTick(function () {
        if (isShow.value) {
          handleScrolling();
        }
      });
    });
    Vue.watch(function () {
      return isShow.value;
    }, function () {
      Vue.nextTick(function () {
        if (isShow.value) {
          state.offset = state.listWidth;
          state.duration = 0;
          handleScrolling();
        }
      });
    });
    return function () {
      if (isShow.value) {
        var renderPrefixIcon = function renderPrefixIcon() {
          var prefixIconContent = renderTNodeJSX("prefixIcon", {
            defaultNode: iconDefault[props2.theme || "info"]
          });
          if (props2.prefixIcon && prefixIconContent) {
            return Vue.createVNode("div", {
              "class": "".concat(noticeBarClass.value, "__prefix-icon"),
              "onClick": function onClick() {
                return handleClick("prefix-icon");
              }
            }, [prefixIconContent]);
          }
        };
        var renderContent = function renderContent() {
          var _slot;
          var renderShowContent = function renderShowContent() {
            var showContent = renderTNodeJSX("content");
            if (!showContent) {
              return null;
            }
            return showContent;
          };
          var renderOperationContent = function renderOperationContent() {
            var operationContent = renderTNodeJSX("operation");
            if (!operationContent) {
              return null;
            }
            return Vue.createVNode("span", {
              "class": "".concat(noticeBarClass.value, "__operation"),
              "onClick": function onClick(event) {
                event.stopPropagation();
                handleClick("operation");
              }
            }, [operationContent]);
          };
          return Vue.createVNode("div", {
            "ref": listDOM,
            "class": "".concat(noticeBarClass.value, "__content-wrap"),
            "onClick": function onClick() {
              return handleClick("content");
            }
          }, [props2.direction === "vertical" && isArray.isArray(props2.content) ? Vue.createVNode("div", null, [Vue.createVNode(swiper_index.Swiper, {
            "autoplay": true,
            "loop": true,
            "direction": "vertical",
            "duration": 2e3,
            "height": 22,
            "interval": props2 === null || props2 === void 0 ? void 0 : props2.interval,
            "onChange": props2 === null || props2 === void 0 ? void 0 : props2.onChange,
            "class": "".concat(noticeBarClass.value, "__content--vertical")
          }, _isSlot(_slot = props2.content.map(function (item, index) {
            return Vue.createVNode(swiper_index.SwiperItem, {
              "key": index
            }, {
              default: function _default() {
                return [Vue.createVNode("div", {
                  "class": "".concat(noticeBarClass.value, "__content--vertical-item")
                }, [item])];
              }
            });
          })) ? _slot : {
            default: function _default() {
              return [_slot];
            }
          })]) : Vue.createVNode("div", {
            "ref": itemDOM,
            "class": ["".concat(noticeBarClass.value, "__content"), !state.scroll.marquee ? "".concat(noticeBarClass.value, "__content-wrapable") : ""],
            "style": state.scroll.marquee ? animateStyle.value : "",
            "onTransitionend": handleTransitionend
          }, [renderShowContent(), renderOperationContent()])]);
        };
        var renderSuffixIconContent = function renderSuffixIconContent() {
          var suffixIconContent = renderTNodeJSX("suffixIcon");
          if (!suffixIconContent) {
            return null;
          }
          return Vue.createVNode("div", {
            "class": "".concat(noticeBarClass.value, "__suffix-icon"),
            "onClick": function onClick() {
              return handleClick("suffix-icon");
            }
          }, [suffixIconContent]);
        };
        return Vue.createVNode("div", {
          "class": rootClasses.value
        }, [renderPrefixIcon(), renderContent(), renderSuffixIconContent()]);
      }
      return null;
    };
  }
});

exports["default"] = _NoticeBar;
