// tdesign-mobile-vue/es/index.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { c as components } from './_chunks/dep-13686d9e.mjs';
import './plugins.mjs';
import './common.mjs';
export { DialogPlugin as Dialog, DialogPlugin } from './dialog/index.mjs';
export { MessagePlugin as Message, MessagePlugin } from './message/index.mjs';
export { ToastPlugin as Toast, ToastPlugin } from './toast/plugin.mjs';
export { default as DrawerPlugin } from './drawer/plugin.mjs';
export { LoadingPlugin } from './loading/plugin.mjs';
export { _ActionSheet as ActionSheet, _ActionSheet } from './action-sheet/index.mjs';
export { Button } from './button/index.mjs';
export { Divider } from './divider/index.mjs';
export { Fab } from './fab/index.mjs';
export { Icon } from './icon/index.mjs';
export { Col, Row } from './layout/index.mjs';
export { Link } from './link/index.mjs';
export { BackTop } from './back-top/index.mjs';
export { Drawer } from './drawer/index.mjs';
export { Indexes, IndexesAnchor } from './indexes/index.mjs';
export { Navbar } from './navbar/index.mjs';
export { SideBar, SideBarItem } from './side-bar/index.mjs';
export { StepItem, Steps } from './steps/index.mjs';
export { TabBar, TabBarItem } from './tab-bar/index.mjs';
export { TabPanel, Tabs } from './tabs/index.mjs';
export { Calendar } from './calendar/index.mjs';
export { Cascader } from './cascader/index.mjs';
export { Checkbox, CheckboxGroup } from './checkbox/index.mjs';
export { ColorPicker } from './color-picker/index.mjs';
export { DateTimePicker } from './date-time-picker/index.mjs';
export { Form, FormItem } from './form/index.mjs';
export { Input } from './input/index.mjs';
export { Cascade, Picker } from './picker/index.mjs';
export { Radio, RadioGroup } from './radio/index.mjs';
export { Rate } from './rate/index.mjs';
export { Search } from './search/index.mjs';
export { Slider } from './slider/index.mjs';
export { Stepper } from './stepper/index.mjs';
export { Switch } from './switch/index.mjs';
export { Textarea } from './textarea/index.mjs';
export { TreeSelect } from './tree-select/index.mjs';
export { Upload } from './upload/index.mjs';
export { Watermark } from './watermark/index.mjs';
export { Avatar, AvatarGroup } from './avatar/index.mjs';
export { Badge } from './badge/index.mjs';
export { Cell, CellGroup } from './cell/index.mjs';
export { Collapse, CollapsePanel } from './collapse/index.mjs';
export { CountDown } from './count-down/index.mjs';
export { Empty } from './empty/index.mjs';
export { Footer } from './footer/index.mjs';
export { Grid, GridItem } from './grid/index.mjs';
export { Image } from './image/index.mjs';
export { ImageViewer } from './image-viewer/index.mjs';
export { List } from './list/index.mjs';
export { Progress } from './progress/index.mjs';
export { Result } from './result/index.mjs';
export { Skeleton } from './skeleton/index.mjs';
export { Sticky } from './sticky/index.mjs';
export { Swiper, SwiperItem } from './swiper/index.mjs';
export { Table } from './table/index.mjs';
export { CheckTag, Tag } from './tag/index.mjs';
export { QRCode } from './qrcode/index.mjs';
export { DropdownItem, DropdownMenu } from './dropdown-menu/index.mjs';
export { Guide } from './guide/index.mjs';
export { Loading } from './loading/index.mjs';
export { NoticeBar } from './notice-bar/index.mjs';
export { Overlay } from './overlay/index.mjs';
export { Popover } from './popover/index.mjs';
export { Popup } from './popup/index.mjs';
export { PullDownRefresh } from './pull-down-refresh/index.mjs';
export { SwipeCell } from './swipe-cell/index.mjs';
export { ConfigProvider } from './config-provider/index.mjs';
import './toast/index.mjs';
import './style/css.mjs';
import './toast/type.mjs';
import './_chunks/dep-f4e31210.mjs';
import './_chunks/dep-4055ef5c.mjs';
import 'vue';
import 'lodash-es';
import './toast/toast.mjs';
import 'tdesign-icons-vue-next';
import './hooks/tnode.mjs';
import './hooks/render-tnode.mjs';
import './_common/js/utils/general.mjs';
import './toast/props.mjs';
import './config.mjs';
import './hooks/useLockScroll.mjs';
import './_util/useTouch.mjs';
import './_util/getScrollParent.mjs';
import './hooks/useClass.mjs';
import './config-provider/useConfig.mjs';
import './config-provider/context.mjs';
import './_common/js/global-config/mobile/default-config.mjs';
import './_common/js/global-config/mobile/locale/zh_CN.mjs';
import './_chunks/dep-4c96ffa8.mjs';
import './_chunks/dep-04a2869f.mjs';
import './_chunks/dep-a8bd494a.mjs';
import './_chunks/dep-d3be051e.mjs';
import './config-provider/type.mjs';
import './overlay/overlay.mjs';
import './shared/dom.mjs';
import './overlay/props.mjs';
import './shared/index.mjs';
import './shared/functions.mjs';
import './shared/util.mjs';
import './_chunks/dep-2ed7e0d5.mjs';
import './_chunks/dep-9961b95b.mjs';
import './shared/component.mjs';
import './shared/render.mjs';
import './shared/render-tnode.mjs';
import './shared/useCountDown/index.mjs';
import './_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import './shared/useCountDown/utils.mjs';
import './shared/useTouch/index.mjs';
import './shared/useScrollParent/index.mjs';
import './shared/useTest/index.mjs';
import './shared/useClickAway/index.mjs';
import './_chunks/dep-fb069758.mjs';
import './shared/hover.mjs';
import './dialog/dialog.mjs';
import './dialog/props.mjs';
import './button/button.mjs';
import './button/props.mjs';
import './form/hooks.mjs';
import './loading/loading.mjs';
import './loading/icon/gradient.mjs';
import './_common/js/loading/circle-adapter.mjs';
import './_common/js/utils/setStyle.mjs';
import './_common/js/utils/helper.mjs';
import './_chunks/dep-a27c5c00.mjs';
import './loading/icon/spinner.mjs';
import './loading/props.mjs';
import './loading/type.mjs';
import './button/type.mjs';
import './popup/popup.mjs';
import './popup/props.mjs';
import './hooks/useVModel.mjs';
import './hooks/useTeleport.mjs';
import './popup/type.mjs';
import './message/message.mjs';
import './message/props.mjs';
import './link/link.mjs';
import './link/props.mjs';
import './link/type.mjs';
import './drawer/drawer.mjs';
import './drawer/props.mjs';
import './action-sheet/action-sheet.mjs';
import './action-sheet/action-sheet-grid.mjs';
import './grid/grid.mjs';
import './grid/props.mjs';
import './grid/grid-item.mjs';
import './grid/grid-item-props.mjs';
import './image/image.mjs';
import './image/props.mjs';
import './image/type.mjs';
import './badge/badge.mjs';
import './badge/props.mjs';
import './badge/type.mjs';
import './grid/type.mjs';
import './swiper/swiper.mjs';
import './swipe-cell/useSwipe.mjs';
import './swiper/props.mjs';
import './swiper/swiper-item.mjs';
import './swiper/type.mjs';
import './action-sheet/action-sheet-list.mjs';
import './action-sheet/props.mjs';
import './action-sheet/type.mjs';
import './divider/divider.mjs';
import './divider/props.mjs';
import './divider/type.mjs';
import './fab/fab.mjs';
import './fab/props.mjs';
import './layout/row.mjs';
import './layout/row-props.mjs';
import './layout/constants.mjs';
import './layout/col.mjs';
import './layout/col-props.mjs';
import './layout/type.mjs';
import './back-top/back-top.mjs';
import './back-top/props.mjs';
import './back-top/type.mjs';
import './drawer/type.mjs';
import './indexes/indexes.mjs';
import './indexes/props.mjs';
import './indexes/indexes-anchor.mjs';
import './indexes/indexes-anchor-props.mjs';
import './indexes/type.mjs';
import './navbar/navbar.mjs';
import './navbar/props.mjs';
import './navbar/type.mjs';
import './side-bar/side-bar.mjs';
import './side-bar/props.mjs';
import './side-bar/side-bar-item.mjs';
import './side-bar/side-bar-item-props.mjs';
import './side-bar/type.mjs';
import './steps/steps.mjs';
import './steps/props.mjs';
import './steps/step-item.mjs';
import './steps/step-item-props.mjs';
import './steps/type.mjs';
import './tab-bar/tab-bar.mjs';
import './tab-bar/props.mjs';
import './hooks/useChildSlots.mjs';
import './tab-bar/tab-bar-item.mjs';
import './tab-bar/useTabBar.mjs';
import './tab-bar/tab-bar-item-props.mjs';
import './tab-bar/type.mjs';
import './tabs/tabs.mjs';
import './tabs/props.mjs';
import './tabs/tab-nav-item.mjs';
import './tabs/tab-panel-props.mjs';
import './hooks/useCommonClassName.mjs';
import './sticky/sticky.mjs';
import './sticky/props.mjs';
import './sticky/type.mjs';
import './tabs/tab-panel.mjs';
import './tabs/type.mjs';
import './calendar/calendar.mjs';
import './calendar/props.mjs';
import './calendar/template.mjs';
import './calendar/template-props.mjs';
import './calendar/utils.mjs';
import './calendar/type.mjs';
import './cascader/cascader.mjs';
import './cascader/props.mjs';
import './radio/radio.mjs';
import './radio/props.mjs';
import './radio/radio-group.mjs';
import './radio/radio-group-props.mjs';
import './radio/type.mjs';
import './cascader/type.mjs';
import './checkbox/checkbox.mjs';
import './checkbox/props.mjs';
import './checkbox/checkbox-group.mjs';
import './checkbox/checkbox-group-props.mjs';
import './_common/js/log/index.mjs';
import './_common/js/log/log.mjs';
import './checkbox/hooks/index.mjs';
import './checkbox/hooks/getOptions.mjs';
import './checkbox/hooks/setCheckAllStatus.mjs';
import './checkbox/type.mjs';
import './color-picker/color-picker.mjs';
import './color-picker/props.mjs';
import './color-picker/helper.mjs';
import './_common/js/color-picker/constants.mjs';
import './_common/js/color-picker/index.mjs';
import './_common/js/color-picker/cmyk.mjs';
import './_common/js/color-picker/color.mjs';
import './_chunks/dep-fba58676.mjs';
import 'tinycolor2';
import './_common/js/color-picker/gradient.mjs';
import './_common/js/color-picker/draggable.mjs';
import './_common/js/color-picker/format.mjs';
import './_common/js/color-picker/types.mjs';
import './color-picker/constants.mjs';
import './color-picker/type.mjs';
import './date-time-picker/date-time-picker.mjs';
import './_chunks/dep-f7dc244a.mjs';
import './date-time-picker/props.mjs';
import './date-time-picker/shared.mjs';
import './picker/picker.mjs';
import './picker/props.mjs';
import './picker/picker-item.mjs';
import './picker/picker.class.mjs';
import './picker/utils.mjs';
import './picker/cascade.mjs';
import './picker/type.mjs';
import './date-time-picker/type.mjs';
import './form/form.mjs';
import './form/props.mjs';
import './form/const.mjs';
import './form/form-item.mjs';
import './form/form-model.mjs';
import './form/form-item-props.mjs';
import './form/type.mjs';
import './input/input.mjs';
import './input/props.mjs';
import './hooks/useLengthLimit.mjs';
import './input/type.mjs';
import './rate/rate.mjs';
import './rate/props.mjs';
import './rate/type.mjs';
import './search/search.mjs';
import './search/props.mjs';
import './_common/js/common.mjs';
import './cell/cell.mjs';
import './cell/props.mjs';
import './search/type.mjs';
import './slider/slider.mjs';
import './slider/props.mjs';
import './slider/tool.mjs';
import './slider/type.mjs';
import './stepper/stepper.mjs';
import './stepper/props.mjs';
import './stepper/type.mjs';
import './switch/switch.mjs';
import './hooks/useToggle.mjs';
import './switch/props.mjs';
import './switch/type.mjs';
import './textarea/textarea.mjs';
import './textarea/props.mjs';
import './_common/js/utils/calcTextareaHeight.mjs';
import './textarea/type.mjs';
import './tree-select/tree-select.mjs';
import './tree-select/props.mjs';
import './tree-select/type.mjs';
import './upload/upload.mjs';
import './upload/props.mjs';
import './upload/hooks/useUpload.mjs';
import './_common/js/upload/main.mjs';
import './_common/js/upload/utils.mjs';
import './_common/js/upload/xhr.mjs';
import './image-viewer/image-viewer.mjs';
import './image-viewer/props.mjs';
import './hooks/useDefaultValue.mjs';
import './hooks/useGesture.mjs';
import '@use-gesture/vanilla';
import './image-viewer/type.mjs';
import './upload/type.mjs';
import './watermark/watermark.mjs';
import './_common/js/watermark/generateBase64Url.mjs';
import './_common/js/watermark/randomMovingStyle.mjs';
import './_common/js/utils/injectStyle.mjs';
import './hooks/index.mjs';
import './hooks/icon.mjs';
import './hooks/useExpose.mjs';
import './hooks/useMutationObservable.mjs';
import './hooks/useVariables.mjs';
import './_common/js/utils/getColorTokenColor.mjs';
import './watermark/hooks/index.mjs';
import './watermark/props.mjs';
import './watermark/type.mjs';
import './avatar/avatar.mjs';
import './avatar/props.mjs';
import './_common/js/avatar/utils.mjs';
import './avatar/avatar-group.mjs';
import './avatar/avatar-group-props.mjs';
import './avatar/type.mjs';
import './cell/cell-group.mjs';
import './cell/cell-group-props.mjs';
import './cell/type.mjs';
import './collapse/collapse.mjs';
import './collapse/props.mjs';
import './collapse/collapse-panel.mjs';
import './collapse/collapse-panel-props.mjs';
import './collapse/util.mjs';
import './collapse/type.mjs';
import './count-down/count-down.mjs';
import './count-down/props.mjs';
import './count-down/type.mjs';
import './empty/empty.mjs';
import './empty/props.mjs';
import './empty/type.mjs';
import './footer/footer.mjs';
import './footer/props.mjs';
import './footer/type.mjs';
import './list/list.mjs';
import './list/props.mjs';
import './list/type.mjs';
import './progress/progress.mjs';
import './progress/utils.mjs';
import './progress/props.mjs';
import './_common/js/progress/const.mjs';
import './_common/js/progress/utils.mjs';
import './progress/type.mjs';
import './result/result.mjs';
import './result/props.mjs';
import './result/type.mjs';
import './skeleton/skeleton.mjs';
import './skeleton/props.mjs';
import './skeleton/type.mjs';
import './table/base-table.mjs';
import './table/base-table-props.mjs';
import './table/hooks/useClassName.mjs';
import './table/hooks/useStyle.mjs';
import './table/hooks/useFixed.mjs';
import './table/utils/index.mjs';
import './table/type.mjs';
import './table/interface.mjs';
import './tag/tag.mjs';
import './tag/props.mjs';
import './tag/check-tag.mjs';
import './tag/check-tag-props.mjs';
import './tag/type.mjs';
import './qrcode/qrcode.mjs';
import './qrcode/props.mjs';
import './_common/js/qrcode/utils.mjs';
import './_common/js/qrcode/qrcodegen.mjs';
import './qrcode/components/qrcode-canvas.mjs';
import './qrcode/components/props.mjs';
import './qrcode/hooks/useQRCode.mjs';
import './qrcode/components/qrcode-svg.mjs';
import './qrcode/components/qrcode-status.mjs';
import './qrcode/type.mjs';
import './dropdown-menu/dropdown-menu.mjs';
import './dropdown-menu/context.mjs';
import './dropdown-menu/dom-utils.mjs';
import './dropdown-menu/props.mjs';
import './dropdown-menu/dropdown-item.mjs';
import './dropdown-menu/dropdown-item-props.mjs';
import './dropdown-menu/type.mjs';
import './guide/guide.mjs';
import './guide/props.mjs';
import './guide/utils/index.mjs';
import './guide/utils/dom.mjs';
import './popover/popover.mjs';
import './popover/props.mjs';
import './popover/type.mjs';
import './guide/type.mjs';
import './notice-bar/notice-bar.mjs';
import './notice-bar/props.mjs';
import './notice-bar/type.mjs';
import './pull-down-refresh/pull-down-refresh.mjs';
import './pull-down-refresh/props.mjs';
import './pull-down-refresh/useTouch.mjs';
import './pull-down-refresh/type.mjs';
import './swipe-cell/swipe-cell.mjs';
import './swipe-cell/props.mjs';
import './swipe-cell/useSureConfirm.mjs';
import './swipe-cell/type.mjs';
import './config-provider/config-provider.mjs';
import './config-provider/props.mjs';

var install = function install(app, option) {
  Object.keys(components).forEach(function (key) {
    app.use(components[key], option);
  });
};
var version = typeof "1.11.1" === "undefined" ? "" : "1.11.1";
var index = {
  install: install,
  version: version
};

export { index as default, install, version };
