// tdesign-mobile-vue/es/plugins.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

export { DialogPlugin as Dialog, DialogPlugin } from './dialog/index.mjs';
export { MessagePlugin as Message, MessagePlugin } from './message/index.mjs';
import './toast/index.mjs';
import './drawer/index.mjs';
import './loading/index.mjs';
export { _ActionSheet as ActionSheet } from './action-sheet/index.mjs';
export { ToastPlugin as Toast, ToastPlugin } from './toast/plugin.mjs';
export { default as DrawerPlugin } from './drawer/plugin.mjs';
export { LoadingPlugin } from './loading/plugin.mjs';
import './_chunks/dep-f4e31210.mjs';
import './_chunks/dep-4055ef5c.mjs';
import 'vue';
import './dialog/dialog.mjs';
import './_chunks/dep-fb069758.mjs';
import './_chunks/dep-9961b95b.mjs';
import 'tdesign-icons-vue-next';
import 'lodash-es';
import './button/index.mjs';
import './button/button.mjs';
import './shared/index.mjs';
import './shared/functions.mjs';
import './shared/util.mjs';
import './_chunks/dep-2ed7e0d5.mjs';
import './hooks/useClass.mjs';
import './config-provider/useConfig.mjs';
import './config-provider/context.mjs';
import './_common/js/global-config/mobile/default-config.mjs';
import './_common/js/global-config/mobile/locale/zh_CN.mjs';
import './_chunks/dep-4c96ffa8.mjs';
import './_chunks/dep-04a2869f.mjs';
import './_chunks/dep-a8bd494a.mjs';
import './_chunks/dep-d3be051e.mjs';
import './config-provider/type.mjs';
import './shared/component.mjs';
import './shared/render.mjs';
import './shared/dom.mjs';
import './shared/render-tnode.mjs';
import './shared/useCountDown/index.mjs';
import './_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import './shared/useCountDown/utils.mjs';
import './shared/useTouch/index.mjs';
import './shared/useScrollParent/index.mjs';
import './shared/useTest/index.mjs';
import './shared/useClickAway/index.mjs';
import './shared/hover.mjs';
import './button/props.mjs';
import './config.mjs';
import './form/hooks.mjs';
import './hooks/tnode.mjs';
import './hooks/render-tnode.mjs';
import './_common/js/utils/general.mjs';
import './loading/loading.mjs';
import './loading/icon/gradient.mjs';
import './_common/js/loading/circle-adapter.mjs';
import './_common/js/utils/setStyle.mjs';
import './_common/js/utils/helper.mjs';
import './_chunks/dep-a27c5c00.mjs';
import './loading/icon/spinner.mjs';
import './loading/props.mjs';
import './style/css.mjs';
import './loading/type.mjs';
import './button/type.mjs';
import './popup/index.mjs';
import './popup/popup.mjs';
import './popup/props.mjs';
import './overlay/index.mjs';
import './overlay/overlay.mjs';
import './overlay/props.mjs';
import './hooks/useVModel.mjs';
import './hooks/useLockScroll.mjs';
import './_util/useTouch.mjs';
import './_util/getScrollParent.mjs';
import './hooks/useTeleport.mjs';
import './popup/type.mjs';
import './dialog/props.mjs';
import './message/message.mjs';
import './link/index.mjs';
import './link/link.mjs';
import './link/props.mjs';
import './link/type.mjs';
import './message/props.mjs';
import './toast/type.mjs';
import './toast/toast.mjs';
import './toast/props.mjs';
import './drawer/drawer.mjs';
import './drawer/props.mjs';
import './drawer/type.mjs';
import './action-sheet/action-sheet.mjs';
import './action-sheet/action-sheet-grid.mjs';
import './grid/index.mjs';
import './grid/grid.mjs';
import './grid/props.mjs';
import './grid/grid-item.mjs';
import './grid/grid-item-props.mjs';
import './image/index.mjs';
import './image/image.mjs';
import './image/props.mjs';
import './image/type.mjs';
import './badge/index.mjs';
import './badge/badge.mjs';
import './badge/props.mjs';
import './badge/type.mjs';
import './grid/type.mjs';
import './swiper/index.mjs';
import './swiper/swiper.mjs';
import './swipe-cell/useSwipe.mjs';
import './swiper/props.mjs';
import './swiper/swiper-item.mjs';
import './swiper/type.mjs';
import './action-sheet/action-sheet-list.mjs';
import './action-sheet/props.mjs';
import './action-sheet/type.mjs';
