// tdesign-mobile-vue/lib/picker/picker.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, toRefs, ref, computed, watch, provide, createVNode, resolveComponent } from 'vue';
import { isFunction, get, isString, isBoolean } from 'lodash-es';
import config from '../config.js';
import PickerProps from './props.js';
import useVModel from '../hooks/useVModel.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import PickerItem from './picker-item.js';
import { getPickerColumns } from './utils.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useConfig } from '../config-provider/useConfig.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import './picker.class.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/classCallCheck';
import '@babel/runtime/helpers/createClass';
import '../shared/dom.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var getIndexFromColumns = function getIndexFromColumns(column, value, keys) {
  if (!value) return 0;
  return column === null || column === void 0 ? void 0 : column.findIndex(function (item) {
    var _keys$value;
    return get(item, (_keys$value = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value !== void 0 ? _keys$value : "value") === value;
  });
};
var _Picker = defineComponent({
  name: "".concat(prefix, "-picker"),
  components: {
    PickerItem: PickerItem
  },
  props: PickerProps,
  emits: ["change", "cancel", "pick", "update:modelValue", "update:value"],
  setup: function setup(props, _ref) {
    var _pickerValue$value;
    var slots = _ref.slots;
    var pickerClass = usePrefixClass("picker");
    var _useConfig = useConfig("picker"),
      globalConfig = _useConfig.globalConfig;
    var renderTNodeJSX = useTNodeJSX();
    var _toRefs = toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      _useVModel2$ = _useVModel2[0],
      pickerValue = _useVModel2$ === void 0 ? ref([]) : _useVModel2$,
      setPickerValue = _useVModel2[1];
    var keys = computed(function () {
      return props.keys;
    });
    var getDefaultText = function getDefaultText(prop, defaultText) {
      if (isString(prop)) return prop;
      if (isBoolean(prop) && prop) return defaultText;
      return "";
    };
    var confirmButtonText = computed(function () {
      return getDefaultText(props.confirmBtn, globalConfig.value.confirm);
    });
    var cancelButtonText = computed(function () {
      return getDefaultText(props.cancelBtn, globalConfig.value.cancel);
    });
    var curValueArray = ref(((_pickerValue$value = pickerValue.value) === null || _pickerValue$value === void 0 ? void 0 : _pickerValue$value.map(function (item) {
      return item;
    })) || []);
    var realColumns = computed(function () {
      if (isFunction(props.columns)) {
        var _columns = props.columns(curValueArray.value);
        return getPickerColumns(_columns);
      }
      return getPickerColumns(props.columns);
    });
    var curIndexArray = realColumns.value.map(function (item, index) {
      var _pickerValue$value2;
      return getIndexFromColumns(item, (_pickerValue$value2 = pickerValue.value) === null || _pickerValue$value2 === void 0 ? void 0 : _pickerValue$value2[index], keys.value);
    });
    var pickerItemInstanceArray = ref([]);
    var setPickerItemRef = function setPickerItemRef(item, index) {
      pickerItemInstanceArray.value[index] = item;
    };
    var handleConfirm = function handleConfirm(e) {
      var _props$onConfirm;
      var target = realColumns.value.map(function (item, index) {
        return item[curIndexArray[index]];
      });
      var label = target.map(function (item) {
        var _keys$value$label, _keys$value2;
        return get(item, (_keys$value$label = (_keys$value2 = keys.value) === null || _keys$value2 === void 0 ? void 0 : _keys$value2.label) !== null && _keys$value$label !== void 0 ? _keys$value$label : "label");
      });
      var value2 = target.map(function (item) {
        var _keys$value$value, _keys$value3;
        return get(item, (_keys$value$value = (_keys$value3 = keys.value) === null || _keys$value3 === void 0 ? void 0 : _keys$value3.value) !== null && _keys$value$value !== void 0 ? _keys$value$value : "value");
      });
      setPickerValue(value2);
      (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, value2, {
        index: curIndexArray,
        label: label,
        e: e
      });
    };
    var handleCancel = function handleCancel(e) {
      var _props$onCancel;
      pickerItemInstanceArray.value.forEach(function (item, index) {
        item === null || item === void 0 || item.setIndex(curIndexArray[index]);
      });
      (_props$onCancel = props.onCancel) === null || _props$onCancel === void 0 || _props$onCancel.call(props, {
        e: e
      });
    };
    var handlePick = function handlePick(context, column) {
      var _realColumns$value, _keys$value$value2, _keys$value4, _props$onPick;
      var index = context.index;
      curIndexArray[column] = index;
      curValueArray.value[column] = get((_realColumns$value = realColumns.value) === null || _realColumns$value === void 0 ? void 0 : _realColumns$value[column][index], (_keys$value$value2 = (_keys$value4 = keys.value) === null || _keys$value4 === void 0 ? void 0 : _keys$value4.value) !== null && _keys$value$value2 !== void 0 ? _keys$value$value2 : "value");
      (_props$onPick = props.onPick) === null || _props$onPick === void 0 || _props$onPick.call(props, curValueArray.value, {
        index: index,
        column: column
      });
    };
    watch(pickerValue, function () {
      curValueArray.value = pickerValue.value.map(function (item) {
        return item;
      });
    });
    watch([realColumns, curValueArray], function () {
      realColumns.value.forEach(function (col, idx) {
        var _pickerItemInstanceAr;
        var index = col.findIndex(function (item) {
          var _keys$value$value3, _keys$value5;
          return get(item, (_keys$value$value3 = (_keys$value5 = keys.value) === null || _keys$value5 === void 0 ? void 0 : _keys$value5.value) !== null && _keys$value$value3 !== void 0 ? _keys$value$value3 : "value") === curValueArray.value[idx];
        });
        curIndexArray[idx] = index > -1 ? index : 0;
        (_pickerItemInstanceAr = pickerItemInstanceArray.value[idx]) === null || _pickerItemInstanceAr === void 0 || _pickerItemInstanceAr.setIndex(curIndexArray[idx]);
      });
    });
    provide("picker", _objectSpread({}, toRefs(props)));
    return function () {
      return createVNode("div", {
        "class": "".concat(pickerClass.value)
      }, [createVNode("div", {
        "class": "".concat(pickerClass.value, "__toolbar")
      }, [cancelButtonText.value && createVNode("div", {
        "class": "".concat(pickerClass.value, "__cancel"),
        "onClick": handleCancel
      }, [cancelButtonText.value]), createVNode("div", {
        "class": "".concat(pickerClass.value, "__title")
      }, [props.title]), confirmButtonText.value && createVNode("div", {
        "class": "".concat(pickerClass.value, "__confirm"),
        "onClick": handleConfirm
      }, [confirmButtonText.value])]), renderTNodeJSX("header"), createVNode("div", {
        "class": "".concat(pickerClass.value, "__main")
      }, [realColumns.value.map(function (item, index) {
        var _pickerValue$value3;
        return createVNode("div", {
          "key": index,
          "class": "".concat(pickerClass.value, "-item__group")
        }, [createVNode(resolveComponent("picker-item"), {
          "ref": function ref(item2) {
            return setPickerItemRef(item2, index);
          },
          "options": item,
          "value": (_pickerValue$value3 = pickerValue.value) === null || _pickerValue$value3 === void 0 ? void 0 : _pickerValue$value3[index],
          "render-label": props.renderLabel,
          "onPick": function onPick($event) {
            return handlePick($event, index);
          },
          "swipeDuration": props.swipeDuration
        }, {
          option: slots.option || props.option
        })]);
      }), createVNode("div", {
        "class": "".concat(pickerClass.value, "__mask ").concat(pickerClass.value, "__mask--top")
      }, null), createVNode("div", {
        "class": "".concat(pickerClass.value, "__mask ").concat(pickerClass.value, "__mask--bottom")
      }, null), createVNode("div", {
        "class": "".concat(pickerClass.value, "__indicator")
      }, null)]), renderTNodeJSX("footer")]);
    };
  }
});

export { _Picker as default };
