// tdesign-mobile-vue/lib/calendar/calendar.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, ref, provide, reactive, onMounted, watch, createVNode, nextTick } from 'vue';
import { Popup } from '../popup/index.js';
import config from '../config.js';
import props from './props.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import TCalendarTemplate from './template.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '../popup/popup.js';
import '@babel/runtime/helpers/defineProperty';
import '@babel/runtime/helpers/slicedToArray';
import 'tdesign-icons-vue-next';
import '../popup/props.js';
import '../overlay/index.js';
import '../overlay/overlay.js';
import '../shared/dom.js';
import 'lodash-es';
import '../overlay/props.js';
import '../shared/component.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../hooks/useVModel.js';
import '../hooks/useLockScroll.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';
import '../hooks/useTeleport.js';
import '../shared/render-tnode.js';
import '../shared/util.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../button/index.js';
import '../button/button.js';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/setStyle.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../loading/icon/spinner.js';
import '../loading/props.js';
import '../loading/plugin.js';
import '../button/props.js';
import '../form/hooks.js';
import '../shared/hover.js';
import './template-props.js';
import './utils.js';

var prefix = config.prefix;
var _Calendar = defineComponent({
  name: "".concat(prefix, "-calendar"),
  props: props,
  emits: ["update:visible"],
  setup: function setup(props2, context) {
    var calendarClass = usePrefixClass("calendar");
    var calendarTemplateRef = ref();
    var renderTNodeJSX = useTNodeJSX();
    provide("templateProps", reactive(props2));
    var selectedValueIntoView = function selectedValueIntoView() {
      var _templateRef$querySel;
      var type = props2.type === "range" ? "start" : "selected";
      var templateRef = calendarTemplateRef.value.templateRef;
      var scrollContainer = templateRef.querySelector(".".concat(calendarClass.value, "__months"));
      var selectedDate = (_templateRef$querySel = templateRef.querySelector(".".concat(calendarClass.value, "__dates-item--").concat(type))) === null || _templateRef$querySel === void 0 || (_templateRef$querySel = _templateRef$querySel.parentNode) === null || _templateRef$querySel === void 0 ? void 0 : _templateRef$querySel.previousElementSibling;
      if (selectedDate) {
        scrollContainer.scrollTop = selectedDate.offsetTop - scrollContainer.offsetTop;
      }
    };
    var onVisibleChange = function onVisibleChange(v) {
      context.emit("update:visible", v);
    };
    var onPopupVisibleChange = function onPopupVisibleChange(v) {
      if (!v) {
        var _props2$onClose;
        (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, "overlay");
      } else {
        nextTick(function () {
          selectedValueIntoView();
        });
      }
      context.emit("update:visible", v);
    };
    onMounted(function () {
      if (!props2.usePopup) selectedValueIntoView();
    });
    watch(function () {
      return props2.value;
    }, function (val) {
      calendarTemplateRef.value.valueRef = val;
    });
    return function () {
      var title = renderTNodeJSX("title");
      var confirmBtn = renderTNodeJSX("confirmBtn");
      return createVNode("div", null, [!props2.usePopup ? createVNode(TCalendarTemplate, {
        "ref": calendarTemplateRef,
        "title": title,
        "confirmBtn": confirmBtn
      }, null) : createVNode(Popup, {
        "visible": props2.visible,
        "placement": "bottom",
        "onVisibleChange": onPopupVisibleChange
      }, {
        default: function _default() {
          return [createVNode(TCalendarTemplate, {
            "ref": calendarTemplateRef,
            "title": title,
            "confirmBtn": confirmBtn,
            "onVisibleChange": onVisibleChange
          }, null)];
        }
      })]);
    };
  }
});

export { _Calendar as default };
