// tdesign-mobile-vue/es/footer/footer.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, createVNode, Fragment, createTextVNode } from 'vue';
import { Image } from '../image/index.mjs';
import FooterProps from './props.mjs';
import config from '../config.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../image/image.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '@vueuse/core';
import 'tdesign-icons-vue-next';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import 'lodash-es';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../hooks/tnode.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../shared/dom.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../image/props.mjs';
import '../image/type.mjs';

var prefix = config.prefix;
var _Footer = defineComponent({
  name: "".concat(prefix, "-footer"),
  props: FooterProps,
  setup: function setup(props) {
    var footerClass = usePrefixClass("footer");
    var footerLinkClass = usePrefixClass("footer__link");
    var renderLogo = function renderLogo() {
      var logo = props.logo;
      if (!logo) {
        return;
      }
      return createVNode("a", {
        "class": "".concat(footerClass.value, "__logo"),
        "href": logo.url,
        "target": logo.target
      }, [logo.icon && createVNode(Image, {
        "class": "".concat(footerClass.value, "__icon"),
        "src": logo.icon
      }, null), logo.title && createVNode("span", {
        "class": "".concat(footerClass.value, "__title")
      }, [logo.title])]);
    };
    var renderText = function renderText() {
      return props.text && createVNode("div", {
        "class": "".concat(footerClass.value, "__text")
      }, [props.text]);
    };
    var renderLink = function renderLink() {
      if (!props.links.length) return null;
      return createVNode("div", {
        "class": "".concat(footerLinkClass.value, "-list")
      }, [props.links.map(function (link, index) {
        return createVNode(Fragment, null, [createVNode("a", {
          "href": link.url,
          "target": link.target,
          "class": "".concat(footerLinkClass.value, "-item")
        }, [link.name]), index !== props.links.length - 1 && createVNode("div", {
          "class": "".concat(footerLinkClass.value, "-line")
        }, [createTextVNode("|")])]);
      })]);
    };
    return function () {
      return createVNode("div", {
        "class": footerClass.value
      }, [renderLogo(), renderLink(), renderText()]);
    };
  }
});

export { _Footer as default };
