// tdesign-mobile-vue/cjs/progress/progress.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
var progress_utils = require('./utils.js');
var progress_props = require('./props.js');
var _common_js_progress_const = require('../_common/js/progress/const.js');
var _common_js_progress_utils = require('../_common/js/progress/utils.js');
var config = require('../config.js');
var isString = require('../_chunks/dep-588e2e70.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('@babel/runtime/helpers/objectWithoutProperties');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _Progress = Vue.defineComponent({
  name: "".concat(prefix, "-progress"),
  props: progress_props["default"],
  setup: function setup(props2) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var progressClass = hooks_useClass.usePrefixClass("progress");
    var computedStatus = Vue.computed(function () {
      if (props2.percentage >= 100) {
        return "success";
      }
      return props2.status || "default";
    });
    var trackBgStyle = Vue.computed(function () {
      var style = {};
      if (props2.strokeWidth) {
        var height = isString.isString(props2.strokeWidth) ? props2.strokeWidth : "".concat(props2.strokeWidth, "px");
        style.height = height;
        style.borderRadius = height;
      }
      if (props2.trackColor) {
        style.backgroundColor = props2.trackColor;
      }
      return style;
    });
    var barStyle = Vue.computed(function () {
      return {
        width: "".concat(props2.percentage, "%"),
        background: props2.color && progress_utils.getBackgroundColor(props2.color)
      };
    });
    var getIconMap = function getIconMap() {
      var CIRCLE_ICONS = {
        success: tdesignIconsVueNext.CheckIcon,
        warning: tdesignIconsVueNext.ErrorIcon,
        error: tdesignIconsVueNext.CloseIcon
      };
      var NORMAL_ICONS = {
        success: tdesignIconsVueNext.CheckCircleFilledIcon,
        warning: tdesignIconsVueNext.ErrorCircleFilledIcon,
        error: tdesignIconsVueNext.ErrorCircleFilledIcon
      };
      return props2.theme === _common_js_progress_const.PRO_THEME.CIRCLE ? CIRCLE_ICONS : NORMAL_ICONS;
    };
    var diameter = Vue.computed(function () {
      return _common_js_progress_utils.getDiameter(props2.size);
    });
    var rPoints = Vue.computed(function () {
      return diameter.value / 2;
    });
    var circleStrokeWidth = Vue.computed(function () {
      return _common_js_progress_utils.getCircleStokeWidth(props2.strokeWidth, props2.size);
    });
    var radius = Vue.computed(function () {
      return rPoints.value - circleStrokeWidth.value / 2;
    });
    var circleStyle = Vue.computed(function () {
      if (props2.theme !== _common_js_progress_const.PRO_THEME.CIRCLE) {
        return {};
      }
      return {
        width: "".concat(diameter.value, "px"),
        height: "".concat(diameter.value, "px")
      };
    });
    var strokeDashArr = Vue.computed(function () {
      var radius2 = (diameter.value - circleStrokeWidth.value) / 2;
      var perimeter = Math.PI * 2 * radius2;
      var percent = props2.percentage / 100;
      return "".concat(perimeter * percent, "  ").concat(perimeter * (1 - percent));
    });
    var circlePathStyle = Vue.computed(function () {
      var strokeColor = isObject.isObject(props2.color) ? "" : props2.color;
      return {
        stroke: strokeColor
      };
    });
    var circleOuterStyle = Vue.computed(function () {
      var strokeColor = isObject.isObject(props2.trackColor) ? "" : props2.trackColor;
      return {
        stroke: strokeColor
      };
    });
    var getLabelContent = function getLabelContent() {
      var labelContent = "".concat(props2.percentage, "%");
      var status = props2.status || "";
      if (_common_js_progress_const.STATUS_ICON.includes(status) && props2.theme !== _common_js_progress_const.PRO_THEME.PLUMP) {
        var components = getIconMap();
        var component = components[status];
        if (component) {
          labelContent = Vue.createVNode(component, {
            "class": ["".concat(progressClass.value, "__icon")]
          }, null);
        }
      }
      return labelContent;
    };
    return function () {
      var labelContent = props2.label && Vue.createVNode("div", {
        "class": "".concat(progressClass.value, "__info")
      }, [renderTNodeJSX("label", getLabelContent())]);
      var separateClasses = props2.percentage > _common_js_progress_const.PLUMP_SEPARATE ? "".concat(progressClass.value, "--over-ten") : "".concat(progressClass.value, "--under-ten");
      return Vue.createVNode("div", {
        "class": progressClass.value
      }, [props2.theme === _common_js_progress_const.PRO_THEME.LINE && Vue.createVNode("div", {
        "class": "".concat(progressClass.value, "--thin ").concat(progressClass.value, "--status-").concat(computedStatus.value)
      }, [Vue.createVNode("div", {
        "class": "".concat(progressClass.value, "__bar"),
        "style": trackBgStyle.value
      }, [Vue.createVNode("div", {
        "class": "".concat(progressClass.value, "__inner"),
        "style": barStyle.value
      }, null)]), labelContent]), props2.theme === _common_js_progress_const.PRO_THEME.PLUMP && Vue.createVNode("div", {
        "class": ["".concat(progressClass.value, "__bar ").concat(progressClass.value, "--plump ").concat(separateClasses), _defineProperty__default["default"]({}, "".concat(progressClass.value, "--status-").concat(computedStatus.value), computedStatus.value)],
        "style": trackBgStyle.value
      }, [Vue.createVNode("div", {
        "class": "".concat(progressClass.value, "__inner"),
        "style": barStyle.value
      }, [props2.percentage > _common_js_progress_const.PLUMP_SEPARATE && labelContent]), props2.percentage <= _common_js_progress_const.PLUMP_SEPARATE && labelContent]), props2.theme === _common_js_progress_const.PRO_THEME.CIRCLE && Vue.createVNode("div", {
        "class": "".concat(progressClass.value, "--circle ").concat(progressClass.value, "--status-").concat(computedStatus.value),
        "style": circleStyle.value
      }, [labelContent, Vue.createVNode("svg", {
        "width": diameter.value,
        "height": diameter.value,
        "viewBox": "0 0 ".concat(diameter.value, " ").concat(diameter.value)
      }, [Vue.createVNode("circle", {
        "cx": rPoints.value,
        "cy": rPoints.value,
        "r": radius.value,
        "stroke-width": circleStrokeWidth.value,
        "fill": "none",
        "class": ["".concat(progressClass.value, "__circle-outer")],
        "style": circleOuterStyle.value
      }, null), props2.percentage > 0 && Vue.createVNode("circle", {
        "cx": rPoints.value,
        "cy": rPoints.value,
        "r": radius.value,
        "stroke-width": circleStrokeWidth.value,
        "fill": "none",
        "stroke-linecap": "round",
        "class": ["".concat(progressClass.value, "__circle-inner")],
        "transform": "matrix(0,-1,1,0,0,".concat(diameter.value, ")"),
        "stroke-dasharray": strokeDashArr.value,
        "style": circlePathStyle.value
      }, null)])])]);
    };
  }
});

exports["default"] = _Progress;
