// tdesign-mobile-vue/cjs/back-top/back-top.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var core = require('@vueuse/core');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var backTop_props = require('./props.js');
var config = require('../config.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var _util_getScrollParent = require('../_util/getScrollParent.js');
var shared_util = require('../shared/util.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-055d28fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _BackTop = Vue.defineComponent({
  name: "".concat(prefix, "-back-top"),
  props: backTop_props["default"],
  setup: function setup(props2) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var backTopClass = hooks_useClass.usePrefixClass("back-top");
    var root = Vue.ref();
    var target = Vue.computed(function () {
      if (!shared_util.isBrowser) return void 0;
      return props2.target ? props2.target() : window.document.documentElement;
    });
    var container = Vue.ref();
    var _useElementBounding = core.useElementBounding(target),
      top = _useElementBounding.top;
    var visible = Vue.ref(false);
    var backTopClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(backTopClass.value), true), "".concat(backTopClass.value, "--fixed"), props2.fixed), "".concat(backTopClass.value, "--").concat(props2.theme), true);
    });
    var handleClick = function handleClick() {
      if (shared_util.isBrowser) {
        var _props2$onToTop;
        container.value.scrollTop = 0 + top.value;
        (_props2$onToTop = props2.onToTop) === null || _props2$onToTop === void 0 || _props2$onToTop.call(props2);
      }
    };
    var getContainer = function getContainer(container2) {
      if (typeof container2 === "function") {
        return container2();
      }
      return _util_getScrollParent["default"](root.value, document.documentElement);
    };
    Vue.onMounted(function () {
      if (shared_util.isBrowser) {
        container.value = getContainer(props2.container);
        container.value.onscroll = function () {
          var _container$value, _container$value2;
          if (((_container$value = container.value) === null || _container$value === void 0 ? void 0 : _container$value.scrollTop) >= props2.visibilityHeight) {
            visible.value = true;
          }
          if (visible.value && ((_container$value2 = container.value) === null || _container$value2 === void 0 ? void 0 : _container$value2.scrollTop) < props2.visibilityHeight) {
            visible.value = false;
          }
        };
      }
    });
    Vue.onUnmounted(function () {
      container.value.onscroll = null;
    });
    var readerIcon = function readerIcon() {
      if (props2.icon === true) {
        return Vue.createVNode(tdesignIconsVueNext.BacktopIcon, {
          "size": "22px"
        }, null);
      }
      return renderTNodeJSX("icon");
    };
    return function () {
      return Vue.withDirectives(Vue.createVNode("div", {
        "class": backTopClasses.value,
        "ref": root,
        "onClick": handleClick
      }, [readerIcon(), props2.text && Vue.createVNode("span", {
        "class": "".concat(backTopClass.value, "__text--").concat(props2.theme)
      }, [props2.text])]), [[Vue.vShow, visible.value]]);
    };
  }
});

exports["default"] = _BackTop;
