// tdesign-mobile-vue/lib/cascader/cascader.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, toRefs, ref, computed, reactive, watch, onMounted, createVNode, Transition, toRaw } from 'vue';
import { CloseIcon, ChevronRightIcon } from 'tdesign-icons-vue-next';
import { get } from 'lodash-es';
import { Popup } from '../popup/index.js';
import { Tabs } from '../tabs/index.js';
import { RadioGroup } from '../radio/index.js';
import config from '../config.js';
import props from './props.js';
import { useConfig } from '../config-provider/useConfig.js';
import useVModel from '../hooks/useVModel.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '../popup/popup.js';
import '../popup/props.js';
import '../overlay/index.js';
import '../overlay/overlay.js';
import '../shared/dom.js';
import '../overlay/props.js';
import '../shared/component.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../hooks/useLockScroll.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';
import '../hooks/useTeleport.js';
import '../shared/render-tnode.js';
import '../shared/util.js';
import '../tabs/tabs.js';
import '../tabs/props.js';
import '../tabs/tab-nav-item.js';
import '../tabs/tab-panel-props.js';
import '../sticky/index.js';
import '../sticky/sticky.js';
import '@vueuse/core';
import '../sticky/props.js';
import '../badge/index.js';
import '../badge/badge.js';
import '../badge/props.js';
import '../hooks/useCommonClassName.js';
import '../tabs/tab-panel.js';
import '../radio/radio.js';
import '../radio/props.js';
import '../form/hooks.js';
import '../radio/radio-group.js';
import '../radio/radio-group-props.js';

var prefix = config.prefix;
var name = "".concat(prefix, "-cascader");
var childrenInfo = {
  value: "",
  level: 0
};
var _Cascader = defineComponent({
  name: name,
  props: props,
  emits: ["update:visible", "update:value", "update:modelValue"],
  setup: function setup(props2, context) {
    var renderTNodeJSX = useTNodeJSX();
    var cascaderClass = usePrefixClass("cascader");
    var _useConfig = useConfig("cascader"),
      globalConfig = _useConfig.globalConfig;
    var _toRefs = toRefs(props2),
      visible = _toRefs.visible,
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      cascaderValue = _useVModel2[0],
      setCascaderValue = _useVModel2[1];
    var open = ref(visible.value || false);
    var placeholder = computed(function () {
      return props2.placeholder || globalConfig.value.placeholder;
    });
    var stepIndex = ref(0);
    var selectedIndexes = reactive([]);
    var selectedValue = reactive([]);
    var items = computed(function () {
      var _props2$options;
      return [(_props2$options = props2.options) !== null && _props2$options !== void 0 ? _props2$options : []];
    });
    var steps = reactive([placeholder.value]);
    var initWithValue = function initWithValue() {
      if (cascaderValue.value != null) {
        steps.length = 0;
        selectedValue.length = 0;
        selectedIndexes.splice(0, selectedIndexes.length);
        var path = _getIndexesByValue(props2.options, cascaderValue.value);
        path === null || path === void 0 || path.forEach(function (e) {
          selectedIndexes.push(e);
        });
        watchSelectedIndexes();
      }
    };
    var watchSelectedIndexes = function watchSelectedIndexes() {
      if (props2.options && props2.options.length > 0) {
        var _items$value;
        (_items$value = items.value).splice.apply(_items$value, [0, items.value.length].concat([props2.options]));
        var keys = props2.keys;
        for (var i = 0, size = selectedIndexes.length; i < size; i += 1) {
          var _items$value$i, _keys$value, _keys$label, _keys$children;
          var index = selectedIndexes[i];
          var next = (_items$value$i = items.value[i]) === null || _items$value$i === void 0 ? void 0 : _items$value$i[index];
          selectedValue.push(get(next, (_keys$value = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value !== void 0 ? _keys$value : "value"));
          steps.push(get(next, (_keys$label = keys === null || keys === void 0 ? void 0 : keys.label) !== null && _keys$label !== void 0 ? _keys$label : "label"));
          if (get(next, (_keys$children = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children !== void 0 ? _keys$children : "children")) {
            var _keys$children2;
            items.value.push(get(next, (_keys$children2 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children2 !== void 0 ? _keys$children2 : "children"));
          }
        }
      }
      if (steps.length < items.value.length) {
        steps.push(placeholder.value);
      }
      stepIndex.value = items.value.length - 1;
    };
    var _getIndexesByValue = function getIndexesByValue(options, value2) {
      var keys = props2.keys;
      for (var i = 0; i < options.length; i++) {
        var _keys$value2, _keys$children3;
        if (get(options[i], (_keys$value2 = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value2 !== void 0 ? _keys$value2 : "value") === value2) {
          return [i];
        }
        if (get(options[i], (_keys$children3 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children3 !== void 0 ? _keys$children3 : "children")) {
          var _keys$children4;
          var res = _getIndexesByValue(get(options[i], (_keys$children4 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children4 !== void 0 ? _keys$children4 : "children"), value2);
          if (res) {
            return [i].concat(_toConsumableArray(res));
          }
        }
      }
    };
    var chooseSelect = function chooseSelect(value2, level, index, item) {
      var _keys$label2, _lodashGet, _keys$children5, _lodashGet2, _keys$children7;
      var keys = props2.keys;
      selectedIndexes[level] = index;
      selectedIndexes.length = level + 1;
      selectedValue[level] = typeof value2 === "number" ? value2 : String(value2);
      selectedValue.length = level + 1;
      steps[level] = get(item, (_keys$label2 = keys === null || keys === void 0 ? void 0 : keys.label) !== null && _keys$label2 !== void 0 ? _keys$label2 : "label");
      if ((_lodashGet = get(item, (_keys$children5 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children5 !== void 0 ? _keys$children5 : "children")) !== null && _lodashGet !== void 0 && _lodashGet.length) {
        var _keys$children6;
        items.value[level + 1] = get(item, (_keys$children6 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children6 !== void 0 ? _keys$children6 : "children");
        items.value.length = level + 2;
        stepIndex.value += 1;
        steps[level + 1] = placeholder.value;
        steps.length = level + 2;
      } else if (((_lodashGet2 = get(item, (_keys$children7 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children7 !== void 0 ? _keys$children7 : "children")) === null || _lodashGet2 === void 0 ? void 0 : _lodashGet2.length) === 0) {
        childrenInfo.value = value2;
        childrenInfo.level = level;
      } else {
        var _keys$value3;
        items.value.length = level + 1;
        steps.length = level + 1;
        stepIndex.value = level;
        setCascaderValue(get(item, (_keys$value3 = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value3 !== void 0 ? _keys$value3 : "value"), items.value.map(function (item2, index2) {
          return toRaw(item2 === null || item2 === void 0 ? void 0 : item2[selectedIndexes[index2]]);
        }));
        close("finish");
      }
    };
    var cancelSelect = function cancelSelect(value2, level, index, item) {
      var _lodashGet3, _keys$children8, _lodashGet4, _keys$children0;
      var keys = props2.keys;
      selectedIndexes[level] = index;
      selectedIndexes.length = level;
      selectedValue.length = level;
      steps[level] = String(placeholder.value);
      steps[level + 1] = placeholder.value;
      steps.length = level + 1;
      if ((_lodashGet3 = get(item, (_keys$children8 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children8 !== void 0 ? _keys$children8 : "children")) !== null && _lodashGet3 !== void 0 && _lodashGet3.length) {
        var _keys$children9;
        items.value[level + 1] = get(item, (_keys$children9 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children9 !== void 0 ? _keys$children9 : "children");
      } else if (((_lodashGet4 = get(item, (_keys$children0 = keys === null || keys === void 0 ? void 0 : keys.children) !== null && _keys$children0 !== void 0 ? _keys$children0 : "children")) === null || _lodashGet4 === void 0 ? void 0 : _lodashGet4.length) === 0) {
        childrenInfo.value = value2;
        childrenInfo.level = level;
      }
    };
    var handleSelect = function handleSelect(value2, level) {
      var _keys$disabled;
      var fromHandler = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
      var keys = props2.keys;
      var index = items.value[level].findIndex(function (item2) {
        var _keys$value4;
        return get(item2, (_keys$value4 = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value4 !== void 0 ? _keys$value4 : "value") === value2;
      });
      var item = items.value[level][index];
      if (get(item, (_keys$disabled = keys === null || keys === void 0 ? void 0 : keys.disabled) !== null && _keys$disabled !== void 0 ? _keys$disabled : "disabled")) {
        return;
      }
      if (fromHandler) {
        var _props2$onPick, _keys$value5, _keys$label3;
        (_props2$onPick = props2.onPick) === null || _props2$onPick === void 0 || _props2$onPick.call(props2, {
          value: get(item, (_keys$value5 = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value5 !== void 0 ? _keys$value5 : "value"),
          label: get(item, (_keys$label3 = keys === null || keys === void 0 ? void 0 : keys.label) !== null && _keys$label3 !== void 0 ? _keys$label3 : "label"),
          level: level,
          index: index
        });
      }
      if (props2.checkStrictly && selectedValue.includes(String(value2)) && fromHandler) {
        cancelSelect(value2, level, index, item);
      } else {
        chooseSelect(value2, level, index, item);
      }
    };
    var close = function close(trigger) {
      var _props2$onClose;
      (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, trigger);
    };
    var handleVisibleChange = function handleVisibleChange(visible2, e) {
      if ((e === null || e === void 0 ? void 0 : e.trigger) !== "overlay") return;
      close("overlay");
    };
    var updateCascaderValue = function updateCascaderValue() {
      setCascaderValue(selectedValue[selectedValue.length - 1], items.value.filter(function (item, index) {
        return !!item && selectedIndexes.length > index;
      }).map(function (item, index) {
        return toRaw(item === null || item === void 0 ? void 0 : item[selectedIndexes[index]]);
      }));
    };
    var onClose = function onClose() {
      open.value = false;
      close("close-btn");
    };
    var onCloseBtn = function onCloseBtn() {
      if (props2.checkStrictly) {
        updateCascaderValue();
        onClose();
      } else {
        onClose();
      }
    };
    var onStepClick = function onStepClick(index) {
      stepIndex.value = index;
    };
    var onTabChange = function onTabChange(value2) {
      stepIndex.value = Number(value2);
    };
    watch(open, function () {
      context.emit("update:visible", open.value);
    });
    watch(visible, function () {
      open.value = visible.value;
      if (visible.value) {
        initWithValue();
      }
    });
    watch(function () {
      return props2.options;
    }, function () {
      initWithValue();
      if (open.value && cascaderValue.value) {
        handleSelect(childrenInfo.value, childrenInfo.level, false);
      }
    }, {
      deep: true
    });
    watch(placeholder, function (newValue, oldValue) {
      var index = steps.indexOf(oldValue);
      if (index !== -1) {
        steps[index] = newValue;
      }
    });
    onMounted(function () {
      initWithValue();
    });
    return function () {
      var title = renderTNodeJSX("title") || globalConfig.value.title;
      var closeBtn = renderTNodeJSX("closeBtn", {
        defaultNode: createVNode(CloseIcon, {
          "size": "24px"
        }, null)
      });
      var PlaceholderNode = renderTNodeJSX("placeholder");
      var readerStep = function readerStep() {
        return createVNode("div", {
          "class": "".concat(cascaderClass.value, "__steps")
        }, [steps.map(function (step, index) {
          return createVNode("div", {
            "class": "".concat(cascaderClass.value, "__step"),
            "onClick": function onClick() {
              onStepClick(index);
            }
          }, [createVNode("div", {
            "class": ["".concat(cascaderClass.value, "__step-dot"), _defineProperty(_defineProperty({}, "".concat(cascaderClass.value, "__step-dot--active"), step !== placeholder.value), "".concat(cascaderClass.value, "__step-dot--last"), index === steps.length - 1)]
          }, null), createVNode("div", {
            "class": ["".concat(cascaderClass.value, "__step-label"), _defineProperty({}, "".concat(cascaderClass.value, "__step-label--active"), index === stepIndex.value)]
          }, [PlaceholderNode && !(typeof PlaceholderNode === "string") && step === placeholder.value ? PlaceholderNode : step]), createVNode(ChevronRightIcon, {
            "class": "".concat(cascaderClass.value, "__step-arrow"),
            "size": "22"
          }, null)]);
        })]);
      };
      var renderTabStep = function renderTabStep() {
        var _steps = steps.map(function (item, index) {
          return {
            value: index,
            label: item
          };
        });
        return createVNode(Tabs, {
          "value": stepIndex.value,
          "spaceEvenly": false,
          "list": _steps,
          "onChange": onTabChange
        }, null);
      };
      var readerSteps = function readerSteps() {
        if (steps.length === 0) {
          return null;
        }
        if (props2.theme === "step") {
          return readerStep();
        }
        if (open.value && props2.theme === "tab") {
          return renderTabStep();
        }
      };
      return createVNode(Popup, {
        "overlayProps": props2.overlayProps,
        "modelValue": open.value,
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return open.value = $event;
        },
        "placement": "bottom",
        "onVisibleChange": handleVisibleChange
      }, {
        default: function _default() {
          return [createVNode("div", {
            "class": "".concat(cascaderClass.value)
          }, [createVNode("div", {
            "class": "".concat(cascaderClass.value, "__title")
          }, [title]), createVNode("div", {
            "class": "".concat(cascaderClass.value, "__close-btn"),
            "onClick": onCloseBtn
          }, [closeBtn]), renderTNodeJSX("header"), createVNode("div", {
            "class": "".concat(cascaderClass.value, "__content")
          }, [readerSteps(), renderTNodeJSX("middleContent"), props2.subTitles && props2.subTitles[stepIndex.value] && createVNode("div", {
            "class": "".concat(cascaderClass.value, "__options-title")
          }, [props2.subTitles[stepIndex.value]]), createVNode("div", {
            "class": "".concat(cascaderClass.value, "__options-container"),
            "style": "width: ".concat(items.value.length + 1, "00vw; transform: translateX(-").concat(stepIndex.value, "00vw);")
          }, [items.value.map(function (options, index) {
            return createVNode("div", {
              "class": "".concat(cascaderClass.value, "__options")
            }, [createVNode(Transition, {
              "appear": true,
              "name": "slide"
            }, {
              default: function _default() {
                return [createVNode("div", {
                  "class": "".concat(cascaderClass.value, "-radio-group-").concat(index)
                }, [createVNode(RadioGroup, {
                  "value": selectedValue[index] || "",
                  "keys": props2.keys,
                  "options": options,
                  "placement": "right",
                  "icon": "line",
                  "borderless": true,
                  "onChange": function onChange(value2, context2) {
                    handleSelect(value2, index);
                  }
                }, null)])];
              }
            })]);
          })])])])];
        }
      });
    };
  }
});

export { _Cascader as default };
