// tdesign-mobile-vue/es/image/image.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, createVNode, ref, computed, watchEffect } from 'vue';
import { useIntersectionObserver } from '@vueuse/core';
import { CloseIcon } from 'tdesign-icons-vue-next';
import { Loading } from '../loading/index.mjs';
import config from '../config.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import props from './props.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import 'lodash-es';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';

var prefix = config.prefix;
var _Image = defineComponent({
  name: "".concat(prefix, "-image"),
  props: props,
  setup: function setup(props2, context) {
    var imageClass = usePrefixClass("image");
    var renderTNodeJSX = useTNodeJSX();
    var closeIcon = createVNode(CloseIcon, {
      "size": "22px"
    }, null);
    var LoadingIcon = createVNode(Loading, {
      "theme": "dots",
      "inheritColor": true
    }, null);
    var isLoading = ref(true);
    var isError = ref(false);
    var imageStyles = computed(function () {
      return {
        objectFit: props2.fit,
        objectPosition: props2.position
      };
    });
    var imageClasses = computed(function () {
      return _defineProperty(_defineProperty({}, "".concat(imageClass.value), true), "".concat(imageClass.value, "--").concat(props2.shape), true);
    });
    var imageDOM = ref();
    var realSrc = ref("");
    watchEffect(function () {
      realSrc.value = props2.lazy ? "" : props2.src;
    });
    var _useIntersectionObser = useIntersectionObserver(imageDOM, function (_ref2, observerElement) {
        var _ref3 = _slicedToArray(_ref2, 1),
          isIntersecting = _ref3[0].isIntersecting;
        if (isIntersecting && props2.lazy) {
          stop();
          realSrc.value = props2.src;
        }
      }),
      stop = _useIntersectionObser.stop;
    var handleImgLoadCompleted = function handleImgLoadCompleted(e) {
      var _props2$onLoad;
      (_props2$onLoad = props2.onLoad) === null || _props2$onLoad === void 0 || _props2$onLoad.call(props2, {
        e: e
      });
      isLoading.value = false;
    };
    var handleImgLoadError = function handleImgLoadError(e) {
      var _props2$onError;
      if (realSrc.value === "") {
        return;
      }
      (_props2$onError = props2.onError) === null || _props2$onError === void 0 || _props2$onError.call(props2, {
        e: e
      });
      isLoading.value = false;
      isError.value = true;
      if (props2.fallback) {
        realSrc.value = props2.fallback;
        isError.value = false;
      }
    };
    var maskContent = computed(function () {
      var statusContent = function statusContent() {
        var _context$slots, _context$slots2, _context$slots3, _context$slots4;
        if ((_context$slots = context.slots) !== null && _context$slots !== void 0 && _context$slots.loading && isLoading.value) {
          return renderTNodeJSX("loading");
        }
        if (!((_context$slots2 = context.slots) !== null && _context$slots2 !== void 0 && _context$slots2.loading) && isLoading.value) {
          return LoadingIcon;
        }
        if ((_context$slots3 = context.slots) !== null && _context$slots3 !== void 0 && _context$slots3.error && isError.value) {
          return renderTNodeJSX("error");
        }
        if (!((_context$slots4 = context.slots) !== null && _context$slots4 !== void 0 && _context$slots4.error) && isError.value) {
          return closeIcon;
        }
        return false;
      };
      var node = function node() {
        if (isLoading.value || isError.value) {
          if (props2.lazy && realSrc.value === "") {
            return;
          }
          return createVNode("div", {
            "class": "".concat(imageClass.value, "__mask")
          }, [statusContent()]);
        }
      };
      return node();
    });
    return function () {
      return createVNode("div", {
        "class": imageClasses.value
      }, [maskContent.value, createVNode("picture", null, [props2.srcset && Object.entries(props2.srcset).map(function (item, index) {
        return createVNode("source", {
          "key": index,
          "type": item[0],
          "srcset": item[1]
        }, null);
      }), createVNode("img", {
        "ref": imageDOM,
        "class": "".concat(imageClass.value, "__img"),
        "style": imageStyles.value,
        "src": realSrc.value,
        "alt": props2.alt,
        "referrerpolicy": props2.referrerpolicy,
        "onLoad": handleImgLoadCompleted,
        "onError": handleImgLoadError
      }, null)])]);
    };
  }
});

export { _Image as default };
