// tdesign-mobile-vue/lib/input/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var InputProps = {
  align: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "center", "right"].includes(val);
    }
  },
  allowInputOverMax: Boolean,
  autocomplete: {
    type: String,
    default: void 0
  },
  autofocus: Boolean,
  borderless: Boolean,
  clearTrigger: {
    type: String,
    default: "always",
    validator: function validator(val) {
      if (!val) return true;
      return ["always", "focus"].includes(val);
    }
  },
  clearable: Boolean,
  cursorColor: {
    type: String,
    default: "#0052d9"
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  enterkeyhint: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["enter", "done", "go", "next", "previous", "search", "send"].includes(val);
    }
  },
  extra: {
    type: Function
  },
  format: {
    type: Function
  },
  label: {
    type: [String, Function]
  },
  layout: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["vertical", "horizontal"].includes(val);
    }
  },
  maxcharacter: {
    type: Number
  },
  maxlength: {
    type: [String, Number]
  },
  name: {
    type: String,
    default: ""
  },
  placeholder: {
    type: String,
    default: void 0
  },
  prefixIcon: {
    type: Function
  },
  readonly: {
    type: Boolean,
    default: void 0
  },
  spellCheck: Boolean,
  status: {
    type: String,
    default: "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "success", "warning", "error"].includes(val);
    }
  },
  suffix: {
    type: [String, Function]
  },
  suffixIcon: {
    type: Function
  },
  tips: {
    type: [String, Function]
  },
  type: {
    type: String,
    default: "text",
    validator: function validator(val) {
      if (!val) return true;
      return ["text", "number", "url", "tel", "password", "search", "submit", "hidden"].includes(val);
    }
  },
  value: {
    type: [String, Number],
    default: void 0
  },
  modelValue: {
    type: [String, Number],
    default: void 0
  },
  defaultValue: {
    type: [String, Number]
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onFocus: Function,
  onValidate: Function
};

export { InputProps as default };
