// tdesign-mobile-vue/esm/action-sheet/action-sheet-list.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isVNode, defineComponent, toRefs, computed, createVNode } from 'vue';
import { Button } from '../button/index.js';
import { Badge } from '../badge/index.js';
import { useTNodeDefault } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '../button/button.js';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/setStyle.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash-es';
import '../config.js';
import '../loading/icon/spinner.js';
import '../loading/props.js';
import '../shared/dom.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/render.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '../shared/hover.js';
import '../loading/style';
import '../loading/type.js';
import '../loading/plugin.js';
import '../button/props.js';
import '../form/hooks.js';
import '../button/style';
import '../button/type.js';
import '../badge/badge.js';
import '../badge/props.js';
import '../badge/style';
import '../badge/type.js';

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var TActionSheetList = defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    },
    align: {
      type: String,
      default: "center"
    }
  },
  emits: ["selected"],
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    var actionSheetClass = usePrefixClass("action-sheet");
    var renderTNodeJSX = useTNodeDefault();
    var _toRefs = toRefs(props),
      align = _toRefs.align,
      items = _toRefs.items;
    var handleSelected = function handleSelected(index) {
      emit("selected", index);
    };
    var itemClasses = computed(function () {
      return _defineProperty(_defineProperty({}, "".concat(actionSheetClass.value, "__list-item"), true), "".concat(actionSheetClass.value, "__list-item--left"), align.value === "left");
    });
    return function () {
      var renderButtonNode = function renderButtonNode() {
        var renderBadgeNode = function renderBadgeNode(item) {
          if (item.badge) {
            var content = function content() {
              if (item.badge.dot || item.badge.count) {
                return createVNode(Badge, {
                  "count": item.badge.count,
                  "max-count": item.badge.maxCount || 99,
                  "dot": item.badge.dot,
                  "content": item.badge.content,
                  "size": item.badge.size,
                  "offset": item.badge.offset || [-16, 20]
                }, {
                  default: function _default() {
                    return [createVNode("span", {
                      "class": "".concat(actionSheetClass.value, "__list-item-text")
                    }, [" ", item.label])];
                  }
                });
              }
              return renderTNodeJSX("badge", {
                defaultNode: createVNode("span", {
                  "class": "".concat(actionSheetClass.value, "__list-item-text")
                }, [item.label])
              });
            };
            return content();
          }
          return createVNode("span", {
            "class": "".concat(actionSheetClass.value, "__list-item-text")
          }, [item.label]);
        };
        var buttonList = items.value.map(function (item, index) {
          var _slot;
          return createVNode(Button, {
            "key": index,
            "variant": "text",
            "block": true,
            "class": [itemClasses.value, _defineProperty({}, "".concat(actionSheetClass.value, "__list-item--disabled"), item.disabled)],
            "disabled": item.disabled,
            "icon": item.icon,
            "style": {
              color: item.color
            },
            "onClick": function onClick() {
              return handleSelected(index);
            }
          }, _isSlot(_slot = renderBadgeNode(item)) ? _slot : {
            default: function _default() {
              return [_slot];
            }
          });
        });
        return buttonList;
      };
      return createVNode("div", {
        "class": "".concat(actionSheetClass.value, "__list")
      }, [renderButtonNode()]);
    };
  }
});

export { TActionSheetList as default };
