// tdesign-mobile-vue/cjs/_common/js/tree/tree-node-model.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _classCallCheck = require('@babel/runtime/helpers/classCallCheck');
var _createClass = require('@babel/runtime/helpers/createClass');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _common_js_log_log = require('../log/log.js');
var isBoolean = require('../../../_chunks/dep-eb605cd5.js');
var omit = require('../../../_chunks/dep-032f49b6.js');
var get = require('../../../_chunks/dep-cfb75f1b.js');
var isUndefined = require('../../../_chunks/dep-9bddd251.js');
require('../../../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../../../_chunks/dep-f84e6e17.js');
require('../../../_chunks/dep-2ab29468.js');
require('../../../_chunks/dep-8d30e8d1.js');
require('../../../_chunks/dep-05850262.js');
require('../../../_chunks/dep-e9a63584.js');
require('../../../_chunks/dep-0c11c462.js');
require('../../../_chunks/dep-0bd122a0.js');
require('../../../_chunks/dep-b999c3d4.js');
require('../../../_chunks/dep-159c7a94.js');
require('../../../_chunks/dep-3cf5587c.js');
require('../../../_chunks/dep-ea269588.js');
require('../../../_chunks/dep-b633ec9e.js');
require('../../../_chunks/dep-7aa6341b.js');
require('../../../_chunks/dep-b8e8d654.js');
require('../../../_chunks/dep-61ce8c0d.js');
require('../../../_chunks/dep-5cf8bf50.js');
require('../../../_chunks/dep-2457b474.js');
require('../../../_chunks/dep-dd93e905.js');
require('../../../_chunks/dep-d1ba5719.js');
require('../../../_chunks/dep-01310648.js');
require('../../../_chunks/dep-42a5275c.js');
require('../../../_chunks/dep-fb9c8646.js');
require('../../../_chunks/dep-09a28f29.js');
require('../../../_chunks/dep-a6a4534e.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _classCallCheck__default = /*#__PURE__*/_interopDefaultLegacy(_classCallCheck);
var _createClass__default = /*#__PURE__*/_interopDefaultLegacy(_createClass);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var nodeKey = "__tdesign_tree-node__";
var TreeNodeModel = /*#__PURE__*/function () {
  function TreeNodeModel(node) {
    _classCallCheck__default["default"](this, TreeNodeModel);
    _defineProperty__default["default"](this, nodeKey, void 0);
    this[nodeKey] = node;
  }
  return _createClass__default["default"](TreeNodeModel, [{
    key: "value",
    get: function get() {
      var node = this[nodeKey];
      return node.value;
    }
  }, {
    key: "label",
    get: function get() {
      var node = this[nodeKey];
      return node.label;
    }
  }, {
    key: "data",
    get: function get() {
      var node = this[nodeKey];
      return node.data;
    }
  }, {
    key: "actived",
    get: function get() {
      var node = this[nodeKey];
      return node.actived;
    }
  }, {
    key: "expanded",
    get: function get() {
      var node = this[nodeKey];
      return node.expanded;
    }
  }, {
    key: "checked",
    get: function get() {
      var node = this[nodeKey];
      return node.checked;
    }
  }, {
    key: "indeterminate",
    get: function get() {
      var node = this[nodeKey];
      return node.indeterminate;
    }
  }, {
    key: "loading",
    get: function get() {
      var node = this[nodeKey];
      return node.loading;
    }
  }, {
    key: "disabled",
    get: function get() {
      var node = this[nodeKey];
      return node.isDisabled();
    }
  }, {
    key: "getLevel",
    value: function getLevel() {
      var node = this[nodeKey];
      return node.getLevel();
    }
  }, {
    key: "getIndex",
    value: function getIndex() {
      var node = this[nodeKey];
      return node.getIndex();
    }
  }, {
    key: "isFirst",
    value: function isFirst() {
      var node = this[nodeKey];
      return node.isFirst();
    }
  }, {
    key: "isLast",
    value: function isLast() {
      var node = this[nodeKey];
      return node.isLast();
    }
  }, {
    key: "isLeaf",
    value: function isLeaf() {
      var node = this[nodeKey];
      return node.isLeaf();
    }
  }, {
    key: "insertBefore",
    value: function insertBefore(newData) {
      var node = this[nodeKey];
      return node.insertBefore(newData);
    }
  }, {
    key: "insertAfter",
    value: function insertAfter(newData) {
      var node = this[nodeKey];
      return node.insertAfter(newData);
    }
  }, {
    key: "appendData",
    value: function appendData(data) {
      var node = this[nodeKey];
      return node.append(data);
    }
  }, {
    key: "getPath",
    value: function getPath() {
      var node = this[nodeKey];
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    }
  }, {
    key: "getParent",
    value: function getParent() {
      var _node$parent;
      var node = this[nodeKey];
      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    }
  }, {
    key: "getParents",
    value: function getParents() {
      var node = this[nodeKey];
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    }
  }, {
    key: "getRoot",
    value: function getRoot() {
      var node = this[nodeKey];
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    }
  }, {
    key: "getSiblings",
    value: function getSiblings() {
      var node = this[nodeKey];
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    }
  }, {
    key: "getChildren",
    value: function getChildren(deep) {
      var node = this[nodeKey];
      var childrenModel = false;
      var children = node.children;
      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (isBoolean.isBoolean(children)) {
        childrenModel = children;
      }
      return childrenModel;
    }
  }, {
    key: "remove",
    value: function remove(value) {
      var node = this[nodeKey];
      if (!value) {
        node.remove();
        return;
      }
      var tree = node.tree;
      var targetNode = tree.getNode(value);
      if (!targetNode) {
        _common_js_log_log["default"].warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }
      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });
      if (parentValues.indexOf(node.value) < 0) {
        _common_js_log_log["default"].warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }
      targetNode.remove();
    }
  }, {
    key: "setData",
    value: function setData(data) {
      var node = this[nodeKey];
      var syncAttrs = ["value", "label", "disabled"];
      var cleanData = omit.omit(data, ["children"].concat(syncAttrs));
      var keys = node.tree.config.keys;
      syncAttrs.forEach(function (attr) {
        var dataAttrValue = get.get(data, (keys === null || keys === void 0 ? void 0 : keys[attr]) || attr);
        if (!isUndefined.isUndefined(dataAttrValue)) cleanData[attr] = dataAttrValue;
      });
      Object.assign(node.data, cleanData);
      Object.assign(node, cleanData);
      node.update();
    }
  }]);
}();
function createNodeModel(node) {
  var model = new TreeNodeModel(node);
  return model;
}

exports.TreeNodeModel = TreeNodeModel;
exports.createNodeModel = createNodeModel;
exports.nodeKey = nodeKey;
