// tdesign-mobile-vue/esm/cell/cell.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, computed, createVNode, withDirectives, resolveDirective } from 'vue';
import { ChevronRightIcon } from 'tdesign-icons-vue-next';
import '../shared/index.js';
import config from '../config.js';
import props from './props.js';
import { useFormDisabled } from '../form/hooks.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useTNodeJSX, useContent } from '../hooks/tnode.js';
import Hover from '../shared/hover.js';
import '../shared/functions.js';
import '../shared/util.js';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash-es';
import '../shared/component.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/dom.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';

var prefix = config.prefix;
var _Cell = defineComponent({
  name: "".concat(prefix, "-cell"),
  directives: {
    Hover: Hover
  },
  props: props,
  setup: function setup(props2) {
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var disabled = useFormDisabled();
    var cellClass = usePrefixClass("cell");
    var cellClasses = computed(function () {
      return ["".concat(cellClass.value), _defineProperty({}, "".concat(cellClass.value, "--borderless"), !props2.bordered)];
    });
    var hoverDisabled = computed(function () {
      return !props2.hover || disabled.value;
    });
    var handleClick = function handleClick(e) {
      if (!disabled.value) {
        var _props2$onClick;
        (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, {
          e: e
        });
      }
    };
    var renderImage = function renderImage() {
      if (typeof props2.image === "string") {
        return createVNode("img", {
          "src": props2.image,
          "class": "".concat(cellClass.value, "__left-image")
        }, null);
      }
      var image = renderTNodeJSX("image");
      return image;
    };
    var renderLeft = function renderLeft() {
      var leftIcon = renderTNodeJSX("leftIcon");
      return createVNode("div", {
        "class": "".concat(cellClass.value, "__left")
      }, [leftIcon && createVNode("div", {
        "class": "".concat(cellClass.value, "__left-icon")
      }, [leftIcon]), renderImage()]);
    };
    var renderTitle = function renderTitle() {
      var title = renderTNodeJSX("title");
      if (!title) {
        return null;
      }
      var description = renderTNodeJSX("description");
      return createVNode("div", {
        "class": "".concat(cellClass.value, "__title")
      }, [createVNode("div", {
        "class": "".concat(cellClass.value, "__title-text")
      }, [title, props2.required && createVNode("span", {
        "class": "".concat(cellClass.value, "--required")
      }, ["\xA0*"])]), description && createVNode("div", {
        "class": "".concat(cellClass.value, "__description")
      }, [description])]);
    };
    var renderRight = function renderRight() {
      var rightIcon = props2.arrow ? createVNode(ChevronRightIcon, null, null) : renderTNodeJSX("rightIcon");
      if (!rightIcon) {
        return null;
      }
      return createVNode("div", {
        "class": ["".concat(cellClass.value, "__right"), "".concat(cellClass.value, "__right--").concat(props2.align)]
      }, [createVNode("div", {
        "class": "".concat(cellClass.value, "__right-icon")
      }, [rightIcon])]);
    };
    return function () {
      var note = renderTNodeContent("default", "note");
      return withDirectives(createVNode("div", {
        "class": cellClasses.value,
        "onClick": handleClick
      }, [renderLeft(), renderTitle(), note && createVNode("div", {
        "class": "".concat(cellClass.value, "__note")
      }, [note]), renderRight()]), [[resolveDirective("hover"), {
        className: "".concat(cellClass.value, "--hover"),
        disabledHover: hoverDisabled.value
      }]]);
    };
  }
});

export { _Cell as default };
