// tdesign-mobile-vue/cjs/hooks/tnode.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var hooks_renderTnode = require('./render-tnode.js');
var _common_js_utils_general = require('../_common/js/utils/general.js');
var camelCase = require('../_chunks/dep-a65592ed.js');
var kebabCase = require('../_chunks/dep-9e13db9a.js');
var isFunction = require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');

function handleSlots(instance, name, params) {
  var _instance$slots$camel, _instance$slots, _instance$slots$kebab, _instance$slots2;
  var node = (_instance$slots$camel = (_instance$slots = instance.slots)[camelCase.camelCase(name)]) === null || _instance$slots$camel === void 0 ? void 0 : _instance$slots$camel.call(_instance$slots, params);
  if (node && node.filter(function (t) {
    var _t$type$toString, _t$type;
    return ((_t$type$toString = (_t$type = t.type).toString) === null || _t$type$toString === void 0 ? void 0 : _t$type$toString.call(_t$type)) !== "Symbol(v-cmt)";
  }).length) return node;
  node = (_instance$slots$kebab = (_instance$slots2 = instance.slots)[kebabCase.kebabCase(name)]) === null || _instance$slots$kebab === void 0 ? void 0 : _instance$slots$kebab.call(_instance$slots2, params);
  if (node && node.filter(function (t) {
    var _t$type$toString2, _t$type2;
    return ((_t$type$toString2 = (_t$type2 = t.type).toString) === null || _t$type$toString2 === void 0 ? void 0 : _t$type$toString2.call(_t$type2)) !== "Symbol(v-cmt)";
  }).length) return node;
  return null;
}
function isEmptyNode(node) {
  if ([void 0, null, ""].includes(node)) return true;
  var innerNodes = node instanceof Array ? node : [node];
  var r = innerNodes.filter(function (node2) {
    var _node2$type;
    return (node2 === null || node2 === void 0 || (_node2$type = node2.type) === null || _node2$type === void 0 ? void 0 : _node2$type.toString()) !== "Symbol(Comment)";
  });
  return !r.length;
}
function isPropExplicitlySet(instance, propName) {
  var vProps = (instance === null || instance === void 0 ? void 0 : instance.vnode.props) || {};
  return _common_js_utils_general.hasOwn(vProps, camelCase.camelCase(propName)) || _common_js_utils_general.hasOwn(vProps, kebabCase.kebabCase(propName));
}
var useTNodeJSX = function useTNodeJSX() {
  var instance = Vue.getCurrentInstance();
  return function (name, options) {
    var renderParams = hooks_renderTnode.getParams(options);
    var defaultNode = hooks_renderTnode.getDefaultNode(options);
    var isSlotFirst = hooks_renderTnode.getSlotFirst(options);
    var renderSlot = instance.slots[camelCase.camelCase(name)] || instance.slots[kebabCase.kebabCase(name)];
    if (isSlotFirst && renderSlot) {
      return handleSlots(instance, name, renderParams);
    }
    if (isPropExplicitlySet(instance, name)) {
      var _instance$type$props$;
      var propsNode2 = instance.props[camelCase.camelCase(name)] || instance.props[kebabCase.kebabCase(name)];
      var types = (_instance$type$props$ = instance.type.props[name]) === null || _instance$type$props$ === void 0 ? void 0 : _instance$type$props$.type;
      if ((types === null || types === void 0 ? void 0 : types.length) > 1) {
        if (types.includes(Boolean) && types.includes(Function)) {
          if (propsNode2 === "" && !renderSlot) return defaultNode;
        }
      }
      if (propsNode2 === false || propsNode2 === null) return;
      if (propsNode2 === true) {
        return handleSlots(instance, name, renderParams) || defaultNode;
      }
      if (isFunction.isFunction(propsNode2)) return propsNode2(Vue.h, renderParams);
      var isPropsEmpty = [void 0, ""].includes(propsNode2);
      if (isPropsEmpty && renderSlot) {
        return handleSlots(instance, name, renderParams);
      }
      return propsNode2;
    }
    if (renderSlot) {
      return handleSlots(instance, name, renderParams);
    }
    var propsNode = instance.props[camelCase.camelCase(name)] || instance.props[kebabCase.kebabCase(name)];
    if (propsNode === false || propsNode === null) return;
    if (propsNode === true) {
      return defaultNode;
    }
    if (isFunction.isFunction(propsNode)) return propsNode(Vue.h, renderParams);
    return propsNode;
  };
};
var useTNodeDefault = function useTNodeDefault() {
  var renderTNodeJSX = useTNodeJSX();
  return function (name, options) {
    var defaultNode = hooks_renderTnode.getDefaultNode(options);
    return renderTNodeJSX(name, options) || defaultNode;
  };
};
var useContent = function useContent() {
  var renderTNodeJSX = useTNodeJSX();
  return function (name1, name2, options) {
    var params = hooks_renderTnode.getParams(options);
    var defaultNode = hooks_renderTnode.getDefaultNode(options);
    var toParams = params ? {
      params: params
    } : void 0;
    var node1 = renderTNodeJSX(name1, toParams);
    var node2 = renderTNodeJSX(name2, toParams);
    var res = isEmptyNode(node1) ? node2 : node1;
    return isEmptyNode(res) ? defaultNode : res;
  };
};

exports.useContent = useContent;
exports.useTNodeDefault = useTNodeDefault;
exports.useTNodeJSX = useTNodeJSX;
