// tdesign-mobile-vue/es/notice-bar/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  content: {
    type: [String, Array, Function]
  },
  direction: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  interval: {
    type: Number,
    default: 2e3
  },
  marquee: {
    type: [Boolean, Object],
    default: false
  },
  operation: {
    type: [String, Function]
  },
  prefixIcon: {
    type: [Boolean, Function],
    default: true
  },
  suffixIcon: {
    type: Function
  },
  theme: {
    type: String,
    default: "info",
    validator: function validator(val) {
      if (!val) return true;
      return ["info", "success", "warning", "error"].includes(val);
    }
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  onChange: Function,
  onClick: Function
};

export { props as default };
