// tdesign-mobile-vue/es/textarea/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  allowInputOverMax: Boolean,
  autofocus: Boolean,
  autosize: {
    type: [Boolean, Object],
    default: false
  },
  bordered: Boolean,
  cursorColor: {
    type: String,
    default: "#0052d9"
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  indicator: Boolean,
  label: {
    type: [String, Function]
  },
  layout: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["vertical", "horizontal"].includes(val);
    }
  },
  maxcharacter: {
    type: Number
  },
  maxlength: {
    type: [String, Number]
  },
  name: {
    type: String,
    default: ""
  },
  placeholder: {
    type: String,
    default: void 0
  },
  readonly: {
    type: Boolean,
    default: void 0
  },
  value: {
    type: [String, Number],
    default: void 0
  },
  modelValue: {
    type: [String, Number],
    default: void 0
  },
  defaultValue: {
    type: [String, Number]
  },
  onBlur: Function,
  onChange: Function,
  onFocus: Function
};

export { props as default };
