// tdesign-mobile-vue/cjs/date-time-picker/date-time-picker.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var dayjs = require('dayjs');
var _commonjsHelpers = require('../_chunks/dep-7220ff76.js');
var customParseFormat = require('../_chunks/dep-8025cea8.js');
var config = require('../config.js');
var dateTimePicker_props = require('./props.js');
var dateTimePicker_shared = require('./shared.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_tnode = require('../hooks/tnode.js');
var picker_index = require('../picker/index.js');
var hooks_useClass = require('../hooks/useClass.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
var isArray = require('../_chunks/dep-3cf5587c.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-f84e6e17.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../picker/picker.js');
require('@babel/runtime/helpers/defineProperty');
require('../picker/props.js');
require('../picker/picker-item.js');
require('../picker/picker.class.js');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('../shared/dom.js');
require('../picker/utils.js');
require('../_chunks/dep-cfb75f1b.js');
require('../_chunks/dep-01310648.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-eb605cd5.js');
require('../picker/cascade.js');
require('../shared/component.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var weekday$1 = {exports: {}};

(function (module, exports) {
  !function (e, t) {
    module.exports = t() ;
  }(_commonjsHelpers.commonjsGlobal, function () {

    return function (e, t) {
      t.prototype.weekday = function (e) {
        var t = this.$locale().weekStart || 0,
          i = this.$W,
          n = (i < t ? i + 7 : i) - t;
        return this.$utils().u(e) ? n : this.subtract(n, "day").add(e, "day");
      };
    };
  });
})(weekday$1);
var weekday = weekday$1.exports;

var objectSupport$1 = {exports: {}};

(function (module, exports) {
  !function (t, n) {
    module.exports = n() ;
  }(_commonjsHelpers.commonjsGlobal, function () {

    return function (t, n, e) {
      var i = n.prototype,
        r = function r(t) {
          var n,
            r = t.date,
            o = t.utc,
            u = {};
          if (!(null === (n = r) || n instanceof Date || n instanceof Array || i.$utils().u(n) || "Object" !== n.constructor.name)) {
            if (!Object.keys(r).length) return new Date();
            var a = o ? e.utc() : e();
            Object.keys(r).forEach(function (t) {
              var n, e;
              u[n = t, e = i.$utils().p(n), "date" === e ? "day" : e] = r[t];
            });
            var c = u.day || (u.year || u.month >= 0 ? 1 : a.date()),
              s = u.year || a.year(),
              d = u.month >= 0 ? u.month : u.year || u.day ? 0 : a.month(),
              f = u.hour || 0,
              b = u.minute || 0,
              h = u.second || 0,
              y = u.millisecond || 0;
            return o ? new Date(Date.UTC(s, d, c, f, b, h, y)) : new Date(s, d, c, f, b, h, y);
          }
          return r;
        },
        o = i.parse;
      i.parse = function (t) {
        t.date = r.bind(this)(t), o.bind(this)(t);
      };
      var u = i.set,
        a = i.add,
        c = i.subtract,
        s = function s(t, n, e, i) {
          void 0 === i && (i = 1);
          var r = Object.keys(n),
            o = this;
          return r.forEach(function (e) {
            o = t.bind(o)(n[e] * i, e);
          }), o;
        };
      i.set = function (t, n) {
        return n = void 0 === n ? t : n, "Object" === t.constructor.name ? s.bind(this)(function (t, n) {
          return u.bind(this)(n, t);
        }, n, t) : u.bind(this)(t, n);
      }, i.add = function (t, n) {
        return "Object" === t.constructor.name ? s.bind(this)(a, t, n) : a.bind(this)(t, n);
      }, i.subtract = function (t, n) {
        return "Object" === t.constructor.name ? s.bind(this)(a, t, n, -1) : c.bind(this)(t, n);
      };
    };
  });
})(objectSupport$1);
var objectSupport = objectSupport$1.exports;

dayjs__default["default"].extend(weekday);
dayjs__default["default"].extend(customParseFormat.customParseFormat);
dayjs__default["default"].extend(objectSupport);
var prefix = config["default"].prefix;
var _DateTimePicker = Vue.defineComponent({
  name: "".concat(prefix, "-date-time-picker"),
  components: {
    TPicker: picker_index.Picker
  },
  props: dateTimePicker_props["default"],
  emits: ["change", "cancel", "confirm", "pick", "update:modelValue", "update:value"],
  setup: function setup(props, _ref) {
    _ref.slots;
    var dateTimePickerClass = hooks_useClass.usePrefixClass("date-time-picker");
    var _useConfig = configProvider_useConfig.useConfig("dateTimePicker"),
      globalConfig = _useConfig.globalConfig;
    var className = Vue.computed(function () {
      return ["".concat(dateTimePickerClass.value)];
    });
    var _toRefs = Vue.toRefs(props),
      value = _toRefs.value;
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _useVModel = hooks_useVModel["default"](value, Vue.ref(props.modelValue), props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerValue = _useVModel2[0],
      setDateTimePickerValue = _useVModel2[1];
    var title = Vue.computed(function () {
      return props.title || globalConfig.value.title;
    });
    var confirmButtonText = Vue.computed(function () {
      return props.confirmBtn || globalConfig.value.confirm;
    });
    var cancelButtonText = Vue.computed(function () {
      return props.cancelBtn || globalConfig.value.cancel;
    });
    var normalize = function normalize(val, defaultDay) {
      return val && dayjs__default["default"](val).isValid() ? dayjs__default["default"](val) : defaultDay;
    };
    var start = Vue.computed(function () {
      return normalize(props.start, dayjs__default["default"]().subtract(10, "year"));
    });
    var end = Vue.computed(function () {
      return normalize(props.end, dayjs__default["default"]().add(10, "year"));
    });
    var meaningColumn = Vue.computed(function () {
      return dateTimePicker_shared.getMeaningColumn(props.mode);
    });
    var isTimeMode = Vue.computed(function () {
      return isArray.isArray(props.mode) && props.mode[0] == null && ["hour", "minute", "second"].includes(props.mode[1]);
    });
    var rationalize = function rationalize(val) {
      if (isTimeMode.value) return val;
      if (val.isBefore(start.value)) return start.value;
      if (val.isAfter(end.value)) return end.value;
      return val;
    };
    var calcDate = function calcDate(currentValue) {
      if (isTimeMode.value) {
        var dateStr = dayjs__default["default"](start.value).format("YYYY-MM-DD");
        currentValue = "".concat(dateStr, " ").concat(currentValue);
      }
      return currentValue && dayjs__default["default"](currentValue).isValid() ? rationalize(dayjs__default["default"](currentValue)) : start.value;
    };
    var curDate = Vue.ref(calcDate(innerValue.value));
    var valueOfPicker = Vue.computed(function () {
      return meaningColumn.value.map(function (item) {
        return curDate.value[item]().toString();
      });
    });
    var columns = Vue.computed(function () {
      var ret = [];
      var getDate = function getDate(date) {
        return [date.year(), date.month() + 1, date.date(), date.hour(), date.minute(), date.second()];
      };
      var _getDate = getDate(curDate.value),
        _getDate2 = _slicedToArray__default["default"](_getDate, 5),
        curYear = _getDate2[0],
        curMonth = _getDate2[1],
        curDay = _getDate2[2],
        curHour = _getDate2[3],
        curMinute = _getDate2[4];
      var _getDate3 = getDate(start.value),
        _getDate4 = _slicedToArray__default["default"](_getDate3, 6),
        minYear = _getDate4[0],
        minMonth = _getDate4[1],
        minDay = _getDate4[2],
        minHour = _getDate4[3],
        minMinute = _getDate4[4],
        minSecond = _getDate4[5];
      var _getDate5 = getDate(end.value),
        _getDate6 = _slicedToArray__default["default"](_getDate5, 6),
        maxYear = _getDate6[0],
        maxMonth = _getDate6[1],
        maxDay = _getDate6[2],
        maxHour = _getDate6[3],
        maxMinute = _getDate6[4],
        maxSecond = _getDate6[5];
      var isInMinYear = curYear === minYear;
      var isInMaxYear = curYear === maxYear;
      var isInMinMonth = isInMinYear && curMonth === minMonth;
      var isInMaxMonth = isInMaxYear && curMonth === maxMonth;
      var isInMinDay = isInMinMonth && curDay === minDay;
      var isInMaxDay = isInMaxMonth && curDay === maxDay;
      var isInMinHour = isInMinDay && curHour === minHour;
      var isInMaxHour = isInMaxDay && curHour === maxHour;
      var isInMinMinute = isInMinHour && curMinute === minMinute;
      var isInMaxMinute = isInMaxHour && curMinute === maxMinute;
      var typeUnit = {
        year: globalConfig.value.yearLabel,
        month: globalConfig.value.monthLabel,
        date: globalConfig.value.dateLabel,
        hour: globalConfig.value.hourLabel,
        minute: globalConfig.value.minuteLabel,
        second: globalConfig.value.secondLabel
      };
      var generateDayWithWeekColumn = function generateDayWithWeekColumn(date) {
        var startOfMonth = date.startOf("month");
        var endOfMonth = date.endOf("month");
        var daysOfWeek = [];
        var type = "date";
        for (var i = 0; i <= endOfMonth.diff(startOfMonth, "days"); i += 1) {
          var currentDate = startOfMonth.add(i, "days");
          var dayName = currentDate.format("ddd");
          daysOfWeek.push({
            value: "".concat(i + 1),
            label: props.renderLabel ? props.renderLabel(type, i) : "".concat(i + 1).concat(typeUnit[type] || "", " ").concat(dayName)
          });
        }
        ret.push(daysOfWeek);
      };
      var generateColumn = function generateColumn(start2, end2, type) {
        var arr = [];
        var step = props.steps[type] || 1;
        for (var i = start2; i <= end2; i += step) {
          var value2 = i.toString();
          arr.push({
            label: props.renderLabel ? props.renderLabel(type, i) : "".concat(value2, " ").concat(typeUnit[type]),
            value: type === "month" ? "".concat(+value2 - 1) : value2
          });
        }
        ret.push(arr);
      };
      if (meaningColumn.value.includes("year")) {
        generateColumn(minYear, maxYear, "year");
      }
      if (meaningColumn.value.includes("month")) {
        var lower = isInMinYear ? minMonth : 1;
        var upper = isInMaxYear ? maxMonth : 12;
        generateColumn(lower, upper, "month");
      }
      if (meaningColumn.value.includes("date")) {
        var _lower = isInMinMonth ? minDay : 1;
        var _upper = isInMaxMonth ? maxDay : dayjs__default["default"]("".concat(curYear, "-").concat(curMonth)).daysInMonth();
        if (props.showWeek) {
          generateDayWithWeekColumn(curDate.value);
        } else {
          generateColumn(_lower, _upper, "date");
        }
      }
      if (meaningColumn.value.includes("hour")) {
        var _lower2 = isInMinDay && !isTimeMode.value ? minHour : 0;
        var _upper2 = isInMaxDay && !isTimeMode.value ? maxHour : 23;
        generateColumn(_lower2, _upper2, "hour");
      }
      if (meaningColumn.value.includes("minute")) {
        var _lower3 = isInMinHour && !isTimeMode.value ? minMinute : 0;
        var _upper3 = isInMaxHour && !isTimeMode.value ? maxMinute : 59;
        generateColumn(_lower3, _upper3, "minute");
      }
      if (meaningColumn.value.includes("second")) {
        var _lower4 = isInMinMinute && !isTimeMode.value ? minSecond : 0;
        var _upper4 = isInMaxMinute && !isTimeMode.value ? maxSecond : 59;
        generateColumn(_lower4, _upper4, "second");
      }
      return ret;
    });
    var onConfirm = function onConfirm(value2) {
      var _props$onConfirm;
      var dayObject = value2.reduce(function (map, cur2, index) {
        var type = meaningColumn.value[index];
        map[type] = cur2;
        return map;
      }, {});
      var cur = dayjs__default["default"](dayObject);
      (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, dayjs__default["default"](cur || curDate.value).format(props.format));
      setDateTimePickerValue(dayjs__default["default"](cur || curDate.value).format(props.format));
    };
    var onCancel = function onCancel(context) {
      var _props$onCancel;
      (_props$onCancel = props.onCancel) === null || _props$onCancel === void 0 || _props$onCancel.call(props, {
        e: context.e
      });
    };
    var onPick = function onPick(value2, context) {
      var _columns$value$column, _props$onPick;
      var column = context.column,
        index = context.index;
      var type = meaningColumn.value[column];
      var val = curDate.value.set(type, parseInt((_columns$value$column = columns.value[column][index]) === null || _columns$value$column === void 0 ? void 0 : _columns$value$column.value, 10));
      curDate.value = rationalize(val);
      (_props$onPick = props.onPick) === null || _props$onPick === void 0 || _props$onPick.call(props, rationalize(val).format(props.format));
    };
    Vue.watch(innerValue, function (val) {
      curDate.value = calcDate(val);
    });
    return function () {
      return Vue.createVNode(Vue.resolveComponent("t-picker"), {
        "class": className.value,
        "value": valueOfPicker.value,
        "title": title.value,
        "confirm-btn": confirmButtonText.value,
        "cancel-btn": cancelButtonText.value,
        "columns": columns.value,
        "onConfirm": onConfirm,
        "onCancel": onCancel,
        "onPick": onPick,
        "header": function header() {
          return renderTNodeJSX("header");
        },
        "footer": function footer() {
          return renderTNodeJSX("footer");
        }
      }, null);
    };
  }
});

exports["default"] = _DateTimePicker;
