// tdesign-mobile-vue/es/table/hooks/useStyle.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-f4e31210.mjs';
import { toRefs, computed } from 'vue';
import useClassName from './useClassName.mjs';
import '../../_chunks/dep-4055ef5c.mjs';
import '../../hooks/useClass.mjs';
import '../../config-provider/useConfig.mjs';
import 'lodash-es';
import '../../config-provider/context.mjs';
import '../../_common/js/global-config/mobile/default-config.mjs';
import '../../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../../_chunks/dep-4c96ffa8.mjs';
import '../../_chunks/dep-04a2869f.mjs';
import '../../_chunks/dep-a8bd494a.mjs';
import '../../_chunks/dep-d3be051e.mjs';
import '../../config-provider/type.mjs';

function formatCSSUnit(unit) {
  if (!unit) return unit;
  return isNaN(Number(unit)) ? unit : "".concat(unit, "px");
}
function useStyle(props) {
  var _toRefs = toRefs(props),
    bordered = _toRefs.bordered,
    stripe = _toRefs.stripe,
    verticalAlign = _toRefs.verticalAlign,
    height = _toRefs.height,
    maxHeight = _toRefs.maxHeight,
    tableContentWidth = _toRefs.tableContentWidth;
  var _useClassName = useClassName(),
    tableBaseClass = _useClassName.tableBaseClass,
    tableAlignClasses = _useClassName.tableAlignClasses;
  var tableClasses = computed(function () {
    return [tableBaseClass.table, [tableAlignClasses[(verticalAlign === null || verticalAlign === void 0 ? void 0 : verticalAlign.value) || "middle"]], _defineProperty(_defineProperty(_defineProperty({}, tableBaseClass.bordered, bordered === null || bordered === void 0 ? void 0 : bordered.value), tableBaseClass.striped, stripe === null || stripe === void 0 ? void 0 : stripe.value), tableBaseClass.loading, props.loading)];
  });
  var tableContentStyles = computed(function () {
    return {
      height: formatCSSUnit(height === null || height === void 0 ? void 0 : height.value),
      maxHeight: formatCSSUnit(maxHeight === null || maxHeight === void 0 ? void 0 : maxHeight.value)
    };
  });
  var tableElementStyles = computed(function () {
    return {
      width: formatCSSUnit(tableContentWidth === null || tableContentWidth === void 0 ? void 0 : tableContentWidth.value)
    };
  });
  return {
    tableClasses: tableClasses,
    tableElementStyles: tableElementStyles,
    tableContentStyles: tableContentStyles
  };
}

export { useStyle as default, formatCSSUnit };
