// tdesign-mobile-vue/es/qrcode/components/qrcode-status.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, toRefs, createVNode, computed, Fragment } from 'vue';
import { RefreshIcon, CheckCircleFilledIcon } from 'tdesign-icons-vue-next';
import { QRCodeStatusProps } from './props.mjs';
import { Loading } from '../../loading/index.mjs';
import '../../loading/loading.mjs';
import '../../_chunks/dep-f4e31210.mjs';
import '../../_chunks/dep-4055ef5c.mjs';
import '../../loading/icon/gradient.mjs';
import '../../_common/js/loading/circle-adapter.mjs';
import '../../_common/js/utils/setStyle.mjs';
import '../../_common/js/utils/helper.mjs';
import '../../_chunks/dep-fb069758.mjs';
import '../../_chunks/dep-9961b95b.mjs';
import '../../_chunks/dep-a27c5c00.mjs';
import '../../_chunks/dep-2ed7e0d5.mjs';
import 'lodash-es';
import '../../config.mjs';
import '../../hooks/useClass.mjs';
import '../../config-provider/useConfig.mjs';
import '../../config-provider/context.mjs';
import '../../_common/js/global-config/mobile/default-config.mjs';
import '../../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../../_chunks/dep-4c96ffa8.mjs';
import '../../_chunks/dep-04a2869f.mjs';
import '../../_chunks/dep-a8bd494a.mjs';
import '../../_chunks/dep-d3be051e.mjs';
import '../../config-provider/type.mjs';
import '../../loading/icon/spinner.mjs';
import '../../loading/props.mjs';
import '../../hooks/tnode.mjs';
import '../../hooks/render-tnode.mjs';
import '../../_common/js/utils/general.mjs';
import '../../shared/dom.mjs';
import '../../shared/index.mjs';
import '../../shared/functions.mjs';
import '../../shared/util.mjs';
import '../../shared/component.mjs';
import '../../shared/render.mjs';
import '../../shared/render-tnode.mjs';
import '../../shared/useCountDown/index.mjs';
import '../../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../../shared/useCountDown/utils.mjs';
import '../../shared/useTouch/index.mjs';
import '../../shared/useScrollParent/index.mjs';
import '../../shared/useTest/index.mjs';
import '../../shared/useClickAway/index.mjs';
import '../../shared/hover.mjs';
import './style/css.mjs';
import '../../loading/type.mjs';
import '../../loading/plugin.mjs';

var QRcodeStatus = defineComponent({
  name: "QRCodeStatus",
  props: QRCodeStatusProps,
  setup: function setup(props) {
    var _locale$value3;
    var _toRefs = toRefs(props),
      classPrefix = _toRefs.classPrefix,
      locale = _toRefs.locale,
      status = _toRefs.status;
    var defaultSpin = createVNode(Loading, {
      "size": "32px"
    }, null);
    var defaultExpiredNode = computed(function () {
      var _locale$value, _locale$value2;
      return createVNode(Fragment, null, [createVNode("p", {
        "class": "".concat(classPrefix.value, "-expired__text")
      }, [(_locale$value = locale.value) === null || _locale$value === void 0 ? void 0 : _locale$value.expiredText]), (props === null || props === void 0 ? void 0 : props.onRefresh) && createVNode("p", {
        "class": "".concat(classPrefix.value, "-expired__button"),
        "onClick": props === null || props === void 0 ? void 0 : props.onRefresh
      }, [createVNode(RefreshIcon, {
        "size": "16"
      }, null), (_locale$value2 = locale.value) === null || _locale$value2 === void 0 ? void 0 : _locale$value2.refreshText])]);
    });
    var defaultScannedNode = createVNode("p", {
      "class": "".concat(classPrefix.value, "-scanned")
    }, [createVNode(CheckCircleFilledIcon, {
      "size": "16",
      "class": "".concat(classPrefix.value, "-scanned__icon")
    }, null), (_locale$value3 = locale.value) === null || _locale$value3 === void 0 ? void 0 : _locale$value3.scannedText]);
    var defaultNodes = computed(function () {
      return {
        expired: defaultExpiredNode.value,
        loading: defaultSpin,
        scanned: defaultScannedNode,
        active: null
      };
    });
    var renderStatus = function renderStatus(info) {
      return defaultNodes.value[info.status];
    };
    return function () {
      return createVNode(Fragment, null, [props.statusRender || renderStatus({
        status: status.value,
        onRefresh: props.onRefresh
      })]);
    };
  }
});

export { QRcodeStatus as default };
