// tdesign-mobile-vue/esm/steps/steps.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, computed, toRefs, reactive, provide, createVNode } from 'vue';
import props from './props.js';
import config from '../config.js';
import useVModel from '../hooks/useVModel.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import 'lodash-es';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _Steps = defineComponent({
  name: "".concat(prefix, "-steps"),
  props: props,
  emits: ["update:current", "update:modelValue", "change"],
  setup: function setup(props2, context) {
    var stepsClass = usePrefixClass("steps");
    var baseClass = computed(function () {
      return [stepsClass.value, "".concat(stepsClass.value, "--").concat(props2.layout), "".concat(stepsClass.value, "--").concat(props2.sequence), _defineProperty({}, "".concat(stepsClass.value, "--readonly"), props2.readonly)];
    });
    var renderTNodeJSX = useTNodeJSX();
    var _toRefs = toRefs(props2),
      current = _toRefs.current,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(current, modelValue, props2.defaultCurrent, props2.onChange, "current"),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      innerCurrent = _useVModel2[0],
      setInnerCurrent = _useVModel2[1];
    var state = reactive({
      children: []
    });
    var relation = function relation(child) {
      child && state.children.push(child);
    };
    var removeRelation = function removeRelation(child) {
      state.children = state.children.filter(function (item) {
        return item !== child;
      });
    };
    var onClickItem = function onClickItem(cur, prev, e) {
      setInnerCurrent(cur, prev, {
        e: e
      });
    };
    provide("stepsProvide", _objectSpread(_objectSpread({}, props2), {}, {
      state: state,
      current: innerCurrent,
      relation: relation,
      removeRelation: removeRelation,
      onClickItem: onClickItem
    }));
    return function () {
      var renderContent = renderTNodeJSX("default") || null;
      return createVNode("div", {
        "class": baseClass.value
      }, [renderContent]);
    };
  }
});

export { _Steps as default };
