// tdesign-mobile-vue/cjs/qrcode/components/qrcode-svg.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var qrcode_components_props = require('./props.js');
var _common_js_qrcode_utils = require('../../_common/js/qrcode/utils.js');
var qrcode_hooks_useQRCode = require('../hooks/useQRCode.js');
require('../../_common/js/qrcode/qrcodegen.js');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/defineProperty');

var QRCodeSVG = Vue.defineComponent({
  name: "QRCodeSVG",
  props: qrcode_components_props.QRCodeSubComponentProps,
  setup: function setup(props) {
    var _useQRCode = qrcode_hooks_useQRCode.useQRCode({
        value: props.value,
        level: props.level,
        minVersion: _common_js_qrcode_utils.DEFAULT_MINVERSION,
        includeMargin: _common_js_qrcode_utils.DEFAULT_NEED_MARGIN,
        marginSize: props.marginSize,
        imageSettings: props.imageSettings,
        size: props.size
      }),
      margin = _useQRCode.margin,
      cells = _useQRCode.cells,
      numCells = _useQRCode.numCells,
      calculatedImageSettings = _useQRCode.calculatedImageSettings;
    var cellsToDraw = Vue.computed(function () {
      var _calculatedImageSetti;
      if (props.imageSettings && ((_calculatedImageSetti = calculatedImageSettings.value) === null || _calculatedImageSetti === void 0 ? void 0 : _calculatedImageSetti.excavation) != null) {
        return _common_js_qrcode_utils.excavateModules(cells.value, calculatedImageSettings.value.excavation);
      }
      return cells.value;
    });
    var imageNode = Vue.computed(function () {
      if (!props.imageSettings || !calculatedImageSettings.value) return null;
      return Vue.createVNode("image", {
        "href": props.imageSettings.src,
        "height": calculatedImageSettings.value.h,
        "width": calculatedImageSettings.value.w,
        "x": calculatedImageSettings.value.x + margin.value,
        "y": calculatedImageSettings.value.y + margin.value,
        "cross-origin": calculatedImageSettings.value.crossOrigin
      }, null);
    });
    return function () {
      var fgPath = _common_js_qrcode_utils.generatePath(cellsToDraw.value, margin.value);
      return Vue.createVNode("svg", {
        "height": props.size,
        "width": props.size,
        "viewBox": "0 0 ".concat(numCells.value, " ").concat(numCells.value),
        "role": "img",
        "style": props.style
      }, [!!props.title && Vue.createVNode("title", null, [props.title]), Vue.createVNode("path", {
        "fill": props.bgColor,
        "d": "M0,0 h".concat(numCells.value, "v").concat(numCells.value, "H0z"),
        "shape-rendering": "crispEdges"
      }, null), Vue.createVNode("path", {
        "fill": props.fgColor,
        "d": fgPath,
        "shape-rendering": "crispEdges"
      }, null), imageNode.value]);
    };
  }
});

exports["default"] = QRCodeSVG;
