// tdesign-mobile-vue/es/calendar/calendar.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, ref, provide, reactive, nextTick, onMounted, watch, createVNode } from 'vue';
import { Popup } from '../popup/index.mjs';
import config from '../config.mjs';
import props from './props.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import TCalendarTemplate from './template.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../popup/popup.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'tdesign-icons-vue-next';
import '../popup/props.mjs';
import '../overlay/index.mjs';
import '../overlay/overlay.mjs';
import '../shared/dom.mjs';
import 'lodash-es';
import '../overlay/props.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../hooks/useVModel.mjs';
import '../hooks/useLockScroll.mjs';
import '../_util/useTouch.mjs';
import '../_util/getScrollParent.mjs';
import '../hooks/useTeleport.mjs';
import '../popup/type.mjs';
import '../button/index.mjs';
import '../button/button.mjs';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../button/props.mjs';
import '../form/hooks.mjs';
import '../button/type.mjs';
import './template-props.mjs';
import './utils.mjs';

var prefix = config.prefix;
var _Calendar = defineComponent({
  name: "".concat(prefix, "-calendar"),
  props: props,
  emits: ["update:visible"],
  setup: function setup(props2, context) {
    var calendarClass = usePrefixClass("calendar");
    var calendarTemplateRef = ref();
    var renderTNodeJSX = useTNodeJSX();
    provide("templateProps", reactive(props2));
    var selectedValueIntoView = function selectedValueIntoView() {
      var _templateRef$querySel;
      var type = props2.type === "range" ? "start" : "selected";
      var templateRef = calendarTemplateRef.value.templateRef;
      var scrollContainer = templateRef.querySelector(".".concat(calendarClass.value, "__months"));
      var selectedDate = (_templateRef$querySel = templateRef.querySelector(".".concat(calendarClass.value, "__dates-item--").concat(type))) === null || _templateRef$querySel === void 0 || (_templateRef$querySel = _templateRef$querySel.parentNode) === null || _templateRef$querySel === void 0 ? void 0 : _templateRef$querySel.previousElementSibling;
      if (selectedDate) {
        scrollContainer.scrollTop = selectedDate.offsetTop - scrollContainer.offsetTop;
      }
    };
    var onVisibleChange = function onVisibleChange(v) {
      context.emit("update:visible", v);
    };
    var onPopupVisibleChange = function onPopupVisibleChange(v) {
      if (!v) {
        var _props2$onClose;
        (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, "overlay");
      } else {
        nextTick(function () {
          selectedValueIntoView();
        });
      }
      context.emit("update:visible", v);
    };
    onMounted(function () {
      if (!props2.usePopup) selectedValueIntoView();
    });
    watch(function () {
      return props2.value;
    }, function (val) {
      calendarTemplateRef.value.valueRef = val;
    });
    return function () {
      var title = renderTNodeJSX("title");
      var confirmBtn = renderTNodeJSX("confirmBtn");
      return createVNode("div", null, [!props2.usePopup ? createVNode(TCalendarTemplate, {
        "ref": calendarTemplateRef,
        "title": title,
        "confirmBtn": confirmBtn
      }, null) : createVNode(Popup, {
        "visible": props2.visible,
        "placement": "bottom",
        "onVisibleChange": onPopupVisibleChange
      }, {
        default: function _default() {
          return [createVNode(TCalendarTemplate, {
            "ref": calendarTemplateRef,
            "title": title,
            "confirmBtn": confirmBtn,
            "onVisibleChange": onVisibleChange
          }, null)];
        }
      })]);
    };
  }
});

export { _Calendar as default };
