// tdesign-mobile-vue/es/calendar/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  allowSameDay: Boolean,
  confirmBtn: {
    type: [String, Object, Function],
    default: ""
  },
  firstDayOfWeek: {
    type: Number,
    default: 0
  },
  format: {
    type: Function
  },
  maxDate: {
    type: [Number, Date]
  },
  minDate: {
    type: [Number, Date]
  },
  readonly: Boolean,
  switchMode: {
    type: String,
    default: "none",
    validator: function validator(val) {
      if (!val) return true;
      return ["none", "month", "year-month"].includes(val);
    }
  },
  title: {
    type: [String, Function]
  },
  type: {
    type: String,
    default: "single",
    validator: function validator(val) {
      if (!val) return true;
      return ["single", "multiple", "range"].includes(val);
    }
  },
  usePopup: {
    type: Boolean,
    default: true
  },
  value: {
    type: [Number, Array, Date],
    default: void 0
  },
  modelValue: {
    type: [Number, Array, Date],
    default: void 0
  },
  defaultValue: {
    type: [Number, Array, Date]
  },
  visible: Boolean,
  onChange: Function,
  onClose: Function,
  onConfirm: Function,
  onPanelChange: Function,
  onScroll: Function,
  onSelect: Function
};

export { props as default };
