// tdesign-mobile-vue/lib/action-sheet/action-sheet.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, computed, toRefs, createVNode, mergeProps, isVNode } from 'vue';
import TActionSheetGrid from './action-sheet-grid.js';
import TActionSheetList from './action-sheet-list.js';
import { Button } from '../button/index.js';
import { Popup } from '../popup/index.js';
import config from '../config.js';
import { usePrefixClass } from '../hooks/useClass.js';
import useVModel from '../hooks/useVModel.js';
import props from './props.js';
import { useConfig } from '../config-provider/useConfig.js';
import '../grid/index.js';
import '../grid/grid.js';
import '../grid/props.js';
import 'lodash-es';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../grid/grid-item.js';
import '../grid/grid-item-props.js';
import '../hooks/tnode.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../image/index.js';
import '../image/image.js';
import '@vueuse/core';
import 'tdesign-icons-vue-next';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/setStyle.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../loading/icon/spinner.js';
import '../loading/props.js';
import '../shared/dom.js';
import '../loading/plugin.js';
import '../shared/component.js';
import '../image/props.js';
import '../badge/index.js';
import '../badge/badge.js';
import '../badge/props.js';
import '../shared/hover.js';
import '../swiper/index.js';
import '../swiper/swiper.js';
import '../swipe-cell/useSwipe.js';
import '../shared/util.js';
import '../swiper/props.js';
import '../swiper/swiper-item.js';
import '../button/button.js';
import '../button/props.js';
import '../form/hooks.js';
import '../popup/popup.js';
import '../popup/props.js';
import '../overlay/index.js';
import '../overlay/overlay.js';
import '../overlay/props.js';
import '../hooks/useLockScroll.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';
import '../hooks/useTeleport.js';
import '../shared/render-tnode.js';

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var prefix = config.prefix;
var ActionSheetVue = defineComponent({
  name: "".concat(prefix, "-action-sheet"),
  props: props,
  emits: ["selected", "update:modelValue", "cancel", "close", "update:visible"],
  setup: function setup(props2, context) {
    var actionSheetClass = usePrefixClass("action-sheet");
    var _useConfig = useConfig("actionSheet"),
      globalConfig = _useConfig.globalConfig;
    var actionItems = computed(function () {
      return props2.items.map(function (item) {
        if (typeof item === "string") {
          return {
            label: item
          };
        }
        return item;
      });
    });
    var _toRefs = toRefs(props2),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(visible, modelValue, props2.defaultVisible, function () {}, "visible"),
      _useVModel2 = _slicedToArray(_useVModel, 1),
      currentVisible = _useVModel2[0];
    var rootClasses = computed(function () {
      return _defineProperty({}, "".concat(actionSheetClass.value, "__content"), true);
    });
    var descriptionClasses = computed(function () {
      return _defineProperty(_defineProperty(_defineProperty({}, "".concat(actionSheetClass.value, "__description"), true), "".concat(actionSheetClass.value, "__description--left"), props2.align === "left"), "".concat(actionSheetClass.value, "__description--grid"), props2.theme === "grid");
    });
    var hide = function hide(trigger) {
      var _props2$onClose;
      context.emit("update:modelValue", false);
      (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, {
        trigger: trigger
      });
    };
    var handleCancel = function handleCancel() {
      var _props2$onCancel;
      (_props2$onCancel = props2.onCancel) === null || _props2$onCancel === void 0 || _props2$onCancel.call(props2);
      context.emit("update:modelValue", false);
    };
    var handleSelected = function handleSelected(index) {
      var _props2$onSelected;
      (_props2$onSelected = props2.onSelected) === null || _props2$onSelected === void 0 || _props2$onSelected.call(props2, props2 === null || props2 === void 0 ? void 0 : props2.items[index], index);
      hide("selected");
    };
    var handleClose = function handleClose() {
      hide("overlay");
    };
    return function () {
      var _slot;
      var root = function root() {
        var description = function description() {
          if (props2.description) {
            return createVNode("p", {
              "class": descriptionClasses.value
            }, [props2.description]);
          }
          return null;
        };
        var list = function list() {
          if (props2.theme === "list") {
            return createVNode(TActionSheetList, {
              "align": props2.align,
              "items": actionItems.value,
              "onSelected": handleSelected
            }, null);
          }
        };
        var grid = function grid() {
          if (props2.theme === "grid") {
            return createVNode(TActionSheetGrid, {
              "items": actionItems.value,
              "count": props2.count,
              "onSelected": handleSelected
            }, null);
          }
        };
        var cancel = function cancel() {
          if (props2.showCancel) {
            return createVNode("div", {
              "class": "".concat(actionSheetClass.value, "__footer")
            }, [createVNode("div", {
              "class": "".concat(actionSheetClass.value, "__gap-").concat(props2.theme)
            }, null), createVNode(Button, {
              "class": "".concat(actionSheetClass.value, "__cancel"),
              "variant": "text",
              "block": true,
              "onClick": handleCancel
            }, {
              default: function _default() {
                return [props2.cancelText || globalConfig.value.cancel];
              }
            })]);
          }
        };
        return createVNode("div", {
          "class": rootClasses.value
        }, [description(), list(), grid(), cancel()]);
      };
      return createVNode(Popup, mergeProps(props2.popupProps, {
        "visible": currentVisible.value,
        "placement": "bottom",
        "destroy-on-close": true,
        "class": actionSheetClass.value,
        "onClose": handleClose,
        "showOverlay": props2.showOverlay
      }), _isSlot(_slot = root()) ? _slot : {
        default: function _default() {
          return [_slot];
        }
      });
    };
  }
});

export { ActionSheetVue as default };
