// tdesign-mobile-vue/lib/action-sheet/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  align: {
    type: String,
    default: "center",
    validator: function validator(val) {
      if (!val) return true;
      return ["center", "left"].includes(val);
    }
  },
  cancelText: {
    type: String,
    default: ""
  },
  count: {
    type: Number,
    default: 8
  },
  description: {
    type: String,
    default: ""
  },
  items: {
    type: Array
  },
  popupProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  showCancel: {
    type: Boolean,
    default: true
  },
  showOverlay: {
    type: Boolean,
    default: true
  },
  theme: {
    type: String,
    default: "list",
    validator: function validator(val) {
      if (!val) return true;
      return ["list", "grid"].includes(val);
    }
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  onCancel: Function,
  onClose: Function,
  onSelected: Function
};

export { props as default };
