// tdesign-mobile-vue/cjs/radio/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  allowUncheck: Boolean,
  block: {
    type: Boolean,
    default: true
  },
  borderless: {
    type: Boolean,
    default: void 0
  },
  checked: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultChecked: Boolean,
  content: {
    type: [String, Function]
  },
  contentDisabled: Boolean,
  default: {
    type: [String, Function]
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  icon: {
    type: [String, Array],
    default: "circle"
  },
  label: {
    type: [String, Function]
  },
  maxContentRow: {
    type: Number,
    default: 5
  },
  maxLabelRow: {
    type: Number,
    default: 3
  },
  name: {
    type: String,
    default: ""
  },
  placement: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right"].includes(val);
    }
  },
  readonly: {
    type: Boolean,
    default: void 0
  },
  value: {
    type: [String, Number, Boolean],
    default: void 0
  },
  onChange: Function
};

exports["default"] = props;
