// tdesign-mobile-vue/cjs/radio/radio.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var radio_props = require('./props.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var form_hooks = require('../form/hooks.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

var prefix = config["default"].prefix;
var _Radio = Vue.defineComponent({
  name: "".concat(prefix, "-radio"),
  props: radio_props["default"],
  emits: ["update:checked", "update:modelValue", "change"],
  setup: function setup(props2, context) {
    var renderTNodeContent = hooks_tnode.useContent();
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var radioClass = hooks_useClass.usePrefixClass("radio");
    var _toRefs = Vue.toRefs(props2),
      checked = _toRefs.checked,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](checked, modelValue, props2.defaultChecked, props2.onChange, "checked"),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerChecked = _useVModel2[0],
      setInnerChecked = _useVModel2[1];
    var rootGroupProps = Vue.inject("rootGroupProps", {});
    var rootGroupValue = Vue.inject("rootGroupValue", {});
    var rootGroupChange = Vue.inject("rootGroupChange", void 0);
    var groupDisabled = Vue.computed(function () {
      return rootGroupProps === null || rootGroupProps === void 0 ? void 0 : rootGroupProps.disabled;
    });
    var isDisabled = form_hooks.useFormDisabled(groupDisabled);
    var radioChecked = Vue.computed(function () {
      return (rootGroupValue === null || rootGroupValue === void 0 ? void 0 : rootGroupValue.value) !== void 0 ? props2.value === (rootGroupValue === null || rootGroupValue === void 0 ? void 0 : rootGroupValue.value) : innerChecked.value;
    });
    var finalBorderless = Vue.computed(function () {
      if (props2.borderless == null && "borderless" in rootGroupProps) return rootGroupProps.borderless;
      return props2.borderless;
    });
    var finalPlacement = Vue.computed(function () {
      if (props2.placement == null && "placement" in rootGroupProps) return rootGroupProps.placement;
      return props2.placement || "left";
    });
    var finalAllowUncheck = Vue.computed(function () {
      return Boolean(props2.allowUncheck || (rootGroupProps === null || rootGroupProps === void 0 ? void 0 : rootGroupProps.allowUncheck));
    });
    var finalReadonly = Vue.computed(function () {
      return Boolean(props2.readonly || (rootGroupProps === null || rootGroupProps === void 0 ? void 0 : rootGroupProps.readonly));
    });
    var inputProps = Vue.computed(function () {
      return {
        name: rootGroupProps.name || props2.name,
        checked: radioChecked.value,
        disabled: isDisabled.value,
        readonly: finalReadonly.value,
        value: props2.value
      };
    });
    var iconContent = Vue.computed(function () {
      var _props2$icon;
      if (((_props2$icon = props2.icon) === null || _props2$icon === void 0 ? void 0 : _props2$icon.length) === 2) {
        var iconIndex = radioChecked.value ? 0 : 1;
        return props2.icon[iconIndex];
      }
      return null;
    });
    var radioClasses = Vue.computed(function () {
      return ["".concat(radioClass.value), "".concat(radioClass.value, "--").concat(finalPlacement.value), _defineProperty__default["default"]({}, "".concat(radioClass.value, "--block"), props2.block)];
    });
    var titleClasses = Vue.computed(function () {
      return ["".concat(radioClass.value, "__title"), _defineProperty__default["default"]({}, "".concat(radioClass.value, "__title--disabled"), isDisabled.value)];
    });
    var iconClass = Vue.computed(function () {
      return ["".concat(radioClass.value, "__icon"), "".concat(radioClass.value, "__icon--").concat(finalPlacement.value), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(radioClass.value, "__icon--checked"), radioChecked.value), "".concat(radioClass.value, "__icon--disabled"), isDisabled.value)];
    });
    var radioContentChange = function radioContentChange(e) {
      e.stopPropagation();
      if (props2.contentDisabled) {
        return;
      }
      radioOrgChange(e);
    };
    var radioOrgChange = function radioOrgChange(e) {
      if (isDisabled.value || finalReadonly.value) {
        return;
      }
      if (rootGroupChange) {
        var value = finalAllowUncheck.value && radioChecked.value ? void 0 : props2.value;
        rootGroupChange(value, {
          e: e,
          name: props2.name
        });
      } else {
        var _value = finalAllowUncheck.value ? !radioChecked.value : true;
        setInnerChecked(_value, {
          e: e
        });
      }
    };
    var readerIcon = function readerIcon() {
      if (radioChecked.value) {
        if (props2.icon === "circle") {
          return Vue.createVNode(tdesignIconsVueNext.CheckCircleFilledIcon, {
            "class": "".concat(radioClass.value, "__icon-wrap")
          }, null);
        }
        if (props2.icon === "line") {
          return Vue.createVNode(tdesignIconsVueNext.CheckIcon, {
            "class": "".concat(radioClass.value, "__icon-wrap")
          }, null);
        }
        if (props2.icon === "dot") {
          return Vue.createVNode("div", {
            "class": ["".concat(radioClass.value, "__icon-").concat(props2.icon), _defineProperty__default["default"]({}, "".concat(radioClass.value, "__icon-").concat(props2.icon, "--disabled"), isDisabled.value)]
          }, null);
        }
      } else {
        if (props2.icon === "circle" || props2.icon === "dot") {
          return Vue.createVNode("div", {
            "class": ["".concat(radioClass.value, "__icon-circle"), _defineProperty__default["default"]({}, "".concat(radioClass.value, "__icon-circle--disabled"), isDisabled.value)]
          }, null);
        }
        if (props2.icon === "line") {
          return Vue.createVNode("div", {
            "class": "placeholder"
          }, null);
        }
      }
    };
    return function () {
      var label = renderTNodeContent("default", "label");
      var content = renderTNodeJSX("content");
      return Vue.createVNode("div", {
        "class": radioClasses.value,
        "onClick": radioOrgChange
      }, [Vue.createVNode("input", Vue.mergeProps({
        "type": "radio",
        "class": "".concat(radioClass.value, "__original")
      }, inputProps.value), null), Vue.createVNode("div", {
        "class": iconClass.value
      }, [readerIcon(), iconContent.value]), Vue.createVNode("div", {
        "class": "".concat(radioClass.value, "__content"),
        "onClick": radioContentChange
      }, [label && Vue.createVNode("span", {
        "class": titleClasses.value,
        "style": "-webkit-line-clamp: ".concat(props2.maxLabelRow)
      }, [label]), content && Vue.createVNode("div", {
        "class": ["".concat(radioClass.value, "__description"), _defineProperty__default["default"]({}, "".concat(radioClass.value, "__description"), isDisabled.value)],
        "style": "-webkit-line-clamp: ".concat(props2.maxContentRow)
      }, [content])]), !finalBorderless.value && Vue.createVNode("div", {
        "class": "".concat(radioClass.value, "__border ").concat(radioClass.value, "__border--").concat(finalPlacement.value)
      }, null)]);
    };
  }
});

exports["default"] = _Radio;
