// tdesign-mobile-vue/cjs/toast/toast.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var hooks_tnode = require('../hooks/tnode.js');
var overlay_index = require('../overlay/index.js');
var toast_props = require('./props.js');
var config = require('../config.js');
var hooks_useLockScroll = require('../hooks/useLockScroll.js');
var hooks_useClass = require('../hooks/useClass.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../overlay/overlay.js');
require('../shared/dom.js');
require('../overlay/props.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../_util/useTouch.js');
require('../_util/getScrollParent.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var vueToast = Vue.defineComponent({
  name: "".concat(prefix, "-toast"),
  props: toast_props["default"],
  setup: function setup(props) {
    var toastTypeIcon = {
      loading: tdesignIconsVueNext.LoadingIcon,
      success: tdesignIconsVueNext.CheckCircleIcon,
      warning: tdesignIconsVueNext.ErrorCircleIcon,
      error: tdesignIconsVueNext.CloseCircleIcon
    };
    var toastClass = hooks_useClass.usePrefixClass("toast");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var renderContent = hooks_tnode.useContent();
    var toastRef = Vue.ref();
    var customOverlayProps = Vue.computed(function () {
      var toastOverlayProps = {
        preventScrollThrough: props.preventScrollThrough,
        visible: props.showOverlay
      };
      return _objectSpread(_objectSpread({}, props.overlayProps), toastOverlayProps);
    });
    var classes = Vue.computed(function () {
      return ["".concat(toastClass.value), "".concat(toastClass.value, "__content"), "".concat(toastClass.value, "__icon"), _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(toastClass.value, "--").concat(props.direction), props.direction), "".concat(toastClass.value, "__content--").concat(props.direction), props.direction), "".concat(toastClass.value, "--loading"), props.theme === "loading")];
    });
    var topOptions = {
      top: "25%",
      bottom: "75%",
      middle: "50%"
    };
    var computedStyle = Vue.computed(function () {
      return {
        top: topOptions[props.placement]
      };
    });
    var iconClasses = Vue.computed(function () {
      return [_defineProperty__default["default"]({}, "".concat(toastClass.value, "__icon--").concat(props.direction), props.direction)];
    });
    var iconContent = Vue.computed(function () {
      var iconNode = renderTNodeJSX("icon");
      if (iconNode === void 0 && props.theme) {
        iconNode = Vue.h(toastTypeIcon[props.theme]);
      }
      return iconNode;
    });
    var renderIconContent = Vue.computed(function () {
      if (iconContent.value) {
        return Vue.createVNode("div", {
          "class": iconClasses.value
        }, [iconContent.value]);
      }
      return "";
    });
    var textClasses = Vue.computed(function () {
      return [_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(toastClass.value, "__text"), !iconContent.value), "".concat(toastClass.value, "__text--").concat(props.direction), props.direction)];
    });
    var messageContent = Vue.computed(function () {
      return renderContent("default", "message");
    });
    var renderMessageContent = Vue.computed(function () {
      if (messageContent.value) {
        return Vue.createVNode("div", {
          "class": textClasses.value
        }, [messageContent.value]);
      }
      return "";
    });
    hooks_useLockScroll.useLockScroll(toastRef, function () {
      return props.preventScrollThrough;
    }, toastClass.value);
    return function () {
      return Vue.createVNode("div", null, [Vue.createVNode(overlay_index.Overlay, customOverlayProps.value, null), Vue.createVNode("div", {
        "class": classes.value,
        "style": computedStyle.value
      }, [renderIconContent.value, renderMessageContent.value])]);
    };
  }
});

exports["default"] = vueToast;
