// tdesign-mobile-vue/es/radio/radio.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { defineComponent, toRefs, inject, computed, createVNode, mergeProps } from 'vue';
import { CheckCircleFilledIcon, CheckIcon } from 'tdesign-icons-vue-next';
import config from '../config.mjs';
import props from './props.mjs';
import useVModel from '../hooks/useVModel.mjs';
import { useFormDisabled } from '../form/hooks.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useContent, useTNodeJSX } from '../hooks/tnode.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'lodash-es';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';

var prefix = config.prefix;
var _Radio = defineComponent({
  name: "".concat(prefix, "-radio"),
  props: props,
  emits: ["update:checked", "update:modelValue", "change"],
  setup: function setup(props2, context) {
    var renderTNodeContent = useContent();
    var renderTNodeJSX = useTNodeJSX();
    var radioClass = usePrefixClass("radio");
    var _toRefs = toRefs(props2),
      checked = _toRefs.checked,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(checked, modelValue, props2.defaultChecked, props2.onChange, "checked"),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      innerChecked = _useVModel2[0],
      setInnerChecked = _useVModel2[1];
    var rootGroupProps = inject("rootGroupProps", {});
    var rootGroupValue = inject("rootGroupValue", {});
    var rootGroupChange = inject("rootGroupChange", void 0);
    var groupDisabled = computed(function () {
      return rootGroupProps === null || rootGroupProps === void 0 ? void 0 : rootGroupProps.disabled;
    });
    var isDisabled = useFormDisabled(groupDisabled);
    var radioChecked = computed(function () {
      return (rootGroupValue === null || rootGroupValue === void 0 ? void 0 : rootGroupValue.value) !== void 0 ? props2.value === (rootGroupValue === null || rootGroupValue === void 0 ? void 0 : rootGroupValue.value) : innerChecked.value;
    });
    var finalBorderless = computed(function () {
      if (props2.borderless == null && "borderless" in rootGroupProps) return rootGroupProps.borderless;
      return props2.borderless;
    });
    var finalPlacement = computed(function () {
      if (props2.placement == null && "placement" in rootGroupProps) return rootGroupProps.placement;
      return props2.placement || "left";
    });
    var finalAllowUncheck = computed(function () {
      return Boolean(props2.allowUncheck || (rootGroupProps === null || rootGroupProps === void 0 ? void 0 : rootGroupProps.allowUncheck));
    });
    var finalReadonly = computed(function () {
      return Boolean(props2.readonly || (rootGroupProps === null || rootGroupProps === void 0 ? void 0 : rootGroupProps.readonly));
    });
    var inputProps = computed(function () {
      return {
        name: rootGroupProps.name || props2.name,
        checked: radioChecked.value,
        disabled: isDisabled.value,
        readonly: finalReadonly.value,
        value: props2.value
      };
    });
    var iconContent = computed(function () {
      var _props2$icon;
      if (((_props2$icon = props2.icon) === null || _props2$icon === void 0 ? void 0 : _props2$icon.length) === 2) {
        var iconIndex = radioChecked.value ? 0 : 1;
        return props2.icon[iconIndex];
      }
      return null;
    });
    var radioClasses = computed(function () {
      return ["".concat(radioClass.value), "".concat(radioClass.value, "--").concat(finalPlacement.value), _defineProperty({}, "".concat(radioClass.value, "--block"), props2.block)];
    });
    var titleClasses = computed(function () {
      return ["".concat(radioClass.value, "__title"), _defineProperty({}, "".concat(radioClass.value, "__title--disabled"), isDisabled.value)];
    });
    var iconClass = computed(function () {
      return ["".concat(radioClass.value, "__icon"), "".concat(radioClass.value, "__icon--").concat(finalPlacement.value), _defineProperty(_defineProperty({}, "".concat(radioClass.value, "__icon--checked"), radioChecked.value), "".concat(radioClass.value, "__icon--disabled"), isDisabled.value)];
    });
    var radioContentChange = function radioContentChange(e) {
      e.stopPropagation();
      if (props2.contentDisabled) {
        return;
      }
      radioOrgChange(e);
    };
    var radioOrgChange = function radioOrgChange(e) {
      if (isDisabled.value || finalReadonly.value) {
        return;
      }
      if (rootGroupChange) {
        var value = finalAllowUncheck.value && radioChecked.value ? void 0 : props2.value;
        rootGroupChange(value, {
          e: e,
          name: props2.name
        });
      } else {
        var _value = finalAllowUncheck.value ? !radioChecked.value : true;
        setInnerChecked(_value, {
          e: e
        });
      }
    };
    var readerIcon = function readerIcon() {
      if (radioChecked.value) {
        if (props2.icon === "circle") {
          return createVNode(CheckCircleFilledIcon, {
            "class": "".concat(radioClass.value, "__icon-wrap")
          }, null);
        }
        if (props2.icon === "line") {
          return createVNode(CheckIcon, {
            "class": "".concat(radioClass.value, "__icon-wrap")
          }, null);
        }
        if (props2.icon === "dot") {
          return createVNode("div", {
            "class": ["".concat(radioClass.value, "__icon-").concat(props2.icon), _defineProperty({}, "".concat(radioClass.value, "__icon-").concat(props2.icon, "--disabled"), isDisabled.value)]
          }, null);
        }
      } else {
        if (props2.icon === "circle" || props2.icon === "dot") {
          return createVNode("div", {
            "class": ["".concat(radioClass.value, "__icon-circle"), _defineProperty({}, "".concat(radioClass.value, "__icon-circle--disabled"), isDisabled.value)]
          }, null);
        }
        if (props2.icon === "line") {
          return createVNode("div", {
            "class": "placeholder"
          }, null);
        }
      }
    };
    return function () {
      var label = renderTNodeContent("default", "label");
      var content = renderTNodeJSX("content");
      return createVNode("div", {
        "class": radioClasses.value,
        "onClick": radioOrgChange
      }, [createVNode("input", mergeProps({
        "type": "radio",
        "class": "".concat(radioClass.value, "__original")
      }, inputProps.value), null), createVNode("div", {
        "class": iconClass.value
      }, [readerIcon(), iconContent.value]), createVNode("div", {
        "class": "".concat(radioClass.value, "__content"),
        "onClick": radioContentChange
      }, [label && createVNode("span", {
        "class": titleClasses.value,
        "style": "-webkit-line-clamp: ".concat(props2.maxLabelRow)
      }, [label]), content && createVNode("div", {
        "class": ["".concat(radioClass.value, "__description"), _defineProperty({}, "".concat(radioClass.value, "__description"), isDisabled.value)],
        "style": "-webkit-line-clamp: ".concat(props2.maxContentRow)
      }, [content])]), !finalBorderless.value && createVNode("div", {
        "class": "".concat(radioClass.value, "__border ").concat(radioClass.value, "__border--").concat(finalPlacement.value)
      }, null)]);
    };
  }
});

export { _Radio as default };
