// tdesign-mobile-vue/lib/message/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  align: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "center"].includes(val);
    }
  },
  closeBtn: {
    type: [String, Boolean, Function],
    default: void 0
  },
  content: {
    type: [String, Function]
  },
  duration: {
    type: Number,
    default: 3e3
  },
  gap: {
    type: [String, Number, Boolean],
    default: 12
  },
  icon: {
    type: [Boolean, Function],
    default: true
  },
  link: {
    type: [String, Object, Function]
  },
  marquee: {
    type: [Boolean, Object],
    default: false
  },
  offset: {
    type: Array
  },
  single: {
    type: Boolean,
    default: true
  },
  theme: {
    type: String,
    default: "info",
    validator: function validator(val) {
      if (!val) return true;
      return ["info", "success", "warning", "error"].includes(val);
    }
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  zIndex: {
    type: Number
  },
  onCloseBtnClick: Function,
  onDurationEnd: Function,
  onLinkClick: Function
};

export { props as default };
