// tdesign-mobile-vue/cjs/calendar/utils.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function getMonthByOffset(date, offset) {
  var _date = new Date(date);
  _date.setMonth(_date.getMonth() + offset);
  _date.setDate(1);
  return _date;
}
function getYearByOffset(date, offset) {
  var _date = new Date(date);
  _date.setFullYear(_date.getFullYear() + offset);
  _date.setDate(1);
  return _date;
}
var getPrevMonth = function getPrevMonth(date) {
  return getMonthByOffset(date, -1);
};
var getNextMonth = function getNextMonth(date) {
  return getMonthByOffset(date, 1);
};
var getPrevYear = function getPrevYear(date) {
  return getYearByOffset(date, -1);
};
var getNextYear = function getNextYear(date) {
  return getYearByOffset(date, 1);
};

exports.getMonthByOffset = getMonthByOffset;
exports.getNextMonth = getNextMonth;
exports.getNextYear = getNextYear;
exports.getPrevMonth = getPrevMonth;
exports.getPrevYear = getPrevYear;
exports.getYearByOffset = getYearByOffset;
