// tdesign-mobile-vue/esm/input/input.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, ref, toRefs, inject, computed, nextTick, watch, createVNode } from 'vue';
import { CloseCircleFilledIcon, BrowseOffIcon, BrowseIcon } from 'tdesign-icons-vue-next';
import { isFunction } from 'lodash-es';
import config from '../config.js';
import InputProps from './props.js';
import '../shared/index.js';
import { FormItemInjectionKey } from '../form/const.js';
import { useFormDisabled } from '../form/hooks.js';
import useVModel from '../hooks/useVModel.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import useLengthLimit from '../hooks/useLengthLimit.js';
import { extendAPI } from '../shared/component.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/dom.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/hover.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../_common/js/log/index.js';
import '../_common/js/log/log.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _Input = defineComponent({
  name: "".concat(prefix, "-input"),
  props: _objectSpread(_objectSpread({}, InputProps), {}, {
    labelAlign: {
      type: String,
      default: "top"
    },
    pattern: {
      type: String
    },
    inputmode: {
      type: String,
      validator: function validator(val) {
        if (!val) return true;
        return ["search", "text", "none", "url", "tel", "email", "numeric", "decimal"].includes(val);
      }
    }
  }),
  setup: function setup(props, context) {
    var renderTNodeJSX = useTNodeJSX();
    var inputClass = usePrefixClass("input");
    var isDisabled = useFormDisabled();
    var inputRef = ref();
    var _toRefs = toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var status = props.status || "default";
    var renderType = ref(props.type);
    var focused = ref(false);
    var formItem = inject(FormItemInjectionKey, void 0);
    var inputClasses = computed(function () {
      return ["".concat(inputClass.value, "__control"), _defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(inputClass.value, "--").concat(props.align), props.align !== "left"), "".concat(inputClass.value, "--").concat(status), status), "".concat(inputClass.value, "__control--disabled"), isDisabled.value), "".concat(inputClass.value, "--password"), renderType.value === "password")];
    });
    var rootClasses = computed(function () {
      return [inputClass.value, _defineProperty(_defineProperty(_defineProperty({}, "".concat(inputClass.value, "--layout-").concat(props.layout), props.layout), "".concat(inputClass.value, "--center"), props.labelAlign === "center"), "".concat(inputClass.value, "--border"), !props.borderless)];
    });
    var showClear = computed(function () {
      if (isDisabled.value || props.readonly === true) return false;
      if (props.clearable && innerValue.value && String(innerValue.value).length > 0) {
        return props.clearTrigger === "always" || props.clearTrigger === "focus" && focused.value;
      }
      return false;
    });
    var limitParams = computed(function () {
      return {
        value: [void 0, null].includes(innerValue.value) ? void 0 : String(innerValue.value),
        maxlength: Number(props.maxlength),
        maxcharacter: props.maxcharacter,
        allowInputOverMax: props.allowInputOverMax,
        status: props.status,
        onValidate: props.onValidate
      };
    });
    var _useLengthLimit = useLengthLimit(limitParams),
      getValueByLimitNumber = _useLengthLimit.getValueByLimitNumber;
    var setInputValue = function setInputValue() {
      var v = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var input = inputRef.value;
      var sV = String(v);
      if (!input) {
        return;
      }
      if (input.value !== sV) {
        input.value = sV;
      }
    };
    var handleInput = function handleInput(e) {
      if (e instanceof InputEvent) {
        var checkInputType = e.inputType && e.inputType === "insertCompositionText";
        if (e.isComposing || checkInputType) return;
      }
      inputValueChangeHandle(e);
    };
    var inputValueChangeHandle = function inputValueChangeHandle(e) {
      var value2 = e.target.value;
      setInnerValue(getValueByLimitNumber(value2));
      nextTick(function () {
        return setInputValue(innerValue.value);
      });
    };
    var focus = function focus() {
      var _inputRef$value;
      focused.value = true;
      (_inputRef$value = inputRef.value) === null || _inputRef$value === void 0 || _inputRef$value.focus();
    };
    var blur = function blur() {
      var _inputRef$value2;
      focused.value = false;
      (_inputRef$value2 = inputRef.value) === null || _inputRef$value2 === void 0 || _inputRef$value2.blur();
    };
    extendAPI({
      focus: focus,
      blur: blur
    });
    var handleClear = function handleClear(e) {
      var _props$onClear;
      e.preventDefault();
      var val = props.type === "number" ? void 0 : "";
      setInnerValue(val);
      focus();
      (_props$onClear = props.onClear) === null || _props$onClear === void 0 || _props$onClear.call(props, {
        e: e
      });
    };
    var handleFocus = function handleFocus(e) {
      var _props$onFocus;
      focused.value = true;
      (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, innerValue.value, {
        e: e
      });
    };
    var handleBlur = function handleBlur(e) {
      var _props$onBlur2;
      focused.value = false;
      if (isFunction(props.format)) {
        setInnerValue(props.format(innerValue.value));
        nextTick(function () {
          var _props$onBlur;
          setInputValue(innerValue.value);
          (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, innerValue.value, {
            e: e
          });
          formItem === null || formItem === void 0 || formItem.handleBlur();
        });
        return;
      }
      (_props$onBlur2 = props.onBlur) === null || _props$onBlur2 === void 0 || _props$onBlur2.call(props, innerValue.value, {
        e: e
      });
      formItem === null || formItem === void 0 || formItem.handleBlur();
    };
    var handleCompositionend = function handleCompositionend(e) {
      inputValueChangeHandle(e);
    };
    var handlePwdIconClick = function handlePwdIconClick() {
      if (isDisabled.value) return;
      renderType.value = renderType.value === "password" ? "text" : "password";
    };
    watch(function () {
      return props.autofocus;
    }, function (v) {
      if (v === true) {
        nextTick(function () {
          focus();
        });
      }
    }, {
      immediate: true
    });
    watch(function () {
      return props.type;
    }, function (v) {
      renderType.value = v;
    }, {
      immediate: true
    });
    return function () {
      var renderPrefix = function renderPrefix() {
        var prefixIcon = renderTNodeJSX("prefixIcon");
        var label = renderTNodeJSX("label");
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--prefix")
        }, [prefixIcon && createVNode("div", {
          "class": "".concat(inputClass.value, "__icon--prefix")
        }, [prefixIcon]), createVNode("div", {
          "class": "".concat(inputClass.value, "__label")
        }, [label])]);
      };
      var renderClearable = function renderClearable() {
        if (showClear.value) {
          return createVNode("div", {
            "class": "".concat(inputClass.value, "__wrap--clearable-icon"),
            "onTouchend": handleClear
          }, [createVNode(CloseCircleFilledIcon, null, null)]);
        }
        return null;
      };
      var renderSuffix = function renderSuffix() {
        var suffix = renderTNodeJSX("suffix");
        if (!suffix) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--suffix")
        }, [suffix]);
      };
      var renderSuffixIcon = function renderSuffixIcon() {
        var suffixIcon = renderTNodeJSX("suffixIcon");
        if (props.type === "password") {
          if (renderType.value === "password") {
            suffixIcon = createVNode(BrowseOffIcon, {
              "onClick": handlePwdIconClick
            }, null);
          } else if (renderType.value === "text") {
            suffixIcon = createVNode(BrowseIcon, {
              "onClick": handlePwdIconClick
            }, null);
          }
        }
        if (!suffixIcon) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--suffix-icon")
        }, [suffixIcon]);
      };
      var renderExtra = function renderExtra() {
        return renderTNodeJSX("extra");
      };
      var renderTips = function renderTips() {
        var tips = renderTNodeJSX("tips");
        if (!tips) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__tips ").concat(inputClass.value, "--").concat(props.align)
        }, [tips]);
      };
      var inputAttrs = {
        ref: inputRef,
        class: inputClasses.value,
        value: innerValue.value,
        name: props.name,
        type: renderType.value,
        disabled: isDisabled.value,
        autocomplete: props.autocomplete ? "On" : "Off",
        placeholder: props.placeholder,
        readonly: props.readonly,
        pattern: props.pattern,
        inputmode: props.inputmode,
        spellcheck: props.spellCheck,
        enterkeyhint: props.enterkeyhint,
        style: {
          "--td-input-cursor-color": props.cursorColor
        },
        onFocus: handleFocus,
        onBlur: handleBlur,
        onInput: handleInput,
        onCompositionend: handleCompositionend
      };
      return createVNode("div", {
        "class": rootClasses.value
      }, [renderPrefix(), createVNode("div", {
        "class": "".concat(inputClass.value, "__wrap")
      }, [createVNode("div", {
        "class": "".concat(inputClass.value, "__content ").concat(inputClass.value, "--").concat(status || "default")
      }, [createVNode("input", inputAttrs, null), renderClearable(), renderSuffix(), renderSuffixIcon()]), renderTips()]), renderExtra()]);
    };
  }
});

export { _Input as default };
