// tdesign-mobile-vue/cjs/swiper/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  animation: {
    type: String,
    default: "slide",
    validator: function validator(val) {
      if (!val) return true;
      return ["slide"].includes(val);
    }
  },
  autoplay: {
    type: Boolean,
    default: true
  },
  current: {
    type: Number,
    default: void 0
  },
  modelValue: {
    type: Number,
    default: void 0
  },
  defaultCurrent: {
    type: Number,
    default: 0
  },
  direction: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  duration: {
    type: Number,
    default: 300
  },
  height: {
    type: [String, Number]
  },
  interval: {
    type: Number,
    default: 5e3
  },
  loop: {
    type: Boolean,
    default: true
  },
  navigation: {
    type: [Object, Function]
  },
  nextMargin: {
    type: [String, Number],
    default: 0
  },
  previousMargin: {
    type: [String, Number],
    default: 0
  },
  type: {
    type: String,
    default: "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "card"].includes(val);
    }
  },
  disabled: {
    type: Boolean,
    default: false
  },
  onChange: Function,
  onClick: Function
};

exports["default"] = props;
