// tdesign-mobile-vue/cjs/collapse/collapse-panel.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var cell_index = require('../cell/index.js');
var collapse_collapsePanelProps = require('./collapse-panel-props.js');
var config = require('../config.js');
var collapse_util = require('./util.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
require('../cell/cell.js');
require('../cell/props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-f84e6e17.js');
require('../shared/hover.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../cell/cell-group.js');
require('../cell/cell-group-props.js');
require('../shared/component.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _CollapsePanel = Vue.defineComponent({
  name: "".concat(prefix, "-collapse-panel"),
  components: {
    TCell: cell_index.Cell
  },
  props: collapse_collapsePanelProps["default"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var renderContent = hooks_tnode.useContent();
    var collapsePanelClass = hooks_useClass.usePrefixClass("collapse-panel");
    var parent = Vue.inject("collapse");
    var renderParentTNode = Vue.inject("renderParentTNode");
    var disabled = Vue.computed(function () {
      return (parent === null || parent === void 0 ? void 0 : parent.disabled.value) || props2.disabled;
    });
    var rootClass = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(collapsePanelClass.value), true), "".concat(collapsePanelClass.value, "--").concat(props2.placement), true), "".concat(collapsePanelClass.value, "--active"), isActive.value), "".concat(collapsePanelClass.value, "--disabled"), disabled.value);
    });
    var isActive = Vue.computed(function () {
      return collapse_util.findIndex(props2.value, parent === null || parent === void 0 ? void 0 : parent.activeValue.value) > -1;
    });
    var updatePanelValue = function updatePanelValue(args) {
      if (props2.value != null) {
        parent === null || parent === void 0 || parent.onPanelChange(props2.value, args);
      }
    };
    var handleClick = function handleClick(_ref3) {
      var e = _ref3.e;
      e === null || e === void 0 || e.stopPropagation();
      if (disabled.value) {
        return;
      }
      updatePanelValue({
        e: e
      });
    };
    Vue.onMounted(function () {
      if (parent !== null && parent !== void 0 && parent.defaultExpandAll) {
        updatePanelValue();
      }
    });
    var renderDefaultIcon = function renderDefaultIcon() {
      if (props2.placement === "bottom") {
        return isActive.value ? Vue.createVNode(tdesignIconsVueNext.ChevronUpIcon, null, null) : Vue.createVNode(tdesignIconsVueNext.ChevronDownIcon, null, null);
      }
      return isActive.value ? Vue.createVNode(tdesignIconsVueNext.ChevronDownIcon, null, null) : Vue.createVNode(tdesignIconsVueNext.ChevronUpIcon, null, null);
    };
    var panelExpandIcon = Vue.computed(function () {
      return slots.expandIcon || props2.expandIcon;
    });
    var renderRightIcon = function renderRightIcon() {
      var tNodeRender = panelExpandIcon.value === void 0 ? renderParentTNode : renderTNodeJSX;
      return Vue.createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__header-icon")
      }, [tNodeRender("expandIcon", renderDefaultIcon())]);
    };
    var renderPanelContent = function renderPanelContent() {
      var panelContent = renderContent("default", "content");
      if (props2.destroyOnCollapse && !isActive.value) {
        return null;
      }
      return Vue.createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__content")
      }, [panelContent]);
    };
    return function () {
      var headerContent = renderTNodeJSX("header");
      var noteContent = renderTNodeJSX("headerRightContent");
      var _leftIcon = renderTNodeJSX("headerLeftIcon");
      return Vue.createVNode("div", {
        "class": rootClass.value
      }, [Vue.createVNode(cell_index.Cell, {
        "class": ["".concat(collapsePanelClass.value, "__title"), "".concat(collapsePanelClass.value, "__header"), "".concat(collapsePanelClass.value, "__header--").concat(props2.placement), _defineProperty__default["default"]({}, "".concat(collapsePanelClass.value, "__header--expanded"), isActive.value)],
        "onClick": handleClick
      }, {
        leftIcon: function leftIcon() {
          return _leftIcon;
        },
        title: function title() {
          return headerContent;
        },
        note: function note() {
          return noteContent;
        },
        rightIcon: function rightIcon() {
          return renderRightIcon();
        }
      }), Vue.createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__body"),
        "style": {
          gridTemplateRows: isActive.value ? "1fr" : "0fr"
        }
      }, [Vue.createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__inner")
      }, [renderPanelContent()])])]);
    };
  }
});

exports["default"] = _CollapsePanel;
