// tdesign-mobile-vue/cjs/shared/util.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var hooks_useClass = require('../hooks/useClass.js');
var isNumber = require('../_chunks/dep-055d28fd.js');
require('vue');
require('../_chunks/dep-605cd804.js');
require('@babel/runtime/helpers/defineProperty');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function toCamel(str) {
  return str.replace(/^\S/, function (m) {
    return m.toUpperCase();
  });
}
var isBrowser = typeof window !== "undefined";
var sleep = function sleep(ms) {
  return new Promise(function (resolve) {
    return setTimeout(resolve, ms);
  });
};
var uniqueFactory = function uniqueFactory(compName) {
  var number = 0;
  var prefixClass = hooks_useClass.usePrefixClass();
  return function () {
    return "".concat(prefixClass.value, "-").concat(compName, "_").concat(number++);
  };
};
var convertUnit = function convertUnit(val) {
  if (val == null) return 0;
  return isNumber.isNumber(val) ? "".concat(val, "px") : val;
};
var reconvertUnit = function reconvertUnit(val) {
  if (val == null) return 0;
  return isNumber.isNumber(val) ? Number(val) : Number(val.slice(0, -2));
};
var formatNumber = function formatNumber(value) {
  var allowDecimal = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var allowNegative = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  if (allowDecimal) {
    var index = value.indexOf(".");
    if (index !== -1) {
      value = "".concat(value.slice(0, index + 1)).concat(value.slice(index).replace(/\./g, ""));
    }
  } else {
    var _value$split = value.split("."),
      _value$split2 = _slicedToArray__default["default"](_value$split, 1),
      _value$split2$ = _value$split2[0],
      splitValue = _value$split2$ === void 0 ? "" : _value$split2$;
    value = splitValue;
  }
  if (allowNegative) {
    var _index = value.indexOf("-");
    if (_index !== -1) {
      value = "".concat(value.slice(0, _index + 1)).concat(value.slice(_index).replace(/-/g, ""));
    }
  } else {
    value = value.replace(/-/g, "");
  }
  return value.replace(/[^\d.-]/g, "");
};

exports.convertUnit = convertUnit;
exports.formatNumber = formatNumber;
exports.isBrowser = isBrowser;
exports.reconvertUnit = reconvertUnit;
exports.sleep = sleep;
exports.toCamel = toCamel;
exports.uniqueFactory = uniqueFactory;
