// tdesign-mobile-vue/esm/guide/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var guideProps = {
  backButtonProps: {
    type: Object
  },
  counter: {
    type: Function
  },
  current: {
    type: Number,
    default: void 0
  },
  modelValue: {
    type: Number,
    default: void 0
  },
  defaultCurrent: {
    type: Number
  },
  finishButtonProps: {
    type: Object
  },
  hideCounter: Boolean,
  hideSkip: Boolean,
  highlightPadding: {
    type: Number,
    default: 8
  },
  mode: {
    type: String,
    default: "popover",
    validator: function validator(val) {
      if (!val) return true;
      return ["popover", "dialog"].includes(val);
    }
  },
  nextButtonProps: {
    type: Object
  },
  showOverlay: {
    type: Boolean,
    default: true
  },
  skipButtonProps: {
    type: Object
  },
  steps: {
    type: Array
  },
  zIndex: {
    type: Number,
    default: 999999
  },
  onBack: Function,
  onChange: Function,
  onFinish: Function,
  onNextStepClick: Function,
  onSkip: Function
};

export { guideProps as default };
