// tdesign-mobile-vue/cjs/picker/picker.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var config = require('../config.js');
var picker_props = require('./props.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_tnode = require('../hooks/tnode.js');
var picker_pickerItem = require('./picker-item.js');
var picker_utils = require('./utils.js');
var hooks_useClass = require('../hooks/useClass.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
var isFunction = require('../_chunks/dep-0bd122a0.js');
var get = require('../_chunks/dep-cfb75f1b.js');
var isString = require('../_chunks/dep-588e2e70.js');
var isBoolean = require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-f84e6e17.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('./picker.class.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('../shared/dom.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-01310648.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var getIndexFromColumns = function getIndexFromColumns(column, value, keys) {
  if (!value) return 0;
  return column === null || column === void 0 ? void 0 : column.findIndex(function (item) {
    var _keys$value;
    return get.get(item, (_keys$value = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value !== void 0 ? _keys$value : "value") === value;
  });
};
var _Picker = Vue.defineComponent({
  name: "".concat(prefix, "-picker"),
  components: {
    PickerItem: picker_pickerItem["default"]
  },
  props: picker_props["default"],
  emits: ["change", "cancel", "pick", "update:modelValue", "update:value"],
  setup: function setup(props, _ref) {
    var _pickerValue$value;
    var slots = _ref.slots;
    var pickerClass = hooks_useClass.usePrefixClass("picker");
    var _useConfig = configProvider_useConfig.useConfig("picker"),
      globalConfig = _useConfig.globalConfig;
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _toRefs = Vue.toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      _useVModel2$ = _useVModel2[0],
      pickerValue = _useVModel2$ === void 0 ? Vue.ref([]) : _useVModel2$,
      setPickerValue = _useVModel2[1];
    var keys = Vue.computed(function () {
      return props.keys;
    });
    var getDefaultText = function getDefaultText(prop, defaultText) {
      if (isString.isString(prop)) return prop;
      if (isBoolean.isBoolean(prop) && prop) return defaultText;
      return "";
    };
    var confirmButtonText = Vue.computed(function () {
      return getDefaultText(props.confirmBtn, globalConfig.value.confirm);
    });
    var cancelButtonText = Vue.computed(function () {
      return getDefaultText(props.cancelBtn, globalConfig.value.cancel);
    });
    var curValueArray = Vue.ref(((_pickerValue$value = pickerValue.value) === null || _pickerValue$value === void 0 ? void 0 : _pickerValue$value.map(function (item) {
      return item;
    })) || []);
    var realColumns = Vue.computed(function () {
      if (isFunction.isFunction(props.columns)) {
        var _columns = props.columns(curValueArray.value);
        return picker_utils.getPickerColumns(_columns);
      }
      return picker_utils.getPickerColumns(props.columns);
    });
    var curIndexArray = realColumns.value.map(function (item, index) {
      var _pickerValue$value2;
      return getIndexFromColumns(item, (_pickerValue$value2 = pickerValue.value) === null || _pickerValue$value2 === void 0 ? void 0 : _pickerValue$value2[index], keys.value);
    });
    var pickerItemInstanceArray = Vue.ref([]);
    var setPickerItemRef = function setPickerItemRef(item, index) {
      pickerItemInstanceArray.value[index] = item;
    };
    var handleConfirm = function handleConfirm(e) {
      var _props$onConfirm;
      var target = realColumns.value.map(function (item, index) {
        return item[curIndexArray[index]];
      });
      var label = target.map(function (item) {
        var _keys$value$label, _keys$value2;
        return get.get(item, (_keys$value$label = (_keys$value2 = keys.value) === null || _keys$value2 === void 0 ? void 0 : _keys$value2.label) !== null && _keys$value$label !== void 0 ? _keys$value$label : "label");
      });
      var value2 = target.map(function (item) {
        var _keys$value$value, _keys$value3;
        return get.get(item, (_keys$value$value = (_keys$value3 = keys.value) === null || _keys$value3 === void 0 ? void 0 : _keys$value3.value) !== null && _keys$value$value !== void 0 ? _keys$value$value : "value");
      });
      setPickerValue(value2);
      (_props$onConfirm = props.onConfirm) === null || _props$onConfirm === void 0 || _props$onConfirm.call(props, value2, {
        index: curIndexArray,
        label: label,
        e: e
      });
    };
    var handleCancel = function handleCancel(e) {
      var _props$onCancel;
      pickerItemInstanceArray.value.forEach(function (item, index) {
        item === null || item === void 0 || item.setIndex(curIndexArray[index]);
      });
      (_props$onCancel = props.onCancel) === null || _props$onCancel === void 0 || _props$onCancel.call(props, {
        e: e
      });
    };
    var handlePick = function handlePick(context, column) {
      var _realColumns$value, _keys$value$value2, _keys$value4, _props$onPick;
      var index = context.index;
      curIndexArray[column] = index;
      curValueArray.value[column] = get.get((_realColumns$value = realColumns.value) === null || _realColumns$value === void 0 ? void 0 : _realColumns$value[column][index], (_keys$value$value2 = (_keys$value4 = keys.value) === null || _keys$value4 === void 0 ? void 0 : _keys$value4.value) !== null && _keys$value$value2 !== void 0 ? _keys$value$value2 : "value");
      (_props$onPick = props.onPick) === null || _props$onPick === void 0 || _props$onPick.call(props, curValueArray.value, {
        index: index,
        column: column
      });
    };
    Vue.watch(pickerValue, function () {
      curValueArray.value = pickerValue.value.map(function (item) {
        return item;
      });
    });
    Vue.watch([realColumns, curValueArray], function () {
      realColumns.value.forEach(function (col, idx) {
        var _pickerItemInstanceAr;
        var index = col.findIndex(function (item) {
          var _keys$value$value3, _keys$value5;
          return get.get(item, (_keys$value$value3 = (_keys$value5 = keys.value) === null || _keys$value5 === void 0 ? void 0 : _keys$value5.value) !== null && _keys$value$value3 !== void 0 ? _keys$value$value3 : "value") === curValueArray.value[idx];
        });
        curIndexArray[idx] = index > -1 ? index : 0;
        (_pickerItemInstanceAr = pickerItemInstanceArray.value[idx]) === null || _pickerItemInstanceAr === void 0 || _pickerItemInstanceAr.setIndex(curIndexArray[idx]);
      });
    });
    Vue.provide("picker", _objectSpread({}, Vue.toRefs(props)));
    return function () {
      return Vue.createVNode("div", {
        "class": "".concat(pickerClass.value)
      }, [Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__toolbar")
      }, [cancelButtonText.value && Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__cancel"),
        "onClick": handleCancel
      }, [cancelButtonText.value]), Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__title")
      }, [props.title]), confirmButtonText.value && Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__confirm"),
        "onClick": handleConfirm
      }, [confirmButtonText.value])]), renderTNodeJSX("header"), Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__main")
      }, [realColumns.value.map(function (item, index) {
        var _pickerValue$value3;
        return Vue.createVNode("div", {
          "key": index,
          "class": "".concat(pickerClass.value, "-item__group")
        }, [Vue.createVNode(Vue.resolveComponent("picker-item"), {
          "ref": function ref(item2) {
            return setPickerItemRef(item2, index);
          },
          "options": item,
          "value": (_pickerValue$value3 = pickerValue.value) === null || _pickerValue$value3 === void 0 ? void 0 : _pickerValue$value3[index],
          "render-label": props.renderLabel,
          "onPick": function onPick($event) {
            return handlePick($event, index);
          },
          "swipeDuration": props.swipeDuration
        }, {
          option: slots.option || props.option
        })]);
      }), Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__mask ").concat(pickerClass.value, "__mask--top")
      }, null), Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__mask ").concat(pickerClass.value, "__mask--bottom")
      }, null), Vue.createVNode("div", {
        "class": "".concat(pickerClass.value, "__indicator")
      }, null)]), renderTNodeJSX("footer")]);
    };
  }
});

exports["default"] = _Picker;
