// tdesign-mobile-vue/cjs/picker/picker-item.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var config = require('../config.js');
var picker_picker_class = require('./picker.class.js');
var hooks_useClass = require('../hooks/useClass.js');
var get = require('../_chunks/dep-cfb75f1b.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('./utils.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-01310648.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var PickerItem = Vue.defineComponent({
  name: "".concat(prefix, "-picker-item"),
  props: {
    options: {
      type: Array
    },
    value: {
      type: [String, Number],
      default: void 0
    },
    renderLabel: {
      type: Function,
      default: void 0
    },
    onPick: {
      type: Function,
      default: void 0
    },
    swipeDuration: {
      type: [String, Number],
      default: 300
    }
  },
  emits: ["pick"],
  setup: function setup(props, context) {
    var pickerItemClass = hooks_useClass.usePrefixClass("picker-item");
    var pickerProps = Vue.inject("picker", void 0);
    var keys = Vue.computed(function () {
      var _pickerProps$keys;
      return pickerProps && ((_pickerProps$keys = pickerProps.keys) === null || _pickerProps$keys === void 0 ? void 0 : _pickerProps$keys.value);
    });
    var picker = null;
    var root = Vue.ref();
    var getIndexByValue = function getIndexByValue(val) {
      var defaultIndex = 0;
      if (val !== void 0) {
        var _props$options;
        defaultIndex = (_props$options = props.options) === null || _props$options === void 0 ? void 0 : _props$options.findIndex(function (item) {
          var _keys$value$value, _keys$value;
          return get.get(item, (_keys$value$value = (_keys$value = keys.value) === null || _keys$value === void 0 ? void 0 : _keys$value.value) !== null && _keys$value$value !== void 0 ? _keys$value$value : "value") === val;
        });
      }
      return defaultIndex < 0 ? 0 : defaultIndex;
    };
    var className = Vue.computed(function () {
      return "".concat(pickerItemClass.value);
    });
    var updatePickerWithNextTick = function updatePickerWithNextTick(index) {
      if (picker) {
        picker.updateItems();
        Vue.nextTick(function () {
          picker.updateIndex(index, {
            isChange: false
          });
        });
      }
    };
    var setIndex = function setIndex(index) {
      updatePickerWithNextTick(index);
    };
    var setValue = function setValue(value) {
      var index = getIndexByValue(value);
      updatePickerWithNextTick(index);
    };
    var setOptions = function setOptions() {
      var _picker;
      (_picker = picker) === null || _picker === void 0 || _picker.update();
    };
    var setUpdateItems = function setUpdateItems() {
      var _picker2;
      (_picker2 = picker) === null || _picker2 === void 0 || _picker2.updateItems();
    };
    context.expose({
      setIndex: setIndex,
      setValue: setValue,
      setOptions: setOptions,
      setUpdateItems: setUpdateItems
    });
    Vue.onMounted(function () {
      if (root.value) {
        picker = new picker_picker_class["default"]({
          el: root.value,
          defaultIndex: getIndexByValue(props.value) || 0,
          keys: keys.value,
          defaultPickerColumns: props.options,
          onChange: function onChange(index) {
            var _keys$value$value2, _keys$value2, _props$onPick;
            var curItem = props.options[index];
            var changeValue = {
              value: get.get(curItem, (_keys$value$value2 = (_keys$value2 = keys.value) === null || _keys$value2 === void 0 ? void 0 : _keys$value2.value) !== null && _keys$value$value2 !== void 0 ? _keys$value$value2 : "value"),
              index: index
            };
            (_props$onPick = props.onPick) === null || _props$onPick === void 0 || _props$onPick.call(props, changeValue);
          },
          swipeDuration: props.swipeDuration
        });
      }
    });
    Vue.watch(function () {
      return props.options;
    }, function () {
      var _picker3, _picker4;
      (_picker3 = picker) === null || _picker3 === void 0 || _picker3.updateOptions(props.options);
      (_picker4 = picker) === null || _picker4 === void 0 || _picker4.updateItems();
    }, {
      flush: "post",
      deep: true
    });
    return function () {
      return Vue.createVNode("ul", {
        "ref": root,
        "class": className.value
      }, [(props.options || []).map(function (option, index) {
        var _keys$value$disabled, _keys$value3, _keys$value$label, _keys$value4;
        return Vue.createVNode("li", {
          "key": index,
          "class": ["".concat(pickerItemClass.value, "__item"), _defineProperty__default["default"]({}, "".concat(pickerItemClass.value, "__item--disabled"), get.get(option, (_keys$value$disabled = (_keys$value3 = keys.value) === null || _keys$value3 === void 0 ? void 0 : _keys$value3.disabled) !== null && _keys$value$disabled !== void 0 ? _keys$value$disabled : "disabled"))]
        }, [context.slots.option ? context.slots.option(option, index) : Vue.createVNode(Vue.Fragment, null, [props.renderLabel ? props.renderLabel(option, index) : get.get(option, (_keys$value$label = (_keys$value4 = keys.value) === null || _keys$value4 === void 0 ? void 0 : _keys$value4.label) !== null && _keys$value$label !== void 0 ? _keys$value$label : "label")])]);
      })]);
    };
  }
});

exports["default"] = PickerItem;
