// tdesign-mobile-vue/cjs/picker/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var PickerProps = {
  cancelBtn: {
    type: [String, Boolean],
    default: true
  },
  columns: {
    type: [Array, Function],
    default: function _default() {
      return [];
    },
    required: true
  },
  confirmBtn: {
    type: [String, Boolean],
    default: true
  },
  footer: {
    type: Function
  },
  header: {
    type: Function
  },
  keys: {
    type: Object
  },
  option: {
    type: Function
  },
  renderLabel: {
    type: Function
  },
  swipeDuration: {
    type: [String, Number],
    default: 300
  },
  title: {
    type: String,
    default: ""
  },
  value: {
    type: Array,
    default: void 0
  },
  modelValue: {
    type: Array,
    default: void 0
  },
  defaultValue: {
    type: Array
  },
  onCancel: Function,
  onChange: Function,
  onConfirm: Function,
  onPick: Function
};

exports["default"] = PickerProps;
