// tdesign-mobile-vue/esm/shared/util.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { isNumber } from 'lodash-es';
import { usePrefixClass } from '../hooks/useClass.js';
import 'vue';
import '../config-provider/useConfig.js';
import '@babel/runtime/helpers/defineProperty';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';

function toCamel(str) {
  return str.replace(/^\S/, function (m) {
    return m.toUpperCase();
  });
}
var isBrowser = typeof window !== "undefined";
var sleep = function sleep(ms) {
  return new Promise(function (resolve) {
    return setTimeout(resolve, ms);
  });
};
var uniqueFactory = function uniqueFactory(compName) {
  var number = 0;
  var prefixClass = usePrefixClass();
  return function () {
    return "".concat(prefixClass.value, "-").concat(compName, "_").concat(number++);
  };
};
var convertUnit = function convertUnit(val) {
  if (val == null) return 0;
  return isNumber(val) ? "".concat(val, "px") : val;
};
var reconvertUnit = function reconvertUnit(val) {
  if (val == null) return 0;
  return isNumber(val) ? Number(val) : Number(val.slice(0, -2));
};
var formatNumber = function formatNumber(value) {
  var allowDecimal = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var allowNegative = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  if (allowDecimal) {
    var index = value.indexOf(".");
    if (index !== -1) {
      value = "".concat(value.slice(0, index + 1)).concat(value.slice(index).replace(/\./g, ""));
    }
  } else {
    var _value$split = value.split("."),
      _value$split2 = _slicedToArray(_value$split, 1),
      _value$split2$ = _value$split2[0],
      splitValue = _value$split2$ === void 0 ? "" : _value$split2$;
    value = splitValue;
  }
  if (allowNegative) {
    var _index = value.indexOf("-");
    if (_index !== -1) {
      value = "".concat(value.slice(0, _index + 1)).concat(value.slice(_index).replace(/-/g, ""));
    }
  } else {
    value = value.replace(/-/g, "");
  }
  return value.replace(/[^\d.-]/g, "");
};

export { convertUnit, formatNumber, isBrowser, reconvertUnit, sleep, toCamel, uniqueFactory };
