// tdesign-mobile-vue/lib/drawer/drawer.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, toRefs, ref, watch, createVNode, h } from 'vue';
import { Popup } from '../popup/index.js';
import config from '../config.js';
import props from './props.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '../popup/popup.js';
import '@babel/runtime/helpers/defineProperty';
import '@babel/runtime/helpers/slicedToArray';
import 'tdesign-icons-vue-next';
import '../popup/props.js';
import '../overlay/index.js';
import '../overlay/overlay.js';
import '../shared/dom.js';
import 'lodash-es';
import '../overlay/props.js';
import '../shared/component.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../hooks/useVModel.js';
import '../hooks/useLockScroll.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';
import '../hooks/useTeleport.js';
import '../shared/render-tnode.js';
import '../shared/util.js';

var prefix = config.prefix;
var _Drawer = defineComponent({
  name: "".concat(prefix, "-drawer"),
  props: props,
  emits: ["update:visible", "itemClick", "overlayClick"],
  setup: function setup(props2, context) {
    var renderTNodeJSX = useTNodeJSX();
    var drawerClass = usePrefixClass("drawer");
    var _toRefs = toRefs(props2),
      visible = _toRefs.visible,
      showOverlay = _toRefs.showOverlay;
    var open = ref(visible.value || false);
    watch(open, function () {
      context.emit("update:visible", open.value);
    });
    watch(visible, function () {
      open.value = visible.value;
    });
    var onItemClick = function onItemClick(index, item, context2) {
      var _props2$onItemClick;
      (_props2$onItemClick = props2.onItemClick) === null || _props2$onItemClick === void 0 || _props2$onItemClick.call(props2, index, item, context2);
    };
    var onVisibleChange = function onVisibleChange(visible2) {
      if (showOverlay.value) {
        var _props2$onOverlayClic;
        (_props2$onOverlayClic = props2.onOverlayClick) === null || _props2$onOverlayClic === void 0 || _props2$onOverlayClic.call(props2, {
          visible: visible2
        });
      }
    };
    var onClose = function onClose() {
      var _props2$onClose;
      (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, "overlay");
    };
    return function () {
      var attach = props2.attach,
        zIndex = props2.zIndex,
        closeOnOverlayClick = props2.closeOnOverlayClick,
        destroyOnClose = props2.destroyOnClose,
        items = props2.items,
        placement = props2.placement,
        showOverlay2 = props2.showOverlay,
        overlayProps = props2.overlayProps;
      var renderTitleNode = function renderTitleNode() {
        var titleNode = renderTNodeJSX("title");
        if (!titleNode) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(drawerClass.value, "__title")
        }, [titleNode]);
      };
      var renderFooterNode = function renderFooterNode() {
        var footerNode = renderTNodeJSX("footer");
        if (!footerNode) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(drawerClass.value, "__footer")
        }, [footerNode]);
      };
      return createVNode(Popup, {
        "modelValue": open.value,
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return open.value = $event;
        },
        "placement": placement,
        "attach": attach,
        "showOverlay": showOverlay2,
        "overlayProps": overlayProps,
        "zIndex": zIndex,
        "closeOnOverlayClick": closeOnOverlayClick,
        "destroyOnClose": destroyOnClose,
        "onVisibleChange": onVisibleChange,
        "onClose": onClose
      }, {
        default: function _default() {
          return [createVNode("div", {
            "class": "".concat(drawerClass.value)
          }, [renderTitleNode(), renderTNodeJSX("default"), createVNode("div", {
            "class": "".concat(drawerClass.value, "__sidebar")
          }, [(items || []).map(function (item, index) {
            return createVNode("div", {
              "key": item.title,
              "class": "".concat(drawerClass.value, "__sidebar-item"),
              "onClick": function onClick(e) {
                return onItemClick(index, item, {
                  e: e
                });
              }
            }, [item.icon && createVNode("span", {
              "class": "".concat(drawerClass.value, "__sidebar-item-icon")
            }, [item.icon(h)]), createVNode("div", {
              "class": "".concat(drawerClass.value, "__sidebar-item-title")
            }, [item.title])]);
          })]), renderFooterNode()])];
        }
      });
    };
  }
});

export { _Drawer as default };
