// tdesign-mobile-vue/lib/drawer/plugin.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { ref, createApp, h, nextTick } from 'vue';
import _Drawer from './drawer.js';
import { isBrowser } from '../shared/util.js';
import '../popup/index.js';
import '../popup/popup.js';
import '@babel/runtime/helpers/slicedToArray';
import 'tdesign-icons-vue-next';
import '../popup/props.js';
import '../overlay/index.js';
import '../overlay/overlay.js';
import '../shared/dom.js';
import 'lodash-es';
import '../config.js';
import '../overlay/props.js';
import '../hooks/useClass.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../hooks/tnode.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../shared/component.js';
import '../hooks/useVModel.js';
import '../hooks/useLockScroll.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';
import '../hooks/useTeleport.js';
import '../shared/render-tnode.js';
import './props.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Drawer = function Drawer(options) {
  if (!isBrowser) return;
  var root = document.createElement("div");
  document.body.appendChild(root);
  var visible = ref(false);
  var props = ref({});
  var destroyOnClose = ref(false);
  createApp(function () {
    return h(_Drawer, _objectSpread(_objectSpread({}, options), {}, {
      visible: visible.value,
      destroyOnClose: destroyOnClose.value
    }, props.value));
  }).mount(root);
  var handler = {
    destroy: function destroy() {
      destroyOnClose.value = true;
      nextTick(function () {
        visible.value = false;
        document.body.removeChild(root);
      });
    },
    hide: function hide() {
      visible.value = false;
    },
    show: function show() {
      visible.value = true;
    },
    update: function update(options2) {
      props.value = options2;
    }
  };
  return handler;
};
Drawer.install = function (app) {
  app.config.globalProperties.$drawer = Drawer;
};
var DrawerPlugin = Drawer;

export { DrawerPlugin as default };
