// tdesign-mobile-vue/esm/side-bar/side-bar-item.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, getCurrentInstance, inject, computed, onUnmounted, createVNode, mergeProps } from 'vue';
import props from './side-bar-item-props.js';
import { Badge } from '../badge/index.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import config from '../config.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '../badge/badge.js';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash-es';
import '../badge/props.js';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/dom.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/hover.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../badge/style';
import '../badge/type.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';

var prefix = config.prefix;
var _SideBarItem = defineComponent({
  name: "".concat(prefix, "-side-bar-item"),
  props: props,
  setup: function setup(props2) {
    var sideBarItemClass = usePrefixClass("side-bar-item");
    var renderTNodeJSX = useTNodeJSX();
    var internalInstance = getCurrentInstance();
    var proxy = internalInstance.proxy;
    var sideBarProvide = inject("sideBarProvide", void 0);
    sideBarProvide.relation(proxy);
    var isActive = computed(function () {
      return proxy.value === sideBarProvide.currentValue.value;
    });
    var rootClassName = computed(function () {
      return [sideBarItemClass.value, _defineProperty({}, "".concat(sideBarItemClass.value, "--active"), isActive.value), _defineProperty({}, "".concat(sideBarItemClass.value, "--disabled"), props2.disabled)];
    });
    var onClick = function onClick(e) {
      if (props2.disabled) return;
      sideBarProvide.onClickItem(proxy.value, props2.label);
    };
    onUnmounted(function () {
      sideBarProvide.removeRelation(proxy);
    });
    return function () {
      var renderIconNode = function renderIconNode() {
        var iconNode = renderTNodeJSX("icon");
        if (!iconNode) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(sideBarItemClass.value, "__icon")
        }, [iconNode]);
      };
      return createVNode("div", {
        "class": rootClassName.value,
        "onClick": onClick
      }, [isActive.value && createVNode("div", null, [createVNode("div", {
        "class": "".concat(sideBarItemClass.value, "__line")
      }, null), createVNode("div", {
        "class": "".concat(sideBarItemClass.value, "__prefix")
      }, null), createVNode("div", {
        "class": "".concat(sideBarItemClass.value, "__suffix")
      }, null)]), renderIconNode(), props2.badgeProps ? createVNode(Badge, mergeProps(props2.badgeProps, {
        "content": props2.label
      }), null) : createVNode("div", null, [props2.label])]);
    };
  }
});

export { _SideBarItem as default };
