// tdesign-mobile-vue/cjs/pull-down-refresh/pull-down-refresh.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var core = require('@vueuse/core');
var pullDownRefresh_props = require('./props.js');
var shared_dom = require('../shared/dom.js');
var config = require('../config.js');
var loading_index = require('../loading/index.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_tnode = require('../hooks/tnode.js');
var pullDownRefresh_useTouch = require('./useTouch.js');
var hooks_useClass = require('../hooks/useClass.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
var debounce = require('../_chunks/dep-4326ca38.js');
var shared_util = require('../shared/util.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/defineProperty');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-055d28fd.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../loading/plugin.js');
require('../shared/component.js');
require('../_chunks/dep-2da1d545.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

var prefix = config["default"].prefix;
var statusName = ["pulling", "loosing", "loading", "success", "initial"];
var _PullDownRefresh = Vue.defineComponent({
  name: "".concat(prefix, "-pull-down-refresh"),
  components: {
    TLoading: loading_index.Loading
  },
  props: pullDownRefresh_props["default"],
  emits: ["refresh", "timeout", "scrolltolower", "update:value", "update:modelValue"],
  setup: function setup(props) {
    var pullDownRefreshClass = hooks_useClass.usePrefixClass("pull-down-refresh");
    var _useConfig = configProvider_useConfig.useConfig("pullDownRefresh"),
      globalConfig = _useConfig.globalConfig;
    var renderContent = hooks_tnode.useContent();
    var timer = null;
    var distance = Vue.ref(0);
    var afterLoading = Vue.ref(false);
    var _toRefs = Vue.toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      loading = _useVModel2[0],
      setLoading = _useVModel2[1];
    var loadingText = Vue.computed(function () {
      var _props$loadingTexts;
      var index = statusName.indexOf(status.value);
      var loadingTexts = ((_props$loadingTexts = props.loadingTexts) === null || _props$loadingTexts === void 0 ? void 0 : _props$loadingTexts.length) > 0 ? props.loadingTexts : globalConfig.value.loadingTexts;
      return index >= 0 ? loadingTexts[index] : "";
    });
    var touch = pullDownRefresh_useTouch.useTouch();
    var loadingBar = Vue.ref();
    var maxBar = Vue.ref();
    var _useElementSize = core.useElementSize(loadingBar),
      loadingBarHeight = _useElementSize.height;
    var _useElementSize2 = core.useElementSize(maxBar),
      maxBarHeight = _useElementSize2.height;
    var actualLoadingBarHeight = Vue.ref(0);
    var touchDir;
    var touchThreshold = 5;
    Vue.watch([loading, loadingBarHeight], function (_ref, _ref2) {
      var _ref3 = _slicedToArray__default["default"](_ref, 1),
        val = _ref3[0];
      var _ref4 = _slicedToArray__default["default"](_ref2, 1),
        prevVal = _ref4[0];
      if (val) {
        distance.value = loadingBarHeight.value;
      }
      if (!val && prevVal) {
        afterLoading.value = true;
      }
    }, {
      immediate: true
    });
    var status = Vue.computed(function () {
      if (afterLoading.value) {
        return "success";
      }
      if (!loading.value && distance.value === 0) {
        return "initial";
      }
      if (distance.value < shared_util.reconvertUnit(props.loadingBarHeight)) {
        return "pulling";
      }
      if (loading.value) {
        return "loading";
      }
      return "loosing";
    });
    Vue.watch(status, function (newVal) {
      if (newVal === "success") {
        setTimeout(function () {
          distance.value = 0;
        }, 300);
      }
    });
    var isTouchable = function isTouchable() {
      return loading.value || props.disabled;
    };
    var onTouchStart = function onTouchStart(e) {
      if (isTouchable()) return;
      if (!pullDownRefresh_useTouch.isReachTop(e)) return;
      clearTimeout(timer);
      timer = null;
      distance.value = 0;
      touch.start(e);
      touchDir = 0;
    };
    var onTouchMove = function onTouchMove(e) {
      if (isTouchable()) return;
      if (!pullDownRefresh_useTouch.isReachTop(e)) return;
      touch.move(e);
      var diffY = touch.diffY,
        diffX = touch.diffX;
      var absX = Math.abs(diffX.value);
      var absY = Math.abs(diffY.value);
      if (!touchDir && absX < touchThreshold && absY < touchThreshold) {
        return;
      }
      if (!touchDir && absX < absY) {
        touchDir = -1;
      } else if (!touchDir && absX >= absY) {
        touchDir = 1;
      }
      if (touchDir === 1) return;
      actualLoadingBarHeight.value = diffY.value;
      var nextDistance = pullDownRefresh_useTouch.easeDistance(diffY.value, loadingBarHeight.value);
      if (diffY.value > 0) {
        shared_dom.preventDefault(e, false);
      }
      if (nextDistance >= 0 && nextDistance < maxBarHeight.value) {
        distance.value = nextDistance;
      }
    };
    var onTouchEnd = function onTouchEnd(e) {
      if (isTouchable()) return;
      if (!pullDownRefresh_useTouch.isReachTop(e)) return;
      if (status.value === "loosing") {
        var _props$onRefresh;
        distance.value = loadingBarHeight.value;
        setLoading(true);
        (_props$onRefresh = props.onRefresh) === null || _props$onRefresh === void 0 || _props$onRefresh.call(props);
        timer = setTimeout(function () {
          if (loading.value) {
            var _props$onTimeout;
            (_props$onTimeout = props.onTimeout) === null || _props$onTimeout === void 0 || _props$onTimeout.call(props);
            setLoading(false);
          }
        }, props.refreshTimeout);
      } else {
        distance.value = 0;
      }
    };
    var onReachBottom = debounce.debounce(function () {
      var scrollTop = document.documentElement.scrollTop || document.body.scrollTop;
      var _document$documentEle = document.documentElement,
        clientHeight = _document$documentEle.clientHeight,
        scrollHeight = _document$documentEle.scrollHeight;
      var distance2 = 20;
      if (scrollTop + clientHeight >= scrollHeight - distance2) {
        var _props$onScrolltolowe;
        (_props$onScrolltolowe = props.onScrolltolower) === null || _props$onScrolltolowe === void 0 || _props$onScrolltolowe.call(props);
      }
    }, 300, {
      leading: true,
      trailing: false
    });
    var onTransitionEnd = function onTransitionEnd() {
      if (afterLoading.value) {
        afterLoading.value = false;
      }
    };
    var trackStyle = Vue.computed(function () {
      return {
        transform: "translate3d(0, ".concat(distance.value, "px, 0)")
      };
    });
    var heightDiff = (shared_util.reconvertUnit(props.maxBarHeight) - shared_util.reconvertUnit(props.loadingBarHeight)) / 2;
    var loadingBarStyles = Vue.computed(function () {
      return {
        transform: "translateY(".concat(heightDiff, "px)"),
        height: "".concat(actualLoadingBarHeight.value, "px"),
        maxHeight: shared_util.convertUnit(props.loadingBarHeight)
      };
    });
    var maxBarStyles = Vue.computed(function () {
      return {
        height: shared_util.convertUnit(props.maxBarHeight)
      };
    });
    Vue.onMounted(function () {
      window.addEventListener("scroll", onReachBottom);
    });
    Vue.onUnmounted(function () {
      clearTimeout(timer);
      window.removeEventListener("scroll", onReachBottom);
    });
    var renderLoading = function renderLoading() {
      if (status.value === "loading") {
        return Vue.createVNode(Vue.resolveComponent("t-loading"), Vue.mergeProps({
          "size": "24px",
          "text": loadingText.value
        }, props.loadingProps), null);
      }
      return Vue.createVNode("div", {
        "class": "".concat(pullDownRefreshClass.value, "__text")
      }, [loadingText.value]);
    };
    return function () {
      var content = renderContent("default", "content");
      var className = "".concat(pullDownRefreshClass.value, "__track");
      if (status.value !== "pulling") {
        className = "".concat(className, " ").concat(pullDownRefreshClass.value, "__track--loosing");
      }
      return Vue.createVNode("div", {
        "class": pullDownRefreshClass.value
      }, [Vue.createVNode("div", {
        "class": className,
        "style": trackStyle.value,
        "onTouchstart": onTouchStart,
        "onTouchmove": onTouchMove,
        "onTouchend": onTouchEnd,
        "onTouchcancel": onTouchEnd,
        "onTransitionend": onTransitionEnd
      }, [Vue.createVNode("div", {
        "ref": maxBar,
        "class": "".concat(pullDownRefreshClass.value, "__tips"),
        "style": maxBarStyles.value
      }, [Vue.createVNode("div", {
        "ref": loadingBar,
        "class": "".concat(pullDownRefreshClass.value, "__loading"),
        "style": loadingBarStyles.value
      }, [renderLoading()])]), content])]);
    };
  }
});

exports["default"] = _PullDownRefresh;
