// tdesign-mobile-vue/es/grid/grid.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, toRefs, computed, provide, createVNode } from 'vue';
import config from '../config.mjs';
import props from './props.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../config-provider/useConfig.mjs';
import 'lodash-es';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';

var prefix = config.prefix;
var _Grid = defineComponent({
  name: "".concat(prefix, "-grid"),
  props: props,
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var gridClass = usePrefixClass("grid");
    var _toRefs = toRefs(props2),
      column = _toRefs.column,
      gutter = _toRefs.gutter,
      border = _toRefs.border,
      align = _toRefs.align,
      hover = _toRefs.hover;
    var rootStyle = computed(function () {
      if (column.value === 0) return [];
      var ans = ["padding: ".concat(gutter.value, "px;"), "grid-template-columns: repeat(".concat(column.value, ", 1fr)"), "grid-gap: ".concat(gutter.value, "px")];
      return ans;
    });
    provide("grid", {
      column: column,
      border: border,
      align: align,
      gutter: gutter,
      hover: hover
    });
    var classes = computed(function () {
      return ["".concat(gridClass.value), _defineProperty(_defineProperty(_defineProperty({}, "".concat(gridClass.value, "--card"), props2.theme === "card"), "".concat(gridClass.value, "--auto-size"), props2.column === 0), "".concat(gridClass.value, "--bordered"), props2.border && !props2.gutter)];
    });
    return function () {
      var _slots$default;
      return createVNode("div", {
        "class": classes.value,
        "style": rootStyle.value
      }, [(_slots$default = slots.default) === null || _slots$default === void 0 ? void 0 : _slots$default.call(slots)]);
    };
  }
});

export { _Grid as default };
