// tdesign-mobile-vue/cjs/steps/step-item.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var steps_stepItemProps = require('./step-item-props.js');
var config = require('../config.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _StepItem = Vue.defineComponent({
  name: "".concat(prefix, "-step-item"),
  props: steps_stepItemProps["default"],
  setup: function setup(props2, context) {
    var stepItemClass = hooks_useClass.usePrefixClass("step-item");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var internalInstance = Vue.getCurrentInstance();
    var proxy = internalInstance.proxy;
    var stepsProvide = Vue.inject("stepsProvide", void 0);
    stepsProvide.relation(proxy);
    var iconDefault = {
      check: Vue.h(tdesignIconsVueNext.CheckIcon),
      close: Vue.h(tdesignIconsVueNext.CloseIcon)
    };
    var index = Vue.computed(function () {
      return stepsProvide.state.children.indexOf(proxy);
    });
    var isLastChild = Vue.computed(function () {
      return index.value === (stepsProvide.sequence === "positive" ? stepsProvide.state.children.length - 1 : 0);
    });
    var theme = Vue.computed(function () {
      return stepsProvide.theme;
    });
    var dot = Vue.computed(function () {
      return theme.value === "dot";
    });
    var icon = Vue.computed(function () {
      return props2.icon || context.slots.icon;
    });
    var current = Vue.computed(function () {
      return stepsProvide.current.value || stepsProvide.defaultCurrent || 0;
    });
    var stepsStatus = Vue.computed(function () {
      return stepsProvide.currentStatus;
    });
    var readonly = Vue.computed(function () {
      return stepsProvide.readonly;
    });
    var currentStatus = Vue.computed(function () {
      var status = props2.status;
      if (status !== "default") return status;
      if (index.value === current.value) return stepsStatus.value;
      if (index.value < current.value) return "finish";
      return status;
    });
    var rootClassName = Vue.computed(function () {
      return [stepItemClass.value, "".concat(stepItemClass.value, "--").concat(stepsProvide.layout), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "--default"), readonly.value), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "--").concat(currentStatus.value), currentStatus.value)];
    });
    var iconWrapperClassName = Vue.computed(function () {
      return ["".concat(stepItemClass.value, "__anchor"), "".concat(stepItemClass.value, "__anchor--").concat(stepsProvide.layout)];
    });
    var dotClass = Vue.computed(function () {
      return ["".concat(stepItemClass.value, "__dot"), "".concat(stepItemClass.value, "__dot--").concat(currentStatus.value)];
    });
    var iconClassName = Vue.computed(function () {
      return [_defineProperty__default["default"]({}, "".concat(stepItemClass.value, "__icon"), icon.value), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "__icon--").concat(currentStatus.value), icon.value), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "__circle"), !icon.value), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "__circle--").concat(currentStatus.value), !icon.value)];
    });
    var contentClass = Vue.computed(function () {
      return ["".concat(stepItemClass.value, "__content"), "".concat(stepItemClass.value, "__content--").concat(stepsProvide.layout), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "__content--").concat(stepsProvide.layout), isLastChild.value), _defineProperty__default["default"]({}, "".concat(stepItemClass.value, "__content--last"), isLastChild.value)];
    });
    var tilteClass = Vue.computed(function () {
      return ["".concat(stepItemClass.value, "__title"), "".concat(stepItemClass.value, "__title--").concat(currentStatus.value), "".concat(stepItemClass.value, "__title--").concat(stepsProvide.layout)];
    });
    var descriptionClass = Vue.computed(function () {
      return ["".concat(stepItemClass.value, "__description"), "".concat(stepItemClass.value, "__description--").concat(currentStatus.value), "".concat(stepItemClass.value, "__description--").concat(stepsProvide.layout)];
    });
    var extraClass = Vue.computed(function () {
      return ["".concat(stepItemClass.value, "__extra"), "".concat(stepItemClass.value, "__extra--").concat(currentStatus.value), "".concat(stepItemClass.value, "__extra--").concat(stepsProvide.layout)];
    });
    var separatorClass = Vue.computed(function () {
      return [stepItemClass.value, "".concat(stepItemClass.value, "__line"), "".concat(stepItemClass.value, "__line--").concat(currentStatus.value), "".concat(stepItemClass.value, "__line--").concat(stepsProvide.sequence), "".concat(stepItemClass.value, "__line--").concat(stepsProvide.layout), "".concat(stepItemClass.value, "__line--").concat(stepsProvide.theme)];
    });
    var onClickIcon = function onClickIcon(e) {
      if (!readonly.value) {
        stepsProvide.onClickItem(index.value, current.value, e);
      }
    };
    Vue.onUnmounted(function () {
      stepsProvide.removeRelation(proxy);
    });
    return function () {
      var renderIconContent = function renderIconContent() {
        if (icon.value) {
          return renderTNodeJSX("icon") || null;
        }
        if (currentStatus.value === "error") {
          return iconDefault.close;
        }
        if (currentStatus.value === "finish") {
          return iconDefault.check;
        }
        return index.value + 1;
      };
      return Vue.createVNode("div", {
        "class": rootClassName.value,
        "onClick": onClickIcon
      }, [Vue.createVNode("div", {
        "class": iconWrapperClassName.value
      }, [dot.value ? Vue.createVNode("div", {
        "class": dotClass.value
      }, null) : Vue.createVNode("div", {
        "class": iconClassName.value
      }, [renderIconContent()])]), Vue.createVNode("div", {
        "class": contentClass.value
      }, [Vue.createVNode("div", {
        "class": tilteClass.value
      }, [renderTNodeJSX("title"), renderTNodeJSX("titleRight")]), Vue.createVNode("div", {
        "class": descriptionClass.value
      }, [renderTNodeJSX("content")]), Vue.createVNode("div", {
        "class": extraClass.value
      }, [renderTNodeJSX("extra")])]), !isLastChild.value && Vue.createVNode("div", {
        "class": separatorClass.value
      }, null)]);
    };
  }
});

exports["default"] = _StepItem;
