// tdesign-mobile-vue/cjs/popover/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var PopoverProps = {
  closeOnClickOutside: {
    type: Boolean,
    default: true
  },
  content: {
    type: [String, Function]
  },
  default: {
    type: [String, Function]
  },
  placement: {
    type: String,
    default: "top",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "left", "right", "bottom", "top-left", "top-right", "bottom-left", "bottom-right", "left-top", "left-bottom", "right-top", "right-bottom"].includes(val);
    }
  },
  showArrow: {
    type: Boolean,
    default: true
  },
  theme: {
    type: String,
    default: "dark",
    validator: function validator(val) {
      if (!val) return true;
      return ["dark", "light", "brand", "success", "warning", "error"].includes(val);
    }
  },
  triggerElement: {
    type: [String, Function]
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  onVisibleChange: Function
};

exports["default"] = PopoverProps;
