// tdesign-mobile-vue/lib/_common/docs/plugins/changelog-to-json/libs.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { WEB_COMPONENT_MAP, MOBILE_COMPONENT_MAP, CHAT_COMPONENT_MAP, NON_PASCAL_CASE_NAMES } from '../../../js/components.js';

var convert2PascalCase = function convert2PascalCase(name) {
  return name.split("-").map(function (part) {
    return part.charAt(0).toUpperCase() + part.slice(1);
  }).join("");
};
var COMPONENT_MAPS = {
  web: WEB_COMPONENT_MAP,
  mobile: MOBILE_COMPONENT_MAP,
  chat: CHAT_COMPONENT_MAP
};
var mapToParentName = function mapToParentName(name, platform) {
  var targetMap = COMPONENT_MAPS[platform];
  var found = Object.entries(targetMap).find(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2);
      _ref2[0];
      var values = _ref2[1];
    return values.includes(name);
  });
  if (found) return NON_PASCAL_CASE_NAMES[found[0]] || convert2PascalCase(found[0]);
  return null;
};

export { COMPONENT_MAPS, convert2PascalCase, mapToParentName };
