// tdesign-mobile-vue/cjs/_chunks/dep-63743cf9.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

var _typeof = require('@babel/runtime/helpers/typeof');
var _commonjsHelpers = require('./dep-7220ff76.js');
var dayjs = require('dayjs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var ar = {exports: {}};

(function (module, exports) {
  !function (e, t) {
    module.exports = t(dayjs__default["default"]) ;
  }(_commonjsHelpers.commonjsGlobal, function (e) {

    function t(e) {
      return e && "object" == _typeof__default["default"](e) && "default" in e ? e : {
        default: e
      };
    }
    var n = t(e),
      r = "يناير_فبراير_مارس_أبريل_مايو_يونيو_يوليو_أغسطس_سبتمبر_أكتوبر_نوفمبر_ديسمبر".split("_"),
      d = {
        1: "١",
        2: "٢",
        3: "٣",
        4: "٤",
        5: "٥",
        6: "٦",
        7: "٧",
        8: "٨",
        9: "٩",
        0: "٠"
      },
      _ = {
        "١": "1",
        "٢": "2",
        "٣": "3",
        "٤": "4",
        "٥": "5",
        "٦": "6",
        "٧": "7",
        "٨": "8",
        "٩": "9",
        "٠": "0"
      },
      o = /[١٢٣٤٥٦٧٨٩٠]/g,
      i = /،/g,
      a = /\d/g,
      s = /,/g,
      u = {
        name: "ar",
        weekdays: "الأحد_الإثنين_الثلاثاء_الأربعاء_الخميس_الجمعة_السبت".split("_"),
        weekdaysShort: "أحد_إثنين_ثلاثاء_أربعاء_خميس_جمعة_سبت".split("_"),
        weekdaysMin: "ح_ن_ث_ر_خ_ج_س".split("_"),
        months: r,
        monthsShort: r,
        weekStart: 6,
        meridiem: function meridiem(e) {
          return e > 12 ? "م" : "ص";
        },
        relativeTime: {
          future: "بعد %s",
          past: "منذ %s",
          s: "ثانية واحدة",
          m: "دقيقة واحدة",
          mm: "%d دقائق",
          h: "ساعة واحدة",
          hh: "%d ساعات",
          d: "يوم واحد",
          dd: "%d أيام",
          M: "شهر واحد",
          MM: "%d أشهر",
          y: "عام واحد",
          yy: "%d أعوام"
        },
        preparse: function preparse(e) {
          return e.replace(o, function (e) {
            return _[e];
          }).replace(i, ",");
        },
        postformat: function postformat(e) {
          return e.replace(a, function (e) {
            return d[e];
          }).replace(s, "،");
        },
        ordinal: function ordinal(e) {
          return e;
        },
        formats: {
          LT: "HH:mm",
          LTS: "HH:mm:ss",
          L: "D/‏M/‏YYYY",
          LL: "D MMMM YYYY",
          LLL: "D MMMM YYYY HH:mm",
          LLLL: "dddd D MMMM YYYY HH:mm"
        }
      };
    return n.default.locale(u, null, !0), u;
  });
})(ar);
