// tdesign-mobile-vue/cjs/_chunks/dep-605cd804.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var configProvider_context = require('./dep-6f53fb1b.js');
var isString = require('./dep-588e2e70.js');
var isFunction = require('./dep-0bd122a0.js');
var _baseClone = require('./dep-8d30e8d1.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

/** Used to compose bitmasks for cloning. */
var CLONE_DEEP_FLAG = 1,
  CLONE_SYMBOLS_FLAG = 4;

/**
 * This method is like `_.clone` except that it recursively clones `value`.
 *
 * @static
 * @memberOf _
 * @since 1.0.0
 * @category Lang
 * @param {*} value The value to recursively clone.
 * @returns {*} Returns the deep cloned value.
 * @see _.clone
 * @example
 *
 * var objects = [{ 'a': 1 }, { 'b': 2 }];
 *
 * var deep = _.cloneDeep(objects);
 * console.log(deep[0] === objects[0]);
 * // => false
 */
function cloneDeep(value) {
  return _baseClone.baseClone(value, CLONE_DEEP_FLAG | CLONE_SYMBOLS_FLAG);
}

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var globalConfigCopy = Vue.ref();
var t = function t(pattern) {
  for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }
  var data = args[0];
  if (isString.isString(pattern)) {
    if (!data) return pattern;
    var regular = /\{\s*([\w-]+)\s*\}/g;
    var translated = pattern.replace(regular, function (match, key) {
      return String(data[key]);
    });
    return translated;
  }
  if (isFunction.isFunction(pattern)) {
    if (!args.length) return pattern(Vue.h);
    return pattern.apply(void 0, args);
  }
  return "";
};
function useConfig() {
  var componentName = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : void 0;
  var componentLocale = arguments.length > 1 ? arguments[1] : undefined;
  var injectGlobalConfig = Vue.getCurrentInstance() ? Vue.inject(configProvider_context.configProviderInjectKey, null) : globalConfigCopy;
  var mergedGlobalConfig = Vue.computed(function () {
    return (injectGlobalConfig === null || injectGlobalConfig === void 0 ? void 0 : injectGlobalConfig.value) || configProvider_context.defaultGlobalConfig;
  });
  var globalConfig = Vue.computed(function () {
    return Object.assign({}, mergedGlobalConfig.value[componentName], componentLocale);
  });
  var classPrefix = Vue.computed(function () {
    return mergedGlobalConfig.value.classPrefix;
  });
  return {
    t: t,
    global: globalConfig,
    globalConfig: globalConfig,
    classPrefix: classPrefix
  };
}
var provideConfig = function provideConfig(props) {
  var defaultData = cloneDeep(configProvider_context.defaultGlobalConfig);
  var mergedGlobalConfig = Vue.computed(function () {
    return _objectSpread({}, configProvider_context.mergeWith(defaultData, props.globalConfig));
  });
  Vue.provide(configProvider_context.configProviderInjectKey, mergedGlobalConfig);
  if (!globalConfigCopy.value) {
    globalConfigCopy.value = mergedGlobalConfig.value;
  }
  return mergedGlobalConfig;
};

exports.cloneDeep = cloneDeep;
exports.provideConfig = provideConfig;
exports.t = t;
exports.useConfig = useConfig;
