// tdesign-mobile-vue/es/checkbox/checkbox-group.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _typeof } from '../_chunks/dep-4055ef5c.mjs';
import { _ as _toConsumableArray } from '../_chunks/dep-fb069758.mjs';
import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { defineComponent, toRefs, computed, provide, createVNode, mergeProps } from 'vue';
import { get } from 'lodash-es';
import config from '../config.mjs';
import props from './checkbox-group-props.mjs';
import '../_common/js/log/index.mjs';
import _Checkbox from './checkbox.mjs';
import useVModel from '../hooks/useVModel.mjs';
import './hooks/index.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { getOptions } from './hooks/getOptions.mjs';
import { setCheckAllStatus } from './hooks/setCheckAllStatus.mjs';
import log from '../_common/js/log/log.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'tdesign-icons-vue-next';
import './props.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../form/hooks.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../hooks/useChildSlots.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _CheckboxGroup = defineComponent({
  name: "".concat(prefix, "-checkbox-group"),
  components: {
    Checkbox: _Checkbox
  },
  props: props,
  emits: ["update:value", "update:modelValue", "change"],
  setup: function setup(props2, context) {
    var checkboxGroupClass = usePrefixClass("checkbox-group");
    var renderTNodeJSX = useTNodeJSX();
    var isArray = Array.isArray;
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var keys = computed(function () {
      return props2.keys;
    });
    var optionList = getOptions(props2, context.slots);
    var checkedSet = computed(function () {
      if (isArray(innerValue.value)) {
        return new Set(innerValue.value);
      }
      return /* @__PURE__ */new Set();
    });
    var checkAllStatus = setCheckAllStatus(optionList, innerValue, checkedSet);
    var maxExceeded = computed(function () {
      return props2.max !== void 0 && innerValue.value.length === props2.max;
    });
    var onCheckedChange = function onCheckedChange(p) {
      var checked = p.checked,
        checkAll = p.checkAll,
        e = p.e;
      if (checkAll) {
        onCheckAllChange(checked, {
          e: e
        });
      } else {
        handleCheckboxChange(p);
      }
    };
    var handleCheckboxChange = function handleCheckboxChange(data) {
      var currentValue = data.option.value;
      if (isArray(innerValue.value) || !innerValue.value) {
        if (currentValue === void 0) {
          return;
        }
        var val = innerValue.value ? _toConsumableArray(innerValue.value) : [];
        if (data.checked) {
          val.push(currentValue);
        } else {
          var i = val.indexOf(currentValue);
          val.splice(i, 1);
        }
        setInnerValue(val, {
          e: data.e,
          current: data.option.value,
          type: data.checked ? "check" : "uncheck"
        });
      } else {
        log.warn("CheckboxGroup", "`value` must be an array, instead of ".concat(_typeof(innerValue.value)));
      }
    };
    var getAllCheckboxValue = function getAllCheckboxValue() {
      var val = /* @__PURE__ */new Set();
      for (var i = 0, len = optionList.value.length; i < len; i++) {
        var item = optionList.value[i];
        if (item.checkAll) continue;
        if (item.value === void 0) continue;
        val.add(item.value);
        if (maxExceeded.value) break;
      }
      return _toConsumableArray(val);
    };
    var onCheckAllChange = function onCheckAllChange(checked, context2) {
      var value2 = checked ? getAllCheckboxValue() : [];
      setInnerValue(value2, {
        e: context2.e,
        type: checked ? "check" : "uncheck",
        current: void 0
      });
    };
    provide("checkboxGroup", _objectSpread(_objectSpread({}, toRefs(props2)), {}, {
      innerValue: innerValue,
      checkAllStatus: checkAllStatus,
      checkedSet: checkedSet,
      maxExceeded: maxExceeded,
      onCheckedChange: onCheckedChange
    }));
    return function () {
      var checkboxNode = function checkboxNode() {
        return createVNode("span", null, [optionList.value.map(function (item, idx) {
          var _keys$value$value, _keys$value, _keys$value$label, _keys$value2, _keys$value$value2, _keys$value3, _keys$value$disabled, _keys$value4;
          return createVNode(_Checkbox, mergeProps(item, {
            "key": "".concat(get(item, (_keys$value$value = (_keys$value = keys.value) === null || _keys$value === void 0 ? void 0 : _keys$value.value) !== null && _keys$value$value !== void 0 ? _keys$value$value : "value", "")).concat(idx),
            "label": get(item, (_keys$value$label = (_keys$value2 = keys.value) === null || _keys$value2 === void 0 ? void 0 : _keys$value2.label) !== null && _keys$value$label !== void 0 ? _keys$value$label : "label", item.text || ""),
            "value": get(item, (_keys$value$value2 = (_keys$value3 = keys.value) === null || _keys$value3 === void 0 ? void 0 : _keys$value3.value) !== null && _keys$value$value2 !== void 0 ? _keys$value$value2 : "value"),
            "disabled": get(item, (_keys$value$disabled = (_keys$value4 = keys.value) === null || _keys$value4 === void 0 ? void 0 : _keys$value4.disabled) !== null && _keys$value$disabled !== void 0 ? _keys$value$disabled : "disabled")
          }), null);
        })]);
      };
      return createVNode("div", {
        "class": "".concat(checkboxGroupClass.value)
      }, [!(props2.options && props2.options.length) ? renderTNodeJSX("default") : checkboxNode()]);
    };
  }
});

export { _CheckboxGroup as default };
