// tdesign-mobile-vue/es/link/link.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, computed, createVNode } from 'vue';
import config from '../config.mjs';
import props from './props.mjs';
import { useTNodeJSX, useContent } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useFormDisabled } from '../form/hooks.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import 'lodash-es';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';

var prefix = config.prefix;
var _Link = defineComponent({
  name: "".concat(prefix, "-link"),
  props: props,
  setup: function setup(props2) {
    var linkClass = usePrefixClass("link");
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var isDisabled = useFormDisabled();
    var linkClasses = computed(function () {
      return [linkClass.value, "".concat(linkClass.value, "--").concat(props2.theme || "default"), "".concat(linkClass.value, "--").concat(props2.size || "medium"), _defineProperty(_defineProperty(_defineProperty({}, "".concat(linkClass.value, "--disabled"), props2.disabled), "".concat(linkClass.value, "--underline"), props2.underline), "".concat(linkClass.value, "--hover"), props2.hover && !props2.disabled)];
    });
    var handleClick = function handleClick(e) {
      var _props2$onClick;
      if (props2.disabled) return;
      (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, e);
    };
    var renderContent = function renderContent() {
      var content = renderTNodeContent("default", "content");
      return content ? createVNode("span", {
        "class": ["".concat(linkClass.value, "__content")]
      }, [content]) : null;
    };
    var renderPrefixIcon = function renderPrefixIcon() {
      var prefixIcon = renderTNodeJSX("prefixIcon");
      return prefixIcon ? createVNode("span", {
        "class": ["".concat(linkClass.value, "__prefix-icon")]
      }, [prefixIcon]) : null;
    };
    var renderSuffixIcon = function renderSuffixIcon() {
      var suffixIcon = renderTNodeJSX("suffixIcon");
      return suffixIcon ? createVNode("span", {
        "class": ["".concat(linkClass.value, "__suffix-icon")]
      }, [suffixIcon]) : null;
    };
    return function () {
      return createVNode("a", {
        "class": linkClasses.value,
        "aria-disabled": isDisabled.value,
        "target": props2.target,
        "href": isDisabled.value || !props2.href ? void 0 : props2.href,
        "onClick": handleClick
      }, [renderPrefixIcon(), renderContent(), renderSuffixIcon()]);
    };
  }
});

export { _Link as default };
