// tdesign-mobile-vue/cjs/shared/useClickAway/index.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var core = require('@vueuse/core');
var shared_util = require('../util.js');
var isArray = require('../../_chunks/dep-3cf5587c.js');
require('@babel/runtime/helpers/slicedToArray');
require('../../hooks/useClass.js');
require('vue');
require('../../_chunks/dep-605cd804.js');
require('@babel/runtime/helpers/defineProperty');
require('../../_chunks/dep-6f53fb1b.js');
require('../../_common/js/global-config/mobile/default-config.js');
require('../../_common/js/global-config/mobile/locale/zh_CN.js');
require('../../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../../_chunks/dep-7220ff76.js');
require('dayjs');
require('../../_chunks/dep-05850262.js');
require('../../_chunks/dep-e9a63584.js');
require('../../_chunks/dep-53383176.js');
require('../../_chunks/dep-f84e6e17.js');
require('../../_chunks/dep-0c11c462.js');
require('../../_chunks/dep-0bd122a0.js');
require('../../_chunks/dep-b999c3d4.js');
require('../../_chunks/dep-159c7a94.js');
require('../../_chunks/dep-ea269588.js');
require('../../_chunks/dep-b633ec9e.js');
require('../../_chunks/dep-7aa6341b.js');
require('../../_chunks/dep-b8e8d654.js');
require('../../_chunks/dep-61ce8c0d.js');
require('../../_chunks/dep-5cf8bf50.js');
require('../../_chunks/dep-2457b474.js');
require('../../_chunks/dep-4b045875.js');
require('../../_chunks/dep-a6a4534e.js');
require('../../_chunks/dep-fef48622.js');
require('../../_chunks/dep-588e2e70.js');
require('../../_chunks/dep-8d30e8d1.js');
require('../../_chunks/dep-dd93e905.js');
require('../../_chunks/dep-d1ba5719.js');
require('../../_chunks/dep-055d28fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);

function useClickAway(target, fn) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  if (!shared_util.isBrowser) return;
  var _options$eventName = options.eventName,
    eventName = _options$eventName === void 0 ? "touchstart" : _options$eventName,
    _options$capture = options.capture,
    capture = _options$capture === void 0 ? true : _options$capture,
    _options$ignore = options.ignore,
    ignore = _options$ignore === void 0 ? [] : _options$ignore,
    _options$detectIframe = options.detectIframe,
    detectIframe = _options$detectIframe === void 0 ? false : _options$detectIframe;
  var listenerOptions = {
    passive: true,
    capture: capture
  };
  var checkElInEvtTarget = function checkElInEvtTarget(el, event) {
    return el && (event.target === el || event.composedPath().includes(el));
  };
  var shouldIgnore = function shouldIgnore(event) {
    return ignore.some(function (ignoreTarget) {
      return checkElInEvtTarget(core.unrefElement(ignoreTarget), event);
    });
  };
  var eventHandler = function eventHandler(event) {
    var targets = isArray.isArray(target) ? _toConsumableArray__default["default"](target) : [target];
    var shouldTrigger = targets.every(function (targetItem) {
      return !checkElInEvtTarget(core.unrefElement(targetItem), event);
    });
    if (!shouldTrigger) return;
    shouldTrigger = !shouldIgnore(event);
    if (!shouldTrigger) return;
    fn(event);
  };
  var cleanup = [core.useEventListener(window, eventName, eventHandler, listenerOptions)];
  if (detectIframe) {
    cleanup.push(core.useEventListener(window, "blur", function (event) {
      setTimeout(function () {
        var targets = isArray.isArray(target) ? _toConsumableArray__default["default"](target) : [target];
        if (targets.every(function (targetItem) {
          var _window$document$acti;
          var el = core.unrefElement(targetItem);
          return ((_window$document$acti = window.document.activeElement) === null || _window$document$acti === void 0 ? void 0 : _window$document$acti.tagName) === "IFRAME" && !(el !== null && el !== void 0 && el.contains(window.document.activeElement));
        })) {
          fn(event);
        }
      }, 0);
    }));
  }
  return function () {
    return cleanup.forEach(function (clean) {
      return clean();
    });
  };
}

exports.useClickAway = useClickAway;
