// tdesign-mobile-vue/es/loading/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  attach: {
    type: [String, Function],
    default: ""
  },
  content: {
    type: [String, Function]
  },
  default: {
    type: [String, Function]
  },
  delay: {
    type: Number,
    default: 0
  },
  duration: {
    type: Number,
    default: 800
  },
  fullscreen: Boolean,
  indicator: {
    type: [Boolean, Function],
    default: true
  },
  inheritColor: Boolean,
  layout: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  loading: {
    type: Boolean,
    default: true
  },
  pause: Boolean,
  reverse: Boolean,
  size: {
    type: String,
    default: "20px"
  },
  text: {
    type: [String, Function]
  },
  theme: {
    type: String,
    default: "circular",
    validator: function validator(val) {
      if (!val) return true;
      return ["circular", "spinner", "dots"].includes(val);
    }
  }
};

export { props as default };
