// tdesign-mobile-vue/cjs/dropdown-menu/dropdown-menu.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var core = require('@vueuse/core');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var dropdownMenu_context = require('./context.js');
var hooks_useExpose = require('../hooks/useExpose.js');
var dropdownMenu_domUtils = require('./dom-utils.js');
var hooks_tnode = require('../hooks/tnode.js');
var dropdownMenu_props = require('./props.js');
var hooks_useClass = require('../hooks/useClass.js');
var get = require('../_chunks/dep-cfb75f1b.js');
var camelCase = require('../_chunks/dep-a65592ed.js');
var isFunction = require('../_chunks/dep-0bd122a0.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-01310648.js');
require('../_chunks/dep-09a28f29.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var _DropdownMenu = Vue.defineComponent({
  name: "".concat(prefix, "-dropdown-menu"),
  components: {
    CaretDownSmallIcon: tdesignIconsVueNext.CaretDownSmallIcon,
    CaretUpSmallIcon: tdesignIconsVueNext.CaretUpSmallIcon
  },
  props: _objectSpread({
    onMenuOpened: Function,
    onMenuClosed: Function
  }, dropdownMenu_props["default"]),
  emits: ["menuOpened", "menuClosed"],
  setup: function setup(props, _ref) {
    var slots = _ref.slots;
    var dropdownMenuClass = hooks_useClass.usePrefixClass("dropdown-menu");
    var renderContent = hooks_tnode.useContent();
    var state = Vue.reactive({
      activeId: null,
      barRect: {},
      childCount: 0,
      itemsLabel: []
    });
    var menuItems = Vue.ref([]);
    var updateItems = function updateItems() {
      if (slots.default) {
        var itemName = "".concat(prefix, "-dropdown-item");
        var children = slots.default();
        menuItems.value = children.filter(function (child) {
          var _child$type;
          var childTypeName = child === null || child === void 0 || (_child$type = child.type) === null || _child$type === void 0 ? void 0 : _child$type.name;
          return (childTypeName === null || childTypeName === void 0 ? void 0 : childTypeName.includes) && childTypeName.includes(itemName);
        });
      }
    };
    Vue.watch(function () {
      var _slots$default;
      return slots === null || slots === void 0 || (_slots$default = slots.default) === null || _slots$default === void 0 ? void 0 : _slots$default.call(slots);
    }, updateItems, {
      deep: true,
      immediate: true
    });
    var menuTitles = Vue.computed(function () {
      var _menuItems$value;
      return (_menuItems$value = menuItems.value) === null || _menuItems$value === void 0 ? void 0 : _menuItems$value.map(function (item, index) {
        var _keys$label2;
        var _item$props = item.props,
          keys = _item$props.keys,
          label = _item$props.label,
          value = _item$props.value,
          modelValue = _item$props.modelValue,
          defaultValue = _item$props.defaultValue,
          disabled = _item$props.disabled,
          options = _item$props.options,
          icon = _item$props.icon;
        var currentValue = value || modelValue || defaultValue;
        var target = options === null || options === void 0 ? void 0 : options.find(function (item2) {
          var _keys$value;
          return get.get(item2, (_keys$value = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value !== void 0 ? _keys$value : "value") === currentValue;
        });
        if (state.itemsLabel.length < index + 1) {
          var _keys$label;
          var targetLabel = target && get.get(target, (_keys$label = keys === null || keys === void 0 ? void 0 : keys.label) !== null && _keys$label !== void 0 ? _keys$label : "label") || "";
          var computedLabel = label || targetLabel;
          state.itemsLabel.push(computedLabel);
          return {
            labelProps: label,
            label: computedLabel,
            disabled: disabled !== void 0 && disabled !== false,
            icon: icon
          };
        }
        return {
          labelProps: label,
          label: label || get.get(target, (_keys$label2 = keys === null || keys === void 0 ? void 0 : keys.label) !== null && _keys$label2 !== void 0 ? _keys$label2 : "label"),
          disabled: disabled !== void 0 && disabled !== false,
          icon: icon
        };
      });
    });
    Vue.provide("dropdownMenuProps", props);
    Vue.provide("dropdownMenuState", state);
    var classes = Vue.computed(function () {
      return ["".concat(dropdownMenuClass.value)];
    });
    var refBar = Vue.ref();
    var styleBarItem = Vue.computed(function () {
      return function (item, idx) {
        return ["".concat(dropdownMenuClass.value, "__item"), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(dropdownMenuClass.value, "__item--disabled"), item.disabled), "".concat(dropdownMenuClass.value, "__item--active"), idx === state.activeId)];
      };
    });
    var styleIcon = Vue.computed(function () {
      return function (item, idx) {
        return ["".concat(dropdownMenuClass.value, "__icon"), _defineProperty__default["default"]({}, "".concat(dropdownMenuClass.value, "__icon--active"), idx === state.activeId)];
      };
    });
    var expandMenu = function expandMenu(item, idx) {
      var _props$onMenuOpened;
      var disabled = item.disabled;
      if (disabled) return;
      if (state.activeId === idx) {
        var _props$onMenuClosed;
        collapseMenu();
        (_props$onMenuClosed = props.onMenuClosed) === null || _props$onMenuClosed === void 0 || _props$onMenuClosed.call(props, {
          trigger: "menu"
        });
        return;
      }
      (_props$onMenuOpened = props.onMenuOpened) === null || _props$onMenuOpened === void 0 || _props$onMenuOpened.call(props, "menuOpened");
      state.activeId = idx;
      state.itemsLabel[idx] = item.label;
      var bar = refBar.value;
      var barRect = dropdownMenu_domUtils.findRelativeRect(bar);
      state.barRect = barRect;
      var container = dropdownMenu_domUtils.findRelativeContainer(bar) || document.body;
      dropdownMenu_context.context.recordMenuExpanded(container, control, dropdownMenu_context.DropdownMenuExpandState.expanded);
    };
    var collapseMenu = function collapseMenu() {
      menuTitles.value.forEach(function (_item, index) {
        menuTitles.value[index].label = state.itemsLabel[index];
      });
      state.activeId = null;
      var bar = refBar.value;
      var container = dropdownMenu_domUtils.findRelativeContainer(bar) || document.body;
      dropdownMenu_context.context.recordMenuExpanded(container, control, dropdownMenu_context.DropdownMenuExpandState.collapsed);
    };
    core.onClickOutside(refBar, function () {
      var _props$onMenuClosed2;
      if (state.activeId === null) return;
      collapseMenu();
      (_props$onMenuClosed2 = props.onMenuClosed) === null || _props$onMenuClosed2 === void 0 || _props$onMenuClosed2.call(props, {
        trigger: "outside"
      });
    });
    var control = {
      expandMenu: expandMenu,
      collapseMenu: collapseMenu,
      emitEvents: function emitEvents(emit, trigger) {
        var eventHandler = props["on".concat(camelCase.camelCase(emit))];
        eventHandler === null || eventHandler === void 0 || eventHandler(trigger);
      }
    };
    Vue.provide("dropdownMenuControl", control);
    hooks_useExpose["default"]({
      toggle: function toggle(idx) {
        if (idx != null) {
          var item = menuTitles.value[idx];
          expandMenu(item, idx);
        } else {
          collapseMenu();
        }
      }
    });
    hooks_useExpose["default"]({
      expandMenu: expandMenu,
      collapseMenu: collapseMenu
    });
    return function () {
      var defaultSlot = renderContent("default", "content");
      var renderDropdownItemIcon = function renderDropdownItemIcon(item, idx) {
        var icon = item.icon;
        var isActive = idx === state.activeId;
        var iconClass = [styleIcon.value(item, idx)];
        if (icon) {
          var isArray = Array.isArray(icon);
          var isTwoIcons = isArray && icon.length === 2;
          var selectedIcon;
          if (isTwoIcons) {
            selectedIcon = isActive ? icon[0] : icon[1];
          } else {
            selectedIcon = isArray ? icon[0] : icon;
          }
          var iconContent = isFunction.isFunction(selectedIcon) ? selectedIcon() : selectedIcon;
          return Vue.createVNode("div", {
            "class": iconClass
          }, [iconContent]);
        }
        return props.direction === "down" ? Vue.createVNode(Vue.resolveComponent("caret-down-small-icon"), {
          "class": iconClass
        }, null) : Vue.createVNode(Vue.resolveComponent("caret-up-small-icon"), {
          "class": iconClass
        }, null);
      };
      return Vue.createVNode("div", {
        "ref": refBar,
        "class": classes.value
      }, [(menuTitles.value || []).map(function (item, idx) {
        return Vue.createVNode("div", {
          "class": styleBarItem.value(item, idx),
          "onClick": function onClick() {
            return expandMenu(item, idx);
          }
        }, [Vue.createVNode("div", {
          "class": "".concat(dropdownMenuClass.value, "__title")
        }, [item.labelProps || item.label]), renderDropdownItemIcon(item, idx)]);
      }), defaultSlot]);
    };
  }
});

exports["default"] = _DropdownMenu;
