// tdesign-mobile-vue/cjs/dropdown-menu/dropdown-item-props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  disabled: Boolean,
  footer: {
    type: Function
  },
  icon: {
    type: [Array, Function],
    default: void 0
  },
  keys: {
    type: Object
  },
  label: {
    type: String,
    default: ""
  },
  multiple: Boolean,
  options: {
    type: Array,
    default: function _default() {
      return [];
    }
  },
  optionsColumns: {
    type: [String, Number],
    default: 1
  },
  placement: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right"].includes(val);
    }
  },
  value: {
    type: [String, Number, Array],
    default: void 0
  },
  modelValue: {
    type: [String, Number, Array],
    default: void 0
  },
  defaultValue: {
    type: [String, Number, Array],
    default: void 0
  },
  onChange: Function,
  onConfirm: Function,
  onReset: Function
};

exports["default"] = props;
