// tdesign-mobile-vue/esm/dialog/index.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { ref, reactive, nextTick, createApp, h } from 'vue';
import Dialog from './dialog.js';
import '../shared/index.js';
import './style';
import { isBrowser } from '../shared/util.js';
import '@babel/runtime/helpers/toConsumableArray';
import 'tdesign-icons-vue-next';
import 'lodash-es';
import '../button/index.js';
import '../button/button.js';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/setStyle.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import '../config.js';
import '../hooks/useClass.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../loading/icon/spinner.js';
import '../loading/props.js';
import '../hooks/tnode.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../shared/dom.js';
import '../loading/style';
import '../loading/type.js';
import '../loading/plugin.js';
import '../shared/component.js';
import '../button/props.js';
import '../form/hooks.js';
import '../shared/hover.js';
import '../button/style';
import '../button/type.js';
import '../shared/functions.js';
import '../shared/render.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '../popup/index.js';
import '../popup/popup.js';
import '../popup/props.js';
import '../overlay/index.js';
import '../overlay/overlay.js';
import '../overlay/props.js';
import '../overlay/style';
import '../hooks/useVModel.js';
import '../hooks/useLockScroll.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';
import '../hooks/useTeleport.js';
import '../popup/style';
import '../popup/type.js';
import './props.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var DialogPropsDefault = {
  title: "",
  content: "",
  confirmBtn: "",
  cancelBtn: "",
  visible: false,
  zIndex: 2500,
  showOverlay: true,
  width: "320px",
  closeOnOverlayClick: false
};
var propsFn = ["onConfirm", "onCancel", "onOverlayClick", "onClose", "onClosed"];
function create(options) {
  if (!isBrowser) return;
  var root = document.createElement("div");
  document.body.appendChild(root);
  var props = ref({});
  var propsObject = _objectSpread(_objectSpread({}, DialogPropsDefault), typeof options === "string" ? {
    content: options
  } : options);
  function callFn(fnType, context) {
    var fn = props.value[fnType] || propsObject[fnType];
    typeof fn === "function" && fn(context);
  }
  var params = reactive(_objectSpread(_objectSpread({}, propsObject), {}, {
    onConfirm: function onConfirm(context) {
      callFn("onConfirm", context);
      params.visible = false;
    },
    onCancel: function onCancel(context) {
      callFn("onCancel", context);
      params.visible = false;
    },
    onOverlayClick: function onOverlayClick(context) {
      callFn("onOverlayClick", context);
      params.visible = false;
    },
    onClose: function onClose(context) {
      callFn("onClose", context);
      params.visible = false;
    },
    onClosed: function onClosed() {
      callFn("onClosed");
      nextTick(function () {
        params.destroyOnClose && app.unmount();
      });
    }
  }));
  var app = createApp(function () {
    return h(Dialog, params);
  });
  app.mount(root);
  var handler = {
    destroy: function destroy() {
      params.destroyOnClose = true;
      nextTick(function () {
        params.visible = false;
        root.remove();
      });
    },
    hide: function hide() {
      params.visible = false;
    },
    show: function show() {
      params.visible = true;
    },
    update: function update(options2) {
      if (typeof options2 === "string") {
        params.content = options2;
      } else {
        for (var key in options2) {
          if (propsFn.includes(key)) {
            props.value[key] = options2[key];
          } else {
            params[key] = options2[key];
          }
        }
      }
    }
  };
  nextTick(function () {
    return params.visible = true;
  });
  return handler;
}
["show", "alert", "confirm"].forEach(function (type) {
  Dialog[type] = function (options) {
    var props = {
      content: ""
    };
    if (typeof options === "string") {
      props.content = options;
    } else {
      props = _objectSpread(_objectSpread({}, props), options);
    }
    if (type === "alert") {
      props.cancelBtn = null;
    }
    return create(props);
  };
});
Dialog.install = function (app) {
  var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "";
  app.component(name || Dialog.name, Dialog);
  app.config.globalProperties.$dialog = Dialog;
  app.provide("$dialog", Dialog);
};
var DialogPlugin = Dialog;

export { DialogPlugin, DialogPlugin as default };
