// tdesign-mobile-vue/cjs/overlay/overlay.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var shared_dom = require('../shared/dom.js');
var config = require('../config.js');
var overlay_props = require('./props.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _Overlay = Vue.defineComponent({
  name: "".concat(prefix, "-overlay"),
  props: overlay_props["default"],
  setup: function setup(props) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var overlayClass = hooks_useClass.usePrefixClass("overlay");
    var overlayClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(overlayClass.value), true), "".concat(overlayClass.value, "--active"), props.visible);
    });
    var overlayStyles = Vue.computed(function () {
      var arr = [];
      if (props.customStyle) {
        arr.push(props.customStyle);
      }
      if (props.zIndex) {
        arr.push("z-index: ".concat(props.zIndex));
      }
      if (props.duration) {
        arr.push("transition-duration: ".concat(props.duration, "ms"));
      }
      if (props.backgroundColor) {
        arr.push("background-color: ".concat(props.backgroundColor));
      }
      return arr.join("; ");
    });
    var handleTouchMove = function handleTouchMove(e) {
      if (props.preventScrollThrough) {
        shared_dom.preventDefault(e, true);
      }
    };
    var handleClick = function handleClick(e) {
      var _props$onClick;
      (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, {
        e: e
      });
    };
    return function () {
      return Vue.createVNode(Vue.Transition, {
        "name": overlayClass.value
      }, {
        default: function _default() {
          return [Vue.withDirectives(Vue.createVNode("div", {
            "class": overlayClasses.value,
            "style": overlayStyles.value,
            "onClick": handleClick,
            "onTouchmove": handleTouchMove
          }, [renderTNodeJSX("default")]), [[Vue.vShow, props.visible]])];
        }
      });
    };
  }
});

exports["default"] = _Overlay;
