// tdesign-mobile-vue/es/dropdown-menu/dropdown-item.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { isVNode, defineComponent, ref, toRefs, inject, computed, onBeforeMount, reactive, nextTick, watch, createVNode } from 'vue';
import { get } from 'lodash-es';
import { RadioGroup, Radio } from '../radio/index.mjs';
import config from '../config.mjs';
import { Button } from '../button/index.mjs';
import { Popup } from '../popup/index.mjs';
import { CheckboxGroup, Checkbox } from '../checkbox/index.mjs';
import '../shared/index.mjs';
import props from './dropdown-item-props.mjs';
import useVModel from '../hooks/useVModel.mjs';
import { useContent, useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { uniqueFactory } from '../shared/util.mjs';
import { useConfig } from '../config-provider/useConfig.mjs';
import { getWindowSize } from '../shared/dom.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../radio/radio.mjs';
import 'tdesign-icons-vue-next';
import '../radio/props.mjs';
import '../form/hooks.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../radio/radio-group.mjs';
import '../radio/radio-group-props.mjs';
import './style/css.mjs';
import '../radio/type.mjs';
import '../shared/component.mjs';
import '../shared/functions.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../button/button.mjs';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../button/props.mjs';
import '../button/type.mjs';
import '../popup/popup.mjs';
import '../popup/props.mjs';
import '../overlay/index.mjs';
import '../overlay/overlay.mjs';
import '../overlay/props.mjs';
import '../hooks/useLockScroll.mjs';
import '../_util/useTouch.mjs';
import '../_util/getScrollParent.mjs';
import '../hooks/useTeleport.mjs';
import '../popup/type.mjs';
import '../checkbox/checkbox.mjs';
import '../checkbox/props.mjs';
import '../checkbox/checkbox-group.mjs';
import '../checkbox/checkbox-group-props.mjs';
import '../_common/js/log/index.mjs';
import '../_common/js/log/log.mjs';
import '../checkbox/hooks/index.mjs';
import '../checkbox/hooks/getOptions.mjs';
import '../hooks/useChildSlots.mjs';
import '../checkbox/hooks/setCheckAllStatus.mjs';
import '../checkbox/type.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var prefix = config.prefix;
var getUniqueID = uniqueFactory("dropdown-popup");
var _DropdownItem = defineComponent({
  name: "".concat(prefix, "-dropdown-item"),
  props: props,
  emits: ["change", "open", "opened", "close", "closed", "update:value", "update:modelValue"],
  setup: function setup(props2) {
    var classPrefix = usePrefixClass();
    var dropdownItemClass = usePrefixClass("dropdown-item");
    var _useConfig = useConfig("dropdownMenu"),
      globalConfig = _useConfig.globalConfig;
    var renderContent = useContent();
    var renderTNodeJSX = useTNodeJSX();
    var popupContent = ref(null);
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      passInValue = _useVModel2[0],
      setValue = _useVModel2[1];
    var menuProps = inject("dropdownMenuProps");
    var menuState = inject("dropdownMenuState");
    var _inject = inject("dropdownMenuControl"),
      expandMenu = _inject.expandMenu,
      collapseMenu = _inject.collapseMenu,
      emitEvents = _inject.emitEvents;
    var classes = computed(function () {
      return ["".concat(dropdownItemClass.value)];
    });
    var itemId = ref(0);
    onBeforeMount(function () {
      itemId.value = menuState.childCount;
      menuState.childCount += 1;
    });
    var state = reactive({
      showOverlay: computed(function () {
        return menuProps.showOverlay;
      }),
      duration: computed(function () {
        return menuProps.duration;
      }),
      isShowItems: false,
      wrapperVisible: false,
      expandStyle: {},
      multiple: computed(function () {
        return props2.multiple;
      }),
      options: computed(function () {
        if (props2.keys) {
          var _props2$options;
          var keys = props2.keys;
          return (_props2$options = props2.options) === null || _props2$options === void 0 ? void 0 : _props2$options.map(function (item) {
            var _keys$value, _keys$label, _keys$disabled;
            return {
              value: get(item, (_keys$value = keys === null || keys === void 0 ? void 0 : keys.value) !== null && _keys$value !== void 0 ? _keys$value : "value"),
              label: get(item, (_keys$label = keys === null || keys === void 0 ? void 0 : keys.label) !== null && _keys$label !== void 0 ? _keys$label : "label"),
              disabled: get(item, (_keys$disabled = keys === null || keys === void 0 ? void 0 : keys.disabled) !== null && _keys$disabled !== void 0 ? _keys$disabled : "disabled")
            };
          });
        }
        return props2.options;
      })
    });
    var isCheckedRadio = function isCheckedRadio(value2) {
      return value2 === radioSelect.value;
    };
    var styleDropRadio = function styleDropRadio(value2) {
      return ["".concat(dropdownItemClass.value, "__radio-item"), _defineProperty(_defineProperty({}, "".concat(classPrefix.value, "-is-tick"), !props2.multiple), "".concat(classPrefix.value, "-is-checked"), isCheckedRadio(value2))];
    };
    var popupStyle = computed(function () {
      return {
        zIndex: menuProps.zIndex && menuProps.zIndex + 1,
        position: "absolute",
        overflow: "hidden"
      };
    });
    var popupId = getUniqueID();
    var setExpand = function setExpand(val) {
      var _props;
      var _menuState$barRect = menuState.barRect,
        bottom = _menuState$barRect.bottom,
        top = _menuState$barRect.top;
      var winHeight = getWindowSize().height;
      menuProps.direction === "up" ? state.expandStyle = {
        zIndex: menuProps.zIndex,
        bottom: "".concat(winHeight - top, "px")
      } : state.expandStyle = {
        zIndex: menuProps.zIndex,
        top: "".concat(bottom, "px")
      };
      var duration2 = menuProps.duration;
      (_props = props2["on".concat(val ? "Open" : "Close")]) === null || _props === void 0 || _props.call(props2);
      if (val) {
        state.wrapperVisible = true;
      }
      nextTick(function () {
        state.isShowItems = val;
      });
      if (!val) {
        setTimeout(function () {
          state.wrapperVisible = false;
        }, Number(duration2));
      }
      setTimeout(function () {
        var _props2;
        (_props2 = props2["on".concat(val ? "Opened" : "Closed")]) === null || _props2 === void 0 || _props2.call(props2);
      }, Number(duration2));
    };
    watch(function () {
      return menuState.activeId === itemId.value;
    }, function (val) {
      return setExpand(val);
    });
    var radioSelect = ref();
    var checkSelect = ref();
    var updateSelectValue = function updateSelectValue(val) {
      if (!props2.multiple) {
        var _ref2;
        var list = state.options;
        var firstChild = list === null || list === void 0 ? void 0 : list[0];
        var newValue = (_ref2 = val !== null && val !== void 0 ? val : firstChild === null || firstChild === void 0 ? void 0 : firstChild.value) !== null && _ref2 !== void 0 ? _ref2 : null;
        radioSelect.value = newValue;
      } else if (props2.multiple) {
        checkSelect.value = val !== null && val !== void 0 ? val : [];
      }
    };
    updateSelectValue(passInValue.value || null);
    watch(function () {
      return passInValue.value;
    }, function (val) {
      if (!val) return;
      updateSelectValue(val);
    });
    var isBtnDisabled = computed(function () {
      return Array.isArray(checkSelect.value) && checkSelect.value.length <= 0;
    });
    var resetSelect = function resetSelect() {
      var _props2$onReset;
      checkSelect.value = [];
      var values = checkSelect.value;
      values = JSON.parse(JSON.stringify(values));
      (_props2$onReset = props2.onReset) === null || _props2$onReset === void 0 || _props2$onReset.call(props2, values);
    };
    var confirmSelect = function confirmSelect() {
      var _props2$onConfirm;
      var values = checkSelect.value;
      values = JSON.parse(JSON.stringify(values));
      (_props2$onConfirm = props2.onConfirm) === null || _props2$onConfirm === void 0 || _props2$onConfirm.call(props2, values);
      setValue(values);
      collapseMenu();
      emitEvents("menuClosed", "confirm");
    };
    watch(checkSelect, function (val) {
      if (!props2.multiple) return;
      if (!state.isShowItems) return;
      if (val) {
        var _props2$onChange;
        var value2 = JSON.stringify(passInValue.value || []);
        var values = JSON.stringify(val);
        if (value2 === values) return;
        (_props2$onChange = props2.onChange) === null || _props2$onChange === void 0 || _props2$onChange.call(props2, JSON.parse(values));
      }
    });
    watch(radioSelect, function (val) {
      var _props2$onChange2;
      if (menuState.activeId !== null) {
        var _state$options;
        var target = (_state$options = state.options) === null || _state$options === void 0 ? void 0 : _state$options.find(function (item) {
          return item.value === val;
        });
        menuState.itemsLabel[menuState.activeId] = target === null || target === void 0 ? void 0 : target.label;
      }
      if (props2.multiple) return;
      if (!state.isShowItems) return;
      var value2 = passInValue.value || [];
      if (value2[0] === val) return;
      (_props2$onChange2 = props2.onChange) === null || _props2$onChange2 === void 0 || _props2$onChange2.call(props2, val);
      setValue(val);
      collapseMenu();
      emitEvents("menuClosed", "content");
    });
    var onVisibleChange = function onVisibleChange(visible) {
      if (menuProps.closeOnClickOverlay && !visible) {
        collapseMenu();
        emitEvents("menuClosed", "overlay");
      }
    };
    var _toRefs2 = toRefs(state),
      showOverlay = _toRefs2.showOverlay,
      duration = _toRefs2.duration,
      isShowItems = _toRefs2.isShowItems,
      wrapperVisible = _toRefs2.wrapperVisible,
      expandStyle = _toRefs2.expandStyle,
      multiple = _toRefs2.multiple,
      options = _toRefs2.options;
    return function () {
      var handleRadioChange = function handleRadioChange(value2) {
        radioSelect.value = value2;
      };
      var handleCheckboxChange = function handleCheckboxChange(value2) {
        checkSelect.value = value2;
      };
      var defaultSlot = function defaultSlot() {
        var _slot2;
        if (!multiple.value) {
          var _slot;
          return createVNode(RadioGroup, {
            "value": radioSelect.value,
            "onChange": handleRadioChange,
            "placement": props2.placement,
            "class": "".concat(dropdownItemClass.value, "__radio-group")
          }, _isSlot(_slot = (options.value || []).map(function (option) {
            return createVNode(Radio, {
              "key": option.value,
              "value": option.value,
              "label": option.label,
              "disabled": option.disabled,
              "class": styleDropRadio(option.value),
              "checked": isCheckedRadio(option.value),
              "icon": "line"
            }, null);
          })) ? _slot : {
            default: function _default() {
              return [_slot];
            }
          });
        }
        return createVNode(CheckboxGroup, {
          "value": checkSelect.value,
          "onChange": handleCheckboxChange,
          "class": "".concat(dropdownItemClass.value, "__checkbox-group"),
          "style": "grid-template-columns: repeat(".concat(props2.optionsColumns, ", 1fr)")
        }, _isSlot(_slot2 = (options.value || []).map(function (option) {
          return createVNode(Checkbox, {
            "key": option.value,
            "class": "".concat(dropdownItemClass.value, "__checkbox-item t-checkbox--tag"),
            "icon": false,
            "borderless": true,
            "value": option.value,
            "label": option.label,
            "disabled": option.disabled
          }, null);
        })) ? _slot2 : {
          default: function _default() {
            return [_slot2];
          }
        });
      };
      var footerSlot = function footerSlot() {
        if (multiple.value) {
          return createVNode("div", {
            "class": "".concat(dropdownItemClass.value, "__footer")
          }, [createVNode(Button, {
            "theme": "light",
            "class": "".concat(dropdownItemClass.value, "__footer-btn ").concat(dropdownItemClass.value, "__reset-btn"),
            "disabled": isBtnDisabled.value,
            "onClick": resetSelect
          }, {
            default: function _default() {
              return [globalConfig.value.reset];
            }
          }), createVNode(Button, {
            "theme": "primary",
            "class": "".concat(dropdownItemClass.value, "__footer-btn ").concat(dropdownItemClass.value, "__confirm-btn"),
            "disabled": isBtnDisabled.value,
            "onClick": confirmSelect
          }, {
            default: function _default() {
              return [globalConfig.value.confirm];
            }
          })]);
        }
      };
      var content = renderContent("default", "content");
      var footer = renderTNodeJSX("footer");
      return wrapperVisible.value && createVNode("div", {
        "id": popupId,
        "class": classes.value,
        "style": _objectSpread({}, expandStyle.value)
      }, [createVNode(Popup, {
        "visible": isShowItems.value,
        "placement": menuProps.direction === "up" ? "bottom" : "top",
        "duration": Number(duration.value),
        "showOverlay": showOverlay.value,
        "style": popupStyle.value,
        "overlayProps": {
          style: "position: absolute"
        },
        "attach": "#".concat(popupId),
        "onVisibleChange": onVisibleChange
      }, {
        default: function _default() {
          return [createVNode("div", {
            "ref": popupContent,
            "class": "".concat(dropdownItemClass.value, "__content")
          }, [createVNode("div", {
            "class": "".concat(dropdownItemClass.value, "__body")
          }, [content || defaultSlot()]), footer || footerSlot()])];
        }
      })]);
    };
  }
});

export { _DropdownItem as default };
