// tdesign-mobile-vue/cjs/stepper/stepper.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var stepper_props = require('./props.js');
var form_hooks = require('../form/hooks.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_useClass = require('../hooks/useClass.js');
var shared_util = require('../shared/util.js');
require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-605cd804.js');
require('@babel/runtime/helpers/defineProperty');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-055d28fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

var prefix = config["default"].prefix;
var _Stepper = Vue.defineComponent({
  name: "".concat(prefix, "-stepper"),
  props: stepper_props["default"],
  setup: function setup(props2, context) {
    var stepperClass = hooks_useClass.usePrefixClass("stepper");
    var _toRefs = Vue.toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue,
      min = _toRefs.min,
      max = _toRefs.max,
      step = _toRefs.step,
      integer = _toRefs.integer;
    var _useVModel = hooks_useVModel["default"](value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      stepperValue = _useVModel2[0],
      setStepperValue = _useVModel2[1];
    var disabled = form_hooks.useFormDisabled();
    var inputStyle = Vue.computed(function () {
      return props2.inputWidth ? {
        width: "".concat(props2.inputWidth, "px")
      } : "";
    });
    var isDisabled = function isDisabled(type) {
      if (disabled.value) return true;
      if (type === "minus" && Number(stepperValue.value) <= min.value) {
        return true;
      }
      if (type === "plus" && Number(stepperValue.value) >= max.value) {
        return true;
      }
      return false;
    };
    var getLen = function getLen(num) {
      var numStr = num.toString();
      return numStr.indexOf(".") === -1 ? 0 : numStr.split(".")[1].length;
    };
    var add = function add(a, b) {
      var maxLen = Math.max(getLen(a), getLen(b));
      var base = Math.pow(10, maxLen);
      return Math.round(a * base + b * base) / base;
    };
    var formatValue = function formatValue(value2) {
      return Math.max(Math.min(max.value, value2, Number.MAX_SAFE_INTEGER), min.value, Number.MIN_SAFE_INTEGER).toFixed(Math.max(getLen(step.value), getLen(value2)));
    };
    var updateValue = function updateValue(value2) {
      setStepperValue(shared_util.formatNumber("".concat(value2), !integer.value));
    };
    var plusValue = function plusValue() {
      if (isDisabled("plus")) {
        var _props2$onOverlimit;
        (_props2$onOverlimit = props2.onOverlimit) === null || _props2$onOverlimit === void 0 || _props2$onOverlimit.call(props2, "plus");
        return;
      }
      updateValue(formatValue(add(Number(stepperValue.value), step.value)));
    };
    var minusValue = function minusValue() {
      if (isDisabled("minus")) {
        var _props2$onOverlimit2;
        (_props2$onOverlimit2 = props2.onOverlimit) === null || _props2$onOverlimit2 === void 0 || _props2$onOverlimit2.call(props2, "minus");
        return;
      }
      updateValue(formatValue(add(Number(stepperValue.value), -step.value)));
    };
    var handleInput = function handleInput(e) {
      var value2 = shared_util.formatNumber(e.target.value, !integer.value);
      setStepperValue(value2);
    };
    var handleChange = function handleChange() {
      var formattedValue = formatValue(Number(stepperValue.value));
      updateValue(formattedValue);
    };
    var handleFocus = function handleFocus() {
      var _props2$onFocus;
      (_props2$onFocus = props2.onFocus) === null || _props2$onFocus === void 0 || _props2$onFocus.call(props2, Number(stepperValue.value));
    };
    var handleBlur = function handleBlur() {
      var _props2$onBlur;
      (_props2$onBlur = props2.onBlur) === null || _props2$onBlur === void 0 || _props2$onBlur.call(props2, Number(stepperValue.value));
    };
    return function () {
      return Vue.createVNode("div", {
        "class": ["".concat(stepperClass.value), "".concat(stepperClass.value, "--").concat(props2.size)]
      }, [Vue.createVNode("div", {
        "class": ["".concat(stepperClass.value, "__minus"), "".concat(stepperClass.value, "__minus--").concat(props2.theme), "".concat(stepperClass.value, "__icon--").concat(props2.size), "".concat(disabled.value || Number(stepperValue.value) <= props2.min ? "".concat(stepperClass.value, "--").concat(props2.theme, "-disabled") : "")],
        "onClick": minusValue
      }, [Vue.createVNode(tdesignIconsVueNext.RemoveIcon, {
        "class": "".concat(stepperClass.value, "__minus-icon")
      }, null)]), Vue.createVNode("input", {
        "value": stepperValue.value,
        "class": ["".concat(stepperClass.value, "__input"), "".concat(stepperClass.value, "__input--").concat(props2.theme), "".concat(stepperClass.value, "__input--").concat(props2.size), "".concat(disabled.value ? "".concat(stepperClass.value, "--").concat(props2.theme, "-disabled") : "")],
        "type": integer.value ? "tel" : "text",
        "inputmode": integer.value ? "numeric" : "decimal",
        "style": inputStyle.value,
        "disabled": props2.disableInput || disabled.value,
        "readonly": props2.disableInput,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onChange": handleChange
      }, null), Vue.createVNode("div", {
        "class": ["".concat(stepperClass.value, "__plus"), "".concat(stepperClass.value, "__plus--").concat(props2.theme), "".concat(stepperClass.value, "__icon--").concat(props2.size), "".concat(disabled.value || Number(stepperValue.value) >= props2.max ? "".concat(stepperClass.value, "--").concat(props2.theme, "-disabled") : "")],
        "onClick": plusValue
      }, [Vue.createVNode(tdesignIconsVueNext.AddIcon, {
        "class": "".concat(stepperClass.value, "__plus-icon")
      }, null)])]);
    };
  }
});

exports["default"] = _Stepper;
