// tdesign-mobile-vue/es/upload/upload.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, ref, computed, createVNode, Fragment, resolveComponent, mergeProps } from 'vue';
import { LoadingIcon, CloseCircleIcon, AddIcon, CloseIcon } from 'tdesign-icons-vue-next';
import { isBoolean } from 'lodash-es';
import { Image } from '../image/index.mjs';
import { ImageViewer } from '../image-viewer/index.mjs';
import UploadProps from './props.mjs';
import config from '../config.mjs';
import useUpload from './hooks/useUpload.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useConfig } from '../config-provider/useConfig.mjs';
import '../image/image.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '@vueuse/core';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../image/props.mjs';
import '../image/type.mjs';
import '../image-viewer/image-viewer.mjs';
import '../image-viewer/props.mjs';
import '../hooks/useDefaultValue.mjs';
import '../hooks/useGesture.mjs';
import '@use-gesture/vanilla';
import '../hooks/useVModel.mjs';
import '../swiper/index.mjs';
import '../swiper/swiper.mjs';
import '../swipe-cell/useSwipe.mjs';
import '../swiper/props.mjs';
import '../swiper/swiper-item.mjs';
import '../swiper/type.mjs';
import '../image-viewer/type.mjs';
import '../_common/js/upload/main.mjs';
import '../_common/js/upload/utils.mjs';
import '../_common/js/log/log.mjs';
import '../_common/js/upload/xhr.mjs';

var prefix = config.prefix;
var _Upload = defineComponent({
  name: "".concat(prefix, "-upload"),
  components: {
    TImage: Image
  },
  props: UploadProps,
  emits: ["update:files", "update:modelValue", "change", "fail", "preview", "progress", "remove", "success", "select-change", "validate", "click-upload"],
  setup: function setup(props, _ref) {
    var expose = _ref.expose;
    var uploadClass = usePrefixClass("upload");
    var _useConfig = useConfig("upload"),
      globalConfig = _useConfig.globalConfig;
    var _useUpload = useUpload(props),
      disabled = _useUpload.disabled,
      displayFiles = _useUpload.displayFiles,
      uploading = _useUpload.uploading,
      inputRef = _useUpload.inputRef,
      uploadFilePercent = _useUpload.uploadFilePercent,
      uploadFiles = _useUpload.uploadFiles,
      onNormalFileChange = _useUpload.onNormalFileChange,
      onInnerRemove = _useUpload.onInnerRemove,
      cancelUpload = _useUpload.cancelUpload;
    var renderTNodeJSX = useTNodeJSX();
    var showViewer = ref(false);
    var initialIndex = ref(0);
    var handlePreview = function handlePreview(e, file, index) {
      var _props$onPreview;
      initialIndex.value = index;
      showViewer.value = props.preview;
      (_props$onPreview = props.onPreview) === null || _props$onPreview === void 0 || _props$onPreview.call(props, {
        e: e,
        file: file,
        index: index
      });
    };
    var triggerUpload = function triggerUpload(e) {
      var _props$onClickUpload;
      var input = inputRef.value;
      if (disabled !== null && disabled !== void 0 && disabled.value) return;
      input.click();
      (_props$onClickUpload = props.onClickUpload) === null || _props$onClickUpload === void 0 || _props$onClickUpload.call(props, {
        e: e
      });
    };
    var previewImgs = computed(function () {
      return displayFiles.value.map(function (item) {
        return item.url;
      });
    });
    var handleImageClose = function handleImageClose(_ref2) {
      var visible = _ref2.visible;
      showViewer.value = visible;
    };
    var renderStatus = function renderStatus(file) {
      if (file.status === "fail" || file.status === "progress") {
        return createVNode("div", {
          "class": "".concat(uploadClass.value, "__progress-mask")
        }, [file.status === "progress" ? createVNode(Fragment, null, [createVNode(LoadingIcon, {
          "class": "".concat(uploadClass.value, "__progress-loading"),
          "size": "24"
        }, null), createVNode("div", {
          "class": "".concat(uploadClass.value, "__progress-text")
        }, [file.percent ? "".concat(file.percent, "%") : globalConfig.value.progress.uploadingText])]) : createVNode(CloseCircleIcon, {
          "size": "24"
        }, null), file.status === "fail" && createVNode("div", {
          "class": "".concat(uploadClass.value, "__progress-text")
        }, [globalConfig.value.progress.failText])]);
      }
    };
    var renderAddContent = function renderAddContent() {
      var _displayFiles$value;
      if (!props.addBtn) return null;
      if (props.max === 0 || props.max > 0 && ((_displayFiles$value = displayFiles.value) === null || _displayFiles$value === void 0 ? void 0 : _displayFiles$value.length) < props.max) {
        var addBtnNode = renderTNodeJSX("addBtn", createVNode(AddIcon, null, null));
        var addContentNode = renderTNodeJSX("addContent");
        return createVNode("div", {
          "class": "".concat(uploadClass.value, "__item ").concat(uploadClass.value, "__item--add"),
          "onClick": triggerUpload
        }, [createVNode("div", {
          "class": "".concat(uploadClass.value, "__add-icon")
        }, [addContentNode || addBtnNode])]);
      }
      return null;
    };
    expose({
      upload: inputRef.value,
      uploading: uploading,
      triggerUpload: triggerUpload,
      uploadFiles: uploadFiles,
      cancelUpload: cancelUpload,
      uploadFilePercent: uploadFilePercent
    });
    return function () {
      return createVNode("div", {
        "class": "".concat(uploadClass.value)
      }, [displayFiles.value.map(function (file, index) {
        return createVNode("div", {
          "key": index,
          "class": "".concat(uploadClass.value, "__item")
        }, [file.url && createVNode(resolveComponent("t-image"), mergeProps({
          "class": "".concat(uploadClass.value, "__image"),
          "shape": "round"
        }, props.imageProps, {
          "src": file.url,
          "onClick": function onClick(e) {
            return handlePreview(e, file, index);
          }
        }), null), renderStatus(file), (isBoolean(file.removeBtn) ? file.removeBtn : props.removeBtn) && createVNode(CloseIcon, {
          "class": "".concat(uploadClass.value, "__delete-btn"),
          "onClick": function onClick(_ref3) {
            var e = _ref3.e;
            return onInnerRemove({
              e: e,
              file: file,
              index: index
            });
          }
        }, null)]);
      }), renderAddContent(), createVNode("input", {
        "ref": inputRef,
        "value": props.files,
        "type": "file",
        "multiple": props.multiple,
        "hidden": true,
        "capture": props.capture,
        "accept": props.accept,
        "onChange": onNormalFileChange
      }, null), createVNode(ImageViewer, {
        "visible": showViewer.value,
        "images": previewImgs.value,
        "index": initialIndex.value,
        "onClose": handleImageClose
      }, null)]);
    };
  }
});

export { _Upload as default };
