// tdesign-mobile-vue/lib/collapse/collapse.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, toRefs, computed, provide, createVNode } from 'vue';
import props from './props.js';
import config from '../config.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import useVModel from '../hooks/useVModel.js';
import 'lodash-es';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../config-provider/useConfig.js';
import '@babel/runtime/helpers/defineProperty';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';

var prefix = config.prefix;
var _Collapse = defineComponent({
  name: "".concat(prefix, "-collapse"),
  props: props,
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var renderTNodeJSX = useTNodeJSX();
    var collapseClass = usePrefixClass("collapse");
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      activeValue = _useVModel2[0],
      setActiveValue = _useVModel2[1];
    var calcActiveValues = function calcActiveValues(activeValues, panelValue, expandMutex) {
      var hit = activeValues.indexOf(panelValue);
      if (hit > -1) {
        return activeValues.filter(function (item) {
          return item !== panelValue;
        });
      }
      return expandMutex ? [panelValue] : activeValues.concat(panelValue);
    };
    var onPanelChange = function onPanelChange(value2, args) {
      if (Array.isArray(activeValue.value)) {
        var val = calcActiveValues(activeValue.value, value2, props2.expandMutex);
        setActiveValue(val, args);
      }
    };
    var disabled = computed(function () {
      return props2.disabled;
    });
    var expandIcon = computed(function () {
      return props2.expandIcon;
    });
    var rootClass = computed(function () {
      return [collapseClass.value, "".concat(collapseClass.value, "--").concat(props2.theme)];
    });
    provide("collapse", {
      activeValue: activeValue,
      disabled: disabled,
      expandIcon: expandIcon,
      onPanelChange: onPanelChange,
      defaultExpandAll: props2.defaultExpandAll
    });
    provide("renderParentTNode", renderTNodeJSX);
    return function () {
      var _slots$default;
      return createVNode("div", {
        "class": rootClass.value
      }, [(_slots$default = slots.default) === null || _slots$default === void 0 ? void 0 : _slots$default.call(slots)]);
    };
  }
});

export { _Collapse as default };
