// tdesign-mobile-vue/es/avatar/avatar-group.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _toConsumableArray } from '../_chunks/dep-fb069758.mjs';
import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, provide, computed, Fragment, createVNode } from 'vue';
import AvatarGroupProps from './avatar-group-props.mjs';
import config from '../config.mjs';
import _Avatar from './avatar.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { isValidSize } from '../_common/js/avatar/utils.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../badge/index.mjs';
import '../badge/badge.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import 'lodash-es';
import '../badge/props.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../badge/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../image/index.mjs';
import '../image/image.mjs';
import 'tdesign-icons-vue-next';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../image/props.mjs';
import '../image/type.mjs';
import './props.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _AvatarGroup = defineComponent({
  name: "".concat(prefix, "-avatar-group"),
  props: AvatarGroupProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var avatarGroupClass = usePrefixClass("avatar-group");
    provide("avatarGroup", _objectSpread({}, props));
    var direction = props.cascading ? props.cascading.split("-")[0] : "right";
    var isCustomSize = computed(function () {
      return !isValidSize(props.size);
    });
    var avatarGroupClasses = computed(function () {
      return ["".concat(avatarGroupClass.value), "".concat(avatarGroupClass.value, "-offset-").concat(direction), "".concat(avatarGroupClass.value, "-offset-").concat(direction, "-").concat(isCustomSize.value ? "medium" : props.size)];
    });
    var onCollapsedItemClick = function onCollapsedItemClick(e) {
      var _props$onCollapsedIte;
      (_props$onCollapsedIte = props.onCollapsedItemClick) === null || _props$onCollapsedIte === void 0 || _props$onCollapsedIte.call(props, {
        e: e
      });
    };
    var readerAvatar = function readerAvatar() {
      var children = renderTNodeJSX("default");
      var allChildren = [];
      children.forEach(function (child) {
        if (child.type === Fragment) {
          allChildren.push.apply(allChildren, _toConsumableArray(child.children));
        } else {
          allChildren.push(child);
        }
      });
      var isShowCollapse = false;
      var avatarList = [];
      if (allChildren.length > props.max) {
        avatarList = allChildren.slice(0, props.max);
        isShowCollapse = true;
      } else {
        avatarList = allChildren;
      }
      if (isShowCollapse) {
        var collapseAvatar = renderTNodeJSX("collapseAvatar");
        avatarList.push(createVNode("div", {
          "class": "".concat(avatarGroupClass.value, "__collapse--default"),
          "onClick": onCollapsedItemClick
        }, [createVNode(_Avatar, {
          "size": avatarList[0].size || props.size,
          "shape": props.shape
        }, {
          default: function _default() {
            return [collapseAvatar || "+".concat(allChildren.length - props.max)];
          }
        })]));
      }
      return avatarList;
    };
    return function () {
      return createVNode("div", {
        "class": avatarGroupClasses.value
      }, [readerAvatar()]);
    };
  }
});

export { _AvatarGroup as default };
