// tdesign-mobile-vue/es/avatar/avatar.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, inject, computed, createVNode, mergeProps, Fragment } from 'vue';
import { Badge } from '../badge/index.mjs';
import { Image } from '../image/index.mjs';
import config from '../config.mjs';
import AvatarProps from './props.mjs';
import { useTNodeJSX, useContent } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { isValidSize } from '../_common/js/avatar/utils.mjs';
import '../badge/badge.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import 'lodash-es';
import '../badge/props.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../badge/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../image/image.mjs';
import 'tdesign-icons-vue-next';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../image/props.mjs';
import '../image/type.mjs';

var prefix = config.prefix;
var _Avatar = defineComponent({
  name: "".concat(prefix, "-avatar"),
  props: AvatarProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var avatarClass = usePrefixClass("avatar");
    var avatarGroupProps = inject("avatarGroup", {});
    var hasAvatarGroupProps = Object.keys(avatarGroupProps).length > 0;
    var shapeValue = computed(function () {
      return props.shape || (avatarGroupProps === null || avatarGroupProps === void 0 ? void 0 : avatarGroupProps.shape) || "circle";
    });
    var sizeValue = computed(function () {
      return props.size || (avatarGroupProps === null || avatarGroupProps === void 0 ? void 0 : avatarGroupProps.size);
    });
    var isCustomSize = computed(function () {
      return !isValidSize(sizeValue.value);
    });
    var avatarClasses = computed(function () {
      return ["".concat(avatarClass.value), "".concat(avatarClass.value, "--").concat(isCustomSize.value ? "medium" : sizeValue.value), "".concat(avatarClass.value, "--").concat(shapeValue.value), hasAvatarGroupProps ? "".concat(avatarClass.value, "--border ").concat(avatarClass.value, "--border-").concat(isCustomSize.value ? "medium" : sizeValue.value) : ""];
    });
    var customSize = computed(function () {
      return isCustomSize.value ? {
        height: sizeValue.value,
        width: sizeValue.value,
        "font-size": "".concat(Number.parseInt(sizeValue.value, 10) / 8 * 3 + 2, "px")
      } : {};
    });
    var handleImgLoadError = function handleImgLoadError(e) {
      var _props$onError;
      (_props$onError = props.onError) === null || _props$onError === void 0 || _props$onError.call(props, e);
    };
    return function () {
      var icon = renderTNodeJSX("icon");
      var TNodeContent = renderTNodeContent("default", "content");
      var readerAvatar = function readerAvatar() {
        if (props.image && !props.hideOnLoadFailed) {
          return createVNode(Image, mergeProps({
            "src": props.image,
            "alt": props.alt
          }, props.imageProps, {
            "onError": handleImgLoadError
          }), null);
        }
        if (icon) {
          return createVNode("div", {
            "class": "".concat(avatarClass.value, "__icon")
          }, [icon]);
        }
        return createVNode(Fragment, null, [TNodeContent]);
      };
      return createVNode("div", {
        "class": "".concat(avatarClass.value, "__wrapper")
      }, [createVNode("div", {
        "class": "".concat(avatarClass.value, "__badge")
      }, [createVNode(Badge, props.badgeProps, {
        default: function _default() {
          return [createVNode("div", {
            "class": avatarClasses.value,
            "style": customSize.value
          }, [readerAvatar()])];
        }
      })])]);
    };
  }
});

export { _Avatar as default };
