// tdesign-mobile-vue/cjs/table/hooks/useStyle.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var table_hooks_useClassName = require('./useClassName.js');
require('../../hooks/useClass.js');
require('../../_chunks/dep-605cd804.js');
require('../../_chunks/dep-6f53fb1b.js');
require('../../_common/js/global-config/mobile/default-config.js');
require('../../_common/js/global-config/mobile/locale/zh_CN.js');
require('../../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../../_chunks/dep-7220ff76.js');
require('dayjs');
require('../../_chunks/dep-05850262.js');
require('../../_chunks/dep-e9a63584.js');
require('../../_chunks/dep-53383176.js');
require('../../_chunks/dep-f84e6e17.js');
require('../../_chunks/dep-0c11c462.js');
require('../../_chunks/dep-0bd122a0.js');
require('../../_chunks/dep-b999c3d4.js');
require('../../_chunks/dep-159c7a94.js');
require('../../_chunks/dep-3cf5587c.js');
require('../../_chunks/dep-ea269588.js');
require('../../_chunks/dep-b633ec9e.js');
require('../../_chunks/dep-7aa6341b.js');
require('../../_chunks/dep-b8e8d654.js');
require('../../_chunks/dep-61ce8c0d.js');
require('../../_chunks/dep-5cf8bf50.js');
require('../../_chunks/dep-2457b474.js');
require('../../_chunks/dep-4b045875.js');
require('../../_chunks/dep-a6a4534e.js');
require('../../_chunks/dep-fef48622.js');
require('../../_chunks/dep-588e2e70.js');
require('../../_chunks/dep-8d30e8d1.js');
require('../../_chunks/dep-dd93e905.js');
require('../../_chunks/dep-d1ba5719.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function formatCSSUnit(unit) {
  if (!unit) return unit;
  return isNaN(Number(unit)) ? unit : "".concat(unit, "px");
}
function useStyle(props) {
  var _toRefs = Vue.toRefs(props),
    bordered = _toRefs.bordered,
    stripe = _toRefs.stripe,
    verticalAlign = _toRefs.verticalAlign,
    height = _toRefs.height,
    maxHeight = _toRefs.maxHeight,
    tableContentWidth = _toRefs.tableContentWidth;
  var _useClassName = table_hooks_useClassName["default"](),
    tableBaseClass = _useClassName.tableBaseClass,
    tableAlignClasses = _useClassName.tableAlignClasses;
  var tableClasses = Vue.computed(function () {
    return [tableBaseClass.table, [tableAlignClasses[(verticalAlign === null || verticalAlign === void 0 ? void 0 : verticalAlign.value) || "middle"]], _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, tableBaseClass.bordered, bordered === null || bordered === void 0 ? void 0 : bordered.value), tableBaseClass.striped, stripe === null || stripe === void 0 ? void 0 : stripe.value), tableBaseClass.loading, props.loading)];
  });
  var tableContentStyles = Vue.computed(function () {
    return {
      height: formatCSSUnit(height === null || height === void 0 ? void 0 : height.value),
      maxHeight: formatCSSUnit(maxHeight === null || maxHeight === void 0 ? void 0 : maxHeight.value)
    };
  });
  var tableElementStyles = Vue.computed(function () {
    return {
      width: formatCSSUnit(tableContentWidth === null || tableContentWidth === void 0 ? void 0 : tableContentWidth.value)
    };
  });
  return {
    tableClasses: tableClasses,
    tableElementStyles: tableElementStyles,
    tableContentStyles: tableContentStyles
  };
}

exports["default"] = useStyle;
exports.formatCSSUnit = formatCSSUnit;
