// tdesign-mobile-vue/lib/shared/useClickAway/index.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import { useEventListener, unrefElement } from '@vueuse/core';
import { isArray } from 'lodash-es';
import { isBrowser } from '../util.js';
import '@babel/runtime/helpers/slicedToArray';
import '../../hooks/useClass.js';
import 'vue';
import '../../config-provider/useConfig.js';
import '@babel/runtime/helpers/defineProperty';
import '../../config-provider/context.js';
import '../../_common/js/global-config/mobile/default-config.js';
import '../../_common/js/global-config/mobile/locale/zh_CN.js';
import '../../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../../_chunks/dep-dc49adb8.js';
import 'dayjs';

function useClickAway(target, fn) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  if (!isBrowser) return;
  var _options$eventName = options.eventName,
    eventName = _options$eventName === void 0 ? "touchstart" : _options$eventName,
    _options$capture = options.capture,
    capture = _options$capture === void 0 ? true : _options$capture,
    _options$ignore = options.ignore,
    ignore = _options$ignore === void 0 ? [] : _options$ignore,
    _options$detectIframe = options.detectIframe,
    detectIframe = _options$detectIframe === void 0 ? false : _options$detectIframe;
  var listenerOptions = {
    passive: true,
    capture: capture
  };
  var checkElInEvtTarget = function checkElInEvtTarget(el, event) {
    return el && (event.target === el || event.composedPath().includes(el));
  };
  var shouldIgnore = function shouldIgnore(event) {
    return ignore.some(function (ignoreTarget) {
      return checkElInEvtTarget(unrefElement(ignoreTarget), event);
    });
  };
  var eventHandler = function eventHandler(event) {
    var targets = isArray(target) ? _toConsumableArray(target) : [target];
    var shouldTrigger = targets.every(function (targetItem) {
      return !checkElInEvtTarget(unrefElement(targetItem), event);
    });
    if (!shouldTrigger) return;
    shouldTrigger = !shouldIgnore(event);
    if (!shouldTrigger) return;
    fn(event);
  };
  var cleanup = [useEventListener(window, eventName, eventHandler, listenerOptions)];
  if (detectIframe) {
    cleanup.push(useEventListener(window, "blur", function (event) {
      setTimeout(function () {
        var targets = isArray(target) ? _toConsumableArray(target) : [target];
        if (targets.every(function (targetItem) {
          var _window$document$acti;
          var el = unrefElement(targetItem);
          return ((_window$document$acti = window.document.activeElement) === null || _window$document$acti === void 0 ? void 0 : _window$document$acti.tagName) === "IFRAME" && !(el !== null && el !== void 0 && el.contains(window.document.activeElement));
        })) {
          fn(event);
        }
      }, 0);
    }));
  }
  return function () {
    return cleanup.forEach(function (clean) {
      return clean();
    });
  };
}

export { useClickAway };
