// tdesign-mobile-vue/cjs/guide/guide.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var popover_index = require('../popover/index.js');
var popup_index = require('../popup/index.js');
var button_index = require('../button/index.js');
var config = require('../config.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var shared_dom = require('../shared/dom.js');
var _common_js_utils_setStyle = require('../_common/js/utils/setStyle.js');
var guide_props = require('./props.js');
var guide_utils_dom = require('./utils/dom.js');
var hooks_useClass = require('../hooks/useClass.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
var isFunction = require('../_chunks/dep-0bd122a0.js');
require('../popover/popover.js');
require('../popover/props.js');
require('../hooks/tnode.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../shared/useClickAway/index.js');
require('@vueuse/core');
require('../shared/util.js');
require('../_chunks/dep-055d28fd.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../popup/popup.js');
require('tdesign-icons-vue-next');
require('../popup/props.js');
require('../overlay/index.js');
require('../overlay/overlay.js');
require('../overlay/props.js');
require('../hooks/useLockScroll.js');
require('../_util/useTouch.js');
require('../_util/getScrollParent.js');
require('../hooks/useTeleport.js');
require('../shared/render-tnode.js');
require('../button/button.js');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../loading/plugin.js');
require('../button/props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../shared/hover.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var _Guide = Vue.defineComponent({
  name: "".concat(prefix, "-guide"),
  components: {
    TPopover: popover_index.Popover,
    TButton: button_index.Button,
    TPopup: popup_index.Popup
  },
  props: guide_props["default"],
  setup: function setup(props, context) {
    var _toRefs = Vue.toRefs(props),
      current = _toRefs.current,
      modelValue = _toRefs.modelValue,
      finishButtonProps = _toRefs.finishButtonProps,
      hideCounter = _toRefs.hideCounter,
      hideSkip = _toRefs.hideSkip,
      steps = _toRefs.steps,
      zIndex = _toRefs.zIndex;
    var _useVModel = hooks_useVModel["default"](current, modelValue, props.defaultCurrent, props.onChange, "current"),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerCurrent = _useVModel2[0],
      setInnerCurrent = _useVModel2[1];
    var _useConfig = configProvider_useConfig.useConfig("guide"),
      globalConfig = _useConfig.globalConfig;
    var guideClass = hooks_useClass.usePrefixClass("guide");
    var LOCK_CLASS = "".concat(guideClass.value, "--lock");
    var overlayLayerRef = Vue.ref();
    var highlightLayerRef = Vue.ref();
    var referenceLayerRef = Vue.ref();
    var currentHighlightLayerElm = Vue.ref();
    var popoverWrapperRef = Vue.ref();
    var actived = Vue.ref(false);
    var stepsTotal = Vue.computed(function () {
      return steps.value.length;
    });
    var currentStepInfo = Vue.computed(function () {
      return steps.value[innerCurrent.value];
    });
    var isPopover = Vue.computed(function () {
      return getCurrentCrossProps("mode") === "popover";
    });
    var popoverVisible = Vue.ref(false);
    var isPopoverCenter = Vue.computed(function () {
      return isPopover.value && currentStepInfo.value.placement === "center";
    });
    var stepProps = Vue.computed(function () {
      if (isPopover.value) {
        return _objectSpread({
          visible: popoverVisible.value,
          placement: isPopoverCenter.value ? "bottom" : currentStepInfo.value.placement,
          theme: "light",
          showArrow: false
        }, currentStepInfo.value.popoverProps);
      }
      return {
        visible: popoverVisible.value,
        zIndex: zIndex.value,
        placement: "center",
        class: "".concat(guideClass.value, "__dialog"),
        overlayProps: {
          zIndex: zIndex.value - 1
        }
      };
    });
    var currentElmIsFixed = Vue.computed(function () {
      return guide_utils_dom.isFixed(currentHighlightLayerElm.value || document.body);
    });
    var getCurrentCrossProps = function getCurrentCrossProps(propsName) {
      var _currentStepInfo$valu;
      return (_currentStepInfo$valu = currentStepInfo.value[propsName]) !== null && _currentStepInfo$valu !== void 0 ? _currentStepInfo$valu : props[propsName];
    };
    var hWithParams = function hWithParams() {
      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
        currentStepInfo: currentStepInfo.value
      };
      var newH = function newH() {
        return Vue.h.apply(void 0, arguments);
      };
      return Object.assign(newH, params);
    };
    var currentCustomHighlightContent = Vue.computed(function () {
      var highlightContent = currentStepInfo.value.highlightContent;
      var node = highlightContent;
      if (isFunction.isFunction(highlightContent)) {
        node = highlightContent(hWithParams());
      } else if (context.slots.highlightContent) {
        node = context.slots.highlightContent(hWithParams());
      } else if (context.slots["highlight-content"]) {
        node = context.slots["highlight-content"](hWithParams());
      }
      if (node) {
        if (!node.props) node.props = {};
        node.props.class = node.props.class || "";
      }
      return node;
    });
    var showCustomHighlightContent = Vue.computed(function () {
      return Boolean(currentCustomHighlightContent.value && isPopover.value);
    });
    var showOverlay = Vue.computed(function () {
      return getCurrentCrossProps("showOverlay");
    });
    var highlightClass = Vue.computed(function () {
      return ["".concat(guideClass.value, "__highlight"), "".concat(guideClass.value, "__highlight--").concat(isPopover.value ? "popover" : "dialog"), "".concat(guideClass.value, "--").concat(currentElmIsFixed.value && isPopover.value ? "fixed" : "absolute")];
    });
    var maskClass = Vue.computed(function () {
      return ["".concat(guideClass.value, "__highlight--").concat(showOverlay.value ? "mask" : "nomask")];
    });
    var wrapperClass = Vue.computed(function () {
      return ["".concat(guideClass.value, "__wrapper"), "".concat(guideClass.value, "--").concat(currentElmIsFixed.value ? "fixed" : "absolute"), _defineProperty__default["default"]({}, "".concat(guideClass.value, "__wrapper--content"), !!currentStepInfo.value.content)];
    });
    var popoverClass = Vue.computed(function () {
      return ["".concat(guideClass.value, "__reference")];
    });
    var contentClass = Vue.computed(function () {
      return ["".concat(guideClass.value, "__content--").concat(isPopover.value ? "popover" : "dialog")];
    });
    var footerClass = Vue.computed(function () {
      return ["".concat(guideClass.value, "__footer"), "".concat(guideClass.value, "__footer--").concat(isPopover.value ? "popover" : "dialog")];
    });
    var isLast = Vue.computed(function () {
      return innerCurrent.value === stepsTotal.value - 1;
    });
    var buttonSize = Vue.computed(function () {
      return isPopover.value ? "extra-small" : "medium";
    });
    var setHighlightLayerPosition = function setHighlightLayerPosition(highlightLayer) {
      var isReference = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var _getRelativePosition = guide_utils_dom.getRelativePosition(currentHighlightLayerElm.value),
        top = _getRelativePosition.top,
        left = _getRelativePosition.left;
      var _currentHighlightLaye = currentHighlightLayerElm.value.getBoundingClientRect(),
        width = _currentHighlightLaye.width,
        height = _currentHighlightLaye.height;
      var highlightPadding = getCurrentCrossProps("highlightPadding");
      if (isPopover.value) {
        width += highlightPadding * 2;
        height += highlightPadding * 2;
        top -= highlightPadding;
        left -= highlightPadding;
      } else {
        var _getWindowScroll = shared_dom.getWindowScroll(),
          scrollTop = _getWindowScroll.scrollTop,
          scrollLeft = _getWindowScroll.scrollLeft;
        top += scrollTop;
        left += scrollLeft;
      }
      var style = {
        top: "".concat(top, "px"),
        left: "".concat(left, "px")
      };
      if (showCustomHighlightContent.value) {
        if (isReference) {
          var _highlightLayerRef$va = highlightLayerRef.value.getBoundingClientRect(),
            width2 = _highlightLayerRef$va.width,
            height2 = _highlightLayerRef$va.height;
          Object.assign(style, {
            width: "".concat(width2, "px"),
            height: "".concat(height2, "px")
          });
        } else {
          Object.assign(style, {
            width: "auto",
            height: "auto"
          });
        }
      } else {
        Object.assign(style, {
          width: "".concat(width, "px"),
          height: "".concat(height, "px")
        });
      }
      _common_js_utils_setStyle["default"](highlightLayer, style);
    };
    var setReferenceFullW = function setReferenceFullW(referenceElements) {
      var style = {
        left: 0,
        width: "100vw"
      };
      referenceElements.forEach(function (elem) {
        return _common_js_utils_setStyle["default"](elem, style);
      });
    };
    var showPopoverGuide = function showPopoverGuide() {
      Vue.nextTick(function () {
        currentHighlightLayerElm.value = guide_utils_dom.getTargetElm(currentStepInfo.value.element);
        if (!currentHighlightLayerElm.value) return;
        guide_utils_dom.scrollToParentVisibleArea(currentHighlightLayerElm.value);
        setHighlightLayerPosition(highlightLayerRef.value);
        setHighlightLayerPosition(popoverWrapperRef.value, true);
        setHighlightLayerPosition(referenceLayerRef.value, true);
        guide_utils_dom.scrollToElm(currentHighlightLayerElm.value);
        isPopoverCenter.value && setReferenceFullW([referenceLayerRef.value, popoverWrapperRef.value]);
      });
    };
    var showDialogGuide = function showDialogGuide() {
      Vue.nextTick(function () {
        currentHighlightLayerElm.value = guide_utils_dom.getTargetElm(currentStepInfo.value.element);
        guide_utils_dom.scrollToParentVisibleArea(currentHighlightLayerElm.value);
        setHighlightLayerPosition(highlightLayerRef.value);
        guide_utils_dom.scrollToElm(currentHighlightLayerElm.value);
      });
    };
    var showGuide = function showGuide() {
      if (isPopover.value) {
        showPopoverGuide();
      } else {
        showDialogGuide();
      }
      Vue.nextTick(function () {
        popoverVisible.value = true;
      });
    };
    var destroyGuide = function destroyGuide() {
      var _highlightLayerRef$va2, _overlayLayerRef$valu;
      (_highlightLayerRef$va2 = highlightLayerRef.value) === null || _highlightLayerRef$va2 === void 0 || _highlightLayerRef$va2.parentNode.removeChild(highlightLayerRef.value);
      (_overlayLayerRef$valu = overlayLayerRef.value) === null || _overlayLayerRef$valu === void 0 || _overlayLayerRef$valu.parentNode.removeChild(overlayLayerRef.value);
      shared_dom.removeClass(document.body, LOCK_CLASS);
    };
    var renderButtonContent = function renderButtonContent(buttonProps, defaultContent) {
      var _ref2 = buttonProps || {},
        content = _ref2.content;
      return isFunction.isFunction(content) ? content(Vue.h) : content || defaultContent;
    };
    var handleSkip = function handleSkip(e) {
      var _props$onSkip;
      var total = stepsTotal.value;
      actived.value = false;
      setInnerCurrent(-1, {
        e: e,
        total: total
      });
      (_props$onSkip = props.onSkip) === null || _props$onSkip === void 0 || _props$onSkip.call(props, {
        e: e,
        current: innerCurrent.value,
        total: total
      });
    };
    var handleNext = function handleNext(e) {
      var _props$onNextStepClic;
      var total = stepsTotal.value;
      setInnerCurrent(innerCurrent.value + 1, {
        e: e,
        total: total
      });
      (_props$onNextStepClic = props.onNextStepClick) === null || _props$onNextStepClic === void 0 || _props$onNextStepClic.call(props, {
        e: e,
        next: innerCurrent.value + 1,
        current: innerCurrent.value,
        total: total
      });
    };
    var handleFinish = function handleFinish(e) {
      var _props$onFinish;
      var total = stepsTotal.value;
      actived.value = false;
      setInnerCurrent(-1, {
        e: e,
        total: total
      });
      (_props$onFinish = props.onFinish) === null || _props$onFinish === void 0 || _props$onFinish.call(props, {
        e: e,
        current: innerCurrent.value,
        total: total
      });
    };
    var handleBack = function handleBack(e) {
      var _props$onBack;
      var total = stepsTotal.value;
      setInnerCurrent(0, {
        e: e,
        total: total
      });
      (_props$onBack = props.onBack) === null || _props$onBack === void 0 || _props$onBack.call(props, {
        e: e,
        current: innerCurrent.value,
        total: total
      });
    };
    var contentProps = Vue.computed(function () {
      return {
        handleSkip: handleSkip,
        handleNext: handleNext,
        handleFinish: handleFinish,
        handleBack: handleBack,
        current: innerCurrent.value,
        total: stepsTotal.value
      };
    });
    var initGuide = function initGuide() {
      if (innerCurrent.value >= 0 && innerCurrent.value < steps.value.length) {
        if (!actived.value) {
          actived.value = true;
          shared_dom.addClass(document.body, LOCK_CLASS);
        }
        showGuide();
      }
    };
    Vue.watch(innerCurrent, function (val) {
      if (val >= 0 && val < stepsTotal.value) {
        isPopover.value && (popoverVisible.value = false);
        initGuide();
      } else {
        actived.value = false;
        destroyGuide();
      }
    });
    Vue.onMounted(function () {
      initGuide();
    });
    return function () {
      var renderStepContent = function renderStepContent() {
        var _finishButtonProps$va;
        var renderTitleNode = function renderTitleNode() {
          var title = currentStepInfo.value.title;
          var renderTitle = null;
          if (isFunction.isFunction(title)) {
            renderTitle = title(hWithParams());
          } else if (context.slots.title) {
            renderTitle = context.slots.title(hWithParams());
          } else if (typeof title === "string") {
            renderTitle = title;
          } else if (title) {
            renderTitle = Vue.h(title);
          }
          return renderTitle;
        };
        var renderBodyNode = function renderBodyNode() {
          var body = currentStepInfo.value.body;
          var renderBody = null;
          if (isFunction.isFunction(body)) {
            renderBody = body(hWithParams());
          } else if (context.slots.body) {
            renderBody = context.slots.body(hWithParams());
          } else if (typeof body === "string") {
            renderBody = body;
          } else if (body) {
            renderBody = Vue.h(body);
          }
          return renderBody;
        };
        var renderCounterNode = function renderCounterNode() {
          var params = {
            total: stepsTotal.value,
            current: innerCurrent.value
          };
          var renderCounter = null;
          var counter = props.counter;
          if (isFunction.isFunction(counter)) {
            renderCounter = counter(Vue.h, params);
          } else if (context.slots.counter) {
            renderCounter = context.slots.counter(hWithParams(params));
          } else if (counter) {
            renderCounter = Vue.h(counter, params);
          }
          return renderCounter || " (".concat(innerCurrent.value + 1, "/").concat(stepsTotal.value, ")");
        };
        return Vue.createVNode("div", {
          "class": contentClass.value
        }, [Vue.createVNode("div", {
          "class": "".concat(guideClass.value, "__tooltip")
        }, [Vue.createVNode("div", {
          "class": "".concat(guideClass.value, "__title")
        }, [renderTitleNode()]), Vue.createVNode("div", {
          "class": "".concat(guideClass.value, "__desc")
        }, [renderBodyNode()])]), Vue.createVNode("div", {
          "class": footerClass.value
        }, [!hideSkip.value && !isLast.value && Vue.createVNode(button_index.Button, Vue.mergeProps({
          "key": "skip",
          "class": "".concat(guideClass.value, "__skip"),
          "theme": "light",
          "size": buttonSize.value,
          "variant": "base",
          "content": globalConfig.value.skip
        }, getCurrentCrossProps("skipButtonProps"), {
          "onClick": handleSkip
        }), null), !isLast.value && Vue.createVNode(button_index.Button, Vue.mergeProps({
          "key": "next",
          "class": "".concat(guideClass.value, "__next"),
          "theme": "primary",
          "size": buttonSize.value,
          "variant": "base"
        }, getCurrentCrossProps("nextButtonProps"), {
          "onClick": handleNext
        }), {
          content: function content() {
            return Vue.createVNode(Vue.Fragment, null, [renderButtonContent(getCurrentCrossProps("nextButtonProps"), globalConfig.value.next), !hideCounter.value && renderCounterNode()]);
          }
        }), isLast.value && Vue.createVNode(button_index.Button, Vue.mergeProps({
          "key": "back",
          "class": "".concat(guideClass.value, "__back"),
          "theme": "light",
          "size": buttonSize.value,
          "variant": "base",
          "content": globalConfig.value.back
        }, getCurrentCrossProps("backButtonProps"), {
          "onClick": handleBack
        }), null), isLast.value && Vue.createVNode(button_index.Button, Vue.mergeProps({
          "key": "finish",
          "class": "".concat(guideClass.value, "__finish"),
          "theme": "primary",
          "size": buttonSize.value,
          "variant": "base"
        }, (_finishButtonProps$va = finishButtonProps.value) !== null && _finishButtonProps$va !== void 0 ? _finishButtonProps$va : {}, {
          "onClick": handleFinish
        }), {
          content: function content() {
            return Vue.createVNode(Vue.Fragment, null, [renderButtonContent(finishButtonProps.value, globalConfig.value.finish), !hideCounter.value && renderCounterNode()]);
          }
        })])]);
      };
      var renderContentNode = function renderContentNode() {
        var content = currentStepInfo.value.content;
        var renderContent;
        if (isFunction.isFunction(content)) {
          renderContent = content(hWithParams(contentProps.value));
        } else if (context.slots.content) {
          renderContent = context.slots.content(hWithParams(contentProps.value));
        } else if (content) {
          renderContent = Vue.h(content, contentProps.value);
        }
        return renderContent;
      };
      var renderPopover = function renderPopover() {
        return Vue.createVNode(popover_index.Popover, stepProps.value, {
          triggerElement: function triggerElement() {
            return isPopover.value && Vue.createVNode("div", {
              "ref": referenceLayerRef,
              "class": _toConsumableArray__default["default"](popoverClass.value)
            }, null);
          },
          content: function content() {
            return isPopover.value && renderContentNode() ? renderContentNode() : renderStepContent();
          }
        });
      };
      var renderPopup = function renderPopup() {
        return Vue.createVNode(popup_index.Popup, stepProps.value, {
          default: function _default() {
            return isPopover.value && renderContentNode() ? renderContentNode() : renderStepContent();
          }
        });
      };
      var renderCurrentCustomHighlightContentNode = function renderCurrentCustomHighlightContentNode() {
        var highlightContent = currentStepInfo.value.highlightContent;
        var node = highlightContent;
        if (isFunction.isFunction(highlightContent)) {
          node = highlightContent(hWithParams());
        } else if (context.slots.highlightContent) {
          node = context.slots.highlightContent(hWithParams());
        } else if (context.slots["highlight-content"]) {
          node = context.slots["highlight-content"](hWithParams());
        }
        if (node) {
          if (!node.props) node.props = {};
          node.props.class = node.props.class || "";
        }
        return node;
      };
      return Vue.createVNode(Vue.Fragment, null, [actived.value && Vue.createVNode(Vue.Teleport, {
        "to": "body"
      }, {
        default: function _default() {
          return [Vue.createVNode("div", {
            "ref": overlayLayerRef,
            "class": "".concat(guideClass.value, "__overlay"),
            "style": {
              zIndex: "".concat(zIndex.value - 2)
            }
          }, null), Vue.createVNode("div", {
            "ref": highlightLayerRef,
            "class": [].concat(_toConsumableArray__default["default"](highlightClass.value), _toConsumableArray__default["default"](maskClass.value)),
            "style": {
              zIndex: "".concat(zIndex.value - 1)
            }
          }, [showCustomHighlightContent.value && renderCurrentCustomHighlightContentNode()]), Vue.createVNode("div", {
            "ref": popoverWrapperRef,
            "class": wrapperClass.value,
            "style": {
              zIndex: zIndex.value
            }
          }, [isPopover.value ? renderPopover() : renderPopup()])];
        }
      })]);
    };
  }
});

exports["default"] = _Guide;
