// tdesign-mobile-vue/es/drawer/plugin.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { ref, createApp, h, nextTick } from 'vue';
import _Drawer from './drawer.mjs';
import '../shared/index.mjs';
import { isBrowser } from '../shared/util.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../popup/index.mjs';
import '../popup/popup.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'tdesign-icons-vue-next';
import '../popup/props.mjs';
import '../overlay/index.mjs';
import '../overlay/overlay.mjs';
import '../shared/dom.mjs';
import 'lodash-es';
import '../config.mjs';
import '../overlay/props.mjs';
import '../hooks/useClass.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../hooks/tnode.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import './style/css.mjs';
import '../shared/component.mjs';
import '../shared/functions.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../hooks/useVModel.mjs';
import '../hooks/useLockScroll.mjs';
import '../_util/useTouch.mjs';
import '../_util/getScrollParent.mjs';
import '../hooks/useTeleport.mjs';
import '../popup/type.mjs';
import './props.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Drawer = function Drawer(options) {
  if (!isBrowser) return;
  var root = document.createElement("div");
  document.body.appendChild(root);
  var visible = ref(false);
  var props = ref({});
  var destroyOnClose = ref(false);
  createApp(function () {
    return h(_Drawer, _objectSpread(_objectSpread({}, options), {}, {
      visible: visible.value,
      destroyOnClose: destroyOnClose.value
    }, props.value));
  }).mount(root);
  var handler = {
    destroy: function destroy() {
      destroyOnClose.value = true;
      nextTick(function () {
        visible.value = false;
        document.body.removeChild(root);
      });
    },
    hide: function hide() {
      visible.value = false;
    },
    show: function show() {
      visible.value = true;
    },
    update: function update(options2) {
      props.value = options2;
    }
  };
  return handler;
};
Drawer.install = function (app) {
  app.config.globalProperties.$drawer = Drawer;
};
var DrawerPlugin = Drawer;

export { DrawerPlugin as default };
