// tdesign-mobile-vue/cjs/image/image.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var core = require('@vueuse/core');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var loading_index = require('../loading/index.js');
var config = require('../config.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var image_props = require('./props.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-055d28fd.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../loading/plugin.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _Image = Vue.defineComponent({
  name: "".concat(prefix, "-image"),
  props: image_props["default"],
  setup: function setup(props2, context) {
    var imageClass = hooks_useClass.usePrefixClass("image");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var closeIcon = Vue.createVNode(tdesignIconsVueNext.CloseIcon, {
      "size": "22px"
    }, null);
    var LoadingIcon = Vue.createVNode(loading_index.Loading, {
      "theme": "dots",
      "inheritColor": true
    }, null);
    var isLoading = Vue.ref(true);
    var isError = Vue.ref(false);
    var imageStyles = Vue.computed(function () {
      return {
        objectFit: props2.fit,
        objectPosition: props2.position
      };
    });
    var imageClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(imageClass.value), true), "".concat(imageClass.value, "--").concat(props2.shape), true);
    });
    var imageDOM = Vue.ref();
    var realSrc = Vue.ref("");
    Vue.watchEffect(function () {
      realSrc.value = props2.lazy ? "" : props2.src;
    });
    var _useIntersectionObser = core.useIntersectionObserver(imageDOM, function (_ref2, observerElement) {
        var _ref3 = _slicedToArray__default["default"](_ref2, 1),
          isIntersecting = _ref3[0].isIntersecting;
        if (isIntersecting && props2.lazy) {
          stop();
          realSrc.value = props2.src;
        }
      }),
      stop = _useIntersectionObser.stop;
    var handleImgLoadCompleted = function handleImgLoadCompleted(e) {
      var _props2$onLoad;
      (_props2$onLoad = props2.onLoad) === null || _props2$onLoad === void 0 || _props2$onLoad.call(props2, {
        e: e
      });
      isLoading.value = false;
    };
    var handleImgLoadError = function handleImgLoadError(e) {
      var _props2$onError;
      if (realSrc.value === "") {
        return;
      }
      (_props2$onError = props2.onError) === null || _props2$onError === void 0 || _props2$onError.call(props2, {
        e: e
      });
      isLoading.value = false;
      isError.value = true;
      if (props2.fallback) {
        realSrc.value = props2.fallback;
        isError.value = false;
      }
    };
    var maskContent = Vue.computed(function () {
      var statusContent = function statusContent() {
        var _context$slots, _context$slots2, _context$slots3, _context$slots4;
        if ((_context$slots = context.slots) !== null && _context$slots !== void 0 && _context$slots.loading && isLoading.value) {
          return renderTNodeJSX("loading");
        }
        if (!((_context$slots2 = context.slots) !== null && _context$slots2 !== void 0 && _context$slots2.loading) && isLoading.value) {
          return LoadingIcon;
        }
        if ((_context$slots3 = context.slots) !== null && _context$slots3 !== void 0 && _context$slots3.error && isError.value) {
          return renderTNodeJSX("error");
        }
        if (!((_context$slots4 = context.slots) !== null && _context$slots4 !== void 0 && _context$slots4.error) && isError.value) {
          return closeIcon;
        }
        return false;
      };
      var node = function node() {
        if (isLoading.value || isError.value) {
          if (props2.lazy && realSrc.value === "") {
            return;
          }
          return Vue.createVNode("div", {
            "class": "".concat(imageClass.value, "__mask")
          }, [statusContent()]);
        }
      };
      return node();
    });
    return function () {
      return Vue.createVNode("div", {
        "class": imageClasses.value
      }, [maskContent.value, Vue.createVNode("picture", null, [props2.srcset && Object.entries(props2.srcset).map(function (item, index) {
        return Vue.createVNode("source", {
          "key": index,
          "type": item[0],
          "srcset": item[1]
        }, null);
      }), Vue.createVNode("img", {
        "ref": imageDOM,
        "class": "".concat(imageClass.value, "__img"),
        "style": imageStyles.value,
        "src": realSrc.value,
        "alt": props2.alt,
        "referrerpolicy": props2.referrerpolicy,
        "onLoad": handleImgLoadCompleted,
        "onError": handleImgLoadError
      }, null)])]);
    };
  }
});

exports["default"] = _Image;
