// tdesign-mobile-vue/cjs/image/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  alt: {
    type: String,
    default: ""
  },
  error: {
    type: [String, Function]
  },
  fallback: {
    type: String,
    default: ""
  },
  fit: {
    type: String,
    default: "fill",
    validator: function validator(val) {
      if (!val) return true;
      return ["contain", "cover", "fill", "none", "scale-down"].includes(val);
    }
  },
  lazy: Boolean,
  loading: {
    type: [String, Function]
  },
  position: {
    type: String,
    default: "center"
  },
  referrerpolicy: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["no-referrer", "no-referrer-when-downgrade", "origin", "origin-when-cross-origin", "same-origin", "strict-origin", "strict-origin-when-cross-origin", "unsafe-url"].includes(val);
    }
  },
  shape: {
    type: String,
    default: "square",
    validator: function validator(val) {
      if (!val) return true;
      return ["circle", "round", "square"].includes(val);
    }
  },
  src: {
    type: String,
    default: ""
  },
  srcset: {
    type: Object
  },
  onError: Function,
  onLoad: Function
};

exports["default"] = props;
