// tdesign-mobile-vue/cjs/shared/useCountDown/utils.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var shared_util = require('../util.js');
require('@babel/runtime/helpers/slicedToArray');
require('../../hooks/useClass.js');
require('vue');
require('../../_chunks/dep-605cd804.js');
require('@babel/runtime/helpers/defineProperty');
require('../../_chunks/dep-6f53fb1b.js');
require('../../_common/js/global-config/mobile/default-config.js');
require('../../_common/js/global-config/mobile/locale/zh_CN.js');
require('../../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../../_chunks/dep-7220ff76.js');
require('dayjs');
require('../../_chunks/dep-05850262.js');
require('../../_chunks/dep-e9a63584.js');
require('../../_chunks/dep-53383176.js');
require('../../_chunks/dep-f84e6e17.js');
require('../../_chunks/dep-0c11c462.js');
require('../../_chunks/dep-0bd122a0.js');
require('../../_chunks/dep-b999c3d4.js');
require('../../_chunks/dep-159c7a94.js');
require('../../_chunks/dep-3cf5587c.js');
require('../../_chunks/dep-ea269588.js');
require('../../_chunks/dep-b633ec9e.js');
require('../../_chunks/dep-7aa6341b.js');
require('../../_chunks/dep-b8e8d654.js');
require('../../_chunks/dep-61ce8c0d.js');
require('../../_chunks/dep-5cf8bf50.js');
require('../../_chunks/dep-2457b474.js');
require('../../_chunks/dep-4b045875.js');
require('../../_chunks/dep-a6a4534e.js');
require('../../_chunks/dep-fef48622.js');
require('../../_chunks/dep-588e2e70.js');
require('../../_chunks/dep-8d30e8d1.js');
require('../../_chunks/dep-dd93e905.js');
require('../../_chunks/dep-d1ba5719.js');
require('../../_chunks/dep-055d28fd.js');

var TimeDataUnit = {
  DD: "\u5929",
  HH: "\u65F6",
  mm: "\u5206",
  ss: "\u79D2",
  SSS: "\u6BEB\u79D2"
};
var getRemainTimes = function getRemainTimes(time) {
  if (!time) {
    return {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
      milliseconds: 0
    };
  }
  var stime = Math.floor(time / 1e3);
  return {
    days: Math.floor(stime / 3600 / 24),
    hours: Math.floor(stime / 3600 % 24),
    minutes: Math.floor(stime / 60 % 60),
    seconds: stime % 60,
    milliseconds: time % 1e3
  };
};
var fillZero = function fillZero(num) {
  var isMillieconds = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  if (isMillieconds) {
    if (num >= 100) {
      return num;
    }
    return num >= 10 ? "0".concat(num) : "00".concat(num);
  }
  return num >= 10 ? num : "0".concat(num);
};
var getMark = function getMark(format, type) {
  var _format$split, _format$split$split;
  return format === null || format === void 0 || (_format$split = format.split) === null || _format$split === void 0 || (_format$split = _format$split.call(format, type)) === null || _format$split === void 0 || (_format$split = _format$split[1]) === null || _format$split === void 0 || (_format$split$split = _format$split.split) === null || _format$split$split === void 0 || (_format$split$split = _format$split$split.call(_format$split, "")) === null || _format$split$split === void 0 ? void 0 : _format$split$split[0];
};
var getShowTimes = function getShowTimes(times, format) {
  var _ref, _ref$toUpperCase, _format, _format2, _format3, _format4, _format5;
  var millisecond = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
  var splitWithUnit = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
  format = (_ref = format || "DD:HH:mm:ss") === null || _ref === void 0 || (_ref$toUpperCase = _ref.toUpperCase) === null || _ref$toUpperCase === void 0 ? void 0 : _ref$toUpperCase.call(_ref);
  millisecond && !format.includes(":SSS") && (format = format.concat(":SSS"));
  var showTimes = [];
  if (((_format = format) === null || _format === void 0 ? void 0 : _format.indexOf("DD")) > -1) {
    showTimes === null || showTimes === void 0 || showTimes.push({
      mark: splitWithUnit ? TimeDataUnit.DD : getMark(format, "DD"),
      value: fillZero(times === null || times === void 0 ? void 0 : times.days)
    });
  }
  if (((_format2 = format) === null || _format2 === void 0 ? void 0 : _format2.indexOf("HH")) > -1) {
    showTimes === null || showTimes === void 0 || showTimes.push({
      mark: splitWithUnit ? TimeDataUnit.HH : getMark(format, "HH"),
      value: fillZero(times === null || times === void 0 ? void 0 : times.hours)
    });
  }
  if (((_format3 = format) === null || _format3 === void 0 ? void 0 : _format3.indexOf("MM")) > -1) {
    showTimes === null || showTimes === void 0 || showTimes.push({
      mark: splitWithUnit ? TimeDataUnit.mm : getMark(format, "MM"),
      value: fillZero(times === null || times === void 0 ? void 0 : times.minutes)
    });
  }
  if (((_format4 = format) === null || _format4 === void 0 ? void 0 : _format4.indexOf("SS")) > -1) {
    showTimes === null || showTimes === void 0 || showTimes.push({
      mark: splitWithUnit ? TimeDataUnit.ss : getMark(format, "SS"),
      value: fillZero(times === null || times === void 0 ? void 0 : times.seconds)
    });
  }
  if (((_format5 = format) === null || _format5 === void 0 ? void 0 : _format5.indexOf("SSS")) > -1 || millisecond) {
    showTimes === null || showTimes === void 0 || showTimes.push({
      mark: splitWithUnit ? TimeDataUnit.SSS : getMark(format, "SSS"),
      value: fillZero(times === null || times === void 0 ? void 0 : times.milliseconds, true)
    });
  }
  return showTimes;
};
var getScreenFps = function (_ref2, _ref2$find) {
  if (!shared_util.isBrowser) return;
  var _window = window,
    requestAnimationFrame = _window.requestAnimationFrame,
    mozRequestAnimationFrame = _window.mozRequestAnimationFrame,
    webkitRequestAnimationFrame = _window.webkitRequestAnimationFrame;
  var nextFrame = (_ref2 = [requestAnimationFrame, mozRequestAnimationFrame, webkitRequestAnimationFrame]) === null || _ref2 === void 0 || (_ref2$find = _ref2.find) === null || _ref2$find === void 0 ? void 0 : _ref2$find.call(_ref2, function (fn) {
    return fn;
  });
  if (!nextFrame) {
    console.error("requestAnimationFrame is not supported!");
    return;
  }
  return function () {
    var targetCount = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 50;
    if (targetCount < 1) {
      return;
    }
    var count = 0;
    var beginDate = Date.now();
    return new Promise(function (resolve) {
      (function log() {
        nextFrame === null || nextFrame === void 0 || nextFrame(function () {
          if (++count >= targetCount) {
            var diffDate = Date.now() - beginDate;
            var fps = count / diffDate * 1e3;
            return resolve(fps);
          }
          log();
        });
      })();
    });
  };
}();

exports.TimeDataUnit = TimeDataUnit;
exports.fillZero = fillZero;
exports.getMark = getMark;
exports.getRemainTimes = getRemainTimes;
exports.getScreenFps = getScreenFps;
exports.getShowTimes = getShowTimes;
