// tdesign-mobile-vue/lib/table/hooks/useStyle.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { toRefs, computed } from 'vue';
import useClassName from './useClassName.js';
import '../../hooks/useClass.js';
import '../../config-provider/useConfig.js';
import 'lodash-es';
import '../../config-provider/context.js';
import '../../_common/js/global-config/mobile/default-config.js';
import '../../_common/js/global-config/mobile/locale/zh_CN.js';
import '../../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../../_chunks/dep-dc49adb8.js';
import 'dayjs';

function formatCSSUnit(unit) {
  if (!unit) return unit;
  return isNaN(Number(unit)) ? unit : "".concat(unit, "px");
}
function useStyle(props) {
  var _toRefs = toRefs(props),
    bordered = _toRefs.bordered,
    stripe = _toRefs.stripe,
    verticalAlign = _toRefs.verticalAlign,
    height = _toRefs.height,
    maxHeight = _toRefs.maxHeight,
    tableContentWidth = _toRefs.tableContentWidth;
  var _useClassName = useClassName(),
    tableBaseClass = _useClassName.tableBaseClass,
    tableAlignClasses = _useClassName.tableAlignClasses;
  var tableClasses = computed(function () {
    return [tableBaseClass.table, [tableAlignClasses[(verticalAlign === null || verticalAlign === void 0 ? void 0 : verticalAlign.value) || "middle"]], _defineProperty(_defineProperty(_defineProperty({}, tableBaseClass.bordered, bordered === null || bordered === void 0 ? void 0 : bordered.value), tableBaseClass.striped, stripe === null || stripe === void 0 ? void 0 : stripe.value), tableBaseClass.loading, props.loading)];
  });
  var tableContentStyles = computed(function () {
    return {
      height: formatCSSUnit(height === null || height === void 0 ? void 0 : height.value),
      maxHeight: formatCSSUnit(maxHeight === null || maxHeight === void 0 ? void 0 : maxHeight.value)
    };
  });
  var tableElementStyles = computed(function () {
    return {
      width: formatCSSUnit(tableContentWidth === null || tableContentWidth === void 0 ? void 0 : tableContentWidth.value)
    };
  });
  return {
    tableClasses: tableClasses,
    tableElementStyles: tableElementStyles,
    tableContentStyles: tableContentStyles
  };
}

export { useStyle as default, formatCSSUnit };
