// tdesign-mobile-vue/es/stepper/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  disableInput: Boolean,
  disabled: {
    type: Boolean,
    default: void 0
  },
  inputWidth: {
    type: Number
  },
  integer: {
    type: Boolean,
    default: true
  },
  max: {
    type: Number,
    default: 100
  },
  min: {
    type: Number,
    default: 0
  },
  size: {
    type: String,
    default: "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  step: {
    type: Number,
    default: 1
  },
  theme: {
    type: String,
    default: "normal",
    validator: function validator(val) {
      if (!val) return true;
      return ["normal", "filled", "outline"].includes(val);
    }
  },
  value: {
    type: [String, Number],
    default: void 0
  },
  modelValue: {
    type: [String, Number],
    default: void 0
  },
  defaultValue: {
    type: [String, Number],
    default: 0
  },
  onBlur: Function,
  onChange: Function,
  onFocus: Function,
  onOverlimit: Function
};

export { props as default };
