// tdesign-mobile-vue/es/stepper/stepper.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { defineComponent, toRefs, computed, createVNode } from 'vue';
import { RemoveIcon, AddIcon } from 'tdesign-icons-vue-next';
import config from '../config.mjs';
import props from './props.mjs';
import '../shared/index.mjs';
import { useFormDisabled } from '../form/hooks.mjs';
import useVModel from '../hooks/useVModel.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { formatNumber } from '../shared/util.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../shared/functions.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import 'lodash-es';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';

var prefix = config.prefix;
var _Stepper = defineComponent({
  name: "".concat(prefix, "-stepper"),
  props: props,
  setup: function setup(props2, context) {
    var stepperClass = usePrefixClass("stepper");
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue,
      min = _toRefs.min,
      max = _toRefs.max,
      step = _toRefs.step,
      integer = _toRefs.integer;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      stepperValue = _useVModel2[0],
      setStepperValue = _useVModel2[1];
    var disabled = useFormDisabled();
    var inputStyle = computed(function () {
      return props2.inputWidth ? {
        width: "".concat(props2.inputWidth, "px")
      } : "";
    });
    var isDisabled = function isDisabled(type) {
      if (disabled.value) return true;
      if (type === "minus" && Number(stepperValue.value) <= min.value) {
        return true;
      }
      if (type === "plus" && Number(stepperValue.value) >= max.value) {
        return true;
      }
      return false;
    };
    var getLen = function getLen(num) {
      var numStr = num.toString();
      return numStr.indexOf(".") === -1 ? 0 : numStr.split(".")[1].length;
    };
    var add = function add(a, b) {
      var maxLen = Math.max(getLen(a), getLen(b));
      var base = Math.pow(10, maxLen);
      return Math.round(a * base + b * base) / base;
    };
    var formatValue = function formatValue(value2) {
      return Math.max(Math.min(max.value, value2, Number.MAX_SAFE_INTEGER), min.value, Number.MIN_SAFE_INTEGER).toFixed(Math.max(getLen(step.value), getLen(value2)));
    };
    var updateValue = function updateValue(value2) {
      setStepperValue(formatNumber("".concat(value2), !integer.value));
    };
    var plusValue = function plusValue() {
      if (isDisabled("plus")) {
        var _props2$onOverlimit;
        (_props2$onOverlimit = props2.onOverlimit) === null || _props2$onOverlimit === void 0 || _props2$onOverlimit.call(props2, "plus");
        return;
      }
      updateValue(formatValue(add(Number(stepperValue.value), step.value)));
    };
    var minusValue = function minusValue() {
      if (isDisabled("minus")) {
        var _props2$onOverlimit2;
        (_props2$onOverlimit2 = props2.onOverlimit) === null || _props2$onOverlimit2 === void 0 || _props2$onOverlimit2.call(props2, "minus");
        return;
      }
      updateValue(formatValue(add(Number(stepperValue.value), -step.value)));
    };
    var handleInput = function handleInput(e) {
      var value2 = formatNumber(e.target.value, !integer.value);
      setStepperValue(value2);
    };
    var handleChange = function handleChange() {
      var formattedValue = formatValue(Number(stepperValue.value));
      updateValue(formattedValue);
    };
    var handleFocus = function handleFocus() {
      var _props2$onFocus;
      (_props2$onFocus = props2.onFocus) === null || _props2$onFocus === void 0 || _props2$onFocus.call(props2, Number(stepperValue.value));
    };
    var handleBlur = function handleBlur() {
      var _props2$onBlur;
      (_props2$onBlur = props2.onBlur) === null || _props2$onBlur === void 0 || _props2$onBlur.call(props2, Number(stepperValue.value));
    };
    return function () {
      return createVNode("div", {
        "class": ["".concat(stepperClass.value), "".concat(stepperClass.value, "--").concat(props2.size)]
      }, [createVNode("div", {
        "class": ["".concat(stepperClass.value, "__minus"), "".concat(stepperClass.value, "__minus--").concat(props2.theme), "".concat(stepperClass.value, "__icon--").concat(props2.size), "".concat(disabled.value || Number(stepperValue.value) <= props2.min ? "".concat(stepperClass.value, "--").concat(props2.theme, "-disabled") : "")],
        "onClick": minusValue
      }, [createVNode(RemoveIcon, {
        "class": "".concat(stepperClass.value, "__minus-icon")
      }, null)]), createVNode("input", {
        "value": stepperValue.value,
        "class": ["".concat(stepperClass.value, "__input"), "".concat(stepperClass.value, "__input--").concat(props2.theme), "".concat(stepperClass.value, "__input--").concat(props2.size), "".concat(disabled.value ? "".concat(stepperClass.value, "--").concat(props2.theme, "-disabled") : "")],
        "type": integer.value ? "tel" : "text",
        "inputmode": integer.value ? "numeric" : "decimal",
        "style": inputStyle.value,
        "disabled": props2.disableInput || disabled.value,
        "readonly": props2.disableInput,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onChange": handleChange
      }, null), createVNode("div", {
        "class": ["".concat(stepperClass.value, "__plus"), "".concat(stepperClass.value, "__plus--").concat(props2.theme), "".concat(stepperClass.value, "__icon--").concat(props2.size), "".concat(disabled.value || Number(stepperValue.value) >= props2.max ? "".concat(stepperClass.value, "--").concat(props2.theme, "-disabled") : "")],
        "onClick": plusValue
      }, [createVNode(AddIcon, {
        "class": "".concat(stepperClass.value, "__plus-icon")
      }, null)])]);
    };
  }
});

export { _Stepper as default };
