// tdesign-mobile-vue/esm/tabs/tabs.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, computed, ref, toRefs, Fragment, onMounted, onBeforeUnmount, onActivated, watch, nextTick, h, provide, readonly, createVNode, mergeProps } from 'vue';
import { isFunction } from 'lodash-es';
import config from '../config.js';
import props from './props.js';
import TTabNavItem from './tab-nav-item.js';
import useVModel from '../hooks/useVModel.js';
import { preventDefault } from '../shared/dom.js';
import { Sticky } from '../sticky/index.js';
import { Badge } from '../badge/index.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useCommonClassName } from '../hooks/useCommonClassName.js';
import './tab-panel-props.js';
import '../sticky/sticky.js';
import '@vueuse/core';
import '../sticky/props.js';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/hover.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../sticky/style';
import '../sticky/type.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../badge/badge.js';
import '../badge/props.js';
import '../badge/style';
import '../badge/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _Tabs = defineComponent({
  name: "".concat(prefix, "-tabs"),
  props: props,
  setup: function setup(props2) {
    var renderTNodeJSX = useTNodeJSX();
    var tabsClass = usePrefixClass("tabs");
    var _useCommonClassName = useCommonClassName(),
      SIZE = _useCommonClassName.SIZE;
    var stickyProps = computed(function () {
      return _objectSpread(_objectSpread({}, props2.stickyProps), {}, {
        disabled: !props2.sticky
      });
    });
    var activeClass = "".concat(tabsClass.value, "__item--active");
    var disabledClass = "".concat(tabsClass.value, "__item--disabled");
    var tabsClasses = computed(function () {
      return ["".concat(tabsClass.value), props2.size && SIZE.value[props2.size]];
    });
    var navClasses = ref(["".concat(tabsClass.value, "__nav")]);
    var startX = ref(0);
    var startY = ref(0);
    var endX = ref(0);
    var endY = ref(0);
    var canMove = ref(true);
    var tabIndex = computed(function () {
      var index = 0;
      for (var i = 0; i < itemProps.value.length; i++) {
        if (itemProps.value[i].value === currentValue.value) {
          index = i;
          break;
        }
      }
      return index;
    });
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      currentValue = _useVModel2[0],
      setCurrentValue = _useVModel2[1];
    var previousValue = ref();
    var itemProps = computed(function () {
      if (props2.list) {
        return props2.list;
      }
      var children = renderTNodeJSX("default");
      var res = [];
      var _label = {};
      var _handler = function handler(children2) {
        children2 === null || children2 === void 0 || children2.forEach(function (child, index) {
          var _child$children;
          if (child.type === Fragment) {
            _handler(child.children);
          } else {
            res.push(child);
          }
          if ((_child$children = child.children) !== null && _child$children !== void 0 && _child$children.label) {
            _label[index] = child.children.label()[0] || null;
          }
        });
      };
      _handler(children);
      children = res.filter(function (child) {
        return child.type.name === "".concat(prefix, "-tab-panel");
      });
      return children.map(function (item, index) {
        return _objectSpread(_objectSpread({}, item.props), {}, {
          label: function label() {
            return _label[index] || item.props.label;
          }
        });
      });
    });
    var valueList = computed(function () {
      return itemProps.value.map(function (v) {
        return v.value;
      });
    });
    var currentIndex = computed(function () {
      return valueList.value.indexOf(currentValue.value);
    });
    var navScroll = ref();
    var navWrap = ref();
    var navLine = ref();
    var lineStyle = ref({
      opacity: 0
    });
    var moveToActiveTab = function moveToActiveTab() {
      if (navWrap.value && navLine.value && props2.showBottomLine) {
        var tab = navWrap.value.querySelector(".".concat(activeClass));
        if (!tab) return;
        var line = navLine.value;
        var tabInner = tab.querySelector(".".concat(prefix, "-badge"));
        var style = {
          opacity: 1
        };
        if (props2.bottomLineMode === "auto") {
          style.width = "".concat(Number(tabInner === null || tabInner === void 0 ? void 0 : tabInner.offsetWidth), "px");
          style.transform = "translateX(".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetLeft) + Number(tabInner === null || tabInner === void 0 ? void 0 : tabInner.offsetLeft), "px)");
        } else if (props2.bottomLineMode === "full") {
          style.width = "".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetWidth), "px");
          style.transform = "translateX(".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetLeft), "px)");
        } else {
          style.transform = "translateX(".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetLeft) + (Number(tab === null || tab === void 0 ? void 0 : tab.offsetWidth) - Number(line === null || line === void 0 ? void 0 : line.offsetWidth)) / 2, "px)");
        }
        var isInit = previousValue.value === void 0;
        previousValue.value = currentValue.value;
        if (isInit) {
          style.transitionDuration = "0s";
        } else if (props2.animation) {
          style.transitionDuration = "".concat(props2.animation.duration, "ms");
        }
        lineStyle.value = style;
      }
      if (navScroll.value) {
        var _tab = navScroll.value.querySelector(".".concat(activeClass));
        if (!_tab) return;
        var tabLeft = _tab === null || _tab === void 0 ? void 0 : _tab.offsetLeft;
        var tabWidth = _tab === null || _tab === void 0 ? void 0 : _tab.offsetWidth;
        var navScrollWidth = navScroll.value.offsetWidth;
        var scrollDistance = tabLeft - navScrollWidth / 2 + tabWidth / 2;
        navScroll.value.scrollTo({
          left: scrollDistance,
          behavior: "smooth"
        });
      }
    };
    onMounted(function () {
      window.addEventListener("resize", moveToActiveTab, false);
      setTimeout(function () {
        moveToActiveTab();
      }, 300);
    });
    onBeforeUnmount(function () {
      window.removeEventListener("resize", moveToActiveTab);
    });
    onActivated(function () {
      moveToActiveTab();
    });
    watch(currentValue, function () {
      nextTick(function () {
        moveToActiveTab();
      });
    });
    watch(itemProps, function () {
      nextTick(function () {
        moveToActiveTab();
      });
    }, {
      deep: true
    });
    var handleTabClick = function handleTabClick(event, item) {
      var _props2$onClick;
      var value2 = item.value,
        disabled = item.disabled;
      if (disabled || currentValue.value === value2) {
        return false;
      }
      var label = isFunction(item.label) ? item.label(h).toString() : item.label;
      setCurrentValue(value2, label);
      (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, value2, label);
      nextTick(function () {
        moveToActiveTab();
      });
    };
    var handlerScroll = function handlerScroll(context) {
      var scrollTop = context.scrollTop,
        isFixed = context.isFixed;
      if (props2.stickyProps) {
        var _props2$onScroll;
        (_props2$onScroll = props2.onScroll) === null || _props2$onScroll === void 0 || _props2$onScroll.call(props2, scrollTop, isFixed);
      }
    };
    var handleTouchstart = function handleTouchstart(e) {
      if (!props2.swipeable) return;
      canMove.value = true;
      startX.value = e.targetTouches[0].clientX;
      startY.value = e.targetTouches[0].clientY;
    };
    var handleTouchmove = function handleTouchmove(e) {
      if (!props2.swipeable) return;
      if (!canMove.value) return;
      endX.value = e.targetTouches[0].clientX;
      endY.value = e.targetTouches[0].clientY;
      var dValueX = Math.abs(startX.value - endX.value);
      var dValueY = Math.abs(startY.value - endY.value);
      if (tabIndex.value >= 0 && tabIndex.value < itemProps.value.length) {
        if (dValueX > dValueY) {
          preventDefault(e, false);
          if (dValueX <= 40) return;
          if (startX.value > endX.value) {
            if (tabIndex.value >= itemProps.value.length - 1) return;
            handleTabClick(e, itemProps.value[tabIndex.value + 1]);
          } else if (startX.value < endX.value) {
            if (tabIndex.value <= 0) return;
            handleTabClick(e, itemProps.value[tabIndex.value - 1]);
          }
        }
      }
    };
    var handleTouchend = function handleTouchend() {
      if (!props2.swipeable) return;
      canMove.value = false;
      startX.value = 0;
      endX.value = 0;
      startY.value = 0;
      endY.value = 0;
    };
    provide("currentValue", readonly(currentValue));
    var readerNav = function readerNav() {
      return itemProps.value.map(function (item, index) {
        var badgeProps = item["badge-props"] || item.badgeProps;
        return createVNode("div", {
          "class": _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(tabsClass.value, "__item ").concat(tabsClass.value, "__item--top"), true), "".concat(tabsClass.value, "__item--evenly"), props2.spaceEvenly), activeClass, item.value === currentValue.value), disabledClass, item.disabled), "".concat(tabsClass.value, "__item--").concat(props2.theme), true),
          "onClick": function onClick(e) {
            return handleTabClick(e, item);
          }
        }, [createVNode(Badge, badgeProps, {
          default: function _default() {
            return [createVNode("div", {
              "class": _defineProperty(_defineProperty({}, "".concat(tabsClass.value, "__item-inner ").concat(tabsClass.value, "__item-inner--").concat(props2.theme), true), "".concat(tabsClass.value, "__item-inner--active"), props2.theme === "tag" && item.value === currentValue.value)
            }, [createVNode(TTabNavItem, {
              "label": item.label,
              "icon": item.icon
            }, null)])];
          }
        }), props2.theme === "card" && index === currentIndex.value - 1 && createVNode("div", {
          "class": "".concat(tabsClass.value, "__item-prefix")
        }, null), props2.theme === "card" && index === currentIndex.value + 1 && createVNode("div", {
          "class": "".concat(tabsClass.value, "__item-suffix")
        }, null)]);
      });
    };
    return function () {
      return createVNode("div", {
        "class": tabsClasses.value
      }, [createVNode(Sticky, mergeProps(stickyProps.value, {
        "onScroll": handlerScroll
      }), {
        default: function _default() {
          return [createVNode("div", {
            "class": navClasses.value
          }, [createVNode("div", {
            "ref": navScroll,
            "class": ["".concat(tabsClass.value, "__scroll"), "".concat(tabsClass.value, "__scroll--").concat(props2.theme), _defineProperty({}, "".concat(tabsClass.value, "__scroll--split"), props2.split)]
          }, [createVNode("div", {
            "ref": navWrap,
            "class": "".concat(tabsClass.value, "__wrapper ").concat(tabsClass.value, "__wrapper--").concat(props2.theme)
          }, [readerNav(), props2.theme === "line" && props2.showBottomLine && createVNode("div", {
            "ref": navLine,
            "class": "".concat(tabsClass.value, "__track ").concat(tabsClass.value, "__track--top"),
            "style": lineStyle.value
          }, null)])])])];
        }
      }), renderTNodeJSX("middle"), createVNode("div", {
        "class": "".concat(tabsClass.value, "__content"),
        "onTouchstart": handleTouchstart,
        "onTouchmove": handleTouchmove,
        "onTouchend": handleTouchend
      }, [renderTNodeJSX("default")])]);
    };
  }
});

export { _Tabs as default };
