// tdesign-mobile-vue/esm/collapse/collapse-panel.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, inject, computed, onMounted, createVNode } from 'vue';
import { ChevronUpIcon, ChevronDownIcon } from 'tdesign-icons-vue-next';
import { Cell } from '../cell/index.js';
import props from './collapse-panel-props.js';
import config from '../config.js';
import { findIndex } from './util.js';
import { useTNodeJSX, useContent } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '../cell/cell.js';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash-es';
import '../shared/component.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/dom.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/hover.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../cell/props.js';
import '../form/hooks.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../cell/cell-group.js';
import '../cell/cell-group-props.js';
import '../cell/style';
import '../cell/type.js';

var prefix = config.prefix;
var _CollapsePanel = defineComponent({
  name: "".concat(prefix, "-collapse-panel"),
  components: {
    TCell: Cell
  },
  props: props,
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var renderTNodeJSX = useTNodeJSX();
    var renderContent = useContent();
    var collapsePanelClass = usePrefixClass("collapse-panel");
    var parent = inject("collapse");
    var renderParentTNode = inject("renderParentTNode");
    var disabled = computed(function () {
      return (parent === null || parent === void 0 ? void 0 : parent.disabled.value) || props2.disabled;
    });
    var rootClass = computed(function () {
      return _defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(collapsePanelClass.value), true), "".concat(collapsePanelClass.value, "--").concat(props2.placement), true), "".concat(collapsePanelClass.value, "--active"), isActive.value), "".concat(collapsePanelClass.value, "--disabled"), disabled.value);
    });
    var isActive = computed(function () {
      return findIndex(props2.value, parent === null || parent === void 0 ? void 0 : parent.activeValue.value) > -1;
    });
    var updatePanelValue = function updatePanelValue(args) {
      if (props2.value != null) {
        parent === null || parent === void 0 || parent.onPanelChange(props2.value, args);
      }
    };
    var handleClick = function handleClick(_ref3) {
      var e = _ref3.e;
      e === null || e === void 0 || e.stopPropagation();
      if (disabled.value) {
        return;
      }
      updatePanelValue({
        e: e
      });
    };
    onMounted(function () {
      if (parent !== null && parent !== void 0 && parent.defaultExpandAll) {
        updatePanelValue();
      }
    });
    var renderDefaultIcon = function renderDefaultIcon() {
      if (props2.placement === "bottom") {
        return isActive.value ? createVNode(ChevronUpIcon, null, null) : createVNode(ChevronDownIcon, null, null);
      }
      return isActive.value ? createVNode(ChevronDownIcon, null, null) : createVNode(ChevronUpIcon, null, null);
    };
    var panelExpandIcon = computed(function () {
      return slots.expandIcon || props2.expandIcon;
    });
    var renderRightIcon = function renderRightIcon() {
      var tNodeRender = panelExpandIcon.value === void 0 ? renderParentTNode : renderTNodeJSX;
      return createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__header-icon")
      }, [tNodeRender("expandIcon", renderDefaultIcon())]);
    };
    var renderPanelContent = function renderPanelContent() {
      var panelContent = renderContent("default", "content");
      if (props2.destroyOnCollapse && !isActive.value) {
        return null;
      }
      return createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__content")
      }, [panelContent]);
    };
    return function () {
      var headerContent = renderTNodeJSX("header");
      var noteContent = renderTNodeJSX("headerRightContent");
      var _leftIcon = renderTNodeJSX("headerLeftIcon");
      return createVNode("div", {
        "class": rootClass.value
      }, [createVNode(Cell, {
        "class": ["".concat(collapsePanelClass.value, "__title"), "".concat(collapsePanelClass.value, "__header"), "".concat(collapsePanelClass.value, "__header--").concat(props2.placement), _defineProperty({}, "".concat(collapsePanelClass.value, "__header--expanded"), isActive.value)],
        "onClick": handleClick
      }, {
        leftIcon: function leftIcon() {
          return _leftIcon;
        },
        title: function title() {
          return headerContent;
        },
        note: function note() {
          return noteContent;
        },
        rightIcon: function rightIcon() {
          return renderRightIcon();
        }
      }), createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__body"),
        "style": {
          gridTemplateRows: isActive.value ? "1fr" : "0fr"
        }
      }, [createVNode("div", {
        "class": "".concat(collapsePanelClass.value, "__inner")
      }, [renderPanelContent()])])]);
    };
  }
});

export { _CollapsePanel as default };
