// tdesign-mobile-vue/cjs/navbar/navbar.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var navbar_props = require('./props.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _Navbar = Vue.defineComponent({
  name: "".concat(prefix, "-navbar"),
  components: {
    TChevronLeftIcon: tdesignIconsVueNext.ChevronLeftIcon
  },
  props: navbar_props["default"],
  emits: ["left-click", "right-click"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var classPrefix = hooks_useClass.usePrefixClass();
    var navbarClass = hooks_useClass.usePrefixClass("navbar");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var animationSuffix = Vue.computed(function () {
      return props2.animation ? "-animation" : "";
    });
    var navClass = Vue.computed(function () {
      return [navbarClass.value, _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(navbarClass.value, "--fixed"), props2.fixed), "".concat(classPrefix.value, "-safe-area-top "), props2.safeAreaInsetTop), props2.visible ? "".concat(navbarClass.value, "--visible").concat(animationSuffix.value) : "".concat(navbarClass.value, "--hide").concat(animationSuffix.value)];
    });
    var navStyle = Vue.computed(function () {
      return "position: ".concat(props2.fixed ? "fixed" : "relative", ";");
    });
    var handleLeftClick = function handleLeftClick() {
      var _props2$onLeftClick;
      (_props2$onLeftClick = props2.onLeftClick) === null || _props2$onLeftClick === void 0 || _props2$onLeftClick.call(props2);
    };
    var handleRightClick = function handleRightClick() {
      var _props2$onRightClick;
      (_props2$onRightClick = props2.onRightClick) === null || _props2$onRightClick === void 0 || _props2$onRightClick.call(props2);
    };
    return function () {
      var fixed = props2.fixed,
        titleMaxLength = props2.titleMaxLength,
        title = props2.title,
        leftArrow = props2.leftArrow;
      var renderRightContent = function renderRightContent() {
        var rightContent = renderTNodeJSX("right");
        if (!rightContent) {
          return null;
        }
        return Vue.createVNode("div", {
          "class": "".concat(navbarClass.value, "__right"),
          "onClick": handleRightClick
        }, [rightContent]);
      };
      var renderCapsuleContent = function renderCapsuleContent() {
        var capsuleContent = renderTNodeJSX("capsule");
        if (!capsuleContent) {
          return null;
        }
        return Vue.createVNode("div", {
          "class": "".concat(navbarClass.value, "__capsule")
        }, [capsuleContent]);
      };
      var renderTitleContent = function renderTitleContent() {
        var isStringTitle = typeof title === "string" && !slots.title;
        var titleContent = renderTNodeJSX("title");
        if (!titleContent) {
          return null;
        }
        if (titleMaxLength != null && title) {
          if (titleMaxLength <= 0) {
            console.warn("titleMaxLength must be greater than 0");
          } else {
            titleContent = title.length <= titleMaxLength ? title : "".concat(title.slice(0, titleMaxLength), "...");
          }
        }
        return isStringTitle ? Vue.createVNode("span", {
          "class": "".concat(navbarClass.value, "__center-title")
        }, [titleContent]) : titleContent;
      };
      return Vue.createVNode("div", {
        "class": navClass.value,
        "style": navStyle.value
      }, [fixed && Vue.createVNode("div", {
        "class": "".concat(navbarClass.value, "____placeholder")
      }, null), Vue.createVNode("div", {
        "class": "".concat(navbarClass.value, "__content")
      }, [Vue.createVNode("div", {
        "class": "".concat(navbarClass.value, "__left"),
        "onClick": handleLeftClick
      }, [leftArrow && Vue.createVNode(Vue.resolveComponent("t-chevron-left-icon"), {
        "class": "".concat(navbarClass.value, "__left-arrow")
      }, null), renderTNodeJSX("left"), renderCapsuleContent()]), Vue.createVNode("div", {
        "class": "".concat(navbarClass.value, "__center")
      }, [renderTitleContent()]), renderRightContent()])]);
    };
  }
});

exports["default"] = _Navbar;
