// tdesign-mobile-vue/cjs/tab-bar/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var TabBarProps = {
  bordered: {
    type: Boolean,
    default: true
  },
  fixed: {
    type: Boolean,
    default: true
  },
  safeAreaInsetBottom: {
    type: Boolean,
    default: true
  },
  shape: {
    type: String,
    default: "normal",
    validator: function validator(val) {
      if (!val) return true;
      return ["normal", "round"].includes(val);
    }
  },
  split: {
    type: Boolean,
    default: true
  },
  theme: {
    type: String,
    default: "normal",
    validator: function validator(val) {
      if (!val) return true;
      return ["normal", "tag"].includes(val);
    }
  },
  value: {
    type: [String, Number, Array],
    default: void 0
  },
  modelValue: {
    type: [String, Number, Array],
    default: void 0
  },
  defaultValue: {
    type: [String, Number, Array]
  },
  onChange: Function
};

exports["default"] = TabBarProps;
