// tdesign-mobile-vue/es/swiper/swiper-item.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, inject, ref, getCurrentInstance, computed, watch, onMounted, onUnmounted, createVNode } from 'vue';
import { useElementBounding } from '@vueuse/core';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import config from '../config.mjs';
import '../config-provider/useConfig.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import 'lodash-es';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';

var prefix = config.prefix;
var _SwiperItem = defineComponent({
  name: "".concat(prefix, "-swiper-item"),
  setup: function setup() {
    var _inject = inject("parent"),
      addChild = _inject.addChild,
      removeChild = _inject.removeChild,
      isVertical = _inject.isVertical,
      root = _inject.root,
      items = _inject.items,
      setContainerHeight = _inject.setContainerHeight;
    var swiperItemClass = usePrefixClass("swiper-item");
    var readerTNodeJSX = useTNodeJSX();
    var selfRef = ref(null);
    var rootStyle = ref("");
    var instance = getCurrentInstance();
    var direction = computed(function () {
      return isVertical.value ? "Y" : "X";
    });
    var calcTranslateStyle = function calcTranslateStyle(index, activeIndex) {
      var _root$value, _root$value2;
      var distance = (_root$value = (_root$value2 = root.value) === null || _root$value2 === void 0 ? void 0 : _root$value2[isVertical.value ? "offsetHeight" : "offsetWidth"]) !== null && _root$value !== void 0 ? _root$value : 0;
      var lastItemIndex = items.value.length - 1;
      var step = index - activeIndex;
      if (activeIndex === lastItemIndex && index === 0) {
        step = 1;
      }
      if (activeIndex === 0 && index === lastItemIndex && index !== 1) {
        step = -1;
      }
      if (activeIndex === index) step = 0;
      rootStyle.value = "transform: translate".concat(direction.value, "(").concat(step * distance, "px)");
    };
    var _useElementBounding = useElementBounding(selfRef),
      height = _useElementBounding.height;
    watch(height, function (val) {
      setContainerHeight(val);
    });
    onMounted(function () {
      addChild({
        proxy: instance === null || instance === void 0 ? void 0 : instance.proxy,
        uid: instance === null || instance === void 0 ? void 0 : instance.uid,
        calcTranslateStyle: calcTranslateStyle
      });
    });
    onUnmounted(function () {
      removeChild(instance === null || instance === void 0 ? void 0 : instance.uid);
    });
    return function () {
      return createVNode("div", {
        "ref": selfRef,
        "class": swiperItemClass.value,
        "style": rootStyle.value
      }, [readerTNodeJSX("default")]);
    };
  }
});

export { _SwiperItem as default };
