// tdesign-mobile-vue/es/shared/index.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

export { NOOP } from './functions.mjs';
export { convertUnit, formatNumber, isBrowser, reconvertUnit, sleep, toCamel, uniqueFactory } from './util.mjs';
export { extendAPI, withInstall } from './component.mjs';
export { renderContent, renderTNode } from './render.mjs';
export { addClass, elementInViewport, getAttach, getSSRAttach, getWindowScroll, getWindowSize, hasClass, preventDefault, removeClass, stopPropagation } from './dom.mjs';
export { default as TNode } from './render-tnode.mjs';
export { useCountDown } from './useCountDown/index.mjs';
export { useTouch } from './useTouch/index.mjs';
export { getScrollParent, useScrollParent } from './useScrollParent/index.mjs';
export { useTestUtils } from './useTest/index.mjs';
export { useClickAway } from './useClickAway/index.mjs';
export { default as Hover } from './hover.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'lodash-es';
import '../hooks/useClass.mjs';
import 'vue';
import '../config-provider/useConfig.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import './useCountDown/utils.mjs';
import '../_chunks/dep-fb069758.mjs';
