// tdesign-mobile-vue/es/dialog/dialog.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _toConsumableArray } from '../_chunks/dep-fb069758.mjs';
import { defineComponent, computed, createVNode, mergeProps } from 'vue';
import { CloseIcon } from 'tdesign-icons-vue-next';
import { get, isString, isObject } from 'lodash-es';
import { Button } from '../button/index.mjs';
import { Popup } from '../popup/index.mjs';
import config from '../config.mjs';
import props from './props.mjs';
import { useTNodeJSX, useContent } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../button/button.mjs';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../button/props.mjs';
import '../form/hooks.mjs';
import '../button/type.mjs';
import '../popup/popup.mjs';
import '../popup/props.mjs';
import '../overlay/index.mjs';
import '../overlay/overlay.mjs';
import '../overlay/props.mjs';
import '../hooks/useVModel.mjs';
import '../hooks/useLockScroll.mjs';
import '../_util/useTouch.mjs';
import '../_util/getScrollParent.mjs';
import '../hooks/useTeleport.mjs';
import '../popup/type.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var Dialog = defineComponent({
  name: "".concat(prefix, "-dialog"),
  props: props,
  emits: ["update:visible", "confirm", "overlay-click", "cancel", "close", "closed"],
  setup: function setup(props2, context) {
    var dialogClass = usePrefixClass("dialog");
    var renderTNodeJSX = useTNodeJSX();
    var renderContent = useContent();
    var isTextStyleBtn = computed(function () {
      return [props2 === null || props2 === void 0 ? void 0 : props2.confirmBtn, props2 === null || props2 === void 0 ? void 0 : props2.cancelBtn].concat(_toConsumableArray((props2 === null || props2 === void 0 ? void 0 : props2.actions) || [])).some(function (item) {
        return get(item, "variant") === "text";
      });
    });
    var footerClass = computed(function () {
      return ["".concat(dialogClass.value, "__footer"), _defineProperty(_defineProperty({}, "".concat(dialogClass.value, "__footer--column"), props2.buttonLayout === "vertical"), "".concat(dialogClass.value, "__footer--full"), isTextStyleBtn.value && get(props2.actions, "length", 0) === 0)];
    });
    var buttonClass = computed(function () {
      return ["".concat(dialogClass.value, "__button"), _defineProperty(_defineProperty({}, "".concat(dialogClass.value, "__button--").concat(props2.buttonLayout), !isTextStyleBtn.value), "".concat(dialogClass.value, "__button--text"), isTextStyleBtn.value)];
    });
    var rootStyles = computed(function () {
      return {
        width: isString(props2.width) ? props2.width : "".concat(props2.width, "px")
      };
    });
    var handleClose = function handleClose(args) {
      var e = args.e;
      context.emit("update:visible", false);
      context.emit("close", {
        e: e,
        trigger: "close-btn"
      });
    };
    var handleClosed = function handleClosed() {
      context.emit("closed");
    };
    var handleConfirm = function handleConfirm(e) {
      var _context$emit;
      context.emit("update:visible", false);
      (_context$emit = context.emit) === null || _context$emit === void 0 || _context$emit.call(context, "confirm", {
        e: e
      });
    };
    var handleCancel = function handleCancel(e) {
      context.emit("update:visible", false);
      context.emit("close", {
        e: e,
        trigger: "cancel"
      });
      context.emit("cancel", {
        e: e
      });
    };
    var handleOverlayClick = function handleOverlayClick(args) {
      var e = args.e;
      if (!props2.closeOnOverlayClick) {
        return;
      }
      context.emit("update:visible", false);
      context.emit("close", {
        e: e,
        trigger: "overlay"
      });
      context.emit("overlay-click", {
        e: e
      });
    };
    var calcBtn = function calcBtn(btn) {
      if (isString(btn)) {
        return {
          content: btn
        };
      }
      if (isObject(btn)) {
        return btn;
      }
      return {};
    };
    var confirmBtnProps = computed(function () {
      return _objectSpread({
        theme: "primary"
      }, calcBtn(props2.confirmBtn));
    });
    var cancelBtnProps = computed(function () {
      return _objectSpread({
        theme: isTextStyleBtn.value ? "default" : "light"
      }, calcBtn(props2.cancelBtn));
    });
    var actionsBtnProps = computed(function () {
      var _props2$actions;
      return (_props2$actions = props2.actions) === null || _props2$actions === void 0 ? void 0 : _props2$actions.map(function (item) {
        return calcBtn(item);
      });
    });
    var renderButtonNode = function renderButtonNode(btnType, btnProps, handleClick) {
      if (actionsBtnProps.value) return null;
      var btnNode = renderTNodeJSX(btnType);
      if (!btnNode) return null;
      if (context.slots[btnType]) {
        return btnNode;
      }
      return createVNode(Button, mergeProps(btnProps, {
        "class": buttonClass.value,
        "onClick": handleClick
      }), null);
    };
    return function () {
      var renderTitleNode = function renderTitleNode() {
        var titleNode = renderTNodeJSX("title");
        if (!titleNode) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(dialogClass.value, "__header")
        }, [titleNode]);
      };
      var renderContentNode = function renderContentNode() {
        var contentNode = renderContent("default", "content");
        if (!contentNode) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(dialogClass.value, "__body")
        }, [createVNode("div", {
          "class": "".concat(dialogClass.value, "__body-text")
        }, [contentNode])]);
      };
      var renderActionsNode = function renderActionsNode() {
        if (actionsBtnProps.value) {
          return actionsBtnProps.value.map(function (item, index) {
            return createVNode(Button, mergeProps({
              "key": index
            }, item, {
              "class": buttonClass.value,
              "onClick": handleCancel
            }), null);
          });
        }
        return renderTNodeJSX("actions");
      };
      var renderCancelBtnNode = function renderCancelBtnNode() {
        return renderButtonNode("cancelBtn", cancelBtnProps.value, handleCancel);
      };
      var renderConfirmBtnNode = function renderConfirmBtnNode() {
        return renderButtonNode("confirmBtn", confirmBtnProps.value, handleConfirm);
      };
      return createVNode(Popup, {
        "class": "".concat(dialogClass.value, "__wrapper"),
        "visible": props2.visible,
        "placement": "center",
        "show-overlay": props2.showOverlay,
        "overlay-props": props2.overlayProps,
        "prevent-scroll-through": props2.preventScrollThrough,
        "destroy-on-close": props2.destroyOnClose,
        "z-index": props2.zIndex,
        "onClose": handleOverlayClick,
        "onClosed": handleClosed
      }, {
        default: function _default() {
          return [createVNode("div", {
            "class": "".concat(dialogClass.value, " ").concat(context.attrs.class || ""),
            "style": rootStyles.value
          }, [renderTNodeJSX("top"), props2.closeBtn && createVNode("div", {
            "class": "".concat(dialogClass.value, "__close-btn")
          }, [createVNode(CloseIcon, {
            "onClick": handleClose
          }, null)]), createVNode("div", {
            "class": "".concat(dialogClass.value, "__content")
          }, [renderTitleNode(), renderContentNode()]), renderTNodeJSX("middle"), createVNode("div", {
            "class": footerClass.value
          }, [renderActionsNode(), renderCancelBtnNode(), renderConfirmBtnNode()])])];
        }
      });
    };
  }
});

export { Dialog as default };
