// tdesign-mobile-vue/es/button/button.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, computed, createVNode, mergeProps, withDirectives, resolveDirective } from 'vue';
import { Loading } from '../loading/index.mjs';
import '../shared/index.mjs';
import ButtonProps from './props.mjs';
import config from '../config.mjs';
import { useFormDisabled } from '../form/hooks.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useTNodeJSX, useContent } from '../hooks/tnode.mjs';
import Hover from '../shared/hover.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import 'lodash-es';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../shared/component.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';

var prefix = config.prefix;
var _Button = defineComponent({
  name: "".concat(prefix, "-button"),
  directives: {
    Hover: Hover
  },
  props: ButtonProps,
  setup: function setup(props) {
    var buttonClass = usePrefixClass("button");
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var isDisabled = useFormDisabled();
    var hoverDisabled = computed(function () {
      return isDisabled.value || props.loading;
    });
    var buttonClasses = computed(function () {
      return ["".concat(buttonClass.value), "".concat(buttonClass.value, "--size-").concat(props.size), "".concat(buttonClass.value, "--").concat(props.variant), _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(buttonClass.value, "--").concat(props.theme), props.theme), "".concat(buttonClass.value, "--").concat(props.shape), props.shape), "".concat(buttonClass.value, "--ghost"), props.ghost), "".concat(buttonClass.value, "--block"), props.block), "".concat(buttonClass.value, "--disabled"), isDisabled.value), "".concat(buttonClass.value, "--loading"), props.loading)];
    });
    var handleClick = function handleClick(e) {
      if (!props.loading && !isDisabled.value) {
        var _props$onClick;
        (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, e);
      } else {
        e.stopPropagation();
      }
    };
    return function () {
      var suffix = renderTNodeJSX("suffix");
      var readerIcon = function readerIcon() {
        if (props.loading) {
          return createVNode(Loading, mergeProps({
            "inherit-color": true
          }, props.loadingProps), null);
        }
        return renderTNodeJSX("icon");
      };
      var readerContent = function readerContent() {
        var content = renderTNodeContent("default", "content");
        return createVNode("span", {
          "class": "".concat(buttonClass.value, "__content")
        }, [content]);
      };
      return withDirectives(createVNode("button", {
        "class": buttonClasses.value,
        "role": "button",
        "type": props.type,
        "disabled": isDisabled.value,
        "aria-disabled": isDisabled.value,
        "onClick": handleClick
      }, [readerIcon(), readerContent(), suffix]), [[resolveDirective("hover"), {
        className: "".concat(buttonClass.value, "--hover"),
        disabledHover: hoverDisabled.value
      }]]);
    };
  }
});

export { _Button as default };
