// tdesign-mobile-vue/cjs/message/message.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var link_index = require('../link/index.js');
var message_props = require('./props.js');
var config = require('../config.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
var isString = require('../_chunks/dep-588e2e70.js');
var shared_util = require('../shared/util.js');
require('../link/link.js');
require('../link/props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-f84e6e17.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../_chunks/dep-055d28fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var iconDefault = {
  info: Vue.h(tdesignIconsVueNext.InfoCircleFilledIcon),
  success: Vue.h(tdesignIconsVueNext.CheckCircleFilledIcon),
  warning: Vue.h(tdesignIconsVueNext.ErrorCircleFilledIcon),
  error: Vue.h(tdesignIconsVueNext.ErrorCircleFilledIcon)
};
var closeBtnDefault = Vue.h(tdesignIconsVueNext.CloseIcon);
var messageIndex = -1;
var Message = Vue.defineComponent({
  name: "".concat(prefix, "-message"),
  props: message_props["default"],
  setup: function setup(props2, context) {
    var messageClass = hooks_useClass.usePrefixClass("message");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var renderContent = hooks_tnode.useContent();
    var messageRef = Vue.ref(null);
    var rect = Vue.ref();
    var state = Vue.reactive({
      duration: 0,
      offset: 0,
      listWidth: 0,
      itemWidth: 0,
      scroll: {
        marquee: false,
        speed: 50,
        loop: -1,
        delay: 300
      }
    });
    var _toRefs = Vue.toRefs(props2),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](visible, modelValue, props2.defaultVisible),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      currentVisible = _useVModel2[0],
      setVisible = _useVModel2[1];
    var rootClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(messageClass.value), true), "".concat(messageClass.value, "--").concat(props2.theme), true), "".concat(messageClass.value, "-align--").concat(props2.align), !!props2.align);
    });
    var textWrapClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(messageClass.value, "__text-wrap"), true), "".concat(messageClass.value, "__text-nowrap"), props2.marquee);
    });
    var getGap = function getGap() {
      var _rect$value;
      if (props2.single) {
        return 0;
      }
      var gap = typeof props2.gap === "boolean" ? 12 : shared_util.reconvertUnit(props2.gap);
      return (gap + (((_rect$value = rect.value) === null || _rect$value === void 0 ? void 0 : _rect$value.height) || 0)) * messageIndex;
    };
    var changeNumToStr = function changeNumToStr() {
      var arr = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
      return arr.map(function (item, index) {
        var value = shared_util.reconvertUnit(item);
        return index === 0 ? "".concat(value + getGap(), "px") : "".concat(value, "px");
      });
    };
    var getMessageStylesOffset = function getMessageStylesOffset(offset) {
      var _messageRef$value;
      rect.value = (_messageRef$value = messageRef.value) === null || _messageRef$value === void 0 ? void 0 : _messageRef$value.getBoundingClientRect();
      var arr = changeNumToStr(offset);
      return {
        top: arr[0],
        right: arr[1],
        left: arr[1]
      };
    };
    var rootStyles = Vue.computed(function () {
      var offset = props2.offset;
      var offsetStyle = offset ? getMessageStylesOffset(offset) : [];
      return _objectSpread({
        zIndex: props2.zIndex
      }, offsetStyle);
    });
    var linkContent = Vue.computed(function () {
      var _context$slots;
      if (typeof props2.link === "function" || (_context$slots = context.slots) !== null && _context$slots !== void 0 && _context$slots.link) {
        return renderTNodeJSX("link");
      }
      if (isObject.isObject(props2.link) || isString.isString(props2.link)) {
        var _link = isObject.isObject(props2.link) ? _objectSpread({}, props2.link) : {
          content: props2.link
        };
        return Vue.h(link_index.Link, _objectSpread({
          theme: "primary"
        }, _link));
      }
      return null;
    });
    var animateStyle = Vue.computed(function () {
      return {
        transform: state.offset ? "translateX(".concat(state.offset, "px)") : "",
        transitionDuration: "".concat(state.duration, "s"),
        transitionTimingFunction: "linear"
      };
    });
    var textWrapDOM = Vue.ref();
    var textDOM = Vue.ref();
    var handleScrolling = function handleScrolling() {
      var _ref3, _props2$marquee, _props2$marquee2, _props2$marquee3, _props2$marquee4, _props2$marquee5, _props2$marquee6;
      if (!(props2 !== null && props2 !== void 0 && props2.marquee) || ((_ref3 = isObject.isObject(props2 === null || props2 === void 0 ? void 0 : props2.marquee) && (props2 === null || props2 === void 0 ? void 0 : props2.marquee)) === null || _ref3 === void 0 ? void 0 : _ref3.loop) === 0) {
        return;
      }
      var _state$scroll = state.scroll,
        loop = _state$scroll.loop,
        speed = _state$scroll.speed,
        delay = _state$scroll.delay;
      state.scroll = {
        marquee: true,
        loop: Math.max(props2.marquee === true || ((_props2$marquee = props2.marquee) === null || _props2$marquee === void 0 ? void 0 : _props2$marquee.loop) == null ? loop : (_props2$marquee2 = props2.marquee) === null || _props2$marquee2 === void 0 ? void 0 : _props2$marquee2.loop, -1),
        speed: Math.max(props2.marquee === true || ((_props2$marquee3 = props2.marquee) === null || _props2$marquee3 === void 0 ? void 0 : _props2$marquee3.speed) == null ? speed : (_props2$marquee4 = props2.marquee) === null || _props2$marquee4 === void 0 ? void 0 : _props2$marquee4.speed, 1),
        delay: Math.max(props2.marquee === true || ((_props2$marquee5 = props2.marquee) === null || _props2$marquee5 === void 0 ? void 0 : _props2$marquee5.delay) == null ? delay : (_props2$marquee6 = props2.marquee) === null || _props2$marquee6 === void 0 ? void 0 : _props2$marquee6.delay, 0)
      };
      state.offset = 0;
      setTimeout(function () {
        var _textWrapDOM$value, _textDOM$value;
        var textWrapDOMWidth = (_textWrapDOM$value = textWrapDOM.value) === null || _textWrapDOM$value === void 0 ? void 0 : _textWrapDOM$value.getBoundingClientRect().width;
        var textDOMWidth = (_textDOM$value = textDOM.value) === null || _textDOM$value === void 0 ? void 0 : _textDOM$value.getBoundingClientRect().width;
        state.offset = -textDOMWidth;
        state.duration = textDOMWidth / state.scroll.speed;
        state.listWidth = textWrapDOMWidth;
        state.itemWidth = textDOMWidth;
      }, state.scroll.delay);
    };
    var handleTransitionend = function handleTransitionend() {
      resetTransition();
      if (state.scroll.loop === -1) {
        return;
      }
      state.scroll.loop = --state.scroll.loop;
      if (state.scroll.loop === 0) {
        state.scroll.marquee = false;
      }
    };
    var resetTransition = function resetTransition() {
      state.duration = 0;
      state.offset = state.listWidth;
      setTimeout(function () {
        state.offset = -state.itemWidth;
        state.duration = (state.itemWidth + state.listWidth) / state.scroll.speed;
      }, 0);
    };
    var onLinkClick = function onLinkClick(e) {
      var _props2$onLinkClick;
      (_props2$onLinkClick = props2.onLinkClick) === null || _props2$onLinkClick === void 0 || _props2$onLinkClick.call(props2, {
        e: e
      });
    };
    var onCloseBtnClick = function onCloseBtnClick(e) {
      var _props2$onCloseBtnCli;
      (_props2$onCloseBtnCli = props2.onCloseBtnClick) === null || _props2$onCloseBtnCli === void 0 || _props2$onCloseBtnCli.call(props2, {
        e: e
      });
      setVisible(false);
    };
    var handleDuration = function handleDuration() {
      if (props2.duration > 0) {
        setTimeout(function () {
          var _props2$onDurationEnd;
          (_props2$onDurationEnd = props2.onDurationEnd) === null || _props2$onDurationEnd === void 0 || _props2$onDurationEnd.call(props2);
          setVisible(false);
        }, props2.duration);
      }
    };
    Vue.onMounted(function () {
      Vue.nextTick(function () {
        if (currentVisible.value) {
          handleScrolling();
        }
      });
    });
    var updateMessageIndex = function updateMessageIndex() {
      var isAdd = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      if (props2.single) return;
      if (isAdd) {
        messageIndex += 1;
      } else {
        messageIndex -= 1;
      }
    };
    Vue.watch(function () {
      return currentVisible.value;
    }, function (val) {
      if (val === false) {
        updateMessageIndex(false);
        return;
      }
      updateMessageIndex(true);
      setVisible(true);
      handleDuration();
      Vue.nextTick(handleScrolling);
    });
    return function () {
      var prefixIconContent = renderTNodeJSX("icon", {
        defaultNode: iconDefault === null || iconDefault === void 0 ? void 0 : iconDefault[props2.theme || "info"]
      });
      var computedContent = renderContent("default", "content");
      var closeBtnContent = renderTNodeJSX("closeBtn", {
        defaultNode: closeBtnDefault
      });
      return Vue.createVNode(Vue.Transition, {
        "name": "message"
      }, {
        default: function _default() {
          return [currentVisible.value && Vue.createVNode("div", {
            "ref": messageRef,
            "class": rootClasses.value,
            "style": rootStyles.value
          }, [prefixIconContent && Vue.createVNode("div", {
            "class": "".concat(messageClass.value, "__icon--left")
          }, [prefixIconContent]), Vue.createVNode("div", {
            "ref": textWrapDOM,
            "class": textWrapClasses.value
          }, [Vue.createVNode("div", {
            "ref": textDOM,
            "class": "".concat(messageClass.value, "__text"),
            "style": state.scroll.marquee ? animateStyle.value : "",
            "onTransitionend": handleTransitionend
          }, [computedContent])]), linkContent.value && Vue.createVNode("div", {
            "class": "".concat(messageClass.value, "__link"),
            "onClick": onLinkClick
          }, [linkContent.value]), closeBtnContent && Vue.createVNode("div", {
            "class": ["".concat(messageClass.value, "__close-wrap"), "".concat(messageClass.value, "__icon--right")],
            "onClick": onCloseBtnClick
          }, [closeBtnContent])])];
        }
      });
    };
  }
});

exports["default"] = Message;
