// tdesign-mobile-vue/cjs/cell/cell.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var config = require('../config.js');
var cell_props = require('./props.js');
var form_hooks = require('../form/hooks.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
var shared_hover = require('../shared/hover.js');
require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var _Cell = Vue.defineComponent({
  name: "".concat(prefix, "-cell"),
  directives: {
    Hover: shared_hover["default"]
  },
  props: cell_props["default"],
  setup: function setup(props2) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var renderTNodeContent = hooks_tnode.useContent();
    var disabled = form_hooks.useFormDisabled();
    var cellClass = hooks_useClass.usePrefixClass("cell");
    var cellClasses = Vue.computed(function () {
      return ["".concat(cellClass.value), _defineProperty__default["default"]({}, "".concat(cellClass.value, "--borderless"), !props2.bordered)];
    });
    var hoverDisabled = Vue.computed(function () {
      return !props2.hover || disabled.value;
    });
    var handleClick = function handleClick(e) {
      if (!disabled.value) {
        var _props2$onClick;
        (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, {
          e: e
        });
      }
    };
    var renderImage = function renderImage() {
      if (typeof props2.image === "string") {
        return Vue.createVNode("img", {
          "src": props2.image,
          "class": "".concat(cellClass.value, "__left-image")
        }, null);
      }
      var image = renderTNodeJSX("image");
      return image;
    };
    var renderLeft = function renderLeft() {
      var leftIcon = renderTNodeJSX("leftIcon");
      return Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__left")
      }, [leftIcon && Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__left-icon")
      }, [leftIcon]), renderImage()]);
    };
    var renderTitle = function renderTitle() {
      var title = renderTNodeJSX("title");
      if (!title) {
        return null;
      }
      var description = renderTNodeJSX("description");
      return Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__title")
      }, [Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__title-text")
      }, [title, props2.required && Vue.createVNode("span", {
        "class": "".concat(cellClass.value, "--required")
      }, ["\xA0*"])]), description && Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__description")
      }, [description])]);
    };
    var renderRight = function renderRight() {
      var rightIcon = props2.arrow ? Vue.createVNode(tdesignIconsVueNext.ChevronRightIcon, null, null) : renderTNodeJSX("rightIcon");
      if (!rightIcon) {
        return null;
      }
      return Vue.createVNode("div", {
        "class": ["".concat(cellClass.value, "__right"), "".concat(cellClass.value, "__right--").concat(props2.align)]
      }, [Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__right-icon")
      }, [rightIcon])]);
    };
    return function () {
      var note = renderTNodeContent("default", "note");
      return Vue.withDirectives(Vue.createVNode("div", {
        "class": cellClasses.value,
        "onClick": handleClick
      }, [renderLeft(), renderTitle(), note && Vue.createVNode("div", {
        "class": "".concat(cellClass.value, "__note")
      }, [note]), renderRight()]), [[Vue.resolveDirective("hover"), {
        className: "".concat(cellClass.value, "--hover"),
        disabledHover: hoverDisabled.value
      }]]);
    };
  }
});

exports["default"] = _Cell;
