// tdesign-mobile-vue/esm/search/search.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, ref, toRefs, computed, nextTick, createVNode, h, mergeProps } from 'vue';
import { SearchIcon, CloseCircleFilledIcon } from 'tdesign-icons-vue-next';
import { useFocus } from '@vueuse/core';
import config from '../config.js';
import { preventDefault } from '../shared/dom.js';
import searchProps from './props.js';
import useVModel from '../hooks/useVModel.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import useLengthLimit from '../hooks/useLengthLimit.js';
import { ENTER_REG } from '../_common/js/common.js';
import _Cell from '../cell/cell.js';
import 'lodash-es';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-197a6425.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-aa2c083c.js';
import 'dayjs';
import '../_chunks/dep-089e3bbb.js';
import '../config-provider/type.js';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../_common/js/log/index.js';
import '../_common/js/log/log.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/render.js';
import '../shared/render-tnode.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../shared/useCountDown/utils.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '../shared/hover.js';
import '../cell/props.js';
import '../form/hooks.js';

var prefix = config.prefix;
var _Search = defineComponent({
  name: "".concat(prefix, "-search"),
  props: searchProps,
  setup: function setup(props, context) {
    var renderTNodeJSX = useTNodeJSX();
    var classPrefix = usePrefixClass();
    var searchClass = usePrefixClass("search");
    var isShowResultList = ref(false);
    var inputRef = ref();
    var _useFocus = useFocus(inputRef, {
        initialValue: props.focus
      }),
      focused = _useFocus.focused;
    var _toRefs = toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      searchValue = _useVModel2[0],
      setSearchValue = _useVModel2[1];
    var boxClasses = computed(function () {
      return ["".concat(searchClass.value, "__input-box"), "".concat(searchClass.value, "__input-box--").concat(props.shape), _defineProperty({}, "".concat(classPrefix.value, "-is-focused"), focused.value)];
    });
    var inputClasses = computed(function () {
      return ["".concat(classPrefix.value, "-input__keyword"), _defineProperty({}, "".concat(searchClass.value, "--center"), props.center)];
    });
    var limitParams = computed(function () {
      return {
        value: [void 0, null].includes(searchValue.value) ? void 0 : String(searchValue.value),
        maxlength: Number(props.maxlength),
        maxcharacter: props.maxcharacter,
        allowInputOverMax: false
      };
    });
    var _useLengthLimit = useLengthLimit(limitParams),
      getValueByLimitNumber = _useLengthLimit.getValueByLimitNumber;
    var setInputValue = function setInputValue(v) {
      var input = inputRef.value;
      var sV = String(v);
      if (!input) {
        return;
      }
      if (input.value !== sV) {
        input.value = sV;
      }
    };
    var inputValueChangeHandle = function inputValueChangeHandle(e) {
      var value2 = e.target.value;
      setSearchValue(getValueByLimitNumber(value2));
      nextTick(function () {
        return setInputValue(searchValue.value);
      });
    };
    var handleInput = function handleInput(e) {
      isShowResultList.value = true;
      if (e instanceof InputEvent) {
        var checkInputType = e.inputType && e.inputType === "insertCompositionText";
        if (e.isComposing || checkInputType) return;
      }
      inputValueChangeHandle(e);
    };
    var handleClear = function handleClear(e) {
      var _props$onClear;
      setSearchValue("");
      focused.value = true;
      (_props$onClear = props.onClear) === null || _props$onClear === void 0 || _props$onClear.call(props, {
        e: e
      });
    };
    var handleFocus = function handleFocus(e) {
      var _props$onFocus;
      (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, {
        value: searchValue.value,
        e: e
      });
    };
    var handleBlur = function handleBlur(e) {
      var _props$onBlur;
      (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, {
        value: searchValue.value,
        e: e
      });
    };
    var handleCompositionend = function handleCompositionend(e) {
      inputValueChangeHandle(e);
    };
    var handleAction = function handleAction(e) {
      var _props$onActionClick;
      (_props$onActionClick = props.onActionClick) === null || _props$onActionClick === void 0 || _props$onActionClick.call(props, {
        e: e
      });
    };
    var handleSearch = function handleSearch(e) {
      if (ENTER_REG.test(e.code) || ENTER_REG.test(e.key)) {
        var _inputRef$value, _props$onSubmit;
        preventDefault(e, false);
        (_inputRef$value = inputRef.value) === null || _inputRef$value === void 0 || _inputRef$value.blur();
        (_props$onSubmit = props.onSubmit) === null || _props$onSubmit === void 0 || _props$onSubmit.call(props, {
          value: searchValue.value,
          e: e
        });
      }
    };
    return function () {
      var readerLeftIcon = function readerLeftIcon() {
        var leftIcon = renderTNodeJSX("leftIcon");
        if (leftIcon === "search") {
          return createVNode(SearchIcon, {
            "size": "24px"
          }, null);
        }
        return renderTNodeJSX("leftIcon");
      };
      var readerClear = function readerClear() {
        if (props.clearable && searchValue.value && (props.clearTrigger === "always" || props.clearTrigger === "focus" && focused.value)) {
          return createVNode("div", {
            "class": "".concat(searchClass.value, "__clear"),
            "onClick": handleClear
          }, [createVNode(CloseCircleFilledIcon, {
            "size": "24"
          }, null)]);
        }
        return null;
      };
      var readerAction = function readerAction() {
        var action = renderTNodeJSX("action");
        if (action) {
          return createVNode("div", {
            "class": "".concat(searchClass.value, "__search-action ").concat(classPrefix.value, "-class-action"),
            "onClick": handleAction
          }, [action]);
        }
        return null;
      };
      var onSelectResultItem = function onSelectResultItem(params) {
        isShowResultList.value = false;
        setSearchValue(params.item);
      };
      var highlightSearchValue = function highlightSearchValue(item, searchValue2) {
        var parts = item.split(new RegExp("(".concat(searchValue2, ")"), "gi"));
        return parts.map(function (part, index) {
          return part.toLowerCase() === searchValue2.toLowerCase() ? createVNode("span", {
            "key": index,
            "class": "".concat(searchClass.value, "__result-item--highLight")
          }, [part]) : part;
        });
      };
      var listNodes = function listNodes(params) {
        return h(_Cell, {
          key: params.index,
          class: "".concat(searchClass.value, "__result-item"),
          onClick: function onClick() {
            return onSelectResultItem({
              item: params.item
            });
          }
        }, {
          title: function title() {
            return highlightSearchValue(params.item, searchValue.value);
          }
        });
      };
      var extraProps = {
        enterkeyhint: "search"
      };
      return createVNode("div", null, [createVNode("div", {
        "class": "".concat(searchClass.value)
      }, [createVNode("div", {
        "class": boxClasses.value
      }, [readerLeftIcon(), createVNode("input", mergeProps({
        "ref": inputRef,
        "value": searchValue.value,
        "type": "search",
        "class": inputClasses.value,
        "autofocus": props.focus,
        "placeholder": props.placeholder,
        "readonly": props.readonly,
        "disabled": props.disabled,
        "onKeypress": handleSearch,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onCompositionend": handleCompositionend
      }, extraProps), null), readerClear()]), readerAction()]), isShowResultList.value && createVNode("div", {
        "class": "".concat(searchClass.value, "__result-list")
      }, [props.resultList.map(function (item, index) {
        return listNodes({
          item: item,
          index: index
        });
      })])]);
    };
  }
});

export { _Search as default };
