// tdesign-mobile-vue/es/message/message.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { h, defineComponent, ref, reactive, toRefs, computed, onMounted, nextTick, watch, createVNode, Transition } from 'vue';
import { InfoCircleFilledIcon, CheckCircleFilledIcon, ErrorCircleFilledIcon, CloseIcon } from 'tdesign-icons-vue-next';
import { isObject, isString } from 'lodash-es';
import { Link } from '../link/index.mjs';
import props from './props.mjs';
import config from '../config.mjs';
import '../shared/index.mjs';
import useVModel from '../hooks/useVModel.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useTNodeJSX, useContent } from '../hooks/tnode.mjs';
import { reconvertUnit } from '../shared/util.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../link/link.mjs';
import '../link/props.mjs';
import '../form/hooks.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../link/type.mjs';
import '../shared/component.mjs';
import '../shared/functions.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var iconDefault = {
  info: h(InfoCircleFilledIcon),
  success: h(CheckCircleFilledIcon),
  warning: h(ErrorCircleFilledIcon),
  error: h(ErrorCircleFilledIcon)
};
var closeBtnDefault = h(CloseIcon);
var messageIndex = -1;
var Message = defineComponent({
  name: "".concat(prefix, "-message"),
  props: props,
  setup: function setup(props2, context) {
    var messageClass = usePrefixClass("message");
    var renderTNodeJSX = useTNodeJSX();
    var renderContent = useContent();
    var messageRef = ref(null);
    var rect = ref();
    var state = reactive({
      duration: 0,
      offset: 0,
      listWidth: 0,
      itemWidth: 0,
      scroll: {
        marquee: false,
        speed: 50,
        loop: -1,
        delay: 300
      }
    });
    var _toRefs = toRefs(props2),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(visible, modelValue, props2.defaultVisible),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      currentVisible = _useVModel2[0],
      setVisible = _useVModel2[1];
    var rootClasses = computed(function () {
      return _defineProperty(_defineProperty(_defineProperty({}, "".concat(messageClass.value), true), "".concat(messageClass.value, "--").concat(props2.theme), true), "".concat(messageClass.value, "-align--").concat(props2.align), !!props2.align);
    });
    var textWrapClasses = computed(function () {
      return _defineProperty(_defineProperty({}, "".concat(messageClass.value, "__text-wrap"), true), "".concat(messageClass.value, "__text-nowrap"), props2.marquee);
    });
    var getGap = function getGap() {
      var _rect$value;
      if (props2.single) {
        return 0;
      }
      var gap = typeof props2.gap === "boolean" ? 12 : reconvertUnit(props2.gap);
      return (gap + (((_rect$value = rect.value) === null || _rect$value === void 0 ? void 0 : _rect$value.height) || 0)) * messageIndex;
    };
    var changeNumToStr = function changeNumToStr() {
      var arr = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
      return arr.map(function (item, index) {
        var value = reconvertUnit(item);
        return index === 0 ? "".concat(value + getGap(), "px") : "".concat(value, "px");
      });
    };
    var getMessageStylesOffset = function getMessageStylesOffset(offset) {
      var _messageRef$value;
      rect.value = (_messageRef$value = messageRef.value) === null || _messageRef$value === void 0 ? void 0 : _messageRef$value.getBoundingClientRect();
      var arr = changeNumToStr(offset);
      return {
        top: arr[0],
        right: arr[1],
        left: arr[1]
      };
    };
    var rootStyles = computed(function () {
      var offset = props2.offset;
      var offsetStyle = offset ? getMessageStylesOffset(offset) : [];
      return _objectSpread({
        zIndex: props2.zIndex
      }, offsetStyle);
    });
    var linkContent = computed(function () {
      var _context$slots;
      if (typeof props2.link === "function" || (_context$slots = context.slots) !== null && _context$slots !== void 0 && _context$slots.link) {
        return renderTNodeJSX("link");
      }
      if (isObject(props2.link) || isString(props2.link)) {
        var _link = isObject(props2.link) ? _objectSpread({}, props2.link) : {
          content: props2.link
        };
        return h(Link, _objectSpread({
          theme: "primary"
        }, _link));
      }
      return null;
    });
    var animateStyle = computed(function () {
      return {
        transform: state.offset ? "translateX(".concat(state.offset, "px)") : "",
        transitionDuration: "".concat(state.duration, "s"),
        transitionTimingFunction: "linear"
      };
    });
    var textWrapDOM = ref();
    var textDOM = ref();
    var handleScrolling = function handleScrolling() {
      var _ref3, _props2$marquee, _props2$marquee2, _props2$marquee3, _props2$marquee4, _props2$marquee5, _props2$marquee6;
      if (!(props2 !== null && props2 !== void 0 && props2.marquee) || ((_ref3 = isObject(props2 === null || props2 === void 0 ? void 0 : props2.marquee) && (props2 === null || props2 === void 0 ? void 0 : props2.marquee)) === null || _ref3 === void 0 ? void 0 : _ref3.loop) === 0) {
        return;
      }
      var _state$scroll = state.scroll,
        loop = _state$scroll.loop,
        speed = _state$scroll.speed,
        delay = _state$scroll.delay;
      state.scroll = {
        marquee: true,
        loop: Math.max(props2.marquee === true || ((_props2$marquee = props2.marquee) === null || _props2$marquee === void 0 ? void 0 : _props2$marquee.loop) == null ? loop : (_props2$marquee2 = props2.marquee) === null || _props2$marquee2 === void 0 ? void 0 : _props2$marquee2.loop, -1),
        speed: Math.max(props2.marquee === true || ((_props2$marquee3 = props2.marquee) === null || _props2$marquee3 === void 0 ? void 0 : _props2$marquee3.speed) == null ? speed : (_props2$marquee4 = props2.marquee) === null || _props2$marquee4 === void 0 ? void 0 : _props2$marquee4.speed, 1),
        delay: Math.max(props2.marquee === true || ((_props2$marquee5 = props2.marquee) === null || _props2$marquee5 === void 0 ? void 0 : _props2$marquee5.delay) == null ? delay : (_props2$marquee6 = props2.marquee) === null || _props2$marquee6 === void 0 ? void 0 : _props2$marquee6.delay, 0)
      };
      state.offset = 0;
      setTimeout(function () {
        var _textWrapDOM$value, _textDOM$value;
        var textWrapDOMWidth = (_textWrapDOM$value = textWrapDOM.value) === null || _textWrapDOM$value === void 0 ? void 0 : _textWrapDOM$value.getBoundingClientRect().width;
        var textDOMWidth = (_textDOM$value = textDOM.value) === null || _textDOM$value === void 0 ? void 0 : _textDOM$value.getBoundingClientRect().width;
        state.offset = -textDOMWidth;
        state.duration = textDOMWidth / state.scroll.speed;
        state.listWidth = textWrapDOMWidth;
        state.itemWidth = textDOMWidth;
      }, state.scroll.delay);
    };
    var handleTransitionend = function handleTransitionend() {
      resetTransition();
      if (state.scroll.loop === -1) {
        return;
      }
      state.scroll.loop = --state.scroll.loop;
      if (state.scroll.loop === 0) {
        state.scroll.marquee = false;
      }
    };
    var resetTransition = function resetTransition() {
      state.duration = 0;
      state.offset = state.listWidth;
      setTimeout(function () {
        state.offset = -state.itemWidth;
        state.duration = (state.itemWidth + state.listWidth) / state.scroll.speed;
      }, 0);
    };
    var onLinkClick = function onLinkClick(e) {
      var _props2$onLinkClick;
      (_props2$onLinkClick = props2.onLinkClick) === null || _props2$onLinkClick === void 0 || _props2$onLinkClick.call(props2, {
        e: e
      });
    };
    var onCloseBtnClick = function onCloseBtnClick(e) {
      var _props2$onCloseBtnCli;
      (_props2$onCloseBtnCli = props2.onCloseBtnClick) === null || _props2$onCloseBtnCli === void 0 || _props2$onCloseBtnCli.call(props2, {
        e: e
      });
      setVisible(false);
    };
    var handleDuration = function handleDuration() {
      if (props2.duration > 0) {
        setTimeout(function () {
          var _props2$onDurationEnd;
          (_props2$onDurationEnd = props2.onDurationEnd) === null || _props2$onDurationEnd === void 0 || _props2$onDurationEnd.call(props2);
          setVisible(false);
        }, props2.duration);
      }
    };
    onMounted(function () {
      nextTick(function () {
        if (currentVisible.value) {
          handleScrolling();
        }
      });
    });
    var updateMessageIndex = function updateMessageIndex() {
      var isAdd = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
      if (props2.single) return;
      if (isAdd) {
        messageIndex += 1;
      } else {
        messageIndex -= 1;
      }
    };
    watch(function () {
      return currentVisible.value;
    }, function (val) {
      if (val === false) {
        updateMessageIndex(false);
        return;
      }
      updateMessageIndex(true);
      setVisible(true);
      handleDuration();
      nextTick(handleScrolling);
    });
    return function () {
      var prefixIconContent = renderTNodeJSX("icon", {
        defaultNode: iconDefault === null || iconDefault === void 0 ? void 0 : iconDefault[props2.theme || "info"]
      });
      var computedContent = renderContent("default", "content");
      var closeBtnContent = renderTNodeJSX("closeBtn", {
        defaultNode: closeBtnDefault
      });
      return createVNode(Transition, {
        "name": "message"
      }, {
        default: function _default() {
          return [currentVisible.value && createVNode("div", {
            "ref": messageRef,
            "class": rootClasses.value,
            "style": rootStyles.value
          }, [prefixIconContent && createVNode("div", {
            "class": "".concat(messageClass.value, "__icon--left")
          }, [prefixIconContent]), createVNode("div", {
            "ref": textWrapDOM,
            "class": textWrapClasses.value
          }, [createVNode("div", {
            "ref": textDOM,
            "class": "".concat(messageClass.value, "__text"),
            "style": state.scroll.marquee ? animateStyle.value : "",
            "onTransitionend": handleTransitionend
          }, [computedContent])]), linkContent.value && createVNode("div", {
            "class": "".concat(messageClass.value, "__link"),
            "onClick": onLinkClick
          }, [linkContent.value]), closeBtnContent && createVNode("div", {
            "class": ["".concat(messageClass.value, "__close-wrap"), "".concat(messageClass.value, "__icon--right")],
            "onClick": onCloseBtnClick
          }, [closeBtnContent])])];
        }
      });
    };
  }
});

export { Message as default };
