// tdesign-mobile-vue/lib/count-down/count-down.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, computed, ref, onMounted, onBeforeUnmount, createVNode, Fragment } from 'vue';
import config from '../config.js';
import CountDownProps from './props.js';
import { useCountDown } from '../shared/useCountDown/index.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/util.js';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash-es';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../config-provider/useConfig.js';
import '@babel/runtime/helpers/defineProperty';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';

var prefix = config.prefix;
var _CountDown = defineComponent({
  name: "".concat(prefix, "-count-down"),
  props: CountDownProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var countDownClass = usePrefixClass("count-down");
    var countDownClasses = computed(function () {
      return [countDownClass.value, "".concat(countDownClass.value, "--").concat(props.theme), "".concat(countDownClass.value, "--").concat(props.size)];
    });
    var visibility = ref(true);
    var visibilitychangeListener = function visibilitychangeListener() {
      visibility.value = !document.hidden;
    };
    onMounted(function () {
      document.addEventListener("visibilitychange", visibilitychangeListener, false);
    });
    onBeforeUnmount(function () {
      document.removeEventListener("visibilitychange", visibilitychangeListener, false);
    });
    var _useCountDown = useCountDown(props, visibility),
      showTimes = _useCountDown.showTimes;
    return function () {
      var renderContent = function renderContent() {
        var content = renderTNodeJSX("content");
        if (content !== "default" && content) {
          return content;
        }
        return showTimes.map(function (item) {
          return createVNode(Fragment, null, [createVNode("span", {
            "class": "".concat(countDownClass.value, "__item")
          }, [item.value]), item.mark && createVNode("span", {
            "class": ["".concat(countDownClass.value, "__split"), "".concat(countDownClass.value, "__split--").concat(props.splitWithUnit ? "text" : "dot")]
          }, [item.mark])]);
        });
      };
      return createVNode("div", {
        "class": countDownClasses.value
      }, [renderContent()]);
    };
  }
});

export { _CountDown as default };
