// tdesign-mobile-vue/es/toast/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var ToastProps = {
  direction: {
    type: String,
    default: "row",
    validator: function validator(val) {
      if (!val) return true;
      return ["row", "column"].includes(val);
    }
  },
  duration: {
    type: Number,
    default: 2e3
  },
  icon: {
    type: [String, Function]
  },
  message: {
    type: [String, Function]
  },
  overlayProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  placement: {
    type: String,
    default: "middle",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "middle", "bottom"].includes(val);
    }
  },
  preventScrollThrough: Boolean,
  showOverlay: Boolean,
  theme: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["loading", "success", "warning", "error"].includes(val);
    }
  },
  onClose: Function,
  onDestroy: Function
};

export { ToastProps as default };
