// tdesign-mobile-vue/lib/tag/check-tag.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, toRefs, computed, createVNode } from 'vue';
import { CloseIcon } from 'tdesign-icons-vue-next';
import config from '../config.js';
import CheckTagProps from './check-tag-props.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useTNodeJSX, useContent } from '../hooks/tnode.js';
import useVModel from '../hooks/useVModel.js';
import '../config-provider/useConfig.js';
import 'lodash-es';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';

var prefix = config.prefix;
var _CheckTag = defineComponent({
  name: "".concat(prefix, "-check-tag"),
  props: CheckTagProps,
  setup: function setup(props) {
    var tagClass = usePrefixClass("tag");
    var renderTNodeJSX = useTNodeJSX();
    var renderContent = useContent();
    var _toRefs = toRefs(props),
      checked = _toRefs.checked,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(checked, modelValue, props.defaultChecked, props.onChange, "checked"),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      innerChecked = _useVModel2[0],
      setInnerChecked = _useVModel2[1];
    var contentIsArray = computed(function () {
      if (Array.isArray(props.content) && props.content.length === 2) {
        return true;
      }
      return false;
    });
    var tagClasses = computed(function () {
      return ["".concat(tagClass.value), "".concat(tagClass.value, "--checkable"), "".concat(tagClass.value, "--").concat(props.shape), "".concat(tagClass.value, "--").concat(innerChecked.value ? "primary" : "default"), "".concat(tagClass.value, "--").concat(props.size), "".concat(tagClass.value, "--").concat(props.variant), _defineProperty(_defineProperty(_defineProperty({}, "".concat(tagClass.value, "--closable"), props.closable), "".concat(tagClass.value, "--disabled"), props.disabled), "".concat(tagClass.value, "--checked"), !props.disabled && innerChecked.value)];
    });
    var handleClick = function handleClick(e) {
      if (!props.disabled) {
        var _props$onClick;
        (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, {
          e: e
        });
        setInnerChecked(!innerChecked.value);
      }
    };
    var handleClose = function handleClose(e) {
      e.stopPropagation();
      if (!props.disabled) {
        var _props$onClose;
        (_props$onClose = props.onClose) === null || _props$onClose === void 0 || _props$onClose.call(props, {
          e: e
        });
      }
    };
    return function () {
      var icon = renderTNodeJSX("icon");
      var readerText = function readerText() {
        if (contentIsArray.value) {
          return innerChecked.value ? props.content[0] : props.content[1];
        }
        var content = renderContent("default", "content");
        return content;
      };
      return createVNode("span", {
        "class": tagClasses.value,
        "aria-disabled": props.disabled,
        "role": "button",
        "onClick": handleClick
      }, [icon && createVNode("span", {
        "class": "".concat(tagClass.value, "__icon")
      }, [icon]), createVNode("span", {
        "class": "".concat(tagClass.value, "__text")
      }, [readerText()]), props.closable && createVNode("span", {
        "class": "".concat(tagClass.value, "__icon-close"),
        "onClick": handleClose
      }, [createVNode(CloseIcon, null, null)])]);
    };
  }
});

export { _CheckTag as default };
