// tdesign-mobile-vue/lib/tag/check-tag-props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var CheckTagProps = {
  checked: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultChecked: Boolean,
  closable: Boolean,
  content: {
    type: [String, Number, Array, Function]
  },
  default: {
    type: [String, Function]
  },
  disabled: Boolean,
  icon: {
    type: Function
  },
  shape: {
    type: String,
    default: "square",
    validator: function validator(val) {
      if (!val) return true;
      return ["square", "round", "mark"].includes(val);
    }
  },
  size: {
    type: String,
    default: "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  variant: {
    type: String,
    default: "dark",
    validator: function validator(val) {
      if (!val) return true;
      return ["dark", "light", "outline", "light-outline"].includes(val);
    }
  },
  onChange: Function,
  onClick: Function,
  onClose: Function
};

export { CheckTagProps as default };
