// tdesign-mobile-vue/cjs/action-sheet/action-sheet.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var actionSheet_actionSheetGrid = require('./action-sheet-grid.js');
var actionSheet_actionSheetList = require('./action-sheet-list.js');
var button_index = require('../button/index.js');
var popup_index = require('../popup/index.js');
var config = require('../config.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var actionSheet_props = require('./props.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
require('../grid/index.js');
require('../grid/grid.js');
require('../grid/props.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../grid/grid-item.js');
require('../grid/grid-item-props.js');
require('../hooks/tnode.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../image/index.js');
require('../image/image.js');
require('@vueuse/core');
require('tdesign-icons-vue-next');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-055d28fd.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../loading/plugin.js');
require('../shared/component.js');
require('../image/props.js');
require('../badge/index.js');
require('../badge/badge.js');
require('../badge/props.js');
require('../shared/hover.js');
require('../swiper/index.js');
require('../swiper/swiper.js');
require('../swipe-cell/useSwipe.js');
require('../shared/util.js');
require('../swiper/props.js');
require('../swiper/swiper-item.js');
require('../button/button.js');
require('../button/props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../popup/popup.js');
require('../popup/props.js');
require('../overlay/index.js');
require('../overlay/overlay.js');
require('../overlay/props.js');
require('../hooks/useLockScroll.js');
require('../_util/useTouch.js');
require('../_util/getScrollParent.js');
require('../hooks/useTeleport.js');
require('../shared/render-tnode.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var prefix = config["default"].prefix;
var ActionSheetVue = Vue.defineComponent({
  name: "".concat(prefix, "-action-sheet"),
  props: actionSheet_props["default"],
  emits: ["selected", "update:modelValue", "cancel", "close", "update:visible"],
  setup: function setup(props2, context) {
    var actionSheetClass = hooks_useClass.usePrefixClass("action-sheet");
    var _useConfig = configProvider_useConfig.useConfig("actionSheet"),
      globalConfig = _useConfig.globalConfig;
    var actionItems = Vue.computed(function () {
      return props2.items.map(function (item) {
        if (typeof item === "string") {
          return {
            label: item
          };
        }
        return item;
      });
    });
    var _toRefs = Vue.toRefs(props2),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](visible, modelValue, props2.defaultVisible, function () {}, "visible"),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 1),
      currentVisible = _useVModel2[0];
    var rootClasses = Vue.computed(function () {
      return _defineProperty__default["default"]({}, "".concat(actionSheetClass.value, "__content"), true);
    });
    var descriptionClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(actionSheetClass.value, "__description"), true), "".concat(actionSheetClass.value, "__description--left"), props2.align === "left"), "".concat(actionSheetClass.value, "__description--grid"), props2.theme === "grid");
    });
    var hide = function hide(trigger) {
      var _props2$onClose;
      context.emit("update:modelValue", false);
      (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, {
        trigger: trigger
      });
    };
    var handleCancel = function handleCancel() {
      var _props2$onCancel;
      (_props2$onCancel = props2.onCancel) === null || _props2$onCancel === void 0 || _props2$onCancel.call(props2);
      context.emit("update:modelValue", false);
    };
    var handleSelected = function handleSelected(index) {
      var _props2$onSelected;
      (_props2$onSelected = props2.onSelected) === null || _props2$onSelected === void 0 || _props2$onSelected.call(props2, props2 === null || props2 === void 0 ? void 0 : props2.items[index], index);
      hide("selected");
    };
    var handleClose = function handleClose() {
      hide("overlay");
    };
    return function () {
      var _slot;
      var root = function root() {
        var description = function description() {
          if (props2.description) {
            return Vue.createVNode("p", {
              "class": descriptionClasses.value
            }, [props2.description]);
          }
          return null;
        };
        var list = function list() {
          if (props2.theme === "list") {
            return Vue.createVNode(actionSheet_actionSheetList["default"], {
              "align": props2.align,
              "items": actionItems.value,
              "onSelected": handleSelected
            }, null);
          }
        };
        var grid = function grid() {
          if (props2.theme === "grid") {
            return Vue.createVNode(actionSheet_actionSheetGrid["default"], {
              "items": actionItems.value,
              "count": props2.count,
              "onSelected": handleSelected
            }, null);
          }
        };
        var cancel = function cancel() {
          if (props2.showCancel) {
            return Vue.createVNode("div", {
              "class": "".concat(actionSheetClass.value, "__footer")
            }, [Vue.createVNode("div", {
              "class": "".concat(actionSheetClass.value, "__gap-").concat(props2.theme)
            }, null), Vue.createVNode(button_index.Button, {
              "class": "".concat(actionSheetClass.value, "__cancel"),
              "variant": "text",
              "block": true,
              "onClick": handleCancel
            }, {
              default: function _default() {
                return [props2.cancelText || globalConfig.value.cancel];
              }
            })]);
          }
        };
        return Vue.createVNode("div", {
          "class": rootClasses.value
        }, [description(), list(), grid(), cancel()]);
      };
      return Vue.createVNode(popup_index.Popup, Vue.mergeProps(props2.popupProps, {
        "visible": currentVisible.value,
        "placement": "bottom",
        "destroy-on-close": true,
        "class": actionSheetClass.value,
        "onClose": handleClose,
        "showOverlay": props2.showOverlay
      }), _isSlot(_slot = root()) ? _slot : {
        default: function _default() {
          return [_slot];
        }
      });
    };
  }
});

exports["default"] = ActionSheetVue;
