// tdesign-mobile-vue/cjs/action-sheet/action-sheet-grid.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var grid_index = require('../grid/index.js');
var swiper_index = require('../swiper/index.js');
var hooks_useClass = require('../hooks/useClass.js');
require('../grid/grid.js');
require('../config.js');
require('../grid/props.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../grid/grid-item.js');
require('../grid/grid-item-props.js');
require('../hooks/tnode.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../image/index.js');
require('../image/image.js');
require('@babel/runtime/helpers/slicedToArray');
require('@vueuse/core');
require('tdesign-icons-vue-next');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-055d28fd.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../loading/plugin.js');
require('../shared/component.js');
require('../image/props.js');
require('../badge/index.js');
require('../badge/badge.js');
require('../badge/props.js');
require('../shared/hover.js');
require('../swiper/swiper.js');
require('../swipe-cell/useSwipe.js');
require('../shared/util.js');
require('../swiper/props.js');
require('../hooks/useVModel.js');
require('../swiper/swiper-item.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var TActionSheetGrid = Vue.defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    },
    count: {
      type: Number,
      default: 8
    },
    selected: {
      type: Function,
      default: void 0
    }
  },
  emits: ["selected"],
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    var actionSheetClass = hooks_useClass.usePrefixClass("action-sheet");
    var gridColumn = Vue.computed(function () {
      return Math.ceil(props.count / 2);
    });
    var pageNum = Vue.computed(function () {
      return Math.ceil(props.items.length / props.count);
    });
    var actionItems = Vue.computed(function () {
      var res = [];
      for (var i = 0; i < pageNum.value; i++) {
        var temp = props.items.slice(i * props.count, (i + 1) * props.count);
        res.push(temp);
      }
      return res;
    });
    var gridClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(actionSheetClass.value, "__grid"), true), "".concat(actionSheetClass.value, "__grid--swiper"), pageNum.value > 1), "".concat(actionSheetClass.value, "__dots"), pageNum.value > 1);
    });
    var handleSelected = function handleSelected(i) {
      emit("selected", i);
    };
    return function () {
      var swiper = function swiper() {
        var swiperItems = actionItems.value.map(function (items, i) {
          var gridItems = items.map(function (item, index) {
            return Vue.createVNode(grid_index.GridItem, {
              "key": index,
              "text": item.label,
              "image": item.icon,
              "badge": item.badge,
              "onClick": function onClick(event) {
                event.preventDefault();
                handleSelected(i * props.count + index);
              }
            }, null);
          });
          return Vue.createVNode(swiper_index.SwiperItem, {
            "key": i
          }, {
            default: function _default() {
              return [Vue.createVNode(grid_index.Grid, {
                "column": gridColumn.value
              }, _isSlot(gridItems) ? gridItems : {
                default: function _default() {
                  return [gridItems];
                }
              })];
            }
          });
        });
        if (actionItems.value.length > 1) {
          return Vue.createVNode(swiper_index.Swiper, {
            "autoplay": false,
            "pagination-position": "bottom",
            "navigation": {
              type: "dots",
              showControls: false
            },
            "loop": false,
            "class": "".concat(actionSheetClass.value, "__swiper-wrap"),
            "height": 192
          }, _isSlot(swiperItems) ? swiperItems : {
            default: function _default() {
              return [swiperItems];
            }
          });
        }
        return null;
      };
      var grid = function grid() {
        if (actionItems.value.length === 1) {
          var items = actionItems.value[0].map(function (item, index) {
            return Vue.createVNode(grid_index.GridItem, {
              "key": index,
              "text": item.label,
              "image": item.icon,
              "badge": item.badge,
              "onClick": function onClick() {
                return handleSelected(index);
              }
            }, null);
          });
          return Vue.createVNode(grid_index.Grid, {
            "column": gridColumn.value
          }, _isSlot(items) ? items : {
            default: function _default() {
              return [items];
            }
          });
        }
        return null;
      };
      return Vue.createVNode("div", {
        "class": gridClasses.value
      }, [swiper(), grid()]);
    };
  }
});

exports["default"] = TActionSheetGrid;
