// tdesign-mobile-vue/cjs/action-sheet/action-sheet-list.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var button_index = require('../button/index.js');
var badge_index = require('../badge/index.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
require('../button/button.js');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-055d28fd.js');
require('../config.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../loading/plugin.js');
require('../shared/component.js');
require('../button/props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../shared/hover.js');
require('../badge/badge.js');
require('../badge/props.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var TActionSheetList = Vue.defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    },
    align: {
      type: String,
      default: "center"
    }
  },
  emits: ["selected"],
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    var actionSheetClass = hooks_useClass.usePrefixClass("action-sheet");
    var renderTNodeJSX = hooks_tnode.useTNodeDefault();
    var _toRefs = Vue.toRefs(props),
      align = _toRefs.align,
      items = _toRefs.items;
    var handleSelected = function handleSelected(index) {
      emit("selected", index);
    };
    var itemClasses = Vue.computed(function () {
      return _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(actionSheetClass.value, "__list-item"), true), "".concat(actionSheetClass.value, "__list-item--left"), align.value === "left");
    });
    return function () {
      var renderButtonNode = function renderButtonNode() {
        var renderBadgeNode = function renderBadgeNode(item) {
          if (item.badge) {
            var content = function content() {
              if (item.badge.dot || item.badge.count) {
                return Vue.createVNode(badge_index.Badge, {
                  "count": item.badge.count,
                  "max-count": item.badge.maxCount || 99,
                  "dot": item.badge.dot,
                  "content": item.badge.content,
                  "size": item.badge.size,
                  "offset": item.badge.offset || [-16, 20]
                }, {
                  default: function _default() {
                    return [Vue.createVNode("span", {
                      "class": "".concat(actionSheetClass.value, "__list-item-text")
                    }, [" ", item.label])];
                  }
                });
              }
              return renderTNodeJSX("badge", {
                defaultNode: Vue.createVNode("span", {
                  "class": "".concat(actionSheetClass.value, "__list-item-text")
                }, [item.label])
              });
            };
            return content();
          }
          return Vue.createVNode("span", {
            "class": "".concat(actionSheetClass.value, "__list-item-text")
          }, [item.label]);
        };
        var buttonList = items.value.map(function (item, index) {
          var _slot;
          return Vue.createVNode(button_index.Button, {
            "key": index,
            "variant": "text",
            "block": true,
            "class": [itemClasses.value, _defineProperty__default["default"]({}, "".concat(actionSheetClass.value, "__list-item--disabled"), item.disabled)],
            "disabled": item.disabled,
            "icon": item.icon,
            "style": {
              color: item.color
            },
            "onClick": function onClick() {
              return handleSelected(index);
            }
          }, _isSlot(_slot = renderBadgeNode(item)) ? _slot : {
            default: function _default() {
              return [_slot];
            }
          });
        });
        return buttonList;
      };
      return Vue.createVNode("div", {
        "class": "".concat(actionSheetClass.value, "__list")
      }, [renderButtonNode()]);
    };
  }
});

exports["default"] = TActionSheetList;
