// tdesign-mobile-vue/cjs/popup/popup.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var popup_props = require('./props.js');
var overlay_index = require('../overlay/index.js');
var config = require('../config.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_useLockScroll = require('../hooks/useLockScroll.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useTeleport = require('../hooks/useTeleport.js');
var shared_renderTnode = require('../shared/render-tnode.js');
var shared_util = require('../shared/util.js');
require('../overlay/overlay.js');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../overlay/props.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../shared/component.js');
require('../_util/useTouch.js');
require('../_util/getScrollParent.js');
require('../_chunks/dep-055d28fd.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var prefix = config["default"].prefix;
var _Popup = Vue.defineComponent({
  name: "".concat(prefix, "-popup"),
  components: {
    TNode: shared_renderTnode["default"],
    TOverlay: overlay_index.Overlay
  },
  inheritAttrs: false,
  props: popup_props["default"],
  emits: ["open", "close", "opened", "closed", "visible-change", "update:visible", "update:modelValue"],
  setup: function setup(props, context) {
    var popupClass = hooks_useClass.usePrefixClass("popup");
    var popupRef = Vue.ref();
    var renderTNodeContent = hooks_tnode.useContent();
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _toRefs = Vue.toRefs(props),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](visible, modelValue, props.defaultVisible, props.onVisibleChange, "visible"),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      currentVisible = _useVModel2[0],
      setCurrentVisible = _useVModel2[1];
    var wrapperVisible = Vue.ref(currentVisible.value);
    var innerVisible = Vue.ref(currentVisible.value);
    var mounted = Vue.ref(!!shared_util.isBrowser);
    Vue.watch(currentVisible, function (v, oldV) {
      if (v) {
        if (v !== oldV) {
          var _props$onVisibleChang;
          (_props$onVisibleChang = props.onVisibleChange) === null || _props$onVisibleChang === void 0 || _props$onVisibleChang.call(props, v, "document");
        }
        wrapperVisible.value = v;
        if (props.destroyOnClose) {
          Vue.nextTick(function () {
            innerVisible.value = v;
          });
        } else {
          innerVisible.value = v;
        }
      } else {
        innerVisible.value = v;
      }
    });
    var rootStyles = Vue.computed(function () {
      var styles = {};
      if (props.zIndex) {
        styles.zIndex = "".concat(props.zIndex);
      }
      return _objectSpread(_objectSpread({}, context.attrs.style), styles);
    });
    var contentClasses = Vue.computed(function () {
      return _defineProperty__default["default"]({}, "".concat(popupClass.value, "--").concat(props.placement), true);
    });
    var contentTransitionName = Vue.computed(function () {
      var transitionName = props.transitionName,
        placement = props.placement;
      if (transitionName) return transitionName;
      if (placement === "center") return "fade-zoom";
      return "slide-".concat(placement);
    });
    var closeBtnNode = Vue.computed(function () {
      return renderTNodeJSX("closeBtn", Vue.h(tdesignIconsVueNext.CloseIcon, {
        size: "24px"
      }));
    });
    var handleCloseClick = function handleCloseClick(e) {
      var _props$onClose;
      (_props$onClose = props.onClose) === null || _props$onClose === void 0 || _props$onClose.call(props, {
        e: e
      });
      setCurrentVisible(false, "close-btn");
    };
    var handleOverlayClick = function handleOverlayClick(args) {
      var _props$onClose2;
      var e = args.e;
      if (!props.closeOnOverlayClick) {
        return;
      }
      (_props$onClose2 = props.onClose) === null || _props$onClose2 === void 0 || _props$onClose2.call(props, {
        e: e
      });
      setCurrentVisible(false, "overlay");
    };
    var afterLeave = function afterLeave() {
      var _props$onClosed;
      wrapperVisible.value = false;
      (_props$onClosed = props.onClosed) === null || _props$onClosed === void 0 || _props$onClosed.call(props);
    };
    var afterEnter = function afterEnter() {
      var _props$onOpened;
      return (_props$onOpened = props.onOpened) === null || _props$onOpened === void 0 ? void 0 : _props$onOpened.call(props);
    };
    var teleportElement = hooks_useTeleport["default"](function () {
      return props.attach;
    });
    Vue.watch(function () {
      return currentVisible.value;
    }, function (val) {
      if (val) {
        var _props$onOpen;
        (_props$onOpen = props.onOpen) === null || _props$onOpen === void 0 || _props$onOpen.call(props);
        mounted.value = val;
      }
    });
    hooks_useLockScroll.useLockScroll(popupRef, function () {
      return wrapperVisible.value && props.preventScrollThrough;
    }, popupClass.value);
    return function () {
      var renderOverlayContent = Vue.createVNode(overlay_index.Overlay, Vue.mergeProps(props.overlayProps, {
        "visible": innerVisible.value && props.showOverlay,
        "onClick": handleOverlayClick
      }), null);
      var renderCloseBtn = closeBtnNode.value && Vue.createVNode("div", {
        "class": "".concat(popupClass.value, "__close"),
        "onClick": handleCloseClick
      }, [closeBtnNode.value]);
      var renderContent = Vue.createVNode(Vue.Transition, {
        "name": contentTransitionName.value,
        "duration": props.duration,
        "onAfterEnter": afterEnter,
        "onAfterLeave": afterLeave
      }, {
        default: function _default() {
          return [Vue.withDirectives(Vue.createVNode("div", Vue.mergeProps({
            "ref": popupRef
          }, context.attrs, {
            "class": [popupClass.value, context.attrs.class, contentClasses.value],
            "style": rootStyles.value
          }), [renderCloseBtn, renderTNodeContent("default", "content")]), [[Vue.vShow, innerVisible.value]])];
        }
      });
      var inner = Vue.createVNode(Vue.Fragment, null, [renderOverlayContent, renderContent]);
      var renderPopupContent = mounted.value ? Vue.createVNode(Vue.Teleport, {
        "to": teleportElement.value,
        "disabled": !teleportElement.value
      }, _isSlot(inner) ? inner : {
        default: function _default() {
          return [inner];
        }
      }) : inner;
      return (!props.destroyOnClose || wrapperVisible.value) && renderPopupContent;
    };
  }
});

exports["default"] = _Popup;
