// tdesign-mobile-vue/cjs/popup/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var popupProps = {
  attach: {
    type: [String, Function],
    default: "body"
  },
  closeBtn: {
    type: [Boolean, Function]
  },
  closeOnOverlayClick: {
    type: Boolean,
    default: true
  },
  destroyOnClose: Boolean,
  duration: {
    type: Number,
    default: 240
  },
  overlayProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  placement: {
    type: String,
    default: "top",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "left", "right", "bottom", "center"].includes(val);
    }
  },
  preventScrollThrough: {
    type: Boolean,
    default: true
  },
  showOverlay: {
    type: Boolean,
    default: true
  },
  transitionName: {
    type: String,
    default: ""
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  zIndex: {
    type: Number
  },
  onClose: Function,
  onClosed: Function,
  onOpen: Function,
  onOpened: Function,
  onVisibleChange: Function
};

exports["default"] = popupProps;
