// tdesign-mobile-vue/lib/qrcode/components/qrcode-status.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, toRefs, createVNode, computed, Fragment } from 'vue';
import { RefreshIcon, CheckCircleFilledIcon } from 'tdesign-icons-vue-next';
import { QRCodeStatusProps } from './props.js';
import { Loading } from '../../loading/index.js';
import '../../loading/loading.js';
import '@babel/runtime/helpers/defineProperty';
import '../../loading/icon/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/setStyle.js';
import '../../_common/js/utils/helper.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import 'lodash-es';
import '../../config.js';
import '../../hooks/useClass.js';
import '../../config-provider/useConfig.js';
import '../../config-provider/context.js';
import '../../_common/js/global-config/mobile/default-config.js';
import '../../_common/js/global-config/mobile/locale/zh_CN.js';
import '../../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../../loading/icon/spinner.js';
import '../../loading/props.js';
import '../../hooks/tnode.js';
import '../../hooks/render-tnode.js';
import '../../_common/js/utils/general.js';
import '../../shared/dom.js';
import '../../loading/plugin.js';
import '../../shared/component.js';

var QRcodeStatus = defineComponent({
  name: "QRCodeStatus",
  props: QRCodeStatusProps,
  setup: function setup(props) {
    var _locale$value3;
    var _toRefs = toRefs(props),
      classPrefix = _toRefs.classPrefix,
      locale = _toRefs.locale,
      status = _toRefs.status;
    var defaultSpin = createVNode(Loading, {
      "size": "32px"
    }, null);
    var defaultExpiredNode = computed(function () {
      var _locale$value, _locale$value2;
      return createVNode(Fragment, null, [createVNode("p", {
        "class": "".concat(classPrefix.value, "-expired__text")
      }, [(_locale$value = locale.value) === null || _locale$value === void 0 ? void 0 : _locale$value.expiredText]), (props === null || props === void 0 ? void 0 : props.onRefresh) && createVNode("p", {
        "class": "".concat(classPrefix.value, "-expired__button"),
        "onClick": props === null || props === void 0 ? void 0 : props.onRefresh
      }, [createVNode(RefreshIcon, {
        "size": "16"
      }, null), (_locale$value2 = locale.value) === null || _locale$value2 === void 0 ? void 0 : _locale$value2.refreshText])]);
    });
    var defaultScannedNode = createVNode("p", {
      "class": "".concat(classPrefix.value, "-scanned")
    }, [createVNode(CheckCircleFilledIcon, {
      "size": "16",
      "class": "".concat(classPrefix.value, "-scanned__icon")
    }, null), (_locale$value3 = locale.value) === null || _locale$value3 === void 0 ? void 0 : _locale$value3.scannedText]);
    var defaultNodes = computed(function () {
      return {
        expired: defaultExpiredNode.value,
        loading: defaultSpin,
        scanned: defaultScannedNode,
        active: null
      };
    });
    var renderStatus = function renderStatus(info) {
      return defaultNodes.value[info.status];
    };
    return function () {
      return createVNode(Fragment, null, [props.statusRender || renderStatus({
        status: status.value,
        onRefresh: props.onRefresh
      })]);
    };
  }
});

export { QRcodeStatus as default };
