// tdesign-mobile-vue/cjs/watermark/watermark.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var _common_js_watermark_generateBase64Url = require('../_common/js/watermark/generateBase64Url.js');
var _common_js_watermark_randomMovingStyle = require('../_common/js/watermark/randomMovingStyle.js');
var _common_js_utils_injectStyle = require('../_common/js/utils/injectStyle.js');
var _common_js_utils_setStyle = require('../_common/js/utils/setStyle.js');
var watermark_hooks_index = require('./hooks/index.js');
var config = require('../config.js');
var watermark_props = require('./props.js');
var hooks_useVariables = require('../hooks/useVariables.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
require('@babel/runtime/helpers/defineProperty');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('../_common/js/common.js');
require('../_common/js/utils/getColorTokenColor.js');
require('../hooks/useMutationObservable.js');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-5f870597.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-4326ca38.js');
require('../_chunks/dep-2da1d545.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');

function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
var prefix = config["default"].prefix;
var _Watermark = Vue.defineComponent({
  name: "".concat(prefix, "-watermark"),
  props: watermark_props["default"],
  setup: function setup(props2) {
    var backgroundImage = Vue.ref("");
    var watermarkRef = Vue.shallowRef();
    var watermarkContentRef = Vue.shallowRef();
    var offset = Vue.reactive(props2.offset || []);
    var gapX = Vue.computed(function () {
      return props2.movable ? 0 : props2.x;
    });
    var gapY = Vue.computed(function () {
      return props2.movable ? 0 : props2.y;
    });
    var rotate = Vue.computed(function () {
      return props2.movable ? 0 : props2.rotate;
    });
    var backgroundRepeat = Vue.computed(function () {
      if (props2.movable) {
        return "no-repeat";
      }
      return props2.isRepeat ? "repeat" : "no-repeat";
    });
    var offsetLeft = Vue.computed(function () {
      return offset[0] || gapX.value / 2;
    });
    var offsetTop = Vue.computed(function () {
      return offset[1] || gapY.value / 2;
    });
    var _useVariables = hooks_useVariables.useVariables({
        fontColor: "--td-text-color-watermark"
      }),
      fontColor = _useVariables.fontColor;
    var bgImageOptions = Vue.computed(function () {
      return {
        width: props2.width,
        height: props2.height,
        rotate: rotate.value,
        lineSpace: props2.lineSpace,
        alpha: props2.alpha,
        gapX: gapX.value,
        gapY: gapY.value,
        watermarkContent: props2.watermarkContent,
        offsetLeft: offsetLeft.value,
        offsetTop: offsetTop.value,
        fontColor: fontColor.value,
        layout: props2.layout
      };
    });
    var removeWaterMark = function removeWaterMark() {
      if (!watermarkContentRef.value) return;
      watermarkContentRef.value.remove();
      watermarkContentRef.value = null;
    };
    var injectWaterMark = function injectWaterMark() {
      _common_js_watermark_generateBase64Url["default"](bgImageOptions.value, function (base64Url, backgroundSize) {
        var _watermarkRef$value;
        removeWaterMark();
        backgroundImage.value = base64Url;
        watermarkContentRef.value = document.createElement("div");
        _common_js_utils_setStyle["default"](watermarkContentRef.value, {
          zIndex: props2.zIndex,
          position: "absolute",
          left: 0,
          right: 0,
          top: 0,
          bottom: 0,
          width: "100%",
          height: "100%",
          backgroundSize: "".concat((backgroundSize === null || backgroundSize === void 0 ? void 0 : backgroundSize.width) || gapX.value + props2.width, "px"),
          pointerEvents: "none",
          backgroundRepeat: backgroundRepeat.value,
          backgroundImage: "url('".concat(backgroundImage.value, "')"),
          animation: props2.movable ? "watermark infinite ".concat(props2.moveInterval * 4 / 60, "s") : "none"
        });
        (_watermarkRef$value = watermarkRef.value) === null || _watermarkRef$value === void 0 || _watermarkRef$value.append(watermarkContentRef.value);
      });
      if (props2.movable) {
        var keyframesStyle = _common_js_watermark_randomMovingStyle["default"]();
        _common_js_utils_injectStyle["default"](keyframesStyle);
      }
    };
    Vue.onMounted(function () {
      injectWaterMark();
      watermark_hooks_index.useMutationObserver(watermarkRef.value, function (mutations) {
        if (props2.removable || !watermarkContentRef.value) return;
        var _iterator = _createForOfIteratorHelper(mutations),
          _step;
        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var mutation = _step.value;
            var isRemoved = Array.from(mutation.removedNodes).includes(watermarkContentRef.value);
            var isModified = mutation.type === "attributes" && watermarkContentRef.value === mutation.target;
            if (isRemoved || isModified) {
              injectWaterMark();
              break;
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      }, {
        attributes: true,
        childList: true,
        characterData: true,
        subtree: true
      });
    });
    Vue.watch(function () {
      return [props2, fontColor.value];
    }, injectWaterMark, {
      deep: true,
      flush: "post"
    });
    return function () {
      var COMPONENT_NAME = hooks_useClass.usePrefixClass("watermark");
      var renderContent = hooks_tnode.useContent();
      return Vue.createVNode("div", {
        "class": COMPONENT_NAME.value,
        "ref": watermarkRef
      }, [renderContent("default", "content")]);
    };
  }
});

exports["default"] = _Watermark;
