// tdesign-mobile-vue/cjs/grid/grid-item.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var config = require('../config.js');
var grid_gridItemProps = require('./grid-item-props.js');
var hooks_tnode = require('../hooks/tnode.js');
var image_index = require('../image/index.js');
var badge_index = require('../badge/index.js');
var hooks_useClass = require('../hooks/useClass.js');
var shared_hover = require('../shared/hover.js');
var isString = require('../_chunks/dep-588e2e70.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
var isFunction = require('../_chunks/dep-0bd122a0.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../image/image.js');
require('@babel/runtime/helpers/slicedToArray');
require('@vueuse/core');
require('tdesign-icons-vue-next');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-055d28fd.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../loading/plugin.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../image/props.js');
require('../badge/badge.js');
require('../badge/props.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !Vue.isVNode(s);
}
var prefix = config["default"].prefix;
var _GridItem = Vue.defineComponent({
  name: "".concat(prefix, "-grid-item"),
  directives: {
    Hover: shared_hover["default"]
  },
  components: {
    TImage: image_index.Image,
    TBadge: badge_index.Badge
  },
  props: grid_gridItemProps["default"],
  setup: function setup(props2, context) {
    var gridItemClass = hooks_useClass.usePrefixClass("grid-item");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _inject = Vue.inject("grid"),
      column = _inject.column,
      border = _inject.border,
      align = _inject.align,
      gutter = _inject.gutter,
      hover = _inject.hover;
    var rootStyle = Vue.computed(function () {
      var percent = column.value > 0 ? "".concat(100 / +column.value, "%") : 0;
      var style = {
        textAlign: ["center", "left"].includes(align.value) ? align.value : "center"
      };
      if (percent !== 0) {
        style.flexBasis = percent;
      }
      return style;
    });
    var size = Vue.computed(function () {
      var col = column.value;
      if (!col || col === 4) return "middle";
      return col > 4 ? "small" : "large";
    });
    var realImage = Vue.computed(function () {
      if (isString.isString(props2.image)) return {
        src: props2.image
      };
      if (isObject.isObject(props2.image) && !isFunction.isFunction(props2.image) && !context.slots.image) {
        return props2.image;
      }
      return null;
    });
    var gridItemClasses = Vue.computed(function () {
      return ["".concat(gridItemClass.value), "".concat(gridItemClass.value, "--").concat(props2.layout), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(gridItemClass.value, "--bordered"), border.value), "".concat(gridItemClass.value, "--surround"), border.value && gutter.value)];
    });
    return function () {
      var _slot;
      var renderImage = function renderImage() {
        if (realImage.value) {
          return Vue.createVNode(Vue.resolveComponent("t-image"), Vue.mergeProps({
            "shape": "round"
          }, realImage.value), null);
        }
        if (props2.icon) {
          return renderTNodeJSX("icon");
        }
        return renderTNodeJSX("image");
      };
      return Vue.withDirectives(Vue.createVNode("div", {
        "class": gridItemClasses.value,
        "style": rootStyle.value
      }, [Vue.createVNode("div", {
        "class": "".concat(gridItemClass.value, "__image ").concat(gridItemClass.value, "__image--").concat(size.value)
      }, [props2.badge ? Vue.createVNode(Vue.resolveComponent("t-badge"), props2.badge, _isSlot(_slot = renderImage()) ? _slot : {
        default: function _default() {
          return [_slot];
        }
      }) : renderImage()]), Vue.createVNode("div", {
        "class": "".concat(gridItemClass.value, "__content ").concat(gridItemClass.value, "__content--").concat(props2.layout)
      }, [Vue.createVNode("div", {
        "class": "".concat(gridItemClass.value, "__title ").concat(gridItemClass.value, "__title--").concat(size.value)
      }, [renderTNodeJSX("text")]), Vue.createVNode("div", {
        "class": "".concat(gridItemClass.value, "__description ").concat(gridItemClass.value, "__description--").concat(props2.layout)
      }, [renderTNodeJSX("description")])])]), [[Vue.resolveDirective("hover"), {
        className: "".concat(gridItemClass.value, "--hover"),
        disabledHover: !hover.value
      }]]);
    };
  }
});

exports["default"] = _GridItem;
