// tdesign-mobile-vue/lib/popup/popup.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, ref, toRefs, watch, nextTick, computed, h, createVNode, mergeProps, Transition, withDirectives, vShow, Fragment, Teleport, isVNode } from 'vue';
import { CloseIcon } from 'tdesign-icons-vue-next';
import popupProps from './props.js';
import { Overlay } from '../overlay/index.js';
import config from '../config.js';
import useVModel from '../hooks/useVModel.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useLockScroll } from '../hooks/useLockScroll.js';
import { useContent, useTNodeJSX } from '../hooks/tnode.js';
import useTeleport from '../hooks/useTeleport.js';
import TNodeComponent from '../shared/render-tnode.js';
import { isBrowser } from '../shared/util.js';
import '../overlay/overlay.js';
import '../shared/dom.js';
import 'lodash-es';
import '../overlay/props.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../shared/component.js';
import '../_util/useTouch.js';
import '../_util/getScrollParent.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var prefix = config.prefix;
var _Popup = defineComponent({
  name: "".concat(prefix, "-popup"),
  components: {
    TNode: TNodeComponent,
    TOverlay: Overlay
  },
  inheritAttrs: false,
  props: popupProps,
  emits: ["open", "close", "opened", "closed", "visible-change", "update:visible", "update:modelValue"],
  setup: function setup(props, context) {
    var popupClass = usePrefixClass("popup");
    var popupRef = ref();
    var renderTNodeContent = useContent();
    var renderTNodeJSX = useTNodeJSX();
    var _toRefs = toRefs(props),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(visible, modelValue, props.defaultVisible, props.onVisibleChange, "visible"),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      currentVisible = _useVModel2[0],
      setCurrentVisible = _useVModel2[1];
    var wrapperVisible = ref(currentVisible.value);
    var innerVisible = ref(currentVisible.value);
    var mounted = ref(!!isBrowser);
    watch(currentVisible, function (v, oldV) {
      if (v) {
        if (v !== oldV) {
          var _props$onVisibleChang;
          (_props$onVisibleChang = props.onVisibleChange) === null || _props$onVisibleChang === void 0 || _props$onVisibleChang.call(props, v, "document");
        }
        wrapperVisible.value = v;
        if (props.destroyOnClose) {
          nextTick(function () {
            innerVisible.value = v;
          });
        } else {
          innerVisible.value = v;
        }
      } else {
        innerVisible.value = v;
      }
    });
    var rootStyles = computed(function () {
      var styles = {};
      if (props.zIndex) {
        styles.zIndex = "".concat(props.zIndex);
      }
      return _objectSpread(_objectSpread({}, context.attrs.style), styles);
    });
    var contentClasses = computed(function () {
      return _defineProperty({}, "".concat(popupClass.value, "--").concat(props.placement), true);
    });
    var contentTransitionName = computed(function () {
      var transitionName = props.transitionName,
        placement = props.placement;
      if (transitionName) return transitionName;
      if (placement === "center") return "fade-zoom";
      return "slide-".concat(placement);
    });
    var closeBtnNode = computed(function () {
      return renderTNodeJSX("closeBtn", h(CloseIcon, {
        size: "24px"
      }));
    });
    var handleCloseClick = function handleCloseClick(e) {
      var _props$onClose;
      (_props$onClose = props.onClose) === null || _props$onClose === void 0 || _props$onClose.call(props, {
        e: e
      });
      setCurrentVisible(false, "close-btn");
    };
    var handleOverlayClick = function handleOverlayClick(args) {
      var _props$onClose2;
      var e = args.e;
      if (!props.closeOnOverlayClick) {
        return;
      }
      (_props$onClose2 = props.onClose) === null || _props$onClose2 === void 0 || _props$onClose2.call(props, {
        e: e
      });
      setCurrentVisible(false, "overlay");
    };
    var afterLeave = function afterLeave() {
      var _props$onClosed;
      wrapperVisible.value = false;
      (_props$onClosed = props.onClosed) === null || _props$onClosed === void 0 || _props$onClosed.call(props);
    };
    var afterEnter = function afterEnter() {
      var _props$onOpened;
      return (_props$onOpened = props.onOpened) === null || _props$onOpened === void 0 ? void 0 : _props$onOpened.call(props);
    };
    var teleportElement = useTeleport(function () {
      return props.attach;
    });
    watch(function () {
      return currentVisible.value;
    }, function (val) {
      if (val) {
        var _props$onOpen;
        (_props$onOpen = props.onOpen) === null || _props$onOpen === void 0 || _props$onOpen.call(props);
        mounted.value = val;
      }
    });
    useLockScroll(popupRef, function () {
      return wrapperVisible.value && props.preventScrollThrough;
    }, popupClass.value);
    return function () {
      var renderOverlayContent = createVNode(Overlay, mergeProps(props.overlayProps, {
        "visible": innerVisible.value && props.showOverlay,
        "onClick": handleOverlayClick
      }), null);
      var renderCloseBtn = closeBtnNode.value && createVNode("div", {
        "class": "".concat(popupClass.value, "__close"),
        "onClick": handleCloseClick
      }, [closeBtnNode.value]);
      var renderContent = createVNode(Transition, {
        "name": contentTransitionName.value,
        "duration": props.duration,
        "onAfterEnter": afterEnter,
        "onAfterLeave": afterLeave
      }, {
        default: function _default() {
          return [withDirectives(createVNode("div", mergeProps({
            "ref": popupRef
          }, context.attrs, {
            "class": [popupClass.value, context.attrs.class, contentClasses.value],
            "style": rootStyles.value
          }), [renderCloseBtn, renderTNodeContent("default", "content")]), [[vShow, innerVisible.value]])];
        }
      });
      var inner = createVNode(Fragment, null, [renderOverlayContent, renderContent]);
      var renderPopupContent = mounted.value ? createVNode(Teleport, {
        "to": teleportElement.value,
        "disabled": !teleportElement.value
      }, _isSlot(inner) ? inner : {
        default: function _default() {
          return [inner];
        }
      }) : inner;
      return (!props.destroyOnClose || wrapperVisible.value) && renderPopupContent;
    };
  }
});

export { _Popup as default };
