// tdesign-mobile-vue/cjs/qrcode/qrcode.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var Vue = require('vue');
var qrcode_props = require('./props.js');
var _common_js_qrcode_utils = require('../_common/js/qrcode/utils.js');
var qrcode_components_qrcodeCanvas = require('./components/qrcode-canvas.js');
var qrcode_components_qrcodeSvg = require('./components/qrcode-svg.js');
var qrcode_components_qrcodeStatus = require('./components/qrcode-status.js');
var hooks_tnode = require('../hooks/tnode.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
var hooks_useClass = require('../hooks/useClass.js');
var isNumber = require('../_chunks/dep-055d28fd.js');
var hooks_useVariables = require('../hooks/useVariables.js');
require('../_common/js/qrcode/qrcodegen.js');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('./components/props.js');
require('./hooks/useQRCode.js');
require('tdesign-icons-vue-next');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../config.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-b999c3d4.js');
require('../loading/plugin.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../_common/js/common.js');
require('../_common/js/utils/getColorTokenColor.js');
require('../hooks/useMutationObservable.js');
require('../_chunks/dep-5f870597.js');
require('../_chunks/dep-4326ca38.js');
require('../_chunks/dep-2da1d545.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var _QRCode = Vue.defineComponent({
  name: "TQRCode",
  props: qrcode_props["default"],
  setup: function setup(props2) {
    var classPrefix = hooks_useClass.usePrefixClass();
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _useConfig = configProvider_useConfig.useConfig("qrcode"),
      globalConfig = _useConfig.globalConfig;
    var _useVariables = hooks_useVariables.useVariables({
        themeFgColor: "--td-text-color-primary",
        themeBgColor: "--td-bg-color-specialcomponent"
      }),
      themeFgColor = _useVariables.themeFgColor,
      themeBgColor = _useVariables.themeBgColor;
    var finalBgColor = Vue.computed(function () {
      return props2.bgColor || themeBgColor.value || "transparent";
    });
    var finalFgColor = Vue.computed(function () {
      return props2.color || themeFgColor.value || _common_js_qrcode_utils.DEFAULT_FRONT_COLOR;
    });
    if (!props2.value) {
      return null;
    }
    var imageSettings = Vue.computed(function () {
      var _props2$iconSize$heig, _props2$iconSize, _props2$iconSize$widt, _props2$iconSize2;
      return {
        src: props2.icon,
        x: void 0,
        y: void 0,
        height: isNumber.isNumber(props2.iconSize) ? props2.iconSize : (_props2$iconSize$heig = (_props2$iconSize = props2.iconSize) === null || _props2$iconSize === void 0 ? void 0 : _props2$iconSize.height) !== null && _props2$iconSize$heig !== void 0 ? _props2$iconSize$heig : 40,
        width: isNumber.isNumber(props2.iconSize) ? props2.iconSize : (_props2$iconSize$widt = (_props2$iconSize2 = props2.iconSize) === null || _props2$iconSize2 === void 0 ? void 0 : _props2$iconSize2.width) !== null && _props2$iconSize$widt !== void 0 ? _props2$iconSize$widt : 40,
        excavate: true,
        crossOrigin: "anonymous"
      };
    });
    var classes = Vue.computed(function () {
      return ["".concat(classPrefix.value, "-qrcode"), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(classPrefix.value, "-borderless"), props2.borderless), "".concat(classPrefix.value, "-qrcode-svg"), props2.type === "svg")];
    });
    var mergedStyle = Vue.computed(function () {
      return {
        backgroundColor: finalBgColor.value,
        width: "".concat(props2.size, "px"),
        height: "".concat(props2.size, "px")
      };
    });
    return function () {
      var QRCodeProps = {
        value: props2.value,
        size: props2.size,
        bgColor: finalBgColor.value,
        fgColor: finalFgColor.value,
        imageSettings: props2.icon ? imageSettings.value : void 0,
        level: props2.level
      };
      return Vue.createVNode("div", Vue.mergeProps({
        "class": classes.value,
        "style": mergedStyle.value
      }, {
        level: props2.level
      }), [props2.status !== "active" && Vue.createVNode("div", {
        "class": ["".concat(classPrefix.value, "-mask"), _defineProperty__default["default"]({}, "".concat(classPrefix.value, "-").concat(props2.status), props2.status !== "loading")]
      }, [Vue.createVNode(qrcode_components_qrcodeStatus["default"], {
        "classPrefix": classPrefix.value,
        "locale": globalConfig.value,
        "status": props2.status,
        "onRefresh": props2.onRefresh,
        "statusRender": renderTNodeJSX("statusRender", {
          params: {
            status: props2.status,
            onRefresh: props2.onRefresh
          }
        })
      }, null)]), props2.type === "canvas" ? Vue.createVNode(qrcode_components_qrcodeCanvas["default"], Vue.mergeProps(QRCodeProps, {
        "size": props2.size
      }), null) : Vue.createVNode(qrcode_components_qrcodeSvg["default"], Vue.mergeProps(QRCodeProps, {
        "size": props2.size
      }), null)]);
    };
  }
});

exports["default"] = _QRCode;
