// tdesign-mobile-vue/es/shared/useCountDown/index.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _asyncToGenerator, r as regenerator } from '../../_chunks/dep-038bf4d0.mjs';
import { ref, reactive, watch } from 'vue';
import { useRafFn } from '@vueuse/core';
import { getShowTimes, getRemainTimes, getScreenFps } from './utils.mjs';
import { isBrowser } from '../util.mjs';
import '../../_chunks/dep-4055ef5c.mjs';
import '../../_chunks/dep-04a2869f.mjs';
import '../../_chunks/dep-2ed7e0d5.mjs';
import '../../_chunks/dep-9961b95b.mjs';
import 'lodash-es';
import '../../hooks/useClass.mjs';
import '../../config-provider/useConfig.mjs';
import '../../_chunks/dep-f4e31210.mjs';
import '../../config-provider/context.mjs';
import '../../_common/js/global-config/mobile/default-config.mjs';
import '../../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../../_chunks/dep-4c96ffa8.mjs';
import '../../_chunks/dep-a8bd494a.mjs';
import '../../_chunks/dep-d3be051e.mjs';
import '../../config-provider/type.mjs';

function useCountDown(props, visibility) {
  var _ref = props || {},
    _ref$time = _ref.time,
    time = _ref$time === void 0 ? 0 : _ref$time,
    autoStart = _ref.autoStart,
    _ref$millisecond = _ref.millisecond,
    millisecond = _ref$millisecond === void 0 ? false : _ref$millisecond,
    _ref$format = _ref.format,
    format = _ref$format === void 0 ? "HH:mm:ss" : _ref$format,
    _ref$splitWithUnit = _ref.splitWithUnit,
    splitWithUnit = _ref$splitWithUnit === void 0 ? false : _ref$splitWithUnit,
    onFinish = _ref.onFinish,
    onChange = _ref.onChange;
  var fps = ref();
  var count = ref(Number(time));
  var showTimes = reactive(getShowTimes(getRemainTimes(time), format, millisecond, splitWithUnit));
  var hiddenTime = 0;
  visibility && watch(visibility, function (val) {
    if (val && autoStart) {
      count.value -= Date.now() - hiddenTime;
      rafFn();
    } else {
      hiddenTime = Date.now();
    }
  });
  var rafFn = /*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator(/*#__PURE__*/regenerator.mark(function _callee() {
      var _getShowTimes, _getShowTimes$forEach;
      var res, times;
      return regenerator.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            if (isBrowser) {
              _context.next = 1;
              break;
            }
            return _context.abrupt("return");
          case 1:
            if (fps.value) {
              _context.next = 3;
              break;
            }
            _context.next = 2;
            return getScreenFps === null || getScreenFps === void 0 ? void 0 : getScreenFps();
          case 2:
            res = _context.sent;
            fps.value = res || 60;
          case 3:
            count.value = parseInt("".concat(Number(count.value) - 1e3 / fps.value), 10);
            if (count.value <= 0) {
              pause === null || pause === void 0 || pause();
              count.value = 0;
            }
            times = getRemainTimes(count.value);
            onChange === null || onChange === void 0 || onChange(times);
            count.value === 0 && (onFinish === null || onFinish === void 0 ? void 0 : onFinish());
            (_getShowTimes = getShowTimes(times, format, millisecond, splitWithUnit)) === null || _getShowTimes === void 0 || (_getShowTimes$forEach = _getShowTimes.forEach) === null || _getShowTimes$forEach === void 0 || _getShowTimes$forEach.call(_getShowTimes, function (i, idx) {
              return showTimes[idx].value = i === null || i === void 0 ? void 0 : i.value;
            });
          case 4:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function rafFn() {
      return _ref2.apply(this, arguments);
    };
  }();
  var _useRafFn = useRafFn(rafFn, {
      immediate: autoStart
    }),
    pause = _useRafFn.pause,
    resume = _useRafFn.resume;
  return {
    time: count,
    showTimes: showTimes,
    pause: pause,
    resume: resume
  };
}

export { useCountDown };
