// tdesign-mobile-vue/es/badge/badge.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { defineComponent, computed, createVNode } from 'vue';
import { isString, isNumber } from 'lodash-es';
import config from '../config.mjs';
import BadgeProps from './props.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useTNodeJSX, useContent } from '../hooks/tnode.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';

var prefix = config.prefix;
var _Badge = defineComponent({
  name: "".concat(prefix, "-badge"),
  props: BadgeProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var badgeClass = usePrefixClass("badge");
    var classPrefix = usePrefixClass();
    var badgeClasses = computed(function () {
      return _defineProperty(_defineProperty({}, "".concat(badgeClass.value), true), "".concat(badgeClass.value, "__ribbon-outer"), props.shape === "ribbon");
    });
    var badgeInnerClasses = computed(function () {
      return _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(badgeClass.value, "--basic"), true), "".concat(badgeClass.value, "--dot"), props.dot), "".concat(badgeClass.value, "--").concat(props.size), true), "".concat(badgeClass.value, "--").concat(props.shape), true), "".concat(badgeClass.value, "--count"), !props.dot && props.count), "".concat(classPrefix.value, "-has-count"), true);
    });
    var isShowBadge = computed(function () {
      if (props.dot) {
        return true;
      }
      var count = Number(props.count);
      if (!props.showZero && count === 0) {
        return false;
      }
      if (props.count == null) return false;
      return true;
    });
    var hasUnit = function hasUnit(unit) {
      return unit.indexOf("px") > 0 || unit.indexOf("rpx") > 0 || unit.indexOf("em") > 0 || unit.indexOf("rem") > 0 || unit.indexOf("%") > 0 || unit.indexOf("vh") > 0 || unit.indexOf("vm") > 0;
    };
    var badgeStyles = computed(function () {
      if (!props.offset) {
        return {
          background: props.color
        };
      }
      var _props$offset = _slicedToArray(props.offset, 2),
        _props$offset$ = _props$offset[0],
        xOffset = _props$offset$ === void 0 ? 0 : _props$offset$,
        _props$offset$2 = _props$offset[1],
        yOffset = _props$offset$2 === void 0 ? 0 : _props$offset$2;
      xOffset = hasUnit(xOffset.toString()) ? xOffset : "".concat(xOffset, "px");
      yOffset = hasUnit(yOffset.toString()) ? yOffset : "".concat(yOffset, "px");
      return {
        background: props.color,
        right: xOffset,
        top: yOffset
      };
    });
    return function () {
      var readerCount = function readerCount() {
        if (props.dot) return null;
        if (isString(props.count) || isNumber(props.count)) {
          if (props.count === 0) {
            return props.showZero ? props.count : null;
          }
          return Number(props.count) > Number(props.maxCount) ? "".concat(props.maxCount, "+") : props.count;
        }
        return renderTNodeJSX("count");
      };
      var readerContent = function readerContent() {
        var content = renderTNodeContent("default", "content");
        if (typeof content === "string") {
          return createVNode("span", {
            "class": "".concat(badgeClass.value, "__content-text")
          }, [content]);
        }
        return content;
      };
      var readerBadge = function readerBadge() {
        if (!isShowBadge.value) {
          return null;
        }
        return createVNode("div", {
          "class": badgeInnerClasses.value,
          "style": badgeStyles.value
        }, [readerCount()]);
      };
      return createVNode("div", {
        "class": badgeClasses.value
      }, [createVNode("div", {
        "class": "".concat(badgeClass.value, "__content")
      }, [readerContent()]), readerBadge()]);
    };
  }
});

export { _Badge as default };
