// tdesign-mobile-vue/cjs/checkbox/checkbox-group.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _typeof = require('@babel/runtime/helpers/typeof');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var config = require('../config.js');
var checkbox_checkboxGroupProps = require('./checkbox-group-props.js');
var _common_js_log_log = require('../_common/js/log/log.js');
var checkbox_checkbox = require('./checkbox.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var checkbox_hooks_getOptions = require('./hooks/getOptions.js');
var checkbox_hooks_setCheckAllStatus = require('./hooks/setCheckAllStatus.js');
var get = require('../_chunks/dep-cfb75f1b.js');
require('tdesign-icons-vue-next');
require('./props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-f84e6e17.js');
require('../shared/render-tnode.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-42a5275c.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../hooks/useChildSlots.js');
require('../_chunks/dep-5f870597.js');
require('../_chunks/dep-df60c8fb.js');
require('../_chunks/dep-01310648.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var _CheckboxGroup = Vue.defineComponent({
  name: "".concat(prefix, "-checkbox-group"),
  components: {
    Checkbox: checkbox_checkbox["default"]
  },
  props: checkbox_checkboxGroupProps["default"],
  emits: ["update:value", "update:modelValue", "change"],
  setup: function setup(props2, context) {
    var checkboxGroupClass = hooks_useClass.usePrefixClass("checkbox-group");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var isArray = Array.isArray;
    var _toRefs = Vue.toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = hooks_useVModel["default"](value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var keys = Vue.computed(function () {
      return props2.keys;
    });
    var optionList = checkbox_hooks_getOptions.getOptions(props2, context.slots);
    var checkedSet = Vue.computed(function () {
      if (isArray(innerValue.value)) {
        return new Set(innerValue.value);
      }
      return /* @__PURE__ */new Set();
    });
    var checkAllStatus = checkbox_hooks_setCheckAllStatus.setCheckAllStatus(optionList, innerValue, checkedSet);
    var maxExceeded = Vue.computed(function () {
      return props2.max !== void 0 && innerValue.value.length === props2.max;
    });
    var onCheckedChange = function onCheckedChange(p) {
      var checked = p.checked,
        checkAll = p.checkAll,
        e = p.e;
      if (checkAll) {
        onCheckAllChange(checked, {
          e: e
        });
      } else {
        handleCheckboxChange(p);
      }
    };
    var handleCheckboxChange = function handleCheckboxChange(data) {
      var currentValue = data.option.value;
      if (isArray(innerValue.value) || !innerValue.value) {
        if (currentValue === void 0) {
          return;
        }
        var val = innerValue.value ? _toConsumableArray__default["default"](innerValue.value) : [];
        if (data.checked) {
          val.push(currentValue);
        } else {
          var i = val.indexOf(currentValue);
          val.splice(i, 1);
        }
        setInnerValue(val, {
          e: data.e,
          current: data.option.value,
          type: data.checked ? "check" : "uncheck"
        });
      } else {
        _common_js_log_log["default"].warn("CheckboxGroup", "`value` must be an array, instead of ".concat(_typeof__default["default"](innerValue.value)));
      }
    };
    var getAllCheckboxValue = function getAllCheckboxValue() {
      var val = /* @__PURE__ */new Set();
      for (var i = 0, len = optionList.value.length; i < len; i++) {
        var item = optionList.value[i];
        if (item.checkAll) continue;
        if (item.value === void 0) continue;
        val.add(item.value);
        if (maxExceeded.value) break;
      }
      return _toConsumableArray__default["default"](val);
    };
    var onCheckAllChange = function onCheckAllChange(checked, context2) {
      var value2 = checked ? getAllCheckboxValue() : [];
      setInnerValue(value2, {
        e: context2.e,
        type: checked ? "check" : "uncheck",
        current: void 0
      });
    };
    Vue.provide("checkboxGroup", _objectSpread(_objectSpread({}, Vue.toRefs(props2)), {}, {
      innerValue: innerValue,
      checkAllStatus: checkAllStatus,
      checkedSet: checkedSet,
      maxExceeded: maxExceeded,
      onCheckedChange: onCheckedChange
    }));
    return function () {
      var checkboxNode = function checkboxNode() {
        return Vue.createVNode("span", null, [optionList.value.map(function (item, idx) {
          var _keys$value$value, _keys$value, _keys$value$label, _keys$value2, _keys$value$value2, _keys$value3, _keys$value$disabled, _keys$value4;
          return Vue.createVNode(checkbox_checkbox["default"], Vue.mergeProps(item, {
            "key": "".concat(get.get(item, (_keys$value$value = (_keys$value = keys.value) === null || _keys$value === void 0 ? void 0 : _keys$value.value) !== null && _keys$value$value !== void 0 ? _keys$value$value : "value", "")).concat(idx),
            "label": get.get(item, (_keys$value$label = (_keys$value2 = keys.value) === null || _keys$value2 === void 0 ? void 0 : _keys$value2.label) !== null && _keys$value$label !== void 0 ? _keys$value$label : "label", item.text || ""),
            "value": get.get(item, (_keys$value$value2 = (_keys$value3 = keys.value) === null || _keys$value3 === void 0 ? void 0 : _keys$value3.value) !== null && _keys$value$value2 !== void 0 ? _keys$value$value2 : "value"),
            "disabled": get.get(item, (_keys$value$disabled = (_keys$value4 = keys.value) === null || _keys$value4 === void 0 ? void 0 : _keys$value4.disabled) !== null && _keys$value$disabled !== void 0 ? _keys$value$disabled : "disabled")
          }), null);
        })]);
      };
      return Vue.createVNode("div", {
        "class": "".concat(checkboxGroupClass.value)
      }, [!(props2.options && props2.options.length) ? renderTNodeJSX("default") : checkboxNode()]);
    };
  }
});

exports["default"] = _CheckboxGroup;
