// tdesign-mobile-vue/lib/search/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var searchProps = {
  action: {
    type: [String, Function],
    default: ""
  },
  autocompleteOptions: {
    type: Array
  },
  center: Boolean,
  clearTrigger: {
    type: String,
    default: "always",
    validator: function validator(val) {
      if (!val) return true;
      return ["always", "focus"].includes(val);
    }
  },
  clearable: {
    type: Boolean,
    default: true
  },
  disabled: Boolean,
  focus: Boolean,
  leftIcon: {
    type: [String, Function],
    default: "search"
  },
  maxcharacter: {
    type: Number
  },
  maxlength: {
    type: [String, Number]
  },
  placeholder: {
    type: String,
    default: ""
  },
  readonly: {
    type: Boolean,
    default: void 0
  },
  resultList: {
    type: Array,
    default: function _default() {
      return [];
    }
  },
  shape: {
    type: String,
    default: "square",
    validator: function validator(val) {
      if (!val) return true;
      return ["square", "round"].includes(val);
    }
  },
  value: {
    type: String,
    default: void 0
  },
  modelValue: {
    type: String,
    default: void 0
  },
  defaultValue: {
    type: String,
    default: ""
  },
  onActionClick: Function,
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onFocus: Function,
  onSearch: Function,
  onSubmit: Function
};

export { searchProps as default };
