// tdesign-mobile-vue/cjs/slider/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  disabled: {
    type: Boolean,
    default: void 0
  },
  label: {
    type: [String, Boolean, Function],
    default: false
  },
  marks: {
    type: [Object, Array]
  },
  max: {
    type: Number,
    default: 100
  },
  min: {
    type: Number,
    default: 0
  },
  range: Boolean,
  showExtremeValue: Boolean,
  step: {
    type: Number,
    default: 1
  },
  theme: {
    type: String,
    default: "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "capsule"].includes(val);
    }
  },
  value: {
    type: [Number, Array],
    default: void 0
  },
  modelValue: {
    type: [Number, Array],
    default: void 0
  },
  defaultValue: {
    type: [Number, Array],
    default: 0
  },
  vertical: Boolean,
  onChange: Function,
  onDragend: Function,
  onDragstart: Function
};

exports["default"] = props;
