// tdesign-mobile-vue/cjs/_common/js/color-picker/index.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _common_js_colorPicker_cmyk = require('./cmyk.js');
var _common_js_colorPicker_color = require('./color.js');
var _common_js_colorPicker_constants = require('./constants.js');
var _common_js_colorPicker_draggable = require('./draggable.js');
var _common_js_colorPicker_format = require('./format.js');
var _common_js_colorPicker_gradient = require('./gradient.js');
require('@babel/runtime/helpers/slicedToArray');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('@babel/runtime/helpers/defineProperty');
require('tinycolor2');
require('../../../_chunks/dep-588e2e70.js');
require('../../../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../../../_chunks/dep-3cf5587c.js');
require('../../../_chunks/dep-f84e6e17.js');
require('../../../_chunks/dep-af0b484e.js');
require('@babel/runtime/helpers/toConsumableArray');



exports.cmyk2rgb = _common_js_colorPicker_cmyk.cmyk2rgb;
exports.cmykInputToColor = _common_js_colorPicker_cmyk.cmykInputToColor;
exports.rgb2cmyk = _common_js_colorPicker_cmyk.rgb2cmyk;
exports.Color = _common_js_colorPicker_color.Color;
exports.genGradientPoint = _common_js_colorPicker_color.genGradientPoint;
exports.genId = _common_js_colorPicker_color.genId;
exports.getColorObject = _common_js_colorPicker_color.getColorObject;
exports.getColorWithoutAlpha = _common_js_colorPicker_color.getColorWithoutAlpha;
exports.gradientColors2string = _common_js_colorPicker_color.gradientColors2string;
exports.ALPHA_FORMAT_MAP = _common_js_colorPicker_constants.ALPHA_FORMAT_MAP;
exports.COLOR_FORMAT_INPUTS = _common_js_colorPicker_constants.COLOR_FORMAT_INPUTS;
exports.COLOR_MODES = _common_js_colorPicker_constants.COLOR_MODES;
exports.DEFAULT_COLOR = _common_js_colorPicker_constants.DEFAULT_COLOR;
exports.DEFAULT_LINEAR_GRADIENT = _common_js_colorPicker_constants.DEFAULT_LINEAR_GRADIENT;
exports.DEFAULT_SYSTEM_SWATCH_COLORS = _common_js_colorPicker_constants.DEFAULT_SYSTEM_SWATCH_COLORS;
exports.FORMATS = _common_js_colorPicker_constants.FORMATS;
exports.GRADIENT_SLIDER_DEFAULT_WIDTH = _common_js_colorPicker_constants.GRADIENT_SLIDER_DEFAULT_WIDTH;
exports.SATURATION_PANEL_DEFAULT_HEIGHT = _common_js_colorPicker_constants.SATURATION_PANEL_DEFAULT_HEIGHT;
exports.SATURATION_PANEL_DEFAULT_WIDTH = _common_js_colorPicker_constants.SATURATION_PANEL_DEFAULT_WIDTH;
exports.SLIDER_DEFAULT_WIDTH = _common_js_colorPicker_constants.SLIDER_DEFAULT_WIDTH;
exports.TD_COLOR_USED_COLORS_MAX_SIZE = _common_js_colorPicker_constants.TD_COLOR_USED_COLORS_MAX_SIZE;
exports.Draggable = _common_js_colorPicker_draggable.Draggable;
exports.getColorFormatInputs = _common_js_colorPicker_format.getColorFormatInputs;
exports.getColorFormatMap = _common_js_colorPicker_format.getColorFormatMap;
exports.getColorFormatOptions = _common_js_colorPicker_format.getColorFormatOptions;
exports.initColorFormat = _common_js_colorPicker_format.initColorFormat;
exports.isGradientColor = _common_js_colorPicker_gradient.isGradientColor;
exports.parseGradientString = _common_js_colorPicker_gradient.parseGradientString;
