// tdesign-mobile-vue/lib/side-bar/side-bar.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { defineComponent, toRefs, ref, provide, createVNode } from 'vue';
import config from '../config.js';
import props from './props.js';
import useVModel from '../hooks/useVModel.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import 'lodash-es';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _SideBar = defineComponent({
  name: "".concat(prefix, "-side-bar"),
  props: props,
  emits: ["update:value", "update:modelValue", "change"],
  setup: function setup(props2, context) {
    var renderTNodeJSX = useTNodeJSX();
    var sideBarClass = usePrefixClass("side-bar");
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      currentValue = _useVModel2[0],
      setCurrentValue = _useVModel2[1];
    var children = ref([]);
    var relation = function relation(child) {
      child && children.value.push(child);
    };
    var removeRelation = function removeRelation(child) {
      children.value = children.value.filter(function (item) {
        return item !== child;
      });
    };
    var onClickItem = function onClickItem(cur, label) {
      var _props2$onClick;
      setCurrentValue(cur);
      (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, cur, label);
    };
    provide("sideBarProvide", _objectSpread(_objectSpread({}, props2), {}, {
      children: children,
      currentValue: currentValue,
      relation: relation,
      removeRelation: removeRelation,
      onClickItem: onClickItem
    }));
    return function () {
      return createVNode("div", {
        "class": "".concat(sideBarClass.value)
      }, [renderTNodeJSX("default"), createVNode("div", {
        "class": "".concat(sideBarClass.value, "__padding")
      }, null)]);
    };
  }
});

export { _SideBar as default };
