// tdesign-mobile-vue/lib/result/result.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, computed, createVNode } from 'vue';
import { InfoCircleIcon, CheckCircleIcon, CloseCircleIcon } from 'tdesign-icons-vue-next';
import config from '../config.js';
import ResultProps from './props.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { Image } from '../image/index.js';
import { useIcon } from '../hooks/icon.js';
import { usePrefixClass } from '../hooks/useClass.js';
import 'lodash-es';
import '../hooks/render-tnode.js';
import '../_common/js/utils/general.js';
import '../image/image.js';
import '@babel/runtime/helpers/slicedToArray';
import '@babel/runtime/helpers/defineProperty';
import '@vueuse/core';
import '../loading/index.js';
import '../loading/loading.js';
import '../loading/icon/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/setStyle.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/toConsumableArray';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../loading/icon/spinner.js';
import '../loading/props.js';
import '../shared/dom.js';
import '../loading/plugin.js';
import '../config-provider/useConfig.js';
import '../config-provider/context.js';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-dc49adb8.js';
import 'dayjs';
import '../shared/component.js';
import '../image/props.js';

var prefix = config.prefix;
var _Result = defineComponent({
  name: "".concat(prefix, "-result"),
  props: ResultProps,
  setup: function setup(props) {
    var resultClass = usePrefixClass("result");
    var renderTNodeJSX = useTNodeJSX();
    var renderIconTNode = useIcon();
    var resultClasses = computed(function () {
      return ["".concat(resultClass.value), "".concat(resultClass.value, "--theme-").concat(props.theme || "default")];
    });
    var renderIcon = function renderIcon() {
      var defaultIcons = {
        default: InfoCircleIcon,
        success: CheckCircleIcon,
        warning: InfoCircleIcon,
        error: CloseCircleIcon
      };
      var iconContent = renderIconTNode("icon", defaultIcons);
      return iconContent ? createVNode(iconContent, {
        "class": "".concat(resultClass.value, "__icon")
      }, null) : null;
    };
    var renderImage = function renderImage() {
      var image = renderTNodeJSX("image");
      if (image) {
        if (typeof image === "string") {
          return createVNode(Image, {
            "src": image
          }, null);
        }
        return image;
      }
      return null;
    };
    var renderThumb = function renderThumb() {
      var image = renderImage();
      return createVNode("div", {
        "class": "".concat(resultClass.value, "__thumb")
      }, [image || renderIcon()]);
    };
    var renderTitle = function renderTitle() {
      var title = renderTNodeJSX("title");
      return title ? createVNode("div", {
        "class": ["".concat(resultClass.value, "__title")]
      }, [title]) : null;
    };
    var renderDescription = function renderDescription() {
      var description = renderTNodeJSX("description");
      return description ? createVNode("div", {
        "class": ["".concat(resultClass.value, "__description")]
      }, [description]) : null;
    };
    return function () {
      return createVNode("div", {
        "class": [resultClasses.value]
      }, [renderThumb(), renderTitle(), renderDescription()]);
    };
  }
});

export { _Result as default };
