// tdesign-mobile-vue/es/fab/fab.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { defineComponent, ref, computed, onMounted, watch, createVNode, mergeProps } from 'vue';
import config from '../config.mjs';
import FabProps from './props.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { Button } from '../button/index.mjs';
import '../shared/index.mjs';
import { reconvertUnit } from '../shared/util.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'lodash-es';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../button/button.mjs';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../shared/component.mjs';
import '../shared/functions.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../button/props.mjs';
import '../form/hooks.mjs';
import '../button/type.mjs';

var prefix = config.prefix;
var _Fab = defineComponent({
  name: "".concat(prefix, "-fab"),
  props: FabProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var fabClass = usePrefixClass("fab");
    var fabRef = ref();
    var fabButtonRef = ref();
    var customNodeRef = ref();
    var handleClick = function handleClick(e) {
      var _props$onClick;
      (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, {
        e: e
      });
    };
    var mounted = ref(false);
    var fabButtonSize = ref({
      width: 48,
      height: 48
    });
    var btnSwitchPos = ref({
      x: 16,
      y: 32
    });
    var switchPos = ref({
      hasMoved: false,
      x: btnSwitchPos.value.x,
      y: btnSwitchPos.value.y,
      startX: 0,
      startY: 0,
      endX: 0,
      endY: 0
    });
    var onTouchStart = function onTouchStart(e) {
      var _props$onDragStart;
      (_props$onDragStart = props.onDragStart) === null || _props$onDragStart === void 0 || _props$onDragStart.call(props, {
        e: e
      });
      switchPos.value.startX = e.touches[0].pageX;
      switchPos.value.startY = e.touches[0].pageY;
    };
    var onTouchMove = function onTouchMove(e) {
      e.stopPropagation();
      e.preventDefault();
      if (!props.draggable) {
        return;
      }
      if (e.touches.length <= 0) {
        return;
      }
      var offsetX = e.touches[0].pageX - switchPos.value.startX;
      var offsetY = e.touches[0].pageY - switchPos.value.startY;
      var x = Math.floor(switchPos.value.x - offsetX);
      var y = Math.floor(switchPos.value.y - offsetY);
      var _getSwitchButtonSafeA = getSwitchButtonSafeAreaXY(x, y);
      var _getSwitchButtonSafeA2 = _slicedToArray(_getSwitchButtonSafeA, 2);
      x = _getSwitchButtonSafeA2[0];
      y = _getSwitchButtonSafeA2[1];
      if (props.draggable !== "vertical") {
        btnSwitchPos.value.x = x;
        switchPos.value.endX = x;
      }
      if (props.draggable !== "horizontal") {
        btnSwitchPos.value.y = y;
        switchPos.value.endY = y;
      }
      switchPos.value.hasMoved = true;
    };
    var getSwitchButtonSafeAreaXY = function getSwitchButtonSafeAreaXY(x, y) {
      var _props$yBounds$, _props$yBounds, _props$yBounds$2, _props$yBounds2;
      var bottomThreshold = reconvertUnit((_props$yBounds$ = (_props$yBounds = props.yBounds) === null || _props$yBounds === void 0 ? void 0 : _props$yBounds[1]) !== null && _props$yBounds$ !== void 0 ? _props$yBounds$ : 0);
      var topThreshold = reconvertUnit((_props$yBounds$2 = (_props$yBounds2 = props.yBounds) === null || _props$yBounds2 === void 0 ? void 0 : _props$yBounds2[0]) !== null && _props$yBounds$2 !== void 0 ? _props$yBounds$2 : 0);
      var docWidth = Math.min(window.innerWidth, document.documentElement.clientWidth, screen.width);
      var docHeight = Math.min(window.innerHeight, document.documentElement.clientHeight, screen.height);
      var maxY = docHeight - fabButtonSize.value.height - topThreshold;
      var maxX = docWidth - fabButtonSize.value.width;
      x = Math.max(0, Math.min(maxX, x));
      y = Math.max(bottomThreshold, Math.min(maxY, y));
      return [x, y];
    };
    var onTouchEnd = function onTouchEnd(e) {
      var _props$onDragEnd;
      if (!switchPos.value.hasMoved) {
        return;
      }
      (_props$onDragEnd = props.onDragEnd) === null || _props$onDragEnd === void 0 || _props$onDragEnd.call(props, {
        e: e
      });
      switchPos.value.startX = 0;
      switchPos.value.startY = 0;
      switchPos.value.hasMoved = false;
      setSwitchPosition(switchPos.value.endX, switchPos.value.endY);
    };
    var setSwitchPosition = function setSwitchPosition(switchX, switchY) {
      var _getSwitchButtonSafeA3 = getSwitchButtonSafeAreaXY(switchX, switchY);
      var _getSwitchButtonSafeA4 = _slicedToArray(_getSwitchButtonSafeA3, 2);
      switchX = _getSwitchButtonSafeA4[0];
      switchY = _getSwitchButtonSafeA4[1];
      switchPos.value.x = switchX;
      switchPos.value.y = switchY;
      if (props.draggable !== "vertical") {
        btnSwitchPos.value.x = switchX;
      }
      if (props.draggable !== "horizontal") {
        btnSwitchPos.value.y = switchY;
      }
    };
    var fabStyle = computed(function () {
      var _btnSwitchPos$value = btnSwitchPos.value,
        x = _btnSwitchPos$value.x,
        y = _btnSwitchPos$value.y;
      return "right: ".concat(x, "px;bottom: ").concat(y, "px;");
    });
    onMounted(function () {
      mounted.value = true;
      resetDraggableParams();
      if (fabButtonRef.value || customNodeRef.value) {
        var _fabButtonRef$value;
        var info = window.getComputedStyle(((_fabButtonRef$value = fabButtonRef.value) === null || _fabButtonRef$value === void 0 ? void 0 : _fabButtonRef$value.$el) || customNodeRef.value);
        fabButtonSize.value.height = +reconvertUnit(info.height);
        fabButtonSize.value.width = +reconvertUnit(info.width);
      }
    });
    var getFabOriginStyle = function getFabOriginStyle() {
      var info = window.getComputedStyle(fabRef.value);
      var _ref = info || {},
        right = _ref.right,
        bottom = _ref.bottom;
      return {
        right: +(reconvertUnit(right) || 0),
        bottom: +(reconvertUnit(bottom) || 0)
      };
    };
    var resetDraggableParams = function resetDraggableParams() {
      var _getFabOriginStyle = getFabOriginStyle(),
        right = _getFabOriginStyle.right,
        bottom = _getFabOriginStyle.bottom;
      btnSwitchPos.value.x = right;
      btnSwitchPos.value.y = bottom;
      switchPos.value.x = right;
      switchPos.value.y = bottom;
    };
    watch(function () {
      return props.style;
    }, function () {
      resetDraggableParams();
    });
    return function () {
      var icon = function icon() {
        return renderTNodeJSX("icon");
      };
      var defaultContent = createVNode(Button, mergeProps({
        "size": "large",
        "theme": "primary",
        "shape": props.text ? "round" : "circle",
        "class": "".concat(fabClass.value, "__button")
      }, props.buttonProps, {
        "icon": icon,
        "ref": fabButtonRef
      }), {
        default: function _default() {
          return [props.text];
        }
      });
      var customNode = renderTNodeJSX("default");
      return createVNode("div", {
        "class": fabClass.value,
        "style": mounted.value && props.draggable ? fabStyle.value : props.style,
        "onClick": handleClick,
        "onTouchstart": onTouchStart,
        "onTouchmove": onTouchMove,
        "onTouchend": onTouchEnd,
        "ref": fabRef
      }, [customNode ? createVNode("div", {
        "ref": customNodeRef
      }, [customNode]) : defaultContent]);
    };
  }
});

export { _Fab as default };
