// tdesign-mobile-vue/es/action-sheet/action-sheet-list.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { isVNode, defineComponent, toRefs, computed, createVNode } from 'vue';
import { Button } from '../button/index.mjs';
import { Badge } from '../badge/index.mjs';
import { useTNodeDefault } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../button/button.mjs';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import 'lodash-es';
import '../config.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../button/props.mjs';
import '../form/hooks.mjs';
import '../button/type.mjs';
import '../badge/badge.mjs';
import '../badge/props.mjs';
import '../badge/type.mjs';

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var TActionSheetList = defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    },
    align: {
      type: String,
      default: "center"
    }
  },
  emits: ["selected"],
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    var actionSheetClass = usePrefixClass("action-sheet");
    var renderTNodeJSX = useTNodeDefault();
    var _toRefs = toRefs(props),
      align = _toRefs.align,
      items = _toRefs.items;
    var handleSelected = function handleSelected(index) {
      emit("selected", index);
    };
    var itemClasses = computed(function () {
      return _defineProperty(_defineProperty({}, "".concat(actionSheetClass.value, "__list-item"), true), "".concat(actionSheetClass.value, "__list-item--left"), align.value === "left");
    });
    return function () {
      var renderButtonNode = function renderButtonNode() {
        var renderBadgeNode = function renderBadgeNode(item) {
          if (item.badge) {
            var content = function content() {
              if (item.badge.dot || item.badge.count) {
                return createVNode(Badge, {
                  "count": item.badge.count,
                  "max-count": item.badge.maxCount || 99,
                  "dot": item.badge.dot,
                  "content": item.badge.content,
                  "size": item.badge.size,
                  "offset": item.badge.offset || [-16, 20]
                }, {
                  default: function _default() {
                    return [createVNode("span", {
                      "class": "".concat(actionSheetClass.value, "__list-item-text")
                    }, [" ", item.label])];
                  }
                });
              }
              return renderTNodeJSX("badge", {
                defaultNode: createVNode("span", {
                  "class": "".concat(actionSheetClass.value, "__list-item-text")
                }, [item.label])
              });
            };
            return content();
          }
          return createVNode("span", {
            "class": "".concat(actionSheetClass.value, "__list-item-text")
          }, [item.label]);
        };
        var buttonList = items.value.map(function (item, index) {
          var _slot;
          return createVNode(Button, {
            "key": index,
            "variant": "text",
            "block": true,
            "class": [itemClasses.value, _defineProperty({}, "".concat(actionSheetClass.value, "__list-item--disabled"), item.disabled)],
            "disabled": item.disabled,
            "icon": item.icon,
            "style": {
              color: item.color
            },
            "onClick": function onClick() {
              return handleSelected(index);
            }
          }, _isSlot(_slot = renderBadgeNode(item)) ? _slot : {
            default: function _default() {
              return [_slot];
            }
          });
        });
        return buttonList;
      };
      return createVNode("div", {
        "class": "".concat(actionSheetClass.value, "__list")
      }, [renderButtonNode()]);
    };
  }
});

export { TActionSheetList as default };
