// tdesign-mobile-vue/es/action-sheet/action-sheet-grid.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { isVNode, defineComponent, computed, createVNode } from 'vue';
import { GridItem, Grid } from '../grid/index.mjs';
import { SwiperItem, Swiper } from '../swiper/index.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../grid/grid.mjs';
import '../config.mjs';
import '../grid/props.mjs';
import '../config-provider/useConfig.mjs';
import 'lodash-es';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../grid/grid-item.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../grid/grid-item-props.mjs';
import '../hooks/tnode.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../image/index.mjs';
import '../image/image.mjs';
import 'tdesign-icons-vue-next';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../image/props.mjs';
import '../image/type.mjs';
import '../badge/index.mjs';
import '../badge/badge.mjs';
import '../badge/props.mjs';
import '../badge/type.mjs';
import '../grid/type.mjs';
import '../swiper/swiper.mjs';
import '../swipe-cell/useSwipe.mjs';
import '../swiper/props.mjs';
import '../hooks/useVModel.mjs';
import '../swiper/swiper-item.mjs';
import '../swiper/type.mjs';

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var TActionSheetGrid = defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    },
    count: {
      type: Number,
      default: 8
    },
    selected: {
      type: Function,
      default: void 0
    }
  },
  emits: ["selected"],
  setup: function setup(props, _ref) {
    var emit = _ref.emit;
    var actionSheetClass = usePrefixClass("action-sheet");
    var gridColumn = computed(function () {
      return Math.ceil(props.count / 2);
    });
    var pageNum = computed(function () {
      return Math.ceil(props.items.length / props.count);
    });
    var actionItems = computed(function () {
      var res = [];
      for (var i = 0; i < pageNum.value; i++) {
        var temp = props.items.slice(i * props.count, (i + 1) * props.count);
        res.push(temp);
      }
      return res;
    });
    var gridClasses = computed(function () {
      return _defineProperty(_defineProperty(_defineProperty({}, "".concat(actionSheetClass.value, "__grid"), true), "".concat(actionSheetClass.value, "__grid--swiper"), pageNum.value > 1), "".concat(actionSheetClass.value, "__dots"), pageNum.value > 1);
    });
    var handleSelected = function handleSelected(i) {
      emit("selected", i);
    };
    return function () {
      var swiper = function swiper() {
        var swiperItems = actionItems.value.map(function (items, i) {
          var gridItems = items.map(function (item, index) {
            return createVNode(GridItem, {
              "key": index,
              "text": item.label,
              "image": item.icon,
              "badge": item.badge,
              "onClick": function onClick(event) {
                event.preventDefault();
                handleSelected(i * props.count + index);
              }
            }, null);
          });
          return createVNode(SwiperItem, {
            "key": i
          }, {
            default: function _default() {
              return [createVNode(Grid, {
                "column": gridColumn.value
              }, _isSlot(gridItems) ? gridItems : {
                default: function _default() {
                  return [gridItems];
                }
              })];
            }
          });
        });
        if (actionItems.value.length > 1) {
          return createVNode(Swiper, {
            "autoplay": false,
            "pagination-position": "bottom",
            "navigation": {
              type: "dots",
              showControls: false
            },
            "loop": false,
            "class": "".concat(actionSheetClass.value, "__swiper-wrap"),
            "height": 192
          }, _isSlot(swiperItems) ? swiperItems : {
            default: function _default() {
              return [swiperItems];
            }
          });
        }
        return null;
      };
      var grid = function grid() {
        if (actionItems.value.length === 1) {
          var items = actionItems.value[0].map(function (item, index) {
            return createVNode(GridItem, {
              "key": index,
              "text": item.label,
              "image": item.icon,
              "badge": item.badge,
              "onClick": function onClick() {
                return handleSelected(index);
              }
            }, null);
          });
          return createVNode(Grid, {
            "column": gridColumn.value
          }, _isSlot(items) ? items : {
            default: function _default() {
              return [items];
            }
          });
        }
        return null;
      };
      return createVNode("div", {
        "class": gridClasses.value
      }, [swiper(), grid()]);
    };
  }
});

export { TActionSheetGrid as default };
