// tdesign-mobile-vue/es/action-sheet/action-sheet.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { isVNode, defineComponent, computed, toRefs, createVNode, mergeProps } from 'vue';
import TActionSheetGrid from './action-sheet-grid.mjs';
import TActionSheetList from './action-sheet-list.mjs';
import { Button } from '../button/index.mjs';
import { Popup } from '../popup/index.mjs';
import config from '../config.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import useVModel from '../hooks/useVModel.mjs';
import props from './props.mjs';
import { useConfig } from '../config-provider/useConfig.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../grid/index.mjs';
import '../grid/grid.mjs';
import '../grid/props.mjs';
import 'lodash-es';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../grid/grid-item.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/dom.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../shared/hover.mjs';
import '../grid/grid-item-props.mjs';
import '../hooks/tnode.mjs';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../image/index.mjs';
import '../image/image.mjs';
import 'tdesign-icons-vue-next';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../image/props.mjs';
import '../image/type.mjs';
import '../badge/index.mjs';
import '../badge/badge.mjs';
import '../badge/props.mjs';
import '../badge/type.mjs';
import '../grid/type.mjs';
import '../swiper/index.mjs';
import '../swiper/swiper.mjs';
import '../swipe-cell/useSwipe.mjs';
import '../swiper/props.mjs';
import '../swiper/swiper-item.mjs';
import '../swiper/type.mjs';
import '../button/button.mjs';
import '../button/props.mjs';
import '../form/hooks.mjs';
import '../button/type.mjs';
import '../popup/popup.mjs';
import '../popup/props.mjs';
import '../overlay/index.mjs';
import '../overlay/overlay.mjs';
import '../overlay/props.mjs';
import '../hooks/useLockScroll.mjs';
import '../_util/useTouch.mjs';
import '../_util/getScrollParent.mjs';
import '../hooks/useTeleport.mjs';
import '../popup/type.mjs';

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}
var prefix = config.prefix;
var ActionSheetVue = defineComponent({
  name: "".concat(prefix, "-action-sheet"),
  props: props,
  emits: ["selected", "update:modelValue", "cancel", "close", "update:visible"],
  setup: function setup(props2, context) {
    var actionSheetClass = usePrefixClass("action-sheet");
    var _useConfig = useConfig("actionSheet"),
      globalConfig = _useConfig.globalConfig;
    var actionItems = computed(function () {
      return props2.items.map(function (item) {
        if (typeof item === "string") {
          return {
            label: item
          };
        }
        return item;
      });
    });
    var _toRefs = toRefs(props2),
      visible = _toRefs.visible,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(visible, modelValue, props2.defaultVisible, function () {}, "visible"),
      _useVModel2 = _slicedToArray(_useVModel, 1),
      currentVisible = _useVModel2[0];
    var rootClasses = computed(function () {
      return _defineProperty({}, "".concat(actionSheetClass.value, "__content"), true);
    });
    var descriptionClasses = computed(function () {
      return _defineProperty(_defineProperty(_defineProperty({}, "".concat(actionSheetClass.value, "__description"), true), "".concat(actionSheetClass.value, "__description--left"), props2.align === "left"), "".concat(actionSheetClass.value, "__description--grid"), props2.theme === "grid");
    });
    var hide = function hide(trigger) {
      var _props2$onClose;
      context.emit("update:modelValue", false);
      (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, {
        trigger: trigger
      });
    };
    var handleCancel = function handleCancel() {
      var _props2$onCancel;
      (_props2$onCancel = props2.onCancel) === null || _props2$onCancel === void 0 || _props2$onCancel.call(props2);
      context.emit("update:modelValue", false);
    };
    var handleSelected = function handleSelected(index) {
      var _props2$onSelected;
      (_props2$onSelected = props2.onSelected) === null || _props2$onSelected === void 0 || _props2$onSelected.call(props2, props2 === null || props2 === void 0 ? void 0 : props2.items[index], index);
      hide("selected");
    };
    var handleClose = function handleClose() {
      hide("overlay");
    };
    return function () {
      var _slot;
      var root = function root() {
        var description = function description() {
          if (props2.description) {
            return createVNode("p", {
              "class": descriptionClasses.value
            }, [props2.description]);
          }
          return null;
        };
        var list = function list() {
          if (props2.theme === "list") {
            return createVNode(TActionSheetList, {
              "align": props2.align,
              "items": actionItems.value,
              "onSelected": handleSelected
            }, null);
          }
        };
        var grid = function grid() {
          if (props2.theme === "grid") {
            return createVNode(TActionSheetGrid, {
              "items": actionItems.value,
              "count": props2.count,
              "onSelected": handleSelected
            }, null);
          }
        };
        var cancel = function cancel() {
          if (props2.showCancel) {
            return createVNode("div", {
              "class": "".concat(actionSheetClass.value, "__footer")
            }, [createVNode("div", {
              "class": "".concat(actionSheetClass.value, "__gap-").concat(props2.theme)
            }, null), createVNode(Button, {
              "class": "".concat(actionSheetClass.value, "__cancel"),
              "variant": "text",
              "block": true,
              "onClick": handleCancel
            }, {
              default: function _default() {
                return [props2.cancelText || globalConfig.value.cancel];
              }
            })]);
          }
        };
        return createVNode("div", {
          "class": rootClasses.value
        }, [description(), list(), grid(), cancel()]);
      };
      return createVNode(Popup, mergeProps(props2.popupProps, {
        "visible": currentVisible.value,
        "placement": "bottom",
        "destroy-on-close": true,
        "class": actionSheetClass.value,
        "onClose": handleClose,
        "showOverlay": props2.showOverlay
      }), _isSlot(_slot = root()) ? _slot : {
        default: function _default() {
          return [_slot];
        }
      });
    };
  }
});

export { ActionSheetVue as default };
