// tdesign-mobile-vue/es/tab-bar/tab-bar.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-f4e31210.mjs';
import { _ as _slicedToArray } from '../_chunks/dep-2ed7e0d5.mjs';
import { defineComponent, toRefs, ref, computed, provide, createVNode } from 'vue';
import TabBarProps from './props.mjs';
import config from '../config.mjs';
import useChildSlots from '../hooks/useChildSlots.mjs';
import useVModel from '../hooks/useVModel.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-fb069758.mjs';
import 'lodash-es';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var _TabBar = defineComponent({
  name: "".concat(prefix, "-tab-bar"),
  props: TabBarProps,
  emits: ["update:value", "update:modelValue", "change"],
  setup: function setup(props, context) {
    var tabBarClass = usePrefixClass("tab-bar");
    var renderTNodeJSX = useTNodeJSX();
    var _toRefs = toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      activeValue = _useVModel2[0],
      setActiveValue = _useVModel2[1];
    var defaultIndex = ref(-1);
    var itemCount = ref(0);
    var updateChild = function updateChild(currentValue) {
      setActiveValue(currentValue);
    };
    var rootClass = computed(function () {
      return ["".concat(tabBarClass.value), _defineProperty(_defineProperty(_defineProperty({}, "".concat(tabBarClass.value, "--bordered"), props.bordered), "".concat(tabBarClass.value, "--fixed"), props.fixed), "".concat(tabBarClass.value, "--safe"), props.safeAreaInsetBottom), "".concat(tabBarClass.value, "--").concat(props.shape)];
    });
    provide("tab-bar", _objectSpread(_objectSpread({}, toRefs(props)), {}, {
      defaultIndex: defaultIndex,
      activeValue: activeValue,
      itemCount: itemCount,
      updateChild: updateChild
    }));
    var updateItemCount = function updateItemCount(vNodes) {
      if (!vNodes || !Array.isArray(vNodes)) {
        itemCount.value = 0;
        return;
      }
      var childSlots = useChildSlots("".concat(prefix, "-tab-bar-item"), vNodes);
      itemCount.value = childSlots.length;
    };
    return function () {
      var vNodes = context.slots.default ? context.slots.default() : [];
      updateItemCount(vNodes);
      return createVNode("div", {
        "class": rootClass.value,
        "role": "tablist"
      }, [renderTNodeJSX("default")]);
    };
  }
});

export { _TabBar as default };
