// tdesign-mobile-vue/es/result/result.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, computed, createVNode } from 'vue';
import { InfoCircleIcon, CheckCircleIcon, CloseCircleIcon } from 'tdesign-icons-vue-next';
import config from '../config.mjs';
import ResultProps from './props.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { Image } from '../image/index.mjs';
import { useIcon } from '../hooks/icon.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import 'lodash-es';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../image/image.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '@vueuse/core';
import '../loading/index.mjs';
import '../loading/loading.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import '../shared/index.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/component.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useScrollParent/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';
import '../config-provider/useConfig.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../image/props.mjs';
import '../image/type.mjs';

var prefix = config.prefix;
var _Result = defineComponent({
  name: "".concat(prefix, "-result"),
  props: ResultProps,
  setup: function setup(props) {
    var resultClass = usePrefixClass("result");
    var renderTNodeJSX = useTNodeJSX();
    var renderIconTNode = useIcon();
    var resultClasses = computed(function () {
      return ["".concat(resultClass.value), "".concat(resultClass.value, "--theme-").concat(props.theme || "default")];
    });
    var renderIcon = function renderIcon() {
      var defaultIcons = {
        default: InfoCircleIcon,
        success: CheckCircleIcon,
        warning: InfoCircleIcon,
        error: CloseCircleIcon
      };
      var iconContent = renderIconTNode("icon", defaultIcons);
      return iconContent ? createVNode(iconContent, {
        "class": "".concat(resultClass.value, "__icon")
      }, null) : null;
    };
    var renderImage = function renderImage() {
      var image = renderTNodeJSX("image");
      if (image) {
        if (typeof image === "string") {
          return createVNode(Image, {
            "src": image
          }, null);
        }
        return image;
      }
      return null;
    };
    var renderThumb = function renderThumb() {
      var image = renderImage();
      return createVNode("div", {
        "class": "".concat(resultClass.value, "__thumb")
      }, [image || renderIcon()]);
    };
    var renderTitle = function renderTitle() {
      var title = renderTNodeJSX("title");
      return title ? createVNode("div", {
        "class": ["".concat(resultClass.value, "__title")]
      }, [title]) : null;
    };
    var renderDescription = function renderDescription() {
      var description = renderTNodeJSX("description");
      return description ? createVNode("div", {
        "class": ["".concat(resultClass.value, "__description")]
      }, [description]) : null;
    };
    return function () {
      return createVNode("div", {
        "class": [resultClasses.value]
      }, [renderThumb(), renderTitle(), renderDescription()]);
    };
  }
});

export { _Result as default };
