// tdesign-mobile-vue/cjs/dialog/dialog.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var Vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var button_index = require('../button/index.js');
var popup_index = require('../popup/index.js');
var config = require('../config.js');
var dialog_props = require('./props.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var get = require('../_chunks/dep-cfb75f1b.js');
var isString = require('../_chunks/dep-588e2e70.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
require('../button/button.js');
require('../loading/index.js');
require('../loading/loading.js');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-055d28fd.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-f84e6e17.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../_chunks/dep-0bd122a0.js');
require('../loading/plugin.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-605cd804.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../button/props.js');
require('../form/hooks.js');
require('../_chunks/dep-eb605cd5.js');
require('../shared/hover.js');
require('../popup/popup.js');
require('../popup/props.js');
require('../overlay/index.js');
require('../overlay/overlay.js');
require('../overlay/props.js');
require('../hooks/useVModel.js');
require('../hooks/useLockScroll.js');
require('../_util/useTouch.js');
require('../_util/getScrollParent.js');
require('../hooks/useTeleport.js');
require('../shared/render-tnode.js');
require('../shared/util.js');
require('../_chunks/dep-01310648.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var Dialog = Vue.defineComponent({
  name: "".concat(prefix, "-dialog"),
  props: dialog_props["default"],
  emits: ["update:visible", "confirm", "overlay-click", "cancel", "close", "closed"],
  setup: function setup(props2, context) {
    var dialogClass = hooks_useClass.usePrefixClass("dialog");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var renderContent = hooks_tnode.useContent();
    var isTextStyleBtn = Vue.computed(function () {
      return [props2 === null || props2 === void 0 ? void 0 : props2.confirmBtn, props2 === null || props2 === void 0 ? void 0 : props2.cancelBtn].concat(_toConsumableArray__default["default"]((props2 === null || props2 === void 0 ? void 0 : props2.actions) || [])).some(function (item) {
        return get.get(item, "variant") === "text";
      });
    });
    var footerClass = Vue.computed(function () {
      return ["".concat(dialogClass.value, "__footer"), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(dialogClass.value, "__footer--column"), props2.buttonLayout === "vertical"), "".concat(dialogClass.value, "__footer--full"), isTextStyleBtn.value && get.get(props2.actions, "length", 0) === 0)];
    });
    var buttonClass = Vue.computed(function () {
      return ["".concat(dialogClass.value, "__button"), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(dialogClass.value, "__button--").concat(props2.buttonLayout), !isTextStyleBtn.value), "".concat(dialogClass.value, "__button--text"), isTextStyleBtn.value)];
    });
    var rootStyles = Vue.computed(function () {
      return {
        width: isString.isString(props2.width) ? props2.width : "".concat(props2.width, "px")
      };
    });
    var handleClose = function handleClose(args) {
      var e = args.e;
      context.emit("update:visible", false);
      context.emit("close", {
        e: e,
        trigger: "close-btn"
      });
    };
    var handleClosed = function handleClosed() {
      context.emit("closed");
    };
    var handleConfirm = function handleConfirm(e) {
      var _context$emit;
      context.emit("update:visible", false);
      (_context$emit = context.emit) === null || _context$emit === void 0 || _context$emit.call(context, "confirm", {
        e: e
      });
    };
    var handleCancel = function handleCancel(e) {
      context.emit("update:visible", false);
      context.emit("close", {
        e: e,
        trigger: "cancel"
      });
      context.emit("cancel", {
        e: e
      });
    };
    var handleOverlayClick = function handleOverlayClick(args) {
      var e = args.e;
      if (!props2.closeOnOverlayClick) {
        return;
      }
      context.emit("update:visible", false);
      context.emit("close", {
        e: e,
        trigger: "overlay"
      });
      context.emit("overlay-click", {
        e: e
      });
    };
    var calcBtn = function calcBtn(btn) {
      if (isString.isString(btn)) {
        return {
          content: btn
        };
      }
      if (isObject.isObject(btn)) {
        return btn;
      }
      return {};
    };
    var confirmBtnProps = Vue.computed(function () {
      return _objectSpread({
        theme: "primary"
      }, calcBtn(props2.confirmBtn));
    });
    var cancelBtnProps = Vue.computed(function () {
      return _objectSpread({
        theme: isTextStyleBtn.value ? "default" : "light"
      }, calcBtn(props2.cancelBtn));
    });
    var actionsBtnProps = Vue.computed(function () {
      var _props2$actions;
      return (_props2$actions = props2.actions) === null || _props2$actions === void 0 ? void 0 : _props2$actions.map(function (item) {
        return calcBtn(item);
      });
    });
    var renderButtonNode = function renderButtonNode(btnType, btnProps, handleClick) {
      if (actionsBtnProps.value) return null;
      var btnNode = renderTNodeJSX(btnType);
      if (!btnNode) return null;
      if (context.slots[btnType]) {
        return btnNode;
      }
      return Vue.createVNode(button_index.Button, Vue.mergeProps(btnProps, {
        "class": buttonClass.value,
        "onClick": handleClick
      }), null);
    };
    return function () {
      var renderTitleNode = function renderTitleNode() {
        var titleNode = renderTNodeJSX("title");
        if (!titleNode) {
          return null;
        }
        return Vue.createVNode("div", {
          "class": "".concat(dialogClass.value, "__header")
        }, [titleNode]);
      };
      var renderContentNode = function renderContentNode() {
        var contentNode = renderContent("default", "content");
        if (!contentNode) {
          return null;
        }
        return Vue.createVNode("div", {
          "class": "".concat(dialogClass.value, "__body")
        }, [Vue.createVNode("div", {
          "class": "".concat(dialogClass.value, "__body-text")
        }, [contentNode])]);
      };
      var renderActionsNode = function renderActionsNode() {
        if (actionsBtnProps.value) {
          return actionsBtnProps.value.map(function (item, index) {
            return Vue.createVNode(button_index.Button, Vue.mergeProps({
              "key": index
            }, item, {
              "class": buttonClass.value,
              "onClick": handleCancel
            }), null);
          });
        }
        return renderTNodeJSX("actions");
      };
      var renderCancelBtnNode = function renderCancelBtnNode() {
        return renderButtonNode("cancelBtn", cancelBtnProps.value, handleCancel);
      };
      var renderConfirmBtnNode = function renderConfirmBtnNode() {
        return renderButtonNode("confirmBtn", confirmBtnProps.value, handleConfirm);
      };
      return Vue.createVNode(popup_index.Popup, {
        "class": "".concat(dialogClass.value, "__wrapper"),
        "visible": props2.visible,
        "placement": "center",
        "show-overlay": props2.showOverlay,
        "overlay-props": props2.overlayProps,
        "prevent-scroll-through": props2.preventScrollThrough,
        "destroy-on-close": props2.destroyOnClose,
        "z-index": props2.zIndex,
        "onClose": handleOverlayClick,
        "onClosed": handleClosed
      }, {
        default: function _default() {
          return [Vue.createVNode("div", {
            "class": "".concat(dialogClass.value, " ").concat(context.attrs.class || ""),
            "style": rootStyles.value
          }, [renderTNodeJSX("top"), props2.closeBtn && Vue.createVNode("div", {
            "class": "".concat(dialogClass.value, "__close-btn")
          }, [Vue.createVNode(tdesignIconsVueNext.CloseIcon, {
            "onClick": handleClose
          }, null)]), Vue.createVNode("div", {
            "class": "".concat(dialogClass.value, "__content")
          }, [renderTitleNode(), renderContentNode()]), renderTNodeJSX("middle"), Vue.createVNode("div", {
            "class": footerClass.value
          }, [renderActionsNode(), renderCancelBtnNode(), renderConfirmBtnNode()])])];
        }
      });
    };
  }
});

exports["default"] = Dialog;
