// tdesign-mobile-vue/cjs/dialog/props.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  actions: {
    type: [Array, Function]
  },
  buttonLayout: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  cancelBtn: {
    type: [String, Object, Function]
  },
  closeBtn: Boolean,
  closeOnOverlayClick: Boolean,
  confirmBtn: {
    type: [String, Object, Function]
  },
  content: {
    type: [String, Function]
  },
  destroyOnClose: Boolean,
  middle: {
    type: Function
  },
  overlayProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  preventScrollThrough: {
    type: Boolean,
    default: true
  },
  showOverlay: {
    type: Boolean,
    default: true
  },
  title: {
    type: [String, Function]
  },
  top: {
    type: Function
  },
  visible: Boolean,
  width: {
    type: [String, Number]
  },
  zIndex: {
    type: Number
  },
  onCancel: Function,
  onClose: Function,
  onClosed: Function,
  onConfirm: Function,
  onOverlayClick: Function
};

exports["default"] = props;
