// tdesign-mobile-vue/es/form/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  colon: Boolean,
  contentAlign: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right"].includes(val);
    }
  },
  data: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  errorMessage: {
    type: Object
  },
  labelAlign: {
    type: String,
    default: "right",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right", "top"].includes(val);
    }
  },
  labelWidth: {
    type: [String, Number],
    default: "81px"
  },
  preventSubmitDefault: {
    type: Boolean,
    default: true
  },
  requiredMark: {
    type: Boolean,
    default: void 0
  },
  requiredMarkPosition: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right"].includes(val);
    }
  },
  resetType: {
    type: String,
    default: "empty",
    validator: function validator(val) {
      if (!val) return true;
      return ["empty", "initial"].includes(val);
    }
  },
  rules: {
    type: Object
  },
  scrollToFirstError: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["", "smooth", "auto"].includes(val);
    }
  },
  showErrorMessage: {
    type: Boolean,
    default: true
  },
  submitWithWarningMessage: Boolean,
  onReset: Function,
  onSubmit: Function,
  onValidate: Function
};

export { props as default };
