// tdesign-mobile-vue/es/list/list.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, ref, createVNode } from 'vue';
import { useWindowSize, useEventListener } from '@vueuse/core';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { Loading } from '../loading/index.mjs';
import config from '../config.mjs';
import props from './props.mjs';
import '../shared/index.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import { useConfig } from '../config-provider/useConfig.mjs';
import { useScrollParent } from '../shared/useScrollParent/index.mjs';
import 'lodash-es';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../loading/loading.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/setStyle.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-fb069758.mjs';
import '../_chunks/dep-9961b95b.mjs';
import '../_chunks/dep-a27c5c00.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../loading/icon/spinner.mjs';
import '../loading/props.mjs';
import '../shared/dom.mjs';
import './style/css.mjs';
import '../loading/type.mjs';
import '../loading/plugin.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';
import '../shared/component.mjs';
import '../shared/functions.mjs';
import '../shared/util.mjs';
import '../shared/render.mjs';
import '../shared/render-tnode.mjs';
import '../shared/useCountDown/index.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../shared/useCountDown/utils.mjs';
import '../shared/useTouch/index.mjs';
import '../shared/useTest/index.mjs';
import '../shared/useClickAway/index.mjs';
import '../shared/hover.mjs';

var prefix = config.prefix;
var _List = defineComponent({
  name: "".concat(prefix, "-list"),
  components: {
    TLoading: Loading
  },
  props: props,
  emits: ["load-more", "scroll"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var listClass = usePrefixClass("list");
    var _useConfig = useConfig("list"),
      globalConfig = _useConfig.globalConfig;
    var renderTNodeJSX = useTNodeJSX();
    var LOADING_TEXT_MAP = {
      loading: globalConfig.value.loading,
      "load-more": globalConfig.value.loadingMoreText
    };
    var root = ref();
    var scrollParent = useScrollParent(root);
    var _useWindowSize = useWindowSize(),
      height = _useWindowSize.height;
    var onLoadMore = function onLoadMore(e) {
      if (props2.asyncLoading === "load-more") {
        var _props2$onLoadMore;
        (_props2$onLoadMore = props2.onLoadMore) === null || _props2$onLoadMore === void 0 || _props2$onLoadMore.call(props2);
      }
    };
    var handleScroll = function handleScroll(e) {
      var _props2$onScroll;
      var scrollHeight = e.target.scrollHeight || Math.max(document.documentElement.scrollHeight, document.body.scrollHeight);
      var scrollTop = e.target.scrollTop || document.documentElement.scrollTop || document.body.scrollTop;
      var offsetHeight = e.target.offsetHeight || height.value;
      (_props2$onScroll = props2.onScroll) === null || _props2$onScroll === void 0 || _props2$onScroll.call(props2, scrollHeight - (scrollTop + offsetHeight), scrollTop);
    };
    var renderLoading = function renderLoading() {
      if (typeof props2.asyncLoading === "string" && ["loading", "load-more"].includes(props2.asyncLoading)) {
        return createVNode(Loading, {
          "indicator": props2.asyncLoading === "loading",
          "text": typeof props2.asyncLoading === "string" ? LOADING_TEXT_MAP[props2.asyncLoading] : "",
          "class": "".concat(listClass.value, "__loading")
        }, null);
      }
      return renderTNodeJSX("asyncLoading");
    };
    useEventListener(scrollParent, "scroll", handleScroll);
    return function () {
      var headerContent = renderTNodeJSX("header");
      var footerContent = renderTNodeJSX("footer");
      return createVNode("div", {
        "ref": root,
        "class": listClass.value,
        "onScroll": handleScroll
      }, [headerContent, slots.default && slots.default(), createVNode("div", {
        "class": "".concat(listClass.value, "__loading--wrapper"),
        "onClick": onLoadMore
      }, [renderLoading()]), footerContent]);
    };
  }
});

export { _List as default };
