// tdesign-mobile-vue/es/cascader/props.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

var props = {
  checkStrictly: Boolean,
  closeBtn: {
    type: [Boolean, Function],
    default: true
  },
  header: {
    type: Function
  },
  keys: {
    type: Object
  },
  middleContent: {
    type: Function
  },
  options: {
    type: Array,
    default: function _default() {
      return [];
    }
  },
  overlayProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  placeholder: {
    type: [String, Function],
    default: "\u9009\u62E9\u9009\u9879"
  },
  subTitles: {
    type: Array,
    default: function _default() {
      return [];
    }
  },
  theme: {
    type: String,
    default: "step",
    validator: function validator(val) {
      if (!val) return true;
      return ["step", "tab"].includes(val);
    }
  },
  title: {
    type: [String, Function]
  },
  value: {
    type: [String, Number],
    default: void 0
  },
  modelValue: {
    type: [String, Number],
    default: void 0
  },
  defaultValue: {
    type: [String, Number]
  },
  visible: Boolean,
  onChange: Function,
  onClose: Function,
  onPick: Function
};

export { props as default };
