// tdesign-mobile-vue/cjs/list/list.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var Vue = require('vue');
var core = require('@vueuse/core');
var hooks_tnode = require('../hooks/tnode.js');
var loading_index = require('../loading/index.js');
var config = require('../config.js');
var list_props = require('./props.js');
var hooks_useClass = require('../hooks/useClass.js');
var configProvider_useConfig = require('../_chunks/dep-605cd804.js');
var shared_useScrollParent_index = require('../shared/useScrollParent/index.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-b999c3d4.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-53383176.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-605741bc.js');
require('../_chunks/dep-d1ba5719.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-7aa6341b.js');
require('../_common/js/utils/general.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/defineProperty');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/setStyle.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@babel/runtime/helpers/slicedToArray');
require('../_chunks/dep-9bddd251.js');
require('../_chunks/dep-af0b484e.js');
require('../_chunks/dep-055d28fd.js');
require('../loading/icon/spinner.js');
require('../loading/props.js');
require('../shared/dom.js');
require('../loading/plugin.js');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../shared/component.js');
require('../shared/util.js');

var prefix = config["default"].prefix;
var _List = Vue.defineComponent({
  name: "".concat(prefix, "-list"),
  components: {
    TLoading: loading_index.Loading
  },
  props: list_props["default"],
  emits: ["load-more", "scroll"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var listClass = hooks_useClass.usePrefixClass("list");
    var _useConfig = configProvider_useConfig.useConfig("list"),
      globalConfig = _useConfig.globalConfig;
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var LOADING_TEXT_MAP = {
      loading: globalConfig.value.loading,
      "load-more": globalConfig.value.loadingMoreText
    };
    var root = Vue.ref();
    var scrollParent = shared_useScrollParent_index.useScrollParent(root);
    var _useWindowSize = core.useWindowSize(),
      height = _useWindowSize.height;
    var onLoadMore = function onLoadMore(e) {
      if (props2.asyncLoading === "load-more") {
        var _props2$onLoadMore;
        (_props2$onLoadMore = props2.onLoadMore) === null || _props2$onLoadMore === void 0 || _props2$onLoadMore.call(props2);
      }
    };
    var handleScroll = function handleScroll(e) {
      var _props2$onScroll;
      var scrollHeight = e.target.scrollHeight || Math.max(document.documentElement.scrollHeight, document.body.scrollHeight);
      var scrollTop = e.target.scrollTop || document.documentElement.scrollTop || document.body.scrollTop;
      var offsetHeight = e.target.offsetHeight || height.value;
      (_props2$onScroll = props2.onScroll) === null || _props2$onScroll === void 0 || _props2$onScroll.call(props2, scrollHeight - (scrollTop + offsetHeight), scrollTop);
    };
    var renderLoading = function renderLoading() {
      if (typeof props2.asyncLoading === "string" && ["loading", "load-more"].includes(props2.asyncLoading)) {
        return Vue.createVNode(loading_index.Loading, {
          "indicator": props2.asyncLoading === "loading",
          "text": typeof props2.asyncLoading === "string" ? LOADING_TEXT_MAP[props2.asyncLoading] : "",
          "class": "".concat(listClass.value, "__loading")
        }, null);
      }
      return renderTNodeJSX("asyncLoading");
    };
    core.useEventListener(scrollParent, "scroll", handleScroll);
    return function () {
      var headerContent = renderTNodeJSX("header");
      var footerContent = renderTNodeJSX("footer");
      return Vue.createVNode("div", {
        "ref": root,
        "class": listClass.value,
        "onScroll": handleScroll
      }, [headerContent, slots.default && slots.default(), Vue.createVNode("div", {
        "class": "".concat(listClass.value, "__loading--wrapper"),
        "onClick": onLoadMore
      }, [renderLoading()]), footerContent]);
    };
  }
});

exports["default"] = _List;
