// tdesign-mobile-vue/cjs/indexes/indexes.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var Vue = require('vue');
var shared_dom = require('../shared/dom.js');
var config = require('../config.js');
var indexes_props = require('./props.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useVModel = require('../hooks/useVModel.js');
var debounce = require('../_chunks/dep-4326ca38.js');
var isObject = require('../_chunks/dep-b999c3d4.js');
require('../_chunks/dep-0bd122a0.js');
require('../_chunks/dep-53383176.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-588e2e70.js');
require('../_chunks/dep-3cf5587c.js');
require('../_chunks/dep-f84e6e17.js');
require('../_chunks/dep-605cd804.js');
require('@babel/runtime/helpers/defineProperty');
require('../_chunks/dep-6f53fb1b.js');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-b3e3ccf8.js');
require('../_chunks/dep-7220ff76.js');
require('dayjs');
require('../_chunks/dep-05850262.js');
require('../_chunks/dep-e9a63584.js');
require('../_chunks/dep-0c11c462.js');
require('../_chunks/dep-159c7a94.js');
require('../_chunks/dep-ea269588.js');
require('../_chunks/dep-b633ec9e.js');
require('../_chunks/dep-7aa6341b.js');
require('../_chunks/dep-b8e8d654.js');
require('../_chunks/dep-61ce8c0d.js');
require('../_chunks/dep-5cf8bf50.js');
require('../_chunks/dep-2457b474.js');
require('../_chunks/dep-4b045875.js');
require('../_chunks/dep-a6a4534e.js');
require('../_chunks/dep-fef48622.js');
require('../_chunks/dep-8d30e8d1.js');
require('../_chunks/dep-dd93e905.js');
require('../_chunks/dep-d1ba5719.js');
require('../hooks/render-tnode.js');
require('../_chunks/dep-a65592ed.js');
require('../_chunks/dep-fb9c8646.js');
require('../_chunks/dep-2ab29468.js');
require('../_chunks/dep-42a5275c.js');
require('../_chunks/dep-09a28f29.js');
require('../_chunks/dep-ebd9fe98.js');
require('../_chunks/dep-9e13db9a.js');
require('../_chunks/dep-605741bc.js');
require('../_common/js/utils/general.js');
require('../_chunks/dep-2da1d545.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);

/** Error message constants. */
var FUNC_ERROR_TEXT = 'Expected a function';

/**
 * Creates a throttled function that only invokes `func` at most once per
 * every `wait` milliseconds. The throttled function comes with a `cancel`
 * method to cancel delayed `func` invocations and a `flush` method to
 * immediately invoke them. Provide `options` to indicate whether `func`
 * should be invoked on the leading and/or trailing edge of the `wait`
 * timeout. The `func` is invoked with the last arguments provided to the
 * throttled function. Subsequent calls to the throttled function return the
 * result of the last `func` invocation.
 *
 * **Note:** If `leading` and `trailing` options are `true`, `func` is
 * invoked on the trailing edge of the timeout only if the throttled function
 * is invoked more than once during the `wait` timeout.
 *
 * If `wait` is `0` and `leading` is `false`, `func` invocation is deferred
 * until to the next tick, similar to `setTimeout` with a timeout of `0`.
 *
 * See [David Corbacho's article](https://css-tricks.com/debouncing-throttling-explained-examples/)
 * for details over the differences between `_.throttle` and `_.debounce`.
 *
 * @static
 * @memberOf _
 * @since 0.1.0
 * @category Function
 * @param {Function} func The function to throttle.
 * @param {number} [wait=0] The number of milliseconds to throttle invocations to.
 * @param {Object} [options={}] The options object.
 * @param {boolean} [options.leading=true]
 *  Specify invoking on the leading edge of the timeout.
 * @param {boolean} [options.trailing=true]
 *  Specify invoking on the trailing edge of the timeout.
 * @returns {Function} Returns the new throttled function.
 * @example
 *
 * // Avoid excessively updating the position while scrolling.
 * jQuery(window).on('scroll', _.throttle(updatePosition, 100));
 *
 * // Invoke `renewToken` when the click event is fired, but not more than once every 5 minutes.
 * var throttled = _.throttle(renewToken, 300000, { 'trailing': false });
 * jQuery(element).on('click', throttled);
 *
 * // Cancel the trailing throttled invocation.
 * jQuery(window).on('popstate', throttled.cancel);
 */
function throttle(func, wait, options) {
  var leading = true,
    trailing = true;
  if (typeof func != 'function') {
    throw new TypeError(FUNC_ERROR_TEXT);
  }
  if (isObject.isObject(options)) {
    leading = 'leading' in options ? !!options.leading : leading;
    trailing = 'trailing' in options ? !!options.trailing : trailing;
  }
  return debounce.debounce(func, wait, {
    'leading': leading,
    'maxWait': wait,
    'trailing': trailing
  });
}

var prefix = config["default"].prefix;
var _Indexes = Vue.defineComponent({
  name: "".concat(prefix, "-indexes"),
  props: indexes_props["default"],
  emits: ["select", "change", "update:current"],
  setup: function setup(props) {
    var _toRefs = Vue.toRefs(props),
      current = _toRefs.current,
      modelValue = _toRefs.modelValue;
    var readerTNodeJSX = hooks_tnode.useTNodeJSX();
    var indexesClass = hooks_useClass.usePrefixClass("indexes");
    var _useVModel = hooks_useVModel["default"](current, modelValue, props.defaultCurrent, function () {}, "current"),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      currentIndex = _useVModel2[0],
      setCurrentIndex = _useVModel2[1];
    var timeOut;
    var indexesRoot = Vue.ref();
    var parentRect = Vue.ref();
    var state = Vue.reactive({
      showSidebarTip: false,
      activeSidebar: "",
      children: []
    });
    var groupTop = [];
    var indexList = Vue.computed(function () {
      if (!props.indexList) {
        var start = "A".charCodeAt(0);
        var alphabet = [];
        for (var i = start, end = start + 26; i < end; i += 1) {
          alphabet.push(String.fromCharCode(i));
        }
        return alphabet;
      }
      return props.indexList;
    });
    var setAnchorOnScroll = function setAnchorOnScroll(scrollTop) {
      if (!groupTop.length) return;
      var sticky = props.sticky,
        stickyOffset = props.stickyOffset;
      var stickyTop = stickyOffset + parentRect.value.top;
      scrollTop += stickyTop;
      var curIndex = groupTop.findIndex(function (group) {
        return scrollTop >= group.top - group.height && scrollTop <= group.top + group.totalHeight - group.height;
      });
      state.activeSidebar = groupTop[0].anchor;
      if (curIndex === -1) return;
      state.activeSidebar = groupTop[curIndex].anchor;
      var curGroup = groupTop[curIndex];
      setCurrentIndex(curGroup.anchor);
      if (sticky) {
        var offset = curGroup.top - scrollTop;
        var betwixt = offset < curGroup.height && offset > 0 && scrollTop > stickyTop;
        state.children.forEach(function (child, index) {
          var $el = child.$el;
          var wrapperClass = "".concat(indexesClass.value, "-anchor__wrapper");
          var headerClass = "".concat(indexesClass.value, "-anchor__header");
          var wrapper = $el.querySelector(".".concat(wrapperClass));
          var header = $el.querySelector(".".concat(headerClass));
          if (index === curIndex) {
            if (scrollTop - parentRect.value.top > stickyOffset) {
              wrapper.classList.add("".concat(wrapperClass, "--sticky"));
            } else {
              wrapper.classList.remove("".concat(wrapperClass, "--sticky"));
            }
            wrapper.classList.add("".concat(wrapperClass, "--active"));
            header.classList.add("".concat(headerClass, "--active"));
            wrapper.style = "transform: translate3d(0, ".concat(betwixt ? offset : 0, "px, 0); top: ").concat(stickyTop, "px");
          } else if (index + 1 === curIndex) {
            wrapper.classList.add("".concat(wrapperClass, "--sticky"));
            wrapper.classList.add("".concat(wrapperClass, "--active"));
            header.classList.add("".concat(headerClass, "--active"));
            wrapper.style = "transform: translate3d(0, ".concat(betwixt ? offset - groupTop[index].height : 0, "px, 0); top: ").concat(stickyTop, "px;");
          } else {
            wrapper.classList.remove("".concat(wrapperClass, "--sticky"));
            wrapper.classList.remove("".concat(wrapperClass, "--active"));
            header.classList.remove("".concat(headerClass, "--active"));
            wrapper.style = "";
          }
        });
      }
    };
    var scrollToByIndex = function scrollToByIndex(index) {
      var curGroup = groupTop.find(function (item) {
        return item.anchor === index;
      });
      if (indexesRoot.value) {
        var _indexesRoot$value$sc, _indexesRoot$value, _curGroup$top;
        (_indexesRoot$value$sc = (_indexesRoot$value = indexesRoot.value).scrollTo) === null || _indexesRoot$value$sc === void 0 || _indexesRoot$value$sc.call(_indexesRoot$value, 0, (_curGroup$top = curGroup.top) !== null && _curGroup$top !== void 0 ? _curGroup$top : 0);
        setCurrentIndex(index);
      }
    };
    var setActiveSidebarAndTip = function setActiveSidebarAndTip(index) {
      state.activeSidebar = index;
      state.showSidebarTip = true;
    };
    var handleSidebarItemClick = function handleSidebarItemClick(index) {
      var _props$onSelect;
      (_props$onSelect = props.onSelect) === null || _props$onSelect === void 0 || _props$onSelect.call(props, index);
      setActiveSidebarAndTip(index);
      scrollToByIndex(index);
    };
    var handleRootScroll = throttle(function (e) {
      var _indexesRoot$value$sc2, _indexesRoot$value2;
      var scrollTop = (_indexesRoot$value$sc2 = (_indexesRoot$value2 = indexesRoot.value) === null || _indexesRoot$value2 === void 0 ? void 0 : _indexesRoot$value2.scrollTop) !== null && _indexesRoot$value$sc2 !== void 0 ? _indexesRoot$value$sc2 : 0;
      setAnchorOnScroll(scrollTop);
    }, 1e3 / 30);
    var clearSidebarTip = function clearSidebarTip() {
      if (state.showSidebarTip && state.activeSidebar) {
        timeOut && clearTimeout(timeOut);
        timeOut = window.setTimeout(function () {
          state.showSidebarTip = false;
        }, 1e3);
      }
    };
    var getAnchorsRect = function getAnchorsRect() {
      return Promise.all(state.children.map(function (child) {
        var $el = child.$el,
          index = child.index;
        var rect = $el.getBoundingClientRect();
        groupTop.push({
          height: rect.height,
          top: rect.top - parentRect.value.top,
          anchor: index,
          totalHeight: 0
        });
        return child;
      }));
    };
    var handleSidebarTouchmove = function handleSidebarTouchmove(event) {
      shared_dom.preventDefault(event, false);
      var touches = event.touches;
      var _touches$ = touches[0],
        clientX = _touches$.clientX,
        clientY = _touches$.clientY;
      var target = document.elementFromPoint(clientX, clientY);
      if (target && target.className === "".concat(indexesClass.value, "__sidebar-item") && target instanceof HTMLElement) {
        var index = target.dataset.index;
        var curIndex = indexList.value.find(function (idx) {
          return String(idx) === index;
        });
        if (curIndex !== void 0 && state.activeSidebar !== curIndex) {
          setActiveSidebarAndTip(curIndex);
          scrollToByIndex(curIndex);
        }
      }
    };
    var relation = function relation(child) {
      child && state.children.push(child);
    };
    Vue.watchEffect(function () {
      if (state.showSidebarTip) {
        clearSidebarTip();
      }
    });
    Vue.watch(function () {
      return state.activeSidebar;
    }, function (val, oldVal) {
      if (val !== oldVal) {
        var _props$onChange;
        (_props$onChange = props.onChange) === null || _props$onChange === void 0 || _props$onChange.call(props, state.activeSidebar);
      }
    });
    var init = function init() {
      Vue.nextTick(function () {
        var _indexesRoot$value3;
        parentRect.value = ((_indexesRoot$value3 = indexesRoot.value) === null || _indexesRoot$value3 === void 0 ? void 0 : _indexesRoot$value3.getBoundingClientRect()) || {
          top: 0
        };
        getAnchorsRect().then(function () {
          groupTop.forEach(function (item, index) {
            var next = groupTop[index + 1];
            item.totalHeight = ((next === null || next === void 0 ? void 0 : next.top) || Infinity) - item.top;
          });
          setAnchorOnScroll(0);
          scrollToByIndex(currentIndex.value);
        });
      });
    };
    Vue.onMounted(init);
    Vue.watch(function () {
      return props.indexList;
    }, init);
    Vue.watch(function () {
      return currentIndex.value;
    }, function (val) {
      scrollToByIndex(val);
    }, {
      immediate: true
    });
    Vue.onBeforeUnmount(function () {
      timeOut && clearTimeout(timeOut);
    });
    Vue.provide("indexesProvide", {
      relation: relation
    });
    return function () {
      return Vue.createVNode("div", {
        "ref": indexesRoot,
        "class": indexesClass.value,
        "onScroll": handleRootScroll
      }, [Vue.createVNode("div", {
        "class": "".concat(indexesClass.value, "__sidebar")
      }, [indexList.value.map(function (item) {
        return Vue.createVNode("div", {
          "class": ["".concat(indexesClass.value, "__sidebar-item"), state.activeSidebar === item ? "".concat(indexesClass.value, "__sidebar-item--active") : ""],
          "data-index": item,
          "onClick": function onClick(e) {
            e.preventDefault();
            handleSidebarItemClick(item);
          },
          "onTouchmove": handleSidebarTouchmove
        }, [item, state.showSidebarTip && state.activeSidebar === item && Vue.createVNode("div", {
          "class": "".concat(indexesClass.value, "__sidebar-tips")
        }, [state.activeSidebar])]);
      })]), readerTNodeJSX("default")]);
    };
  }
});

exports["default"] = _Indexes;
