// tdesign-mobile-vue/lib/plugins.js
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

export { DialogPlugin as Dialog, DialogPlugin } from './dialog/index.js';
export { MessagePlugin as Message, MessagePlugin } from './message/index.js';
export { _ActionSheet as ActionSheet } from './action-sheet/index.js';
export { ToastPlugin as Toast, ToastPlugin } from './toast/plugin.js';
export { default as DrawerPlugin } from './drawer/plugin.js';
export { LoadingPlugin } from './loading/plugin.js';
import '@babel/runtime/helpers/defineProperty';
import 'vue';
import './dialog/dialog.js';
import '@babel/runtime/helpers/toConsumableArray';
import 'tdesign-icons-vue-next';
import 'lodash-es';
import './button/index.js';
import './button/button.js';
import './loading/index.js';
import './loading/loading.js';
import './loading/icon/gradient.js';
import './_common/js/loading/circle-adapter.js';
import './_common/js/utils/setStyle.js';
import './_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@babel/runtime/helpers/slicedToArray';
import './config.js';
import './hooks/useClass.js';
import './config-provider/useConfig.js';
import './config-provider/context.js';
import './_common/js/global-config/mobile/default-config.js';
import './_common/js/global-config/mobile/locale/zh_CN.js';
import './_chunks/dep-bbb6edd3.js';
import '@babel/runtime/helpers/typeof';
import './_chunks/dep-dc49adb8.js';
import 'dayjs';
import './loading/icon/spinner.js';
import './loading/props.js';
import './hooks/tnode.js';
import './hooks/render-tnode.js';
import './_common/js/utils/general.js';
import './shared/dom.js';
import './shared/component.js';
import './button/props.js';
import './form/hooks.js';
import './shared/hover.js';
import './popup/index.js';
import './popup/popup.js';
import './popup/props.js';
import './overlay/index.js';
import './overlay/overlay.js';
import './overlay/props.js';
import './hooks/useVModel.js';
import './hooks/useLockScroll.js';
import './_util/useTouch.js';
import './_util/getScrollParent.js';
import './hooks/useTeleport.js';
import './shared/render-tnode.js';
import './shared/util.js';
import './dialog/props.js';
import './message/message.js';
import './link/index.js';
import './link/link.js';
import './link/props.js';
import './message/props.js';
import './action-sheet/action-sheet.js';
import './action-sheet/action-sheet-grid.js';
import './grid/index.js';
import './grid/grid.js';
import './grid/props.js';
import './grid/grid-item.js';
import './grid/grid-item-props.js';
import './image/index.js';
import './image/image.js';
import '@vueuse/core';
import './image/props.js';
import './badge/index.js';
import './badge/badge.js';
import './badge/props.js';
import './swiper/index.js';
import './swiper/swiper.js';
import './swipe-cell/useSwipe.js';
import './swiper/props.js';
import './swiper/swiper-item.js';
import './action-sheet/action-sheet-list.js';
import './action-sheet/props.js';
import './toast/toast.js';
import './toast/props.js';
import './drawer/drawer.js';
import './drawer/props.js';
