// tdesign-mobile-vue/es/count-down/count-down.mjs
/**
 * tdesign v1.11.1
 * (c) 2025 TDesign Group
 * @license MIT
 */

import { defineComponent, computed, ref, onMounted, onBeforeUnmount, createVNode, Fragment } from 'vue';
import config from '../config.mjs';
import CountDownProps from './props.mjs';
import { useCountDown } from '../shared/useCountDown/index.mjs';
import { useTNodeJSX } from '../hooks/tnode.mjs';
import { usePrefixClass } from '../hooks/useClass.mjs';
import '../_chunks/dep-038bf4d0.mjs';
import '../_chunks/dep-4055ef5c.mjs';
import '../_chunks/dep-04a2869f.mjs';
import '@vueuse/core';
import '../shared/useCountDown/utils.mjs';
import '../shared/util.mjs';
import '../_chunks/dep-2ed7e0d5.mjs';
import '../_chunks/dep-9961b95b.mjs';
import 'lodash-es';
import '../hooks/render-tnode.mjs';
import '../_common/js/utils/general.mjs';
import '../config-provider/useConfig.mjs';
import '../_chunks/dep-f4e31210.mjs';
import '../config-provider/context.mjs';
import '../_common/js/global-config/mobile/default-config.mjs';
import '../_common/js/global-config/mobile/locale/zh_CN.mjs';
import '../_chunks/dep-4c96ffa8.mjs';
import '../_chunks/dep-a8bd494a.mjs';
import '../_chunks/dep-d3be051e.mjs';
import '../config-provider/type.mjs';

var prefix = config.prefix;
var _CountDown = defineComponent({
  name: "".concat(prefix, "-count-down"),
  props: CountDownProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var countDownClass = usePrefixClass("count-down");
    var countDownClasses = computed(function () {
      return [countDownClass.value, "".concat(countDownClass.value, "--").concat(props.theme), "".concat(countDownClass.value, "--").concat(props.size)];
    });
    var visibility = ref(true);
    var visibilitychangeListener = function visibilitychangeListener() {
      visibility.value = !document.hidden;
    };
    onMounted(function () {
      document.addEventListener("visibilitychange", visibilitychangeListener, false);
    });
    onBeforeUnmount(function () {
      document.removeEventListener("visibilitychange", visibilitychangeListener, false);
    });
    var _useCountDown = useCountDown(props, visibility),
      showTimes = _useCountDown.showTimes;
    return function () {
      var renderContent = function renderContent() {
        var content = renderTNodeJSX("content");
        if (content !== "default" && content) {
          return content;
        }
        return showTimes.map(function (item) {
          return createVNode(Fragment, null, [createVNode("span", {
            "class": "".concat(countDownClass.value, "__item")
          }, [item.value]), item.mark && createVNode("span", {
            "class": ["".concat(countDownClass.value, "__split"), "".concat(countDownClass.value, "__split--").concat(props.splitWithUnit ? "text" : "dot")]
          }, [item.mark])]);
        });
      };
      return createVNode("div", {
        "class": countDownClasses.value
      }, [renderContent()]);
    };
  }
});

export { _CountDown as default };
